{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "from collections import Counter\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import ListedColormap\n",
    "from datetime import datetime, date, timedelta\n",
    "\n",
    "import io\n",
    "from pysam import VariantFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "context = 'paper' # talk or paper\n",
    "\n",
    "if context == 'talk':\n",
    "    sns.set(style=\"darkgrid\", context=\"talk\", rc={\"lines.linewidth\": 2, \"legend.fontsize\":25})\n",
    "    plt.style.use(\"dark_background\")\n",
    "    plt.rcParams.update({\"grid.linewidth\":0.5, \"grid.alpha\":0.5,'font.size': 30})\n",
    "    sns.set_palette(\"deep\")\n",
    "else:\n",
    "    sns.set(context=\"paper\", font_scale=4, rc={\"lines.linewidth\": 2, \"legend.fontsize\":25})\n",
    "    plt.rcParams.update({'font.size': 35})\n",
    "    sns.set_palette(\"deep\")\n",
    "\n",
    "lc = 'w' if context == 'talk' else 'k'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CRC deep WGS patients\n",
    "# MSS: 986, 809, 1014, 519, 1279\n",
    "# MS status not done: 512, 1531\n",
    "\n",
    "#patient = 1014\n",
    "#patient = 986\n",
    "#patient = 809\n",
    "patient = 1388"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tumor burden"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tf_file['patient'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# tumor burden\n",
    "tf_file_1 = pd.read_csv('../data/tumor_burden/tumor_burden_ichorcna_CCG.txt', header=None)\n",
    "tf_file_1['patient'] = tf_file_1[0].str.split('.').str[0].str.split('_').str[-1]\n",
    "tf_file_1['date'] = tf_file_1[0].str.split('.').str[1]\n",
    "tf_file_1['patient_date'] = tf_file_1['patient']  + '_' + tf_file_1['date'] \n",
    "tf_file_1.set_index('patient_date', inplace=True)\n",
    "tf_file_1.drop(0, axis=1, inplace=True)\n",
    "tf_file_1.rename(columns={1: 'tumor_burden'}, inplace=True)\n",
    "\n",
    "tf_file_2 = pd.read_csv('../data/tumor_burden/tumor_burden_ichorcna_CRC.txt', header=None)\n",
    "tf_file_2['patient'] = tf_file_2[0].str.split('-').str[1].str.split('_').str[0]\n",
    "tf_file_2['date'] = tf_file_2[0].str.split('_').str[1]\n",
    "tf_file_2['patient_date'] = tf_file_2['patient'] + '_' + tf_file_2['date'] \n",
    "tf_file_2.set_index('patient_date', inplace=True)\n",
    "tf_file_2.drop(0, axis=1, inplace=True)\n",
    "tf_file_2.rename(columns={1: 'tumor_burden'}, inplace=True)\n",
    "\n",
    "# print(tf_file_1.shape, tf_file_2.shape)\n",
    "\n",
    "tf_file = pd.concat([tf_file_1, tf_file_2])\n",
    "tf_file['patient'] = tf_file['patient'].astype(int)\n",
    "tf_file['date'] = pd.to_datetime(tf_file['date'], format='%d%m%y')\n",
    "tf_file.reset_index(inplace=True)\n",
    "tf_file.drop('patient_date', axis=1, inplace=True)\n",
    "\n",
    "tumorburden_dates = tf_file[tf_file['patient'] == patient]['date'].sort_values().astype(str).unique()\n",
    "print(tumorburden_dates)\n",
    "\n",
    "print(tf_file[tf_file['patient'] == patient].shape[0])\n",
    "tf_file[tf_file['patient'] == patient]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Treatment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "treatment_file = pd.read_csv('../data/treatment/patient_treatment_total_std_201109.txt', sep='\\t')\n",
    "treatment_file['patient'] = treatment_file['patient'].astype(int)\n",
    "treatment_file['date'] = pd.to_datetime(treatment_file['date'], format='%Y-%m-%d')\n",
    "treatment_file.rename(columns={'value':'treatment'}, inplace=True)\n",
    "treatment_file = treatment_file[['patient', 'date', 'treatment']]\n",
    "treatment_file[treatment_file['patient'] == patient]['date'].astype(str).unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Patient timeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "treatment_patient = treatment_file[treatment_file['patient'] == patient].sort_values('date')\n",
    "treatment_patient['date'] = treatment_patient['date'].astype(str)\n",
    "tf_patient = tf_file[tf_file['patient'] == patient].sort_values('date')\n",
    "tf_patient['date'] = tf_patient['date'].astype(str)\n",
    "df_patient = pd.concat([treatment_patient, tf_patient])\n",
    "df_patient = df_patient.sort_values('date')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print([str(a) for a in tf_patient['date'].unique() if a not in df_patient['date'].unique()])\n",
    "print([str(b) for b in df_patient['date'].unique() if b not in tf_patient['date'].unique()])\n",
    "alldates = sorted(list(set(list(df_patient['date'].values) + list(tf_patient['date'].values))))\n",
    "print(len(alldates))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "daytimes = [(datetime.strptime(alldates[i], '%Y-%m-%d') - datetime.strptime(alldates[0], '%Y-%m-%d')).days\n",
    "            for i in range(len(alldates))]\n",
    "print(daytimes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "\n",
    "fig, ax2 = plt.subplots(figsize=(40,10))\n",
    "\n",
    "# make a plot with different y-axis using second axis object\n",
    "sns.stripplot(y='treatment', x='date', hue='treatment', data=df_patient, s=10, ax=ax2)\n",
    "ax2.grid(False)\n",
    "plt.legend((), ())\n",
    "        \n",
    "# twin object for two different y-axis on the sample plot\n",
    "ax=ax2.twinx()\n",
    "ele0 = ax.plot(df_patient['date'], df_patient['tumor_burden'], lc+'.', marker='s', markersize=10, label='ichorCNA tumor burden')\n",
    "\n",
    "ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'-', linewidth=4)\n",
    "ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'.', marker='s', markersize=10)\n",
    "ax.set_ylabel('fraction (tumor burden or VAF)', fontsize=30)\n",
    "ax.set_ylim(-0.01, 1)\n",
    "\n",
    "fig.legend([ele0], ['ichorCNA tumor burden'], loc='upper left')\n",
    "ax.legend(loc='upper left')\n",
    "\n",
    "labels = [ad if ad in tumorburden_dates else '' for ad in alldates]\n",
    "ax2.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "ax.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "\n",
    "plt.title('Patient {}: VAF evolution across timepoints'.format(patient))\n",
    "\n",
    "if not os.path.exists('../figures/oncosg_timeline_patient'+str(patient)+'.png'):\n",
    "    plt.savefig('../figures/oncosg_timeline_patient'+str(patient)+'.png', bbox_inches='tight')\n",
    "    \n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# check tf dates\n",
    "\n",
    "date_lowtftimepoints = list(tf_patient[tf_patient['tumor_burden'] == 0]['date'].unique())\n",
    "print(date_lowtftimepoints)\n",
    "tf_patient[['date', 'tumor_burden']].T"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    res = pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    return res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# if  os.path.exists('../data/variant_calls/'+str(patient)+'_reGeno.VEP.vcf'):\n",
    "#    mutation_df = read_vcf('../data/variant_calls/'+str(patient)+'_reGeno.VEP.vcf')\n",
    "#    mutation_df = mutation_df[mutation_df['FILTER'] == 'PASS']\n",
    "#    print(mutation_df.columns)\n",
    "#    \n",
    "#    col = ['CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO', 'FORMAT']\n",
    "#    for it in list(date_lowtftimepoints):\n",
    "#        aux = '986_'+ it.split('-')[-1]+it.split('-')[1]+it.split('-')[0][-2:]\n",
    "#        if aux in mutation_df.columns:\n",
    "#            col.append(aux)\n",
    "#        else:\n",
    "#            print(aux)\n",
    "#    print(col)\n",
    "#    mutation_lowtftimepoints = mutation_df[col]\n",
    "#    mutation_lowtftimepoints\n",
    "#    foo2 = lambda x: pd.Series(float(x.split(':')[-1])/float(x.split(':')[-2]))\n",
    "#    mutation_lowtftimepoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_file = pd.read_csv('../data/variant_calls/total_variants_mmpm_puradj.txt', sep='\\t')\n",
    "print(mutation_file['GENE'].nunique())\n",
    "mutation_file = mutation_file[mutation_file['pid'] == patient]\n",
    "mutation_file.groupby('GENE').count()\n",
    "mutation_lowtftimepoints = mutation_file[(mutation_file['ichor_tf'] == 0)][['date', 'GENE', 'REF', 'ALT', 'VAF', 'TIERS']]\n",
    "mutation_lowtftimepoints = mutation_file[(mutation_file['ichor_tf'] < 0.05)][['date', 'GENE', 'REF', 'ALT', 'VAF', 'TIERS']]\n",
    "mutation_lowtftimepoints = mutation_lowtftimepoints[(mutation_lowtftimepoints['TIERS'] != 'Germline') & (mutation_lowtftimepoints['TIERS'] != 'Artifact')]\n",
    "\n",
    "mutation_lowtftimepoints = mutation_lowtftimepoints[(mutation_lowtftimepoints['TIERS'] == 'Trusted') ]\n",
    "mutation_lowtftimepoints"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lowtftimepoints_dict = {'date': date_lowtftimepoints,\n",
    "                        'median VAF': [],\n",
    "                        '# mutated genes' : [],\n",
    "                        'median VAF within mutated genes': [],\n",
    "                        '# mutated genes TRUSTED': [],\n",
    "                        'median VAF within mutated genes TRUSTED': []\n",
    "                       }\n",
    "\n",
    "for date in date_lowtftimepoints:\n",
    "    nmut = mutation_lowtftimepoints[(date_lowtftimepoints == date) & (mutation_lowtftimepoints['VAF'] != 0)].shape[0]\n",
    "    nmuttrust = mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date)  & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')].shape[0]\n",
    "    medianvaf = np.median(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date)]['VAF'].values)\n",
    "    medianvafn = np.median(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)]['VAF'].values)\n",
    "    medianvafntrust = np.median(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0) & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]['VAF'].values)\n",
    "    lowtftimepoints_dict['# mutated genes'].append(nmut)\n",
    "    lowtftimepoints_dict['# mutated genes TRUSTED'].append(nmuttrust)\n",
    "    lowtftimepoints_dict['median VAF'].append(medianvaf)\n",
    "    lowtftimepoints_dict['median VAF within mutated genes'].append(medianvafn)\n",
    "    lowtftimepoints_dict['median VAF within mutated genes TRUSTED'].append(medianvafntrust)\n",
    "    \n",
    "\n",
    "lowtftimepoints_pd = pd.DataFrame.from_dict(lowtftimepoints_dict)\n",
    "lowtftimepoints_pd.set_index('date', inplace=True)\n",
    "lowtftimepoints_pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bed file for genes to check manually"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "location on Aquila\n",
    "\n",
    "/mnt/projects/zwpoh/cfDNA/bulk/ccg/ccg_batch2/lpwgs/hg19_bam/patient/986*.bam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genes2checkmanually = []\n",
    "results_df = None\n",
    "for date in mutation_lowtftimepoints['date'].unique():\n",
    "    print(date)\n",
    "    if results_df is None:\n",
    "        results_df = mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]\n",
    "    else:\n",
    "        results_df = pd.concat([results_df, mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]])\n",
    "    ll = list(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]['GENE'].values)\n",
    "    for g in ll:\n",
    "        genes2checkmanually.append(g)\n",
    "genes2checkmanually = list(np.unique(genes2checkmanually))\n",
    "# print genes to check munually\n",
    "print(genes2checkmanually)\n",
    "\n",
    "# need to get gene location in hg19\n",
    "print(mutation_lowtftimepoints['date'].unique())\n",
    "gene_df = None\n",
    "for date in  mutation_lowtftimepoints['date'].unique():\n",
    "    for i, r in mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')].iterrows():\n",
    "        if gene_df is None:\n",
    "            gene_df = mutation_file[['date', '#CHROM', 'POS', 'REF', 'ALT', 'GENE']][(mutation_file['date'] == r['date']) & (mutation_file['GENE'] == r['GENE'])]\n",
    "        else:\n",
    "            gene_df = pd.concat([gene_df, mutation_file[['date', '#CHROM', 'POS', 'REF', 'ALT', 'GENE']][(mutation_file['date'] == r['date']) & (mutation_file['GENE'] == r['GENE'])]])\n",
    "gene_df.drop('date', axis=1, inplace=True)\n",
    "gene_df.drop_duplicates(inplace=True)\n",
    "gene_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_df[['#CHROM', 'POS', 'GENE', 'REF', 'ALT']].to_csv('../data/variant_calls/positionfile'+str(patient)+'.txt', header=False, index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 4 mutations across time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutations_acrosstime = mutation_file[mutation_file['GENE'].isin(genes2checkmanually)][['date', 'GENE', 'REF', 'ALT', 'VAF', 'TIERS']]\n",
    "mutations_acrosstime = mutations_acrosstime[(mutations_acrosstime['TIERS'] == 'Trusted')]\n",
    "mutations_acrosstime = mutations_acrosstime.pivot_table(values='VAF', index='GENE', columns='date', aggfunc='first')\n",
    "mutations_acrosstime = mutations_acrosstime.T\n",
    "mutations_acrosstime.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "\n",
    "fig, ax2 = plt.subplots(figsize=(40,10))\n",
    "\n",
    "# make a plot with different y-axis using second axis object\n",
    "sns.stripplot(y='treatment', x='date', hue='treatment', data=df_patient, s=10, ax=ax2)\n",
    "labels = ax2.axes.get_xticklabels()\n",
    "ax2.grid(False)\n",
    "plt.legend((), ())\n",
    "        \n",
    "# twin object for two different y-axis on the sample plot\n",
    "ax=ax2.twinx()\n",
    "ele0 = ax.plot(df_patient['date'], df_patient['tumor_burden'], lc+'.', marker='s', markersize=10, label='ichorCNA tumor burden')\n",
    "\n",
    "ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'-', linewidth=4)\n",
    "ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'.', marker='s', markersize=10)\n",
    "ax.set_ylabel('fraction (tumor burden or VAF)', fontsize=30)\n",
    "ax.set_ylim(-0.01, 1)\n",
    "\n",
    "xacrosstime = [i for i in mutations_acrosstime.index if i in df_patient['date'].values]\n",
    "\n",
    "eles = [ele0]\n",
    "collist = ['r', 'b', 'g', 'c', 'm', 'y']\n",
    "\n",
    "for gi, gene in enumerate(genes2checkmanually):\n",
    "\n",
    "    yacrosstime = [m for i,m in enumerate(mutations_acrosstime[gene].values) if mutations_acrosstime.index[i] in df_patient['date'].values]\n",
    "    elei = ax.plot(xacrosstime, yacrosstime, collist[gi]+'-', linewidth=2, label=gene)\n",
    "    ax.plot(xacrosstime, yacrosstime, collist[gi]+'.', marker='D', markersize=10)\n",
    "    eles.append(elei)\n",
    "\n",
    "fig.legend(eles, ['ichorCNA tumor burden'] + genes2checkmanually, loc='upper left')\n",
    "ax.legend(loc='upper left')\n",
    "\n",
    "labels = [ad if ad in tumorburden_dates else '' for ad in alldates]\n",
    "ax2.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "ax.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "\n",
    "plt.title('Patient {}: VAF evolution across timepoints'.format(patient))\n",
    "\n",
    "if not os.path.exists('../figures/oncosg_timeline_mutations_patient'+str(patient)+'.png'):\n",
    "    plt.savefig('../figures/oncosg_timeline_mutations_patient'+str(patient)+'.png', bbox_inches='tight')\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Generate plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mutation_file = pd.read_csv('../data/variant_calls/total_variants_mmpm_puradj.txt', sep='\\t')\n",
    "a = mutation_file['pid'].unique()\n",
    "b = tf_file['patient'].unique()\n",
    "print(a)\n",
    "set(a)-set(b)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for patient in mutation_file['pid'].unique():\n",
    "    print(patient)\n",
    "    # tumor burden\n",
    "    tf_file_1 = pd.read_csv('../data/tumor_burden/tumor_burden_ichorcna_CCG.txt', header=None)\n",
    "    tf_file_1['patient'] = tf_file_1[0].str.split('.').str[0].str.split('_').str[-1]\n",
    "    tf_file_1['date'] = tf_file_1[0].str.split('.').str[1]\n",
    "    tf_file_1['patient_date'] = tf_file_1['patient']  + '_' + tf_file_1['date'] \n",
    "    tf_file_1.set_index('patient_date', inplace=True)\n",
    "    tf_file_1.drop(0, axis=1, inplace=True)\n",
    "    tf_file_1.rename(columns={1: 'tumor_burden'}, inplace=True)\n",
    "\n",
    "    tf_file_2 = pd.read_csv('../data/tumor_burden/tumor_burden_ichorcna_CRC.txt', header=None)\n",
    "    tf_file_2['patient'] = tf_file_2[0].str.split('-').str[1].str.split('_').str[0]\n",
    "    tf_file_2['date'] = tf_file_2[0].str.split('_').str[1]\n",
    "    tf_file_2['patient_date'] = tf_file_2['patient'] + '_' + tf_file_2['date'] \n",
    "    tf_file_2.set_index('patient_date', inplace=True)\n",
    "    tf_file_2.drop(0, axis=1, inplace=True)\n",
    "    tf_file_2.rename(columns={1: 'tumor_burden'}, inplace=True)\n",
    "\n",
    "    # print(tf_file_1.shape, tf_file_2.shape)\n",
    "\n",
    "    tf_file = pd.concat([tf_file_1, tf_file_2])\n",
    "    tf_file['patient'] = tf_file['patient'].astype(int)\n",
    "    tf_file['date'] = pd.to_datetime(tf_file['date'], format='%d%m%y')\n",
    "    tf_file.reset_index(inplace=True)\n",
    "    tf_file.drop('patient_date', axis=1, inplace=True)\n",
    "\n",
    "    tumorburden_dates = tf_file[tf_file['patient'] == patient]['date'].sort_values().astype(str).unique()\n",
    "    \n",
    "    treatment_file = pd.read_csv('../data/treatment/patient_treatment_total_std_201109.txt', sep='\\t')\n",
    "    treatment_file['patient'] = treatment_file['patient'].astype(int)\n",
    "    treatment_file['date'] = pd.to_datetime(treatment_file['date'], format='%Y-%m-%d')\n",
    "    treatment_file.rename(columns={'value':'treatment'}, inplace=True)\n",
    "    treatment_file = treatment_file[['patient', 'date', 'treatment']]\n",
    "    treatment_file[treatment_file['patient'] == patient]['date'].astype(str).unique()\n",
    "    \n",
    "    treatment_patient = treatment_file[treatment_file['patient'] == patient].sort_values('date')\n",
    "    treatment_patient['date'] = treatment_patient['date'].astype(str)\n",
    "    tf_patient = tf_file[tf_file['patient'] == patient].sort_values('date')\n",
    "    tf_patient['date'] = tf_patient['date'].astype(str)\n",
    "    df_patient = pd.concat([treatment_patient, tf_patient])\n",
    "    df_patient = df_patient.sort_values('date')\n",
    "\n",
    "    alldates = sorted(list(set(list(df_patient['date'].values) + list(tf_patient['date'].values))))\n",
    "    \n",
    "\n",
    "    daytimes = [(datetime.strptime(alldates[i], '%Y-%m-%d') - datetime.strptime(alldates[0], '%Y-%m-%d')).days\n",
    "                for i in range(len(alldates))]\n",
    "    \n",
    "\n",
    "    plt.figure()\n",
    "\n",
    "    fig, ax2 = plt.subplots(figsize=(40,10))\n",
    "\n",
    "    # make a plot with different y-axis using second axis object\n",
    "    sns.stripplot(y='treatment', x='date', hue='treatment', data=df_patient, s=10, ax=ax2)\n",
    "    ax2.grid(False)\n",
    "    plt.legend((), ())\n",
    "\n",
    "    # twin object for two different y-axis on the sample plot\n",
    "    ax=ax2.twinx()\n",
    "    ele0 = ax.plot(df_patient['date'], df_patient['tumor_burden'], lc+'.', marker='s', markersize=10, label='ichorCNA tumor burden')\n",
    "\n",
    "    ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'-', linewidth=4)\n",
    "    ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'.', marker='s', markersize=10)\n",
    "    ax.set_ylabel('fraction (tumor burden or VAF)', fontsize=30)\n",
    "    ax.set_ylim(-0.01, 1)\n",
    "\n",
    "    fig.legend([ele0], ['ichorCNA tumor burden'], loc='upper left')\n",
    "    ax.legend(loc='upper left')\n",
    "\n",
    "    labels = [ad if ad in tumorburden_dates else '' for ad in alldates]\n",
    "    ax2.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "    ax.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "\n",
    "    plt.title('Patient {}: VAF evolution across timepoints'.format(patient))\n",
    "\n",
    "    if not os.path.exists('../figures/low_tb_timepoints/oncosg_timeline_patient'+str(patient)+'.png'):\n",
    "        plt.savefig('../figures/low_tb_timepoints/oncosg_timeline_patient'+str(patient)+'.png', bbox_inches='tight')\n",
    "\n",
    "\n",
    "    date_lowtftimepoints = list(tf_patient[tf_patient['tumor_burden'] == 0]['date'].unique())\n",
    "    \n",
    "    tf_patient[['date', 'tumor_burden']].T\n",
    "\n",
    "    mutation_file = pd.read_csv('../data/variant_calls/total_variants_mmpm_puradj.txt', sep='\\t')\n",
    "    \n",
    "    mutation_file = mutation_file[mutation_file['pid'] == patient]\n",
    "    mutation_file.groupby('GENE').count()\n",
    "    mutation_lowtftimepoints = mutation_file[(mutation_file['ichor_tf'] <= max(0, mutation_file['ichor_tf'].min()))][['date', 'GENE', 'REF', 'ALT', 'VAF', 'TIERS']]\n",
    "    mutation_lowtftimepoints = mutation_lowtftimepoints[(mutation_lowtftimepoints['TIERS'] != 'Germline') & (mutation_lowtftimepoints['TIERS'] != 'Artifact')]\n",
    "\n",
    "    mutation_lowtftimepoints = mutation_lowtftimepoints[(mutation_lowtftimepoints['TIERS'] == 'Trusted') ]\n",
    "    mutation_lowtftimepoints\n",
    "\n",
    "    lowtftimepoints_dict = {'date': date_lowtftimepoints,\n",
    "                            'median VAF': [],\n",
    "                            '# mutated genes' : [],\n",
    "                            'median VAF within mutated genes': [],\n",
    "                            '# mutated genes TRUSTED': [],\n",
    "                            'median VAF within mutated genes TRUSTED': []\n",
    "                           }\n",
    "\n",
    "    for date in date_lowtftimepoints:\n",
    "        nmut = mutation_lowtftimepoints[(date_lowtftimepoints == date) & (mutation_lowtftimepoints['VAF'] != 0)].shape[0]\n",
    "        nmuttrust = mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date)  & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')].shape[0]\n",
    "        medianvaf = np.median(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date)]['VAF'].values)\n",
    "        medianvafn = np.median(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)]['VAF'].values)\n",
    "        medianvafntrust = np.median(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0) & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]['VAF'].values)\n",
    "        lowtftimepoints_dict['# mutated genes'].append(nmut)\n",
    "        lowtftimepoints_dict['# mutated genes TRUSTED'].append(nmuttrust)\n",
    "        lowtftimepoints_dict['median VAF'].append(medianvaf)\n",
    "        lowtftimepoints_dict['median VAF within mutated genes'].append(medianvafn)\n",
    "        lowtftimepoints_dict['median VAF within mutated genes TRUSTED'].append(medianvafntrust)\n",
    "\n",
    "\n",
    "    lowtftimepoints_pd = pd.DataFrame.from_dict(lowtftimepoints_dict)\n",
    "    lowtftimepoints_pd.set_index('date', inplace=True)\n",
    "    lowtftimepoints_pd\n",
    "\n",
    "    genes2checkmanually = []\n",
    "    results_df = None\n",
    "    for date in mutation_lowtftimepoints['date'].unique():\n",
    "        if results_df is None:\n",
    "            results_df = mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]\n",
    "        else:\n",
    "            results_df = pd.concat([results_df, mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]])\n",
    "        ll = list(mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')]['GENE'].values)\n",
    "        for g in ll:\n",
    "            genes2checkmanually.append(g)\n",
    "    genes2checkmanually = list(np.unique(genes2checkmanually))\n",
    "    # print genes to check munually\n",
    "\n",
    "    if genes2checkmanually:\n",
    "        # need to get gene location in hg19\n",
    "        gene_df = None\n",
    "        for date in  mutation_lowtftimepoints['date'].unique():\n",
    "            for i, r in mutation_lowtftimepoints[(mutation_lowtftimepoints['date'] == date) & (mutation_lowtftimepoints['VAF'] != 0)  & (mutation_lowtftimepoints['TIERS'] == 'Trusted')].iterrows():\n",
    "                if gene_df is None:\n",
    "                    gene_df = mutation_file[['date', '#CHROM', 'POS', 'REF', 'ALT', 'GENE']][(mutation_file['date'] == r['date']) & (mutation_file['GENE'] == r['GENE'])]\n",
    "                else:\n",
    "                    gene_df = pd.concat([gene_df, mutation_file[['date', '#CHROM', 'POS', 'REF', 'ALT', 'GENE']][(mutation_file['date'] == r['date']) & (mutation_file['GENE'] == r['GENE'])]])\n",
    "        gene_df.drop('date', axis=1, inplace=True)\n",
    "        gene_df.drop_duplicates(inplace=True)\n",
    "\n",
    "\n",
    "        mutations_acrosstime = mutation_file[mutation_file['GENE'].isin(genes2checkmanually)][['date', 'GENE', 'REF', 'ALT', 'VAF', 'TIERS']]\n",
    "        mutations_acrosstime = mutations_acrosstime[(mutations_acrosstime['TIERS'] == 'Trusted')]\n",
    "        mutations_acrosstime = mutations_acrosstime.pivot_table(values='VAF', index='GENE', columns='date', aggfunc='first')\n",
    "        mutations_acrosstime = mutations_acrosstime.T\n",
    "\n",
    "\n",
    "        plt.figure()\n",
    "\n",
    "        fig, ax2 = plt.subplots(figsize=(40,10))\n",
    "\n",
    "        # make a plot with different y-axis using second axis object\n",
    "        sns.stripplot(y='treatment', x='date', hue='treatment', data=df_patient, s=10, ax=ax2)\n",
    "        labels = ax2.axes.get_xticklabels()\n",
    "        ax2.grid(False)\n",
    "        plt.legend((), ())\n",
    "\n",
    "        # twin object for two different y-axis on the sample plot\n",
    "        ax=ax2.twinx()\n",
    "        ele0 = ax.plot(df_patient['date'], df_patient['tumor_burden'], lc+'.', marker='s', markersize=10, label='ichorCNA tumor burden')\n",
    "\n",
    "        ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'-', linewidth=4)\n",
    "        ax.plot(df_patient['date'][~df_patient['tumor_burden'].isna()], df_patient['tumor_burden'][~df_patient['tumor_burden'].isna()], lc+'.', marker='s', markersize=10)\n",
    "        ax.set_ylabel('fraction (tumor burden or VAF)', fontsize=30)\n",
    "        ax.set_ylim(-0.01, 1)\n",
    "\n",
    "        xacrosstime = [i for i in mutations_acrosstime.index if i in df_patient['date'].values]\n",
    "\n",
    "        eles = [ele0]\n",
    "        collist = ['r', 'b', 'g', 'c', 'm', 'y', 'tab:pink', 'tab:orange', 'tab:olive', 'tab:purple', 'grey']\n",
    "\n",
    "        for gi, gene in enumerate(genes2checkmanually):\n",
    "\n",
    "            yacrosstime = [m for i,m in enumerate(mutations_acrosstime[gene].values) if mutations_acrosstime.index[i] in df_patient['date'].values]\n",
    "            elei = ax.plot(xacrosstime, yacrosstime, color=collist[gi], linewidth=2, label=gene)\n",
    "            ax.plot(xacrosstime, yacrosstime, color=collist[gi], ls='-.', marker='D', markersize=10)\n",
    "            eles.append(elei)\n",
    "\n",
    "        fig.legend(eles, ['ichorCNA tumor burden'] + genes2checkmanually, loc='upper left')\n",
    "        ax.legend(loc='upper left')\n",
    "\n",
    "        labels = [ad if ad in tumorburden_dates else '' for ad in alldates]\n",
    "        ax2.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "        ax.set_xticklabels(labels, rotation=90, fontsize=30)\n",
    "\n",
    "        plt.title('Patient {}: VAF evolution across timepoints'.format(patient))\n",
    "\n",
    "        if not os.path.exists('../figures/low_tb_timepoints/oncosg_timeline_mutations_patient'+str(patient)+'.png'):\n",
    "            plt.savefig('../figures/low_tb_timepoints/oncosg_timeline_mutations_patient'+str(patient)+'.png', bbox_inches='tight')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
