{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "lucky-border",
   "metadata": {},
   "source": [
    "# Ground truth analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "naked-driving",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: /Users/hanae/Repositories/cfdna_snv_benchmark\n"
     ]
    }
   ],
   "source": [
    "# Imports\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "import warnings\n",
    "from tqdm.notebook import tqdm\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, average_precision_score\n",
    "warnings.filterwarnings('ignore')\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "# set working directory\n",
    "if not os.getcwd().endswith('cfdna_snv_benchmark'):\n",
    "    os.chdir('../')\n",
    "print('Current working directory: {}'.format(os.getcwd()))\n",
    "\n",
    "from utils.config import Config\n",
    "from utils.viz import *\n",
    "from utils.table import *\n",
    "from utils.metrics import *\n",
    "from utils.calltable import *\n",
    "from utils.calltableseries import *\n",
    "from utils.groundtruth import *\n",
    "from utils.metricsseries import *\n",
    "from utils.venn import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "assigned-jaguar",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "paper\n",
      "['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan']\n"
     ]
    }
   ],
   "source": [
    "# Config and Display paramaters\n",
    "\n",
    "config = Config(\"config/\", \"config_viz.yaml\")\n",
    "set_display_params(config)\n",
    "print(config.methods)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "purple-adult",
   "metadata": {},
   "source": [
    "# Get call sets for cfDNA, T1W, M1W on exome calling"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "documented-destination",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "986_100215_T1-E\n",
      "freebayes\n",
      "mutect2\n",
      "retrieving mutect2 104 calls with MinAF tags out of 126\n",
      "strelka2\n",
      "retrieving strelka2 345 calls with MinAF tags out of 119\n",
      "vardict\n",
      "retrieving vardict 0 calls with f0.01;REJECT;REJECT tags out of 197\n",
      "varscan\n",
      "# calls before using germline calls from GATK Haplotype: 674 SNV, 121 INDEL, 0 SNP\n",
      "# calls after using germline calls from GATK Haplotype: 671 SNV, 119 INDEL, 5 SNP\n",
      "final shape SNV: (671, 30)\n",
      "final shape INDEL: (119, 30)\n",
      "final shape SNP: (5, 30)\n",
      "986_100215_M1-E\n",
      "freebayes\n",
      "mutect2\n",
      "retrieving mutect2 52 calls with MinAF tags out of 174\n",
      "strelka2\n",
      "retrieving strelka2 180 calls with MinAF tags out of 176\n",
      "vardict\n",
      "retrieving vardict 0 calls with f0.01;REJECT;REJECT tags out of 241\n",
      "varscan\n",
      "# calls before using germline calls from GATK Haplotype: 617 SNV, 104 INDEL, 0 SNP\n",
      "# calls after using germline calls from GATK Haplotype: 612 SNV, 99 INDEL, 10 SNP\n",
      "final shape SNV: (612, 30)\n",
      "final shape INDEL: (99, 30)\n",
      "final shape SNP: (10, 30)\n",
      "986_100215_M2-E\n",
      "freebayes\n",
      "mutect2\n",
      "retrieving mutect2 185 calls with MinAF tags out of 132\n",
      "strelka2\n",
      "retrieving strelka2 1472 calls with MinAF tags out of 146\n",
      "vardict\n",
      "retrieving vardict 0 calls with f0.01;REJECT;REJECT tags out of 208\n",
      "varscan\n",
      "# calls before using germline calls from GATK Haplotype: 1883 SNV, 109 INDEL, 0 SNP\n",
      "# calls after using germline calls from GATK Haplotype: 1880 SNV, 105 INDEL, 7 SNP\n",
      "final shape SNV: (1880, 30)\n",
      "final shape INDEL: (105, 30)\n",
      "final shape SNP: (7, 30)\n"
     ]
    }
   ],
   "source": [
    "T1W_WES_snv_table, _, _ = get_calltable('data/matchedtissue_ultradeep/986_100215_T1-E', config.methods_tissue, save=True, filter='all')\n",
    "M1W_WES_snv_table, _, _ = get_calltable('data/matchedtissue_ultradeep/986_100215_M1-E', config.methods_tissue, save=True, filter='all')\n",
    "M2W_WES_snv_table, _, _ = get_calltable('data/matchedtissue_ultradeep/986_100215_M2-E', config.methods_tissue, save=True, filter='all')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "documentary-illinois",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NCC_CRC-986_100215-M2W\n",
      "mutect2\n"
     ]
    }
   ],
   "source": [
    "#T1W_WGS_snv_table, _, _ = get_calltable('data/matchedtissue/NCC_CRC-986_100215-T1W', config.methods_tissue, save=True, filter='PASS')\n",
    "#M1W_WGS_snv_table, _, _ = get_calltable('data/matchedtissue/NCC_CRC-986_100215-M1W', config.methods_tissue, save=True, filter='PASS')\n",
    "M2W_WGS_snv_table, _, _ = get_calltable('data/matchedtissue/NCC_CRC-986_100215-M2W', ['mutect2', 'strelka2'], save=True, filter='PASS')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "binding-conducting",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>freebayes</th>\n",
       "      <th>freebayes_score</th>\n",
       "      <th>vardict</th>\n",
       "      <th>vardict_score</th>\n",
       "      <th>varscan</th>\n",
       "      <th>varscan_score</th>\n",
       "      <th>freebayes_altcov</th>\n",
       "      <th>freebayes_totcov</th>\n",
       "      <th>freebayes_vaf</th>\n",
       "      <th>vardict_altcov</th>\n",
       "      <th>vardict_totcov</th>\n",
       "      <th>vardict_vaf</th>\n",
       "      <th>varscan_altcov</th>\n",
       "      <th>varscan_totcov</th>\n",
       "      <th>varscan_vaf</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrom_pos_ref_alt</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10_10004273_C_A</th>\n",
       "      <td>10</td>\n",
       "      <td>10004273</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0.78177</td>\n",
       "      <td>True</td>\n",
       "      <td>0.826180</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100081798_A_G</th>\n",
       "      <td>10</td>\n",
       "      <td>100081798</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.954714</td>\n",
       "      <td>True</td>\n",
       "      <td>0.99998</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>45.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100099019_C_T</th>\n",
       "      <td>10</td>\n",
       "      <td>100099019</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.960858</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0.999039</td>\n",
       "      <td>0.0</td>\n",
       "      <td>59.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100111597_A_C</th>\n",
       "      <td>10</td>\n",
       "      <td>100111597</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>0.585060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10_100153430_C_A</th>\n",
       "      <td>10</td>\n",
       "      <td>100153430</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.895333</td>\n",
       "      <td>True</td>\n",
       "      <td>0.89262</td>\n",
       "      <td>True</td>\n",
       "      <td>0.887120</td>\n",
       "      <td>0.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y_9999388_G_T</th>\n",
       "      <td>Y</td>\n",
       "      <td>9999388</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.964881</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y_9999399_C_G</th>\n",
       "      <td>Y</td>\n",
       "      <td>9999399</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.964729</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y_9999421_G_A</th>\n",
       "      <td>Y</td>\n",
       "      <td>9999421</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.963905</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y_9999808_G_T</th>\n",
       "      <td>Y</td>\n",
       "      <td>9999808</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.908102</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Y_9999818_T_A</th>\n",
       "      <td>Y</td>\n",
       "      <td>9999818</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>True</td>\n",
       "      <td>0.900538</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>169581 rows × 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  chrom        pos ref alt type  freebayes  freebayes_score  \\\n",
       "chrom_pos_ref_alt                                                             \n",
       "10_10004273_C_A      10   10004273   C   A  SNV      False              NaN   \n",
       "10_100081798_A_G     10  100081798   A   G  SNV       True         0.954714   \n",
       "10_100099019_C_T     10  100099019   C   T  SNV       True         0.960858   \n",
       "10_100111597_A_C     10  100111597   A   C  SNV      False              NaN   \n",
       "10_100153430_C_A     10  100153430   C   A  SNV       True         0.895333   \n",
       "...                 ...        ...  ..  ..  ...        ...              ...   \n",
       "Y_9999388_G_T         Y    9999388   G   T  SNV       True         0.964881   \n",
       "Y_9999399_C_G         Y    9999399   C   G  SNV       True         0.964729   \n",
       "Y_9999421_G_A         Y    9999421   G   A  SNV       True         0.963905   \n",
       "Y_9999808_G_T         Y    9999808   G   T  SNV       True         0.908102   \n",
       "Y_9999818_T_A         Y    9999818   T   A  SNV       True         0.900538   \n",
       "\n",
       "                   vardict  vardict_score  varscan  varscan_score  \\\n",
       "chrom_pos_ref_alt                                                   \n",
       "10_10004273_C_A       True        0.78177     True       0.826180   \n",
       "10_100081798_A_G      True        0.99998    False            NaN   \n",
       "10_100099019_C_T     False            NaN     True       0.999039   \n",
       "10_100111597_A_C     False            NaN     True       0.585060   \n",
       "10_100153430_C_A      True        0.89262     True       0.887120   \n",
       "...                    ...            ...      ...            ...   \n",
       "Y_9999388_G_T        False            NaN    False            NaN   \n",
       "Y_9999399_C_G        False            NaN    False            NaN   \n",
       "Y_9999421_G_A        False            NaN    False            NaN   \n",
       "Y_9999808_G_T        False            NaN    False            NaN   \n",
       "Y_9999818_T_A        False            NaN    False            NaN   \n",
       "\n",
       "                   freebayes_altcov  freebayes_totcov  freebayes_vaf  \\\n",
       "chrom_pos_ref_alt                                                      \n",
       "10_10004273_C_A                 NaN               NaN            NaN   \n",
       "10_100081798_A_G                0.0              49.0            0.0   \n",
       "10_100099019_C_T                0.0              59.0            0.0   \n",
       "10_100111597_A_C                NaN               NaN            NaN   \n",
       "10_100153430_C_A                0.0              41.0            0.0   \n",
       "...                             ...               ...            ...   \n",
       "Y_9999388_G_T                   0.0               6.0            0.0   \n",
       "Y_9999399_C_G                   0.0               6.0            0.0   \n",
       "Y_9999421_G_A                   0.0               6.0            0.0   \n",
       "Y_9999808_G_T                   0.0               5.0            0.0   \n",
       "Y_9999818_T_A                   0.0               4.0            0.0   \n",
       "\n",
       "                   vardict_altcov  vardict_totcov  vardict_vaf  \\\n",
       "chrom_pos_ref_alt                                                \n",
       "10_10004273_C_A               0.0            47.0          0.0   \n",
       "10_100081798_A_G              0.0            45.0          0.0   \n",
       "10_100099019_C_T              NaN             NaN          NaN   \n",
       "10_100111597_A_C              NaN             NaN          NaN   \n",
       "10_100153430_C_A              0.0            35.0          0.0   \n",
       "...                           ...             ...          ...   \n",
       "Y_9999388_G_T                 NaN             NaN          NaN   \n",
       "Y_9999399_C_G                 NaN             NaN          NaN   \n",
       "Y_9999421_G_A                 NaN             NaN          NaN   \n",
       "Y_9999808_G_T                 NaN             NaN          NaN   \n",
       "Y_9999818_T_A                 NaN             NaN          NaN   \n",
       "\n",
       "                   varscan_altcov  varscan_totcov  varscan_vaf  \n",
       "chrom_pos_ref_alt                                               \n",
       "10_10004273_C_A               0.0            45.0          0.0  \n",
       "10_100081798_A_G              NaN             NaN          NaN  \n",
       "10_100099019_C_T              0.0            54.0          0.0  \n",
       "10_100111597_A_C              0.0            30.0          0.0  \n",
       "10_100153430_C_A              0.0            36.0          0.0  \n",
       "...                           ...             ...          ...  \n",
       "Y_9999388_G_T                 NaN             NaN          NaN  \n",
       "Y_9999399_C_G                 NaN             NaN          NaN  \n",
       "Y_9999421_G_A                 NaN             NaN          NaN  \n",
       "Y_9999808_G_T                 NaN             NaN          NaN  \n",
       "Y_9999818_T_A                 NaN             NaN          NaN  \n",
       "\n",
       "[169581 rows x 20 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "M2W_WGS_snv_table = pd.read_csv('data/matchedtissue/NCC_CRC-986_100215-M2W/calls/NCC_CRC-986_100215-M2W_snv_calls_PASS.csv', index_col=0)\n",
    "M2W_WGS_snv_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "premier-laptop",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "3f640068e7d64e79af3012005e4b623f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "HBox(children=(FloatProgress(value=0.0, max=169581.0), HTML(value='')))"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    }
   ],
   "source": [
    "# load exome regions\n",
    "exomebed = pd.read_csv('data/extdata/xgen-exome-research-panel-targets_nochr.bed', sep='\\t', header=None)\n",
    "exomebed.columns = ['chrom', 'startpos', 'endpos', 'a', 'b', 'c']\n",
    "exomebed = exomebed[['chrom', 'startpos', 'endpos']]\n",
    "exomebed['chrom'] = exomebed['chrom'].astype(str)\n",
    "\n",
    "# exome filtering\n",
    "noncodingregions = []\n",
    "for pos in tqdm(list(M2W_WGS_snv_table.index)):\n",
    "    chrom, startpos, ref, alt = pos.split('_')\n",
    "    if str(chrom) not in exomebed['chrom'].unique().astype(str):\n",
    "        noncodingregions.append(pos)\n",
    "    else:\n",
    "        exomechrom = exomebed[exomebed['chrom'] == str(chrom)]\n",
    "        count = False\n",
    "        for ind in list(exomechrom.index):\n",
    "            # print(exomechrom['endpos'][ind], exomechrom['startpos'][ind])\n",
    "            if exomechrom['endpos'][ind] >= int(startpos):\n",
    "                if exomechrom['startpos'][ind] <= int(startpos):\n",
    "                    count = True\n",
    "        if not count:\n",
    "            noncodingregions.append(pos)\n",
    "        #print(pos, count)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "green-statistics",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "168158 169581\n"
     ]
    }
   ],
   "source": [
    "print(len(noncodingregions), M2W_WGS_snv_table.shape[0])\n",
    "\n",
    "#T1W_exome_snv_table = T1W_WGS_snv_table.drop(noncodingregions)\n",
    "M2W_exome_snv_table = M2W_WGS_snv_table.drop(noncodingregions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "amino-broad",
   "metadata": {},
   "outputs": [],
   "source": [
    "#T1W_exome_snv_table.to_csv(os.path.join('data/matchedtissue/NCC_CRC-986_100215-T1W', 'calls', 'NCC_CRC-986_100215-T1W_snv_calls_PASS_exome.csv'))\n",
    "#M1W_exome_snv_table.to_csv(os.path.join('data/matchedtissue/NCC_CRC-986_100215-M1W', 'calls', 'NCC_CRC-986_100215-M1W_snv_calls_PASS_exome.csv'))\n",
    "M2W_exome_snv_table.to_csv(os.path.join('data/matchedtissue/NCC_CRC-986_100215-M2W', 'calls', 'NCC_CRC-986_100215-M2W_snv_calls_PASS_exome.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "thorough-series",
   "metadata": {},
   "outputs": [],
   "source": [
    "M1W_exome_snv_table = pd.read_csv(os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-M1W', 'calls', 'NCC_CRC-986_100215-M1W_snv_calls_PASS_exome.csv'), index_col=0)\n",
    "M1W_exome_snv_table = M1W_exome_snv_table[M1W_exome_snv_table['chrom'] != 'X']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "thick-batman",
   "metadata": {},
   "outputs": [],
   "source": [
    "M2W_exome_snv_table = pd.read_csv(os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-M2W', 'calls', 'NCC_CRC-986_100215-M2W_snv_calls_PASS_exome.csv'), index_col=0)\n",
    "M2W_exome_snv_table = M2W_exome_snv_table[M2W_exome_snv_table['chrom'] != 'X']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incredible-notion",
   "metadata": {},
   "outputs": [],
   "source": [
    "T1W_exome_snv_table = pd.read_csv(os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-T1W', 'calls', 'NCC_CRC-986_100215-T1W_snv_calls_PASS_exome.csv'), index_col=0)\n",
    "T1W_exome_snv_table = T1W_exome_snv_table[T1W_exome_snv_table['chrom'] != 'X']\n",
    "T1W_exome_snv_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "average-serum",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>type</th>\n",
       "      <th>freebayes</th>\n",
       "      <th>freebayes_score</th>\n",
       "      <th>mutect2</th>\n",
       "      <th>mutect2_score</th>\n",
       "      <th>strelka2</th>\n",
       "      <th>...</th>\n",
       "      <th>vardict_vaf</th>\n",
       "      <th>varscan_altcov</th>\n",
       "      <th>varscan_totcov</th>\n",
       "      <th>varscan_vaf</th>\n",
       "      <th>abemus_altcov</th>\n",
       "      <th>abemus_totcov</th>\n",
       "      <th>abemus_vaf</th>\n",
       "      <th>sinvict_altcov</th>\n",
       "      <th>sinvict_totcov</th>\n",
       "      <th>sinvict_vaf</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>chrom_pos_ref_alt</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_100436166_C_A</th>\n",
       "      <td>1</td>\n",
       "      <td>100436166</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_100684255_T_A</th>\n",
       "      <td>1</td>\n",
       "      <td>100684255</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1007218_G_A</th>\n",
       "      <td>1</td>\n",
       "      <td>1007218</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>0.021978</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_1007226_A_G</th>\n",
       "      <td>1</td>\n",
       "      <td>1007226</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.054795</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_10166570_G_T</th>\n",
       "      <td>1</td>\n",
       "      <td>10166570</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>76.0</td>\n",
       "      <td>0.026316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51159340_A_C</th>\n",
       "      <td>22</td>\n",
       "      <td>51159340</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>0.038462</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51159637_C_T</th>\n",
       "      <td>22</td>\n",
       "      <td>51159637</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>0.020000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51159801_C_T</th>\n",
       "      <td>22</td>\n",
       "      <td>51159801</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>0.032787</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51159957_G_A</th>\n",
       "      <td>22</td>\n",
       "      <td>51159957</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>0.025316</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22_51160025_G_T</th>\n",
       "      <td>22</td>\n",
       "      <td>51160025</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>SNV</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.021277</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27151 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                   chrom        pos ref alt type  freebayes  freebayes_score  \\\n",
       "chrom_pos_ref_alt                                                              \n",
       "1_100436166_C_A        1  100436166   C   A  SNV      False              NaN   \n",
       "1_100684255_T_A        1  100684255   T   A  SNV      False              NaN   \n",
       "1_1007218_G_A          1    1007218   G   A  SNV      False              NaN   \n",
       "1_1007226_A_G          1    1007226   A   G  SNV      False              NaN   \n",
       "1_10166570_G_T         1   10166570   G   T  SNV      False              NaN   \n",
       "...                  ...        ...  ..  ..  ...        ...              ...   \n",
       "22_51159340_A_C       22   51159340   A   C  SNV      False              NaN   \n",
       "22_51159637_C_T       22   51159637   C   T  SNV      False              NaN   \n",
       "22_51159801_C_T       22   51159801   C   T  SNV      False              NaN   \n",
       "22_51159957_G_A       22   51159957   G   A  SNV      False              NaN   \n",
       "22_51160025_G_T       22   51160025   G   T  SNV      False              NaN   \n",
       "\n",
       "                   mutect2  mutect2_score  strelka2  ...  vardict_vaf  \\\n",
       "chrom_pos_ref_alt                                    ...                \n",
       "1_100436166_C_A      False            NaN     False  ...          0.0   \n",
       "1_100684255_T_A      False            NaN     False  ...          0.0   \n",
       "1_1007218_G_A        False            NaN     False  ...          NaN   \n",
       "1_1007226_A_G        False            NaN     False  ...          NaN   \n",
       "1_10166570_G_T       False            NaN     False  ...          NaN   \n",
       "...                    ...            ...       ...  ...          ...   \n",
       "22_51159340_A_C      False            NaN     False  ...          NaN   \n",
       "22_51159637_C_T      False            NaN     False  ...          NaN   \n",
       "22_51159801_C_T      False            NaN     False  ...          NaN   \n",
       "22_51159957_G_A      False            NaN     False  ...          NaN   \n",
       "22_51160025_G_T      False            NaN     False  ...          NaN   \n",
       "\n",
       "                   varscan_altcov  varscan_totcov  varscan_vaf  abemus_altcov  \\\n",
       "chrom_pos_ref_alt                                                               \n",
       "1_100436166_C_A               NaN             NaN          NaN            NaN   \n",
       "1_100684255_T_A               NaN             NaN          NaN            NaN   \n",
       "1_1007218_G_A                 NaN             NaN          NaN            2.0   \n",
       "1_1007226_A_G                 NaN             NaN          NaN            4.0   \n",
       "1_10166570_G_T                NaN             NaN          NaN            2.0   \n",
       "...                           ...             ...          ...            ...   \n",
       "22_51159340_A_C               NaN             NaN          NaN            2.0   \n",
       "22_51159637_C_T               NaN             NaN          NaN            2.0   \n",
       "22_51159801_C_T               NaN             NaN          NaN            2.0   \n",
       "22_51159957_G_A               NaN             NaN          NaN            2.0   \n",
       "22_51160025_G_T               NaN             NaN          NaN            2.0   \n",
       "\n",
       "                   abemus_totcov  abemus_vaf  sinvict_altcov  sinvict_totcov  \\\n",
       "chrom_pos_ref_alt                                                              \n",
       "1_100436166_C_A              NaN         NaN             NaN             NaN   \n",
       "1_100684255_T_A              NaN         NaN             NaN             NaN   \n",
       "1_1007218_G_A               91.0    0.021978             NaN             NaN   \n",
       "1_1007226_A_G               73.0    0.054795             NaN             NaN   \n",
       "1_10166570_G_T              76.0    0.026316             NaN             NaN   \n",
       "...                          ...         ...             ...             ...   \n",
       "22_51159340_A_C             52.0    0.038462             NaN             NaN   \n",
       "22_51159637_C_T            100.0    0.020000             NaN             NaN   \n",
       "22_51159801_C_T             61.0    0.032787             NaN             NaN   \n",
       "22_51159957_G_A             79.0    0.025316             NaN             NaN   \n",
       "22_51160025_G_T             94.0    0.021277             NaN             NaN   \n",
       "\n",
       "                   sinvict_vaf  \n",
       "chrom_pos_ref_alt               \n",
       "1_100436166_C_A            NaN  \n",
       "1_100684255_T_A            NaN  \n",
       "1_1007218_G_A              NaN  \n",
       "1_1007226_A_G              NaN  \n",
       "1_10166570_G_T             NaN  \n",
       "...                        ...  \n",
       "22_51159340_A_C            NaN  \n",
       "22_51159637_C_T            NaN  \n",
       "22_51159801_C_T            NaN  \n",
       "22_51159957_G_A            NaN  \n",
       "22_51160025_G_T            NaN  \n",
       "\n",
       "[27151 rows x 40 columns]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cfdna1_snv_table_list = []\n",
    "for i in range(1,23):\n",
    "    cfdna1_chri_snv_table = pd.read_csv(os.path.join('data', 'mixtures', 'mixtures_chr'+str(i), 'mixtures_chr'+str(i)+'_CRC-986_100215-CW-T_CRC-986_300316-CW-T',\n",
    "                                                'mixture_chr'+str(i)+'_CRC-986_100215-CW-T_70x_CRC-986_300316-CW-T_0x', 'calls',\n",
    "                                                'mixture_chr'+str(i)+'_CRC-986_100215-CW-T_70x_CRC-986_300316-CW-T_0x_snv_calls_all.csv'), index_col=0)\n",
    "    cfdna1_snv_table_list.append(cfdna1_chri_snv_table)\n",
    "cfdna1_snv_table = pd.concat(cfdna1_snv_table_list)\n",
    "cfdna1_snv_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "flush-azerbaijan",
   "metadata": {},
   "outputs": [],
   "source": [
    "calltabledict = {\n",
    "    'cfdna1': cfdna1_snv_table,\n",
    "    'M2W': M2W_exome_snv_table,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "official-repeat",
   "metadata": {},
   "outputs": [],
   "source": [
    "calltabledict = {\n",
    "    'cfdna1': cfdna1_snv_table,\n",
    "    'T1W': T1W_exome_snv_table,\n",
    "    'M1W': M1W_exome_snv_table,\n",
    "    'M2W': M2W_exome_snv_table,\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "palestinian-shareware",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "cfdna1\n",
      "['abemus' 'freebayes' 'mutect2' 'sinvict' 'strelka2' 'vardict' 'varscan']\n",
      "27151\n",
      "1 27151 27151\n",
      "2 1533 1533\n",
      "3 375 375\n",
      "4 159 159\n",
      "5 92 92\n",
      "6 40 40\n",
      "7 15 15\n",
      "M2W\n",
      "['freebayes' 'vardict' 'varscan']\n",
      "1394\n",
      "1 1394 229\n",
      "2 318 85\n",
      "3 71 21\n",
      "2\n",
      "cfdna1\n",
      "['abemus' 'freebayes' 'mutect2' 'sinvict' 'strelka2' 'vardict' 'varscan']\n",
      "27151\n",
      "1 27151 1533\n",
      "2 1533 1533\n",
      "3 375 375\n",
      "4 159 159\n",
      "5 92 92\n",
      "6 40 40\n",
      "7 15 15\n",
      "M2W\n",
      "['freebayes' 'vardict' 'varscan']\n",
      "1394\n",
      "1 1394 105\n",
      "2 318 63\n",
      "3 71 19\n",
      "3\n",
      "cfdna1\n",
      "['abemus' 'freebayes' 'mutect2' 'sinvict' 'strelka2' 'vardict' 'varscan']\n",
      "27151\n",
      "1 27151 375\n",
      "2 1533 375\n",
      "3 375 375\n",
      "4 159 159\n",
      "5 92 92\n",
      "6 40 40\n",
      "7 15 15\n",
      "M2W\n",
      "['freebayes' 'vardict' 'varscan']\n",
      "1394\n",
      "1 1394 91\n",
      "2 318 61\n",
      "3 71 17\n",
      "4\n",
      "cfdna1\n",
      "['abemus' 'freebayes' 'mutect2' 'sinvict' 'strelka2' 'vardict' 'varscan']\n",
      "27151\n",
      "1 27151 159\n",
      "2 1533 159\n",
      "3 375 159\n",
      "4 159 159\n",
      "5 92 92\n",
      "6 40 40\n",
      "7 15 15\n",
      "M2W\n",
      "['freebayes' 'vardict' 'varscan']\n",
      "1394\n",
      "1 1394 82\n",
      "2 318 56\n",
      "3 71 15\n",
      "5\n",
      "cfdna1\n",
      "['abemus' 'freebayes' 'mutect2' 'sinvict' 'strelka2' 'vardict' 'varscan']\n",
      "27151\n",
      "1 27151 92\n",
      "2 1533 92\n",
      "3 375 92\n",
      "4 159 92\n",
      "5 92 92\n",
      "6 40 40\n",
      "7 15 15\n",
      "M2W\n",
      "['freebayes' 'vardict' 'varscan']\n",
      "1394\n",
      "1 1394 69\n",
      "2 318 44\n",
      "3 71 12\n",
      "6\n",
      "cfdna1\n",
      "['abemus' 'freebayes' 'mutect2' 'sinvict' 'strelka2' 'vardict' 'varscan']\n",
      "27151\n",
      "1 27151 40\n",
      "2 1533 40\n",
      "3 375 40\n",
      "4 159 40\n",
      "5 92 40\n",
      "6 40 40\n",
      "7 15 15\n",
      "M2W\n",
      "['freebayes' 'vardict' 'varscan']\n",
      "1394\n",
      "1 1394 36\n",
      "2 318 19\n",
      "3 71 5\n",
      "7\n",
      "cfdna1\n",
      "['abemus' 'freebayes' 'mutect2' 'sinvict' 'strelka2' 'vardict' 'varscan']\n",
      "27151\n",
      "1 27151 15\n",
      "2 1533 15\n",
      "3 375 15\n",
      "4 159 15\n",
      "5 92 15\n",
      "6 40 15\n",
      "7 15 15\n",
      "M2W\n",
      "['freebayes' 'vardict' 'varscan']\n",
      "1394\n",
      "1 1394 15\n",
      "2 318 5\n",
      "3 71 0\n"
     ]
    }
   ],
   "source": [
    "res =  compare_groundtruth(calltabledict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "falling-warner",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "318\n",
      "{'01': '262', '10': '103', '11': '56'}\n"
     ]
    }
   ],
   "source": [
    "calls_cfdna1 = res['cfdna1_4_4']\n",
    "calls_M2W = res['M2W_1_2']\n",
    "print(len(calls_M2W))\n",
    "labels = get_labels([calls_cfdna1, calls_M2W])\n",
    "print(labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "dynamic-article",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(calls_M2W).tofile('data/matchedtissue/venndiagram/calls_M2W_atleast2callers.csv', sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "nominated-orientation",
   "metadata": {},
   "outputs": [],
   "source": [
    "calls_cfdna1 = res['cfdna1_4_4']\n",
    "print(len(calls_cfdna1))\n",
    "calls_T1W = res['T1W_1_3']\n",
    "print(len(calls_T1W))\n",
    "calls_M1W = res['M1W_1_3']\n",
    "print(len(calls_M1W))\n",
    "labels = get_labels([calls_cfdna1, calls_T1W, calls_M1W])\n",
    "print(labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "color-principal",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(calls_cfdna1).tofile('data/matchedtissue/venndiagram/calls_cfdna1_atleast4callers.csv', sep = ',')\n",
    "np.array(calls_T1W).tofile('data/matchedtissue/venndiagram/calls_T1W_atleast3callers.csv', sep = ',')\n",
    "np.array(calls_M1W).tofile('data/matchedtissue/venndiagram/calls_M1W_atleast3callers.csv', sep = ',')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ordinary-defendant",
   "metadata": {},
   "outputs": [],
   "source": [
    "calls_cfdna1 = [str(i)[1:-1] for i in list(pd.read_csv(\"data/matchedtissue/venndiagram/calls_cfdna1_atleast4callers.csv\", header=None).values)[0]]\n",
    "calls_T1W =  [str(i)[1:-1] for i in list(pd.read_csv(\"data/matchedtissue/venndiagram/calls_T1W_atleast3callers.csv\", header=None).values)[0]]\n",
    "calls_M1W = [str(i)[1:-1] for i in list(pd.read_csv(\"data/matchedtissue/venndiagram/calls_M1W_atleast3callers.csv\", header=None).values)[0]]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "liquid-finnish",
   "metadata": {},
   "source": [
    "# Table to compare call sets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "specific-caribbean",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cfdna</th>\n",
       "      <th>T1W</th>\n",
       "      <th>M1W</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1_10687378_T_G</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_11898693_G_A</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_147380601_C_T</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_173878724_G_T</th>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1_183086809_G_A</th>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_114393727_G_A</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_13221458_A_G</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_140009153_G_A</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_43915893_G_C</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9_90749760_A_G</th>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>309 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 cfdna    T1W    M1W\n",
       "1_10687378_T_G    True  False  False\n",
       "1_11898693_G_A    True  False  False\n",
       "1_147380601_C_T   True   True   True\n",
       "1_173878724_G_T   True   True   True\n",
       "1_183086809_G_A   True  False  False\n",
       "...                ...    ...    ...\n",
       "9_114393727_G_A  False  False   True\n",
       "9_13221458_A_G   False  False   True\n",
       "9_140009153_G_A  False  False   True\n",
       "9_43915893_G_C   False  False   True\n",
       "9_90749760_A_G   False  False   True\n",
       "\n",
       "[309 rows x 3 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cfdna_df = pd.DataFrame(index=calls_cfdna1)\n",
    "cfdna_df['cfdna'] = True\n",
    "T1W_df = pd.DataFrame(index=calls_T1W)\n",
    "T1W_df['T1W'] = True\n",
    "M1W_df = pd.DataFrame(index=calls_M1W)\n",
    "M1W_df['M1W'] = True\n",
    "res = pd.concat([cfdna_df, T1W_df, M1W_df], axis=1)\n",
    "res.fillna(False, inplace=True)\n",
    "res"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "received-democrat",
   "metadata": {},
   "source": [
    "# Jaccard similarity score of call set pairs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "aerial-species",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.29,0.30,0.35\n",
      "0.29,0.30,0.35\n"
     ]
    }
   ],
   "source": [
    "from sklearn.metrics import jaccard_score\n",
    "print(\"{:.2f},{:.2f},{:.2f}\".format(jaccard_score(res['cfdna'], res['T1W']), jaccard_score(res['cfdna'], res['M1W']), jaccard_score(res['T1W'], res['M1W'])))\n",
    "print(\"{:.2f},{:.2f},{:.2f}\".format(jaccard_score(res['T1W'], res['cfdna']), jaccard_score(res['M1W'], res['cfdna']), jaccard_score(res['M1W'], res['T1W'])))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "physical-restoration",
   "metadata": {},
   "source": [
    "# Fisher exact test on call sets pairs\n",
    "\n",
    "## NB: would need to evaluate number of base pairs with sufficient coverage evaluated. Here, took N = 20K (approx number of protein-coding genes with decent coverage)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "supreme-exhaust",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "############\n",
      "cfdna VS T1W\n",
      "############\n",
      "Contingency table:\n",
      "               cfdna & T1W  !cfdna & T1W\n",
      "cfdna & !T1W            68            73\n",
      "!cfdna & !T1W           91         19768\n",
      "Fisher exact test (two-sided): odd_ratio = 202.35194942044257, p_value = 1.0190037260737785e-109\n",
      "############\n",
      "cfdna VS M1W\n",
      "############\n",
      "Contingency table:\n",
      "               cfdna & M1W  !cfdna & M1W\n",
      "cfdna & !M1W            75            95\n",
      "!cfdna & !M1W           84         19746\n",
      "Fisher exact test (two-sided): odd_ratio = 185.58270676691728, p_value = 5.148787917418059e-118\n",
      "############\n",
      "T1W VS M1W\n",
      "############\n",
      "Contingency table:\n",
      "             T1W & M1W  !T1W & M1W\n",
      "T1W & !M1W          81          89\n",
      "!T1W & !M1W         60       19770\n",
      "Fisher exact test (two-sided): odd_ratio = 299.8820224719101, p_value = 6.450909216876732e-138\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>p_value_adjusted</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>cfdna VS T1W</th>\n",
       "      <td>3.057011e-109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cfdna VS M1W</th>\n",
       "      <td>1.544636e-117</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>T1W VS M1W</th>\n",
       "      <td>1.935273e-137</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              p_value_adjusted\n",
       "cfdna VS T1W     3.057011e-109\n",
       "cfdna VS M1W     1.544636e-117\n",
       "T1W VS M1W       1.935273e-137"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "# contingency table and Fisher exact test\n",
    "\n",
    "testres = {}\n",
    "for pairs in [('cfdna', 'T1W'), ('cfdna', 'M1W'), ('T1W', 'M1W')]:\n",
    "    a, b =  pairs\n",
    "    print(\"############\")\n",
    "    print(a, 'VS', b)\n",
    "    print(\"############\")\n",
    "    ctable = [[res[(res[a]==True) & (res[b]==True)].shape[0], res[(res[a]==False) & (res[b]==True)].shape[0]],\n",
    "              [res[(res[a]==True) & (res[b]==False)].shape[0], int(20000 - res[(res[a]==True) | (res[b]==True)].shape[0])]]\n",
    "    \n",
    "    #p_adjusted = multipletests(Column6, method='bonferroni')\n",
    "    odd_ratio, p_value = fisher_exact(ctable, alternative='two-sided')\n",
    "    print('Contingency table:')\n",
    "    print(pd.DataFrame(ctable, columns=[a+' & '+b, '!'+a+' & '+b], index=[a+' & !'+b, '!'+a+' & !'+b]))\n",
    "    print('Fisher exact test (two-sided): odd_ratio = {}, p_value = {}'.format(odd_ratio, p_value))\n",
    "    testres[a+' VS '+b] = [p_value]\n",
    "\n",
    "testres = pd.DataFrame.from_dict(testres).T\n",
    "testres.columns = ['p_value']\n",
    "testres['p_value_adjusted'] = multipletests(testres['p_value'].values, method='bonferroni')[1]\n",
    "testres[['p_value_adjusted']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "continuing-morrison",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
