{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Series analysis\n",
    "\n",
    "\n",
    "Analysis updates:\n",
    "- retrieve calls rejected only because VAF<0.01 (bcbio lower acceptable threshold) for Mutect2, Strelka2 and Vardict\n",
    "- correct for germline mutations using GATK haplotype calls\n",
    "- get PR curves for sinvict by encoding the 6 files as thresholds. Assumption: linear filters\n",
    "- plotting: PR curves stops at 10e-2 on the left\n",
    "- ground truths:\n",
    "\n",
    "    1) Consensus: build using\n",
    "        majority of 5/8 callers for SNV and 3/5 callers for INDELS\n",
    "\n",
    "\n",
    "    2) Ranked mutations: metascore built using\n",
    "        weigthed sum of normalised scores between 0 and 1 for each caller\n",
    "        with weigths = inversially proportional to number of calls made by caller (if a caller calls few mutations higher weight, it many calls low weight)\n",
    "        threshold = 1/ncallers\n",
    "        interpretation: if 1 caller is sure (score = 1) of calling this position, add it to GT\n",
    "        interpretation: if 2 callers are quite sure of calling this position (score > 0,5 each), add it to GT\n",
    "        \n",
    "- integrate VAF approx\n",
    "\n",
    "    1) mixture with VAF instead of tumor burden\n",
    "    \n",
    "    2) correct for mutations non present in diluted samples (vaf = 0)\n",
    "    \n",
    "    3) pool patients together using VAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "import warnings\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, average_precision_score\n",
    "warnings.filterwarnings('ignore')\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "# set working directory\n",
    "if not os.getcwd().endswith('cfdna_snv_benchmark'):\n",
    "    os.chdir('../')\n",
    "print('Current working directory: {}'.format(os.getcwd()))\n",
    "\n",
    "from utils.config import Config\n",
    "from utils.viz import *\n",
    "from utils.table import *\n",
    "from utils.metrics import *\n",
    "from utils.calltable import *\n",
    "from utils.calltableseries import *\n",
    "from utils.groundtruth import *\n",
    "from utils.metricsseries import *\n",
    "from utils.venn import venn6, get_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Config and Display paramaters\n",
    "\n",
    "config = Config(\"config/\", \"config_viz.yaml\")\n",
    "set_display_params(config)\n",
    "print(config.methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chomosome\n",
    "\n",
    "mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "reload = False\n",
    "save = False\n",
    "fixedvars=['coverage', 'ctdna']\n",
    "filterparam = 'all'\n",
    "\n",
    "markers = ['o', '^', 'X']\n",
    "linestyles = ['-', '-', '-']\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "\n",
    "muttypes = ['snv', 'indel']\n",
    "metrics = ['auprc', 'precision', 'recall']\n",
    "\n",
    "chrom = 'all'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part I: (1) Load/Generate call tables, (2) Generate Ground truths and (3) Compute/Save metrics per patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixedvar = 'coverage'\n",
    "#for fixedvar in fixedvars:\n",
    "if fixedvar == 'coverage':\n",
    "    seriesorder = [(1000, 0), (1000, 1000), (750, 1250), (500, 1500), (400, 1600), (200, 1800), (100, 1900)] \n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    seriesorder = [(1000, 0), (1000, 1000), (1000, 1900)]\n",
    "    xaxis = 'coverage'\n",
    "#for mixtureid in mixtureids:\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, 'all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='mixture_wes', concat='tf', bcbiovaf=0.01)\n",
    "#calltable_indel, aux = get_calltableseries(config, mixtureid, 'all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='mixture_wes', bcbiovaf=0.01)\n",
    "#calltable_snp, aux = get_calltableseries(config, mixtureid, 'all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='mixture_wes', bcbiovaf=0.01)\n",
    "#print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltable_snv = calltable_snv[calltable_snv['chrom'] != 'X']\n",
    "calltable_snv = calltable_snv[calltable_snv['chrom'] != 'Y']\n",
    "calltable_snv = calltable_snv[calltable_snv['chrom'].isin([str(i) for i in range(1, 23)])]\n",
    "#calltable_indel = calltable_indel[calltable_indel['chrom'] != 'X']\n",
    "#calltable_indel = calltable_indel[calltable_indel['chrom'] != 'Y']\n",
    "#calltable_indel = calltable_indel[calltable_indel['chrom'].isin([str(i) for i in range(1, 23)])]\n",
    "#calltable_snp = calltable_snp[calltable_snp['chrom'] != 'X']\n",
    "#calltable_snp = calltable_snp[calltable_snp['chrom'] != 'Y']\n",
    "#calltable_snp = calltable_snp[calltable_snp['chrom'].isin([str(i) for i in range(1, 23)])]\n",
    "calltable_snv['chrom'] = calltable_snv['chrom'].astype(int)\n",
    "#calltable_indel['chrom'] = calltable_indel['chrom'].astype(int)\n",
    "#calltable_snp['chrom'] = calltable_snp['chrom'].astype(int)\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(plasmasample)\n",
    "healthysample = '_'.join(mixtureid.split('_')[2:])\n",
    "print(healthysample)\n",
    "#calltables['snv'].append(calltable_snv)\n",
    "#calltables['indel'].append(calltable_indel)\n",
    "#calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "#calltables['snv'] = pd.concat(calltables['snv'])\n",
    "#calltables['indel'] = pd.concat(calltables['indel'])\n",
    "#calltables['snp'] = pd.concat(calltables['snp'])\n",
    "calltables['snv'] = calltable_snv.copy() ## as only one mixtureid\n",
    "dilutionseries = aux.T[['mixture_chrall_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "#for muttype in muttypes:\n",
    "muttype = 'snv'\n",
    "if muttype == 'snv':\n",
    "    gtm = 5\n",
    "else:  # elif muttype == 'indel':\n",
    "    gtm = 3\n",
    "print(max(aux['tf']))\n",
    "refsample = 'undiluted'\n",
    "#if mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "#    gtm = 3\n",
    "#    refsample = 'tissue'\n",
    "    #calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "    #                                    matchedtissuepath=os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-T1W', 'calls', 'NCC_CRC-986_100215-T1W_snv_calls_PASS_exome.csv'))\n",
    "#    calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "#                                        matchedtissuepath=os.path.join('data', 'matchedtissue_ultradeep', '986_100215_T1-E', 'calls', '986_100215_T1-E_snv_calls_all.csv'))#\n",
    "#else:\n",
    "calltablesseries = generate_groundtruth(config, calltable_snv, aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "#calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "#                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "print(calltablesseries)\n",
    "#results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "#                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall', ground_truth_method=gtm,\n",
    "#                                       refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision', ground_truth_method=gtm,\n",
    "#                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method=gtm,\n",
    "#                    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save, diltype='mixture_wes')\n",
    "results_maxrecallatlast0_05precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_05precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "results_maxrecallatlast0_10precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_10precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_snv_calls_all.csv'), index_col=0)\n",
    "calltablesseries = generate_groundtruth(config, calltablesseries, aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "aux = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_tf_cov.csv'), index_col=0)\n",
    "print(calltablesseries)\n",
    "save=True\n",
    "#seriesorder = [(1000, 0), (1000, 1000), (1000, 1900)]\n",
    "#fixedvar = 'ctdna'\n",
    "#xaxis = 'coverage'\n",
    "seriesorder = [(1000, 0), (1000, 1000), (750, 1250), (500, 1500), (400, 1600), (200, 1800), (100, 1900)] \n",
    "fixedvar = 'coverage'\n",
    "xaxis = 'tumor burden'\n",
    "dilutionseries = aux.T[['mixture_chrall_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "#results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "#                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall', ground_truth_method=gtm,\n",
    "#                                       refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision', ground_truth_method=gtm,\n",
    "#                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method=gtm,\n",
    "#                    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save, diltype='mixture_wes')\n",
    "#results_maxrecallatlast0_05precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_01precision', ground_truth_method=gtm,\n",
    "#                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "results_maxrecallatlast0_10precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_025precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calltablesseries = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_snv_calls_all.csv'), index_col=0)\n",
    "aux = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_tf_cov.csv'),index_col=0)\n",
    "gtm = 5\n",
    "muttype = 'snv'\n",
    "refsample = 'undiluted'\n",
    "fixedvar = 'coverage'\n",
    "xaxis = 'tumor burden'\n",
    "#calltablesseries = generate_groundtruth(config, calltablesseries, aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "#                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, aux, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall', ground_truth_method=gtm,\n",
    "#                                       refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision', ground_truth_method=gtm,\n",
    "#                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wes')\n",
    "#figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method=gtm,\n",
    "#                    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save, diltype='mixture_wes')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdf = calltablesseries[calltablesseries['truth'] == True]\n",
    "gtdf['median_vaf'] = gtdf[['{:.2f}_{}_vaf'.format(aux['tf'].max(), m) for m in config.methods if '{:.2f}_{}_vaf'.format(aux['tf'].max(), m) in gtdf.columns]].median(axis=1, skipna=True)\n",
    "sns.histplot(data=gtdf, x=\"median_vaf\", bins=30, stat='frequency')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdf['median_vaf'].to_csv('figures/figure2b/gt_986_exome_2000x_atleast5callersinundilutedsample_snv.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Awithoutab = ['12_40882016_G_T',\n",
    " '18_9122611_T_C',\n",
    " '19_9015323_C_A',\n",
    " '11_1085845_C_T',\n",
    " '15_23104346_C_G',\n",
    " '6_32712967_C_G',\n",
    " '7_100681363_C_A',\n",
    " '19_9018508_T_G',\n",
    " '7_152522195_C_T',\n",
    " '18_47364009_G_A',\n",
    " '19_9015324_A_G',\n",
    " '11_124253181_T_C',\n",
    " '4_103832610_C_T',\n",
    " '19_9012501_A_G',\n",
    " '11_71847122_C_T',\n",
    " '7_72412591_G_A',\n",
    " '11_46369127_G_A',\n",
    " '16_71163579_C_T',\n",
    " '4_190883030_G_T',\n",
    " '12_52431751_C_G',\n",
    " '7_112092224_G_T',\n",
    " '19_6222411_T_G',\n",
    " '12_11286164_G_A',\n",
    " '15_22471481_G_T',\n",
    " '17_21114408_T_C',\n",
    " '12_11244356_A_T',\n",
    " '18_47363917_A_G',\n",
    " '14_105939672_C_G',\n",
    " '19_41354606_G_T',\n",
    " '19_41354661_T_C',\n",
    " '18_13761856_G_C',\n",
    " '4_103832597_A_G',\n",
    " '12_11286149_G_A',\n",
    " '4_103832665_T_C',\n",
    " '11_1017806_G_T',\n",
    " '14_105410352_T_C',\n",
    " '12_11244378_C_G',\n",
    " '19_9018476_G_A',\n",
    " '17_39383069_A_G',\n",
    " '7_157959929_G_A',\n",
    " '4_103832643_A_C',\n",
    " '19_14877799_G_C',\n",
    " '15_22471522_C_T',\n",
    " '11_67759316_C_T',\n",
    " '15_50785016_A_G',\n",
    " '5_69372353_T_C',\n",
    " '17_5036748_G_T',\n",
    " '3_195506271_A_T',\n",
    " '19_9021141_G_A',\n",
    " '12_94543436_G_A',\n",
    " '3_195509006_C_G',\n",
    " '18_9122591_T_C',\n",
    " '12_11546742_T_C',\n",
    " '15_22471542_G_A',\n",
    " '15_23686113_C_G',\n",
    " '12_11546686_C_T',\n",
    " '16_88788786_G_A',\n",
    " '13_45968389_G_C',\n",
    " '6_31323233_A_G',\n",
    " '19_55795866_A_C',\n",
    " '12_52699033_G_A',\n",
    " '14_105173873_A_C',\n",
    " '18_31325884_T_C',\n",
    " '5_256514_G_C',\n",
    " '6_32712961_G_A',\n",
    " '16_88909572_G_A',\n",
    " '19_53116939_C_G',\n",
    " '19_1110748_G_C',\n",
    " '12_40880634_T_G',\n",
    " '17_5036740_C_G',\n",
    " '4_103832611_G_A',\n",
    " '7_143453661_T_G',\n",
    " '19_9012515_G_A',\n",
    " '11_1085849_A_C',\n",
    " '3_125297866_G_A',\n",
    " '4_103832656_C_T',\n",
    " '7_151896464_A_G',\n",
    " '19_9008248_G_C',\n",
    " '4_103832620_A_T',\n",
    " '7_151932923_G_A',\n",
    " '18_74208553_C_G',\n",
    " '19_1881355_T_C',\n",
    " '14_105173879_A_C',\n",
    " '17_63200307_G_A',\n",
    " '19_9008264_C_T',\n",
    " '9_117800_G_T',\n",
    " '15_32738694_G_A',\n",
    " '10_50534954_T_C',\n",
    " '19_9009652_C_T',\n",
    " '11_4608262_A_C',\n",
    " '17_74676686_C_A',\n",
    " '15_79057989_C_T',\n",
    " '11_89531533_G_T',\n",
    " '5_145613200_T_C',\n",
    " '15_22471573_C_T',\n",
    " '11_47647265_A_G',\n",
    " '3_195505897_G_A',\n",
    " '12_11546749_T_C',\n",
    " '12_40877466_T_C',\n",
    " '9_151046_A_G',\n",
    " '19_9018479_G_T',\n",
    " '12_11244390_T_C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = pd.read_csv('figures/figure2b/gt_986_exome_150x_atleast5callersinundilutedsample_snv.csv', index_col=0)\n",
    "B = pd.read_csv('figures/figure2b/gt_986_exome_2000x_atleast5callersinundilutedsample_snv.csv', index_col=0)\n",
    "print(A.shape[0], B.shape[0])\n",
    "ab = list(set(set(list(A.index)) & set(list(B.index))))\n",
    "len(ab)\n",
    "Awithoutab = A.loc[list(set(A.index) - set(ab))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(calltablesseries.loc[[a for a in list(Awithoutab.index) if a in calltablesseries.index]][['{:.2f}_{}'.format(aux['tf'].max(), m) for m in config.methods if ('{:.2f}_{}'.format(aux['tf'].max(), m) in calltablesseries.columns) and (m != 'smurf')]].sum(axis=1).value_counts())\n",
    "\n",
    "plt.figure(figsize=(20,5))\n",
    "ac = calltablesseries.loc[[a for a in list(Awithoutab.index) if a in calltablesseries.index]][['{:.2f}_{}_vaf'.format(aux['tf'].max(), m) for m in config.methods if ('{:.2f}_{}'.format(aux['tf'].max(), m) in calltablesseries.columns) and (m != 'smurf')]].median(skipna=True, axis=1)\n",
    "print(ac.shape)\n",
    "sns.histplot(x=ac, binwidth=0.01, binrange=[0,1], alpha=0.5, label='150x only looked up in 2000x')\n",
    "plt.legend()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "comp = pd.concat([Awithoutab, ac], axis=1)\n",
    "comp.columns = ['vaf 150x', 'vaf 2000x']\n",
    "plt.figure(figsize=(10,10))\n",
    "ac = calltablesseries.loc[[a for a in list(Awithoutab.index) if a in calltablesseries.index]][['{:.2f}_{}_vaf'.format(aux['tf'].max(), m) for m in config.methods if ('{:.2f}_{}'.format(aux['tf'].max(), m) in calltablesseries.columns) and (m != 'smurf')]].median(skipna=True, axis=1)\n",
    "print(ac.shape)\n",
    "sns.histplot(x='vaf 150x', y='vaf 2000x', data=comp, binwidth=0.01, binrange=[0,0.5], alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_snv_calls_all.csv'), index_col=0)\n",
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltables[muttype].shape, test.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test = generate_groundtruth(config,test, aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "test[(test['chrom'] != 'X') & (test['chrom'] != 'Y') & (test[['{:.2f}_{}'.format(max(aux['tf']), m) for m in config.methods]].sum(axis=1) >= 5)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries['41.90_freebayes'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux['tf'].max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries[['{:.2f}_{}'.format(max(aux['tf']), m) for m in config.methods]]#.sum(axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "max(aux['tf'])\n",
    "calltables[muttype].columns\n",
    "dilutionseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux_dict = {'tf': [41.9000, 20.9496, 15.7102, 10.4725, 8.3783, 4.18723, 2.0935], 'vaf': [np.nan] * 7, 'cov': [1000.0] + 6*[2000.0], 'ichorcna': [np.nan] * 7, 'samplename': [np.nan] * 7}\n",
    "print(aux_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux_dict = {'tf': [41.9000, 20.9496, 15.7102, 10.4725, 8.3783, 4.18723, 2.0935], 'vaf': [np.nan] * 7, 'cov': [1000.0] + 6*[2000.0], 'ichorcna': [np.nan] * 7, 'samplename': [np.nan] * 7, 'sampleid':\n",
    "            ['mixture_chrall_CRC-986_100215-CW-T_1000x_CRC-986_300316-CW-T_0x', 'mixture_chrall_CRC-986_100215-CW-T_1000x_CRC-986_300316-CW-T_1000x', 'mixture_chrall_CRC-986_100215-CW-T_750x_CRC-986_300316-CW-T_1250x',\n",
    "    'mixture_chrall_CRC-986_100215-CW-T_500x_CRC-986_300316-CW-T_1500x', 'mixture_chrall_CRC-986_100215-CW-T_400x_CRC-986_300316-CW-T_1600x', 'mixture_chrall_CRC-986_100215-CW-T_200x_CRC-986_300316-CW-T_1800x',\n",
    "    'mixture_chrall_CRC-986_100215-CW-T_100x_CRC-986_300316-CW-T_1900x']}\n",
    "print(aux_dict)\n",
    "aux = pd.DataFrame.from_dict(aux_dict).set_index('sampleid')\n",
    "aux.to_csv('data/mixtures_ultradeep/mixtures_chrall/mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T/calls/CRC-986_100215-CW-T_CRC-986_300316-CW-T_tf_cov.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries['truth'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = list(test[(test['chrom'] != 'X') & (test['chrom'] != 'Y') & (test[['{:.2f}_{}'.format(max(aux['tf']), m) for m in config.methods]].sum(axis=1) >= 5)].index)\n",
    "len(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = list(calltablesseries[calltablesseries['truth'] == True][['{:.2f}_{}_score'.format(aux['tf'].max(), m) for m in config.methods]].index)\n",
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "B = ['4_66217311_G_A',\n",
    " '4_103832597_A_G',\n",
    " '4_103832610_C_T',\n",
    " '4_103832611_G_A',\n",
    " '4_103832620_A_T',\n",
    " '4_103832643_A_C',\n",
    " '4_103832656_C_T',\n",
    " '4_103832665_T_C',\n",
    " '4_144922436_T_G',\n",
    " '4_190883030_G_T',\n",
    " '5_256514_G_C',\n",
    " '5_15937234_G_A',\n",
    " '5_31297469_C_T',\n",
    " '5_69372353_T_C',\n",
    " '5_89949151_G_A',\n",
    " '5_112128143_C_T',\n",
    " '5_140502997_A_C',\n",
    " '5_140503002_C_A',\n",
    " '5_140736267_T_C',\n",
    " '5_145613200_T_C',\n",
    " '5_175387011_G_A',\n",
    " '5_177733904_G_A',\n",
    " '6_26205178_C_T',\n",
    " '6_26368486_C_A',\n",
    " '6_29407980_G_A',\n",
    " '6_31323233_A_G',\n",
    " '6_32712961_G_A',\n",
    " '6_32712967_C_G',\n",
    " '6_116783166_T_C',\n",
    " '7_5352438_G_A',\n",
    " '7_44840986_A_C',\n",
    " '7_72412591_G_A',\n",
    " '7_72418618_A_T',\n",
    " '7_89790175_C_G',\n",
    " '7_100681211_G_C',\n",
    " '7_100681266_C_A',\n",
    " '7_100681293_G_A',\n",
    " '7_100681345_T_G',\n",
    " '7_100681363_C_A',\n",
    " '7_112092224_G_T',\n",
    " '7_117188736_C_A',\n",
    " '7_117188797_A_G',\n",
    " '7_143094423_C_T',\n",
    " '7_143140675_G_A',\n",
    " '7_143453661_T_G',\n",
    " '7_151896464_A_G',\n",
    " '7_151932923_G_A',\n",
    " '7_151945101_G_C',\n",
    " '7_152522195_C_T',\n",
    " '7_153750014_G_A',\n",
    " '7_157959929_G_A',\n",
    " '9_117800_G_T',\n",
    " '9_151046_A_G',\n",
    " '9_151071_T_C',\n",
    " '9_38571752_G_A',\n",
    " '9_38571802_A_G',\n",
    " '9_73233803_G_A',\n",
    " '9_111625513_C_T',\n",
    " '9_130507136_G_A',\n",
    " '9_140772453_G_C',\n",
    " '10_19856538_G_A',\n",
    " '10_22023011_C_T',\n",
    " '10_29779847_A_C',\n",
    " '10_50534954_T_C',\n",
    " '10_79569328_C_T',\n",
    " '10_103587994_C_T',\n",
    " '10_117278812_C_T',\n",
    " '10_124583002_C_T',\n",
    " '11_320606_G_T',\n",
    " '11_320649_G_A',\n",
    " '11_1017806_G_T',\n",
    " '11_1085845_C_T',\n",
    " '11_1085849_A_C',\n",
    " '11_4608262_A_C',\n",
    " '11_46369127_G_A',\n",
    " '11_47647265_A_G',\n",
    " '11_48347424_T_A',\n",
    " '11_64545497_C_T',\n",
    " '11_67759316_C_T',\n",
    " '11_71847122_C_T',\n",
    " '11_89531533_G_T',\n",
    " '11_93826745_G_A',\n",
    " '11_124253181_T_C',\n",
    " '11_132016219_C_T',\n",
    " '12_7945629_T_A',\n",
    " '12_11244356_A_T',\n",
    " '12_11244378_C_G',\n",
    " '12_11244390_T_C',\n",
    " '12_11244432_A_G',\n",
    " '12_11244646_T_C',\n",
    " '12_11244687_G_C',\n",
    " '12_11286149_G_A',\n",
    " '12_11286164_G_A',\n",
    " '12_11546686_C_T',\n",
    " '12_11546742_T_C',\n",
    " '12_11546749_T_C',\n",
    " '12_40877466_T_C',\n",
    " '12_40880634_T_G',\n",
    " '12_40882016_G_T',\n",
    " '12_40882735_A_G',\n",
    " '12_52431751_C_G',\n",
    " '12_52699033_G_A',\n",
    " '12_94543436_G_A',\n",
    " '13_20066994_T_C',\n",
    " '13_20067011_A_C',\n",
    " '13_45968389_G_C',\n",
    " '13_47470350_C_T',\n",
    " '13_73357855_C_T',\n",
    " '13_73636650_C_T',\n",
    " '13_88330033_A_C',\n",
    " '14_58063540_C_T',\n",
    " '14_77229439_C_T',\n",
    " '14_105173873_A_C',\n",
    " '14_105173879_A_C',\n",
    " '14_105410352_T_C',\n",
    " '14_105939672_C_G',\n",
    " '15_22471481_G_T',\n",
    " '15_22471522_C_T',\n",
    " '15_22471542_G_A',\n",
    " '15_22471573_C_T',\n",
    " '15_23104346_C_G',\n",
    " '15_23104657_G_C',\n",
    " '15_23686113_C_G',\n",
    " '15_28950629_C_T',\n",
    " '15_32738694_G_A',\n",
    " '15_41192469_C_T',\n",
    " '15_50785016_A_G',\n",
    " '15_51634218_A_C',\n",
    " '15_74425533_A_G',\n",
    " '15_79057989_C_T',\n",
    " '15_79067099_G_C',\n",
    " '15_84488603_G_T',\n",
    " '15_90213399_C_A',\n",
    " '16_2522557_C_T',\n",
    " '16_30718920_G_A',\n",
    " '16_67183982_C_T',\n",
    " '16_71163579_C_T',\n",
    " '16_88788786_G_A',\n",
    " '16_88909572_G_A',\n",
    " '17_1394895_G_C',\n",
    " '17_5036740_C_G',\n",
    " '17_5036748_G_T',\n",
    " '17_7386217_T_C',\n",
    " '17_7386234_G_A',\n",
    " '17_7578394_T_C',\n",
    " '17_11772531_C_T',\n",
    " '17_21114408_T_C',\n",
    " '17_28887667_T_C',\n",
    " '17_39383069_A_G',\n",
    " '17_56569055_A_G',\n",
    " '17_58234877_G_A',\n",
    " '17_63200307_G_A',\n",
    " '17_74676686_C_A',\n",
    " '18_335173_G_T',\n",
    " '18_9122591_T_C',\n",
    " '18_9122611_T_C',\n",
    " '18_9122638_G_C',\n",
    " '18_13761856_G_C',\n",
    " '18_31325884_T_C',\n",
    " '18_47363917_A_G',\n",
    " '18_47364009_G_A',\n",
    " '18_74208553_C_G',\n",
    " '19_1110748_G_C',\n",
    " '19_1440068_G_C',\n",
    " '19_1881355_T_C',\n",
    " '19_1881408_A_G',\n",
    " '19_6222411_T_G',\n",
    " '19_7677678_C_T',\n",
    " '19_9008248_G_C',\n",
    " '19_9008264_C_T',\n",
    " '19_9009652_C_T',\n",
    " '19_9012501_A_G',\n",
    " '19_9012515_G_A',\n",
    " '19_9015323_C_A',\n",
    " '19_9015324_A_G',\n",
    " '19_9018476_G_A',\n",
    " '19_9018479_G_T',\n",
    " '19_9018508_T_G',\n",
    " '19_9021141_G_A',\n",
    " '19_14877799_G_C',\n",
    " '19_15540781_C_A',\n",
    " '19_15918177_G_T',\n",
    " '19_33698018_C_T',\n",
    " '19_41354606_G_T',\n",
    " '19_41354661_T_C',\n",
    " '19_53116939_C_G',\n",
    " '19_55795866_A_C']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "len([a for a in A if a.startswith('9_')])\n",
    "len([a for a in B if a.startswith('9_')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(A), len(B))\n",
    "A = ['_'.join(a.split('_')[:2]) for a in A]\n",
    "B = ['_'.join(b.split('_')[:2]) for b in B]\n",
    "print(len(set(A) & set(B)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux = calltablesseries[[c for c in list(calltablesseries.columns) if c.startswith('20') or c=='truth']]\n",
    "#aux[aux['truth'] == True][['20.95_mutect2','20.95_mutect2_score', '20.95_freebayes_score', '20.95_strelka2_score', '20.95_vardict_score', '20.95_varscan_score', '20.95_abemus_score', '20.95_smurf_score']].head(50)\n",
    "\n",
    "print(aux[(aux['truth'] != True) & (aux['20.95_mutect2'] == True)].shape)\n",
    "aux[(aux['truth'] != True) & (aux['20.95_mutect2'] == True)]['20.95_mutect2_score'].head(50)#.isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part II: (4) Load back metric results and plot combined metric plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mixtureids =  ['CRC-1014_180816-CW-T_CRC-1014_090516-CW-T', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T', 'CRC-123_310715-CW-T_CRC-123_121115-CW-T']\n",
    "\n",
    "\n",
    "#for fixedvar in fixedvars:\n",
    "fixedvar = 'coverage'\n",
    "if fixedvar == 'coverage':\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    xaxis = 'coverage'\n",
    "#for mt in muttypes:\n",
    "mt = 'snv'\n",
    "if mt == 'snv':\n",
    "    if mixtureid == 'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        gtm = 3\n",
    "        refname = 'intissuesamplebyatleast'+str(gtm)+'callers'\n",
    "    else:\n",
    "        gtm = 3\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "else:  # elif mt == 'indel':\n",
    "    gtm = 2\n",
    "    refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "print(refname)\n",
    "# for metric in metrics:\n",
    "metric = 'auprc'\n",
    "# load results tables\n",
    "restables = {'snv': [], 'indel': []}\n",
    "# for mixtureid in mixtureids:\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(mixtureid, plasmasample)\n",
    "xa = xaxis if xaxis != 'tumor burden' else 'tb'\n",
    "print(xa)\n",
    "restable = pd.read_csv(os.path.join(*config.mixturefolderultradeep, 'mixtures_allchr', 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar+'_'+ xa +'.csv'), index_col=0)\n",
    "restable['plasma sample'] = plasmasample\n",
    "restables[mt].append(restable)\n",
    "restables[mt] = pd.concat(restables[mt])\n",
    "res1 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=False, save=False)\n",
    "#res2 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=True, save=True)\n",
    "#res3 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True)\n",
    "#res4 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=True, save=True)\n",
    "plt.grid(linewidth=1)\n",
    "plt.ylim([0, 0.5])\n",
    "plt.xlim([22, 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "mixtureids =  ['CRC-1014_180816-CW-T_CRC-1014_090516-CW-T', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T', 'CRC-123_310715-CW-T_CRC-123_121115-CW-T']\n",
    "\n",
    "\n",
    "#for fixedvar in fixedvars:\n",
    "fixedvar = 'tumor burden'\n",
    "if fixedvar == 'coverage':\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    xaxis = 'coverage'\n",
    "#for mt in muttypes:\n",
    "mt = 'snv'\n",
    "if mt == 'snv':\n",
    "    if mixtureid == 'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        gtm = 3\n",
    "        refname = 'intissuesamplebyatleast'+str(gtm)+'callers'\n",
    "    else:\n",
    "        gtm = 3\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "else:  # elif mt == 'indel':\n",
    "    gtm = 2\n",
    "    refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "print(refname)\n",
    "# for metric in metrics:\n",
    "metric = 'auprc'\n",
    "# load results tables\n",
    "restables = {'snv': [], 'indel': []}\n",
    "# for mixtureid in mixtureids:\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(mixtureid, plasmasample)\n",
    "xa = xaxis if xaxis != 'tumor burden' else 'tb'\n",
    "print(xa)\n",
    "restable = pd.read_csv(os.path.join(*config.mixturefolderultradeep, 'mixtures_allchr', 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar+'_'+ xa +'.csv'), index_col=0)\n",
    "restable['plasma sample'] = plasmasample\n",
    "restables[mt].append(restable)\n",
    "restables[mt] = pd.concat(restables[mt])\n",
    "res1 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=False, save=False)\n",
    "#res2 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=True, save=True)\n",
    "#res3 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True)\n",
    "#res4 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=True, save=True)\n",
    "plt.grid(linewidth=1)\n",
    "#plt.ylim([0, 0.5])\n",
    "#plt.xlim([22, 0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "resx = np.array([rx.values for rx in res1['x']])\n",
    "resy = np.array([ry.values for ry in res1['y']])\n",
    "res = {m: [] for m in config.methods}\n",
    "for mi, m in enumerate(config.methods):\n",
    "    if len(resy.shape) > 2:\n",
    "        resmean = np.mean([resy[mi], resy[mi+5], resy[mi+10]], axis=0)\n",
    "        resstd = np.std([resy[mi], resy[mi+5], resy[mi+10]], axis=0)\n",
    "        #reslabel = m\n",
    "    else:\n",
    "        resmean = list(resy[mi])\n",
    "        resstd = [0 for i in range(7)]\n",
    "    if len(resx.shape) == 2:\n",
    "        res[m] = [resmean, resstd, resx[0]]\n",
    "    else:\n",
    "        res[m] = [resmean, resstd, resx.mean(axis=0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "plt.figure(figsize=(15, 10))\n",
    "plt.grid(linewidth=0)\n",
    "for m in config.methods:\n",
    "    #plt.plot(res['freebayes'][2], res['freebayes'][0], c=color_dict['freebayes'], markersize=15, lw=2)\n",
    "    plt.plot(res[m][2], res[m][0], marker=config.markers[0], label=m,  markersize=15, lw=2)\n",
    "    plt.gca().invert_xaxis()\n",
    "ax = plt.gca()\n",
    "xlab = 'coverage or added noise'\n",
    "xlab='tumor burden'\n",
    "hand, labl = ax.get_legend_handles_labels()\n",
    "ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.xlabel(xlab)\n",
    "plt.ylabel(metric.upper()+' score')\n",
    "plt.title(metric.upper() + \" score for {} calling in chr{} with ref {}\".format(mt.upper(), chrom, refname))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for m in config.methods:\n",
    "    res[m][2] = res[m][2][::-1]\n",
    "    res[m][0] = res[m][0][::-1]\n",
    "    res[m][1] = res[m][1][::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15, 10))\n",
    "#plt.grid(linewidth=1)\n",
    "#res[m][2] = res[m][2][::-1]\n",
    "#res[m][0] = res[m][0][::-1]\n",
    "#res[m][1] = res[m][1][::-1]\n",
    "rankres = []\n",
    "for m in config.methods:\n",
    "    xstart = 0\n",
    "    ystart = 0\n",
    "    sum_all = []\n",
    "    for i in range(len(res[m][2])):\n",
    "        #if i >0:\n",
    "        #print(res[m][2][i], xstart)\n",
    "        #print((res[m][2][i]-xstart) * res[m][0][i] )\n",
    "        #print(((res[m][2][i]-xstart) * (res[m][0][i]-ystart)/2))\n",
    "        sum_all.append(((res[m][2][i]-xstart) * res[m][0][i]) - ((res[m][2][i]-xstart) * (res[m][0][i]-ystart)/2))\n",
    "        xstart = res[m][2][i]\n",
    "        ystart = res[m][0][i]\n",
    "    print(m, np.sum(sum_all) )\n",
    "    rankres.append(np.sum(sum_all))\n",
    "    #plt.plot(res['freebayes'][2], res['freebayes'][0], c=color_dict['freebayes'], markersize=15, lw=2)\n",
    "    #plt.errorbar(res[m][2], res[m][0], xerr = resx.std(axis=0), yerr=res[m][1], marker=config.markers[0], label=m,  markersize=15, lw=2, fmt='-o')\n",
    "    #plt.gca().invert_xaxis()\n",
    "#ax = plt.gca()\n",
    "#xlab = 'coverage or added noise'\n",
    "#xlab='tumor burden'\n",
    "#hand, labl = ax.get_legend_handles_labels()\n",
    "#ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "#plt.xlabel(xlab)\n",
    "#plt.ylabel(metric.upper()+' score')\n",
    "#plt.title(metric.upper() + \" score for {} calling in chr{} with ref {}\".format(mt.upper(), chrom, refname))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patches as mpatches\n",
    "rankres_sort = np.argsort(rankres)[::-1]\n",
    "rankres_sort\n",
    "plt.figure(figsize=(14, 5))\n",
    "method_order =[]\n",
    "rank_order =[]\n",
    "for i in rankres_sort:\n",
    "    method_order.append(config.methods[i])\n",
    "    rank_order.append(rankres[i])\n",
    "plt.bar(method_order, rank_order, color=['b', 'b', 'b', 'b', 'b', 'r', 'r'])\n",
    "blue_patch = mpatches.Patch(color='b', label='DNA methods')\n",
    "red_patch = mpatches.Patch(color='r', label='cfDNA methods')\n",
    "plt.legend(handles=[blue_patch, red_patch])\n",
    "#plt.legend(['DNA methods', 'cfDNA methods'])\n",
    "plt.ylabel('area under AUPRC curve')\n",
    "plt.title('Caller ranking mixtures with decreasing tumor burden')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
