{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Series analysis\n",
    "\n",
    "\n",
    "Analysis updates:\n",
    "- retrieve calls rejected only because VAF<0.01 (bcbio lower acceptable threshold) for Mutect2, Strelka2 and Vardict\n",
    "- correct for germline mutations using GATK haplotype calls\n",
    "- get PR curves for sinvict by encoding the 6 files as thresholds. Assumption: linear filters\n",
    "- plotting: PR curves stops at 10e-2 on the left\n",
    "- ground truths:\n",
    "\n",
    "    1) Consensus: build using\n",
    "        majority of 5/8 callers for SNV and 3/5 callers for INDELS\n",
    "\n",
    "\n",
    "    2) Ranked mutations: metascore built using\n",
    "        weigthed sum of normalised scores between 0 and 1 for each caller\n",
    "        with weigths = inversially proportional to number of calls made by caller (if a caller calls few mutations higher weight, it many calls low weight)\n",
    "        threshold = 1/ncallers\n",
    "        interpretation: if 1 caller is sure (score = 1) of calling this position, add it to GT\n",
    "        interpretation: if 2 callers are quite sure of calling this position (score > 0,5 each), add it to GT\n",
    "        \n",
    "- integrate VAF approx\n",
    "\n",
    "    1) mixture with VAF instead of tumor burden\n",
    "    \n",
    "    2) correct for mutations non present in diluted samples (vaf = 0)\n",
    "    \n",
    "    3) pool patients together using VAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "import warnings\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, average_precision_score\n",
    "warnings.filterwarnings('ignore')\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "# set working directory\n",
    "if not os.getcwd().endswith('cfdna_snv_benchmark'):\n",
    "    os.chdir('../')\n",
    "print('Current working directory: {}'.format(os.getcwd()))\n",
    "\n",
    "from utils.config import Config\n",
    "from utils.viz import *\n",
    "from utils.table import *\n",
    "from utils.metrics import *\n",
    "from utils.calltable import *\n",
    "from utils.calltableseries import *\n",
    "from utils.groundtruth import *\n",
    "from utils.metricsseries import *\n",
    "from utils.venn import venn6, get_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Config and Display paramaters\n",
    "\n",
    "config = Config(\"config/\", \"config_viz.yaml\")\n",
    "set_display_params(config)\n",
    "print(config.methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chomosome\n",
    "\n",
    "mixtureids =  ['CRC-1014_180816-CW-T_CRC-1014_090516-CW-T', 'CRC-123_310715-CW-T_CRC-123_121115-CW-T', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T']\n",
    "#mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "#mixtureid = 'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T'\n",
    "mixtureid = 'CRC-123_310715-CW-T_CRC-123_121115-CW-T'\n",
    "#mixtureid = 'BRA-412_240820-CW-T_BRA-412_060220-CW-T'\n",
    "reload = False\n",
    "save = False\n",
    "fixedvars=['coverage', 'ctdna']\n",
    "filterparam = 'all'\n",
    "\n",
    "markers = ['o', '^', 'X']\n",
    "linestyles = ['-', '-', '-']\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "\n",
    "muttypes = ['snv', 'indel']\n",
    "metrics = ['auprc', 'precision', 'recall']\n",
    "\n",
    "chrom = 'all'\n",
    "\n",
    "# INDELS: patient 986 = 96, patient 123 = 357, patient 1014 = 277 (â‰¥ 4/6 callers without SMURF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part I: (1) Load/Generate call tables, (2) Generate Ground truths and (3) Compute/Save metrics per patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixedvar = 'coverage'\n",
    "for fixedvar in fixedvars:\n",
    "    if fixedvar == 'coverage':\n",
    "        seriesorder = [(70, 0), (70, 80), (50, 100), (30, 120), (20, 130), (10, 140), (5, 145)]\n",
    "        xaxis = 'tumor burden'\n",
    "    elif fixedvar == 'ctdna':\n",
    "        seriesorder = [(70, 0), (70, 80), (70, 180)]\n",
    "        xaxis = 'coverage'\n",
    "    #for mixtureid in mixtureids:\n",
    "    #mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "    print('############# {} ############'.format(mixtureid))\n",
    "    if mixtureid ==  'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T':\n",
    "        chroms = [str(c) for c in range(1,23) if c != 2 and c!=6 and c !=17 and c!=19 and c!=20 and c!=21]\n",
    "        #chroms = [str(c) for c in range(1,9) if c != 2 and c!=6]\n",
    "    elif mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        chroms = [str(c) for c in range(1,23) if c !=1 and c!= 2 and c !=8 and c!=20 and c!=21 and c!=22] \n",
    "    else:\n",
    "        chroms = [str(c) for c in range(1,23) if c !=6 and c!=19 and c!=20]  # c !=1 and c!= 2 and\n",
    "    calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "    aux_all = []\n",
    "    calltable_snv, aux = get_calltableseries(config, mixtureid, chroms, muttype='snv', filterparam=filterparam, reload=reload, save=save)\n",
    "    calltable_indel, aux = get_calltableseries(config, mixtureid, chroms, muttype='indel', filterparam=filterparam, reload=reload, save=save)\n",
    "    calltable_snp, aux = get_calltableseries(config, mixtureid, chroms, muttype='snp', filterparam=filterparam, reload=reload, save=save)\n",
    "    print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "    print(aux)\n",
    "    plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "    print(plasmasample)\n",
    "    healthysample = '_'.join(mixtureid.split('_')[2:])\n",
    "    print(healthysample)\n",
    "    calltables['snv'].append(calltable_snv)\n",
    "    calltables['indel'].append(calltable_indel)\n",
    "    calltables['snp'].append(calltable_snp)\n",
    "    calltables['sampleid'] = mixtureid \n",
    "    calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "    calltables['snv'] = pd.concat(calltables['snv'])\n",
    "    calltables['indel'] = pd.concat(calltables['indel'])\n",
    "    calltables['snp'] = pd.concat(calltables['snp'])\n",
    "    dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "    #for muttype in muttypes:\n",
    "    muttype = 'snv'\n",
    "    refsample = 'undiluted'\n",
    "    if muttype == 'snv':\n",
    "        gtm = 5\n",
    "    else:  # elif muttype == 'indel':\n",
    "        gtm = 4\n",
    "    print(max(aux['tf']))\n",
    "    #if mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T' and muttype == 'snv':\n",
    "    #    gtm = 3\n",
    "    #    refsample = 'tissue'\n",
    "    #    #calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "    #    #                                    matchedtissuepath=os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-T1W', 'calls', 'NCC_CRC-986_100215-T1W_snv_calls_PASS_exome.csv'))\n",
    "    #    calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "    #                                        matchedtissuepath=os.path.join('data', 'matchedtissue_ultradeep', '986_100215_T1-E', 'calls', '986_100215_T1-E_snv_calls_all.csv'))\n",
    "    #if mixtureid !=  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    #calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "    #                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict'])\n",
    "    #else:\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                            matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "    results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "                                     refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall', ground_truth_method=gtm,\n",
    "    #                                       refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision', ground_truth_method=gtm,\n",
    "    #                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #results_maxf1_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1', ground_truth_method=gtm,\n",
    "    #                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #results_maxrecallatlast0_05precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_03precision', ground_truth_method=gtm,\n",
    "    #                                  refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #results_maxrecallatlast0_05precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_04precision', ground_truth_method=gtm,\n",
    "    #                                  refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    results_maxrecallatlast0_03precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_03precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall', ground_truth_method=gtm,\n",
    "    #                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision', ground_truth_method=gtm,\n",
    "    #                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "    #figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method=gtm,\n",
    "    #                    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries = pd.read_csv(os.path.join('data', 'mixtures', 'mixtures_allchr', 'BRA-412_240820-CW-T_BRA-412_060220-CW-T_snv_calls_all.csv'), index_col=0)\n",
    "aux = pd.read_csv(os.path.join('data', 'mixtures', 'mixtures_allchr', 'BRA-412_240820-CW-T_BRA-412_060220-CW-T_tf_cov.csv'),index_col=0)\n",
    "gtm = 5\n",
    "muttype = 'snv'\n",
    "refsample = 'undiluted'\n",
    "fixedvar = 'coverage'\n",
    "if fixedvar == 'coverage':\n",
    "    seriesorder = [(100, 0), (70, 30), (50, 50), (30, 70), (20, 80), (10, 90)] \n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    seriesorder = [(100, 0), (100, 50), (100, 100)]\n",
    "    xaxis = 'coverage'\n",
    "dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "calltablesseries = generate_groundtruth(config, calltablesseries, aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture')\n",
    "results_maxrecallatlast0_03precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_03precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_maxrecallatlast0_05precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_03precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(calltablesseries['truth'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "3*10e6 == 30000000"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(calltablesseries.shape[0])\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "print(calltablesseries[calltablesseries['truth'] == True].shape[0] / calltablesseries.shape[0])\n",
    "print(calltablesseries[calltablesseries['truth'] == True].shape[0] / (3*10e6))\n",
    "print(72815 * calltablesseries[calltablesseries['truth'] == True].shape[0] / (3*10e6))\n",
    "print(0.03 / (calltablesseries[calltablesseries['truth'] == True].shape[0] /(3*10e6)) )\n",
    "print(0.03 / (calltablesseries[calltablesseries['truth'] == True].shape[0] / calltablesseries.shape[0] ))\n",
    "\n",
    "print(100000 * (calltablesseries[calltablesseries['truth'] == True].shape[0] /(3*10e6)) )\n",
    "print(500 * (calltablesseries[calltablesseries['truth'] == True].shape[0] / calltablesseries.shape[0] ))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(config.methods)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# GT analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Config and Display paramaters\n",
    "\n",
    "config = Config(\"config/\", \"config_viz.yaml\")\n",
    "set_display_params(config)\n",
    "print(config.methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "#mixtureid = 'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T'\n",
    "#mixtureid = 'CRC-123_310715-CW-T_CRC-123_121115-CW-T'\n",
    "\n",
    "fixedvar = 'coverage'\n",
    "if fixedvar == 'coverage':\n",
    "    seriesorder = [(70, 0), (70, 80), (50, 100), (30, 120), (20, 130), (10, 140), (5, 145)]\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    seriesorder = [(70, 0), (70, 80), (70, 180)]\n",
    "    xaxis = 'coverage'\n",
    "#for mixtureid in mixtureids:\n",
    "#mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "if mixtureid ==  'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T':\n",
    "    chroms = [str(c) for c in range(1,23) if c != 2 and c!=6 and c !=17 and c!=19 and c!=20 and c!=21]\n",
    "    #chroms = [str(c) for c in range(1,9) if c != 2 and c!=6]\n",
    "elif mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    chroms = [str(c) for c in range(1,23) if c !=1 and c!= 2 and c !=8 and c!=20 and c!=21 and c!=22] \n",
    "else:\n",
    "    chroms = [str(c) for c in range(1,23) if c !=6 and c!=19 and c!=20]  # c !=1 and c!= 2 and\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "aux_all = []\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chroms, muttype='snv', filterparam=filterparam, reload=reload, save=save)\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chroms, muttype='indel', filterparam=filterparam, reload=reload, save=save)\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chroms, muttype='snp', filterparam=filterparam, reload=reload, save=save)\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(plasmasample)\n",
    "healthysample = '_'.join(mixtureid.split('_')[2:])\n",
    "print(healthysample)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "#for muttype in muttypes:\n",
    "muttype = 'snv'\n",
    "refsample = 'undiluted'\n",
    "if muttype == 'snv':\n",
    "    gtm = 5\n",
    "else:  # elif muttype == 'indel':\n",
    "    gtm = 3\n",
    "print(max(aux['tf']))\n",
    "#if mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T' and muttype == 'snv':\n",
    "#    gtm = 3\n",
    "#    refsample = 'tissue'\n",
    "#    #calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "#    #                                    matchedtissuepath=os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-T1W', 'calls', 'NCC_CRC-986_100215-T1W_snv_calls_PASS_exome.csv'))\n",
    "#    calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "#                                        matchedtissuepath=os.path.join('data', 'matchedtissue_ultradeep', '986_100215_T1-E', 'calls', '986_100215_T1-E_snv_calls_all.csv'))\n",
    "if mixtureid !=  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict'])\n",
    "else:\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdf = calltablesseries[calltablesseries['truth'] == True]\n",
    "gtdf['median_vaf'] = gtdf[['{:.2f}_{}_vaf'.format(aux['tf'].max(), m) for m in config.methods if '{:.2f}_{}_vaf'.format(aux['tf'].max(), m) in gtdf.columns]].median(axis=1, skipna=True)\n",
    "sns.histplot(data=gtdf, x=\"median_vaf\", bins=30, stat='frequency')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtdf['median_vaf'].to_csv('figures/figure2b/gt_986_exome_150x_atleast5callersinundilutedsample_snv.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries.loc[ac.loc[ac.index.duplicated()].index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Bwithoutab.loc[Bwithoutab.loc[Bwithoutab.index.duplicated()].index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = pd.read_csv('figures/figure2b/gt_986_exome_150x_atleast5callersinundilutedsample_snv.csv', index_col=0)\n",
    "B = pd.read_csv('figures/figure2b/gt_986_exome_2000x_atleast5callersinundilutedsample_snv.csv', index_col=0)\n",
    "print(A.shape[0], B.shape[0])\n",
    "ab = list(set(set(list(A.index)) & set(list(B.index))))\n",
    "len(ab)\n",
    "Awithoutab = A.loc[list(set(A.index) - set(ab))]\n",
    "Bwithoutab = B.loc[list(set(B.index) - set(ab))]\n",
    "\n",
    "plt.figure(figsize=(10,10))\n",
    "\n",
    "#calltablesseries = pd.read_csv(os.path.join('data', 'mixtures', 'mixtures_allchr', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_snv_calls_all.csv'), index_col=0, memory_map=True)\n",
    "aux = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_tf_cov.csv'), index_col=0)\n",
    "ac = pd.DataFrame(calltablesseries.loc[[a for a in list(Bwithoutab.index) if a in calltablesseries.index]][['{:.2f}_{}_vaf'.format(aux['tf'].max(), m) for m in config.methods if ('{:.2f}_{}'.format(aux['tf'].max(), m) in calltablesseries.columns) and (m != 'smurf')]].median(skipna=True, axis=1))\n",
    "ac = ac[~ac.index.duplicated()]\n",
    "Bwithoutab = Bwithoutab[~Bwithoutab.index.duplicated()]\n",
    "comp = pd.concat([ac, Bwithoutab], axis=1)\n",
    "comp.columns = ['vaf 150x', 'vaf 2000x']\n",
    "comp.fillna(0, inplace=True)\n",
    "list_only2000x = list(comp.index)\n",
    "print(Bwithoutab.shape[0], ac.shape[0])\n",
    "sns.histplot(x='vaf 150x', y='vaf 2000x', data=comp, binwidth=0.01, binrange=[0,.5], alpha=1, color='tab:blue', label='2000x only')\n",
    "\n",
    "comp = pd.concat([A.loc[ab], B.loc[ab]], axis=1)\n",
    "comp.columns = ['vaf 150x', 'vaf 2000x']\n",
    "comp.fillna(0, inplace=True)\n",
    "list_both = list(comp.index)\n",
    "sns.histplot(x='vaf 150x', y='vaf 2000x', data=comp, binwidth=0.01, binrange=[0,.5], alpha=1, color='tab:red', label='both')\n",
    "\n",
    "#table_ultradeep = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_snv_calls_all.csv'), index_col=0, memory_map=True)\n",
    "#aux_ultradeep = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_tf_cov.csv'), index_col=0)\n",
    "#table_ultradeep = generate_groundtruth(config, table_ultradeep, aux_ultradeep['tf'], ground_truth_method=5, muttype='snv', matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "\n",
    "ac = calltablesseries.loc[[a for a in list(Awithoutab.index) if a in table_ultradeep.index]][['{:.2f}_{}_vaf'.format(aux_ultradeep['tf'].max(), m) for m in config.methods if ('{:.2f}_{}'.format(aux_ultradeep['tf'].max(), m) in table_ultradeep.columns) and (m != 'smurf')]].median(skipna=True, axis=1)\n",
    "comp = pd.concat([Awithoutab, ac], axis=1)\n",
    "comp.columns = ['vaf 150x', 'vaf 2000x']\n",
    "comp.fillna(0, inplace=True)\n",
    "list_only150x = list(comp.index)\n",
    "print(Awithoutab.shape[0], ac.shape[0])\n",
    "sns.histplot(x='vaf 150x', y='vaf 2000x', data=comp, binwidth=0.01, binrange=[0,.5], alpha=1, color='tab:olive', label='150x only')\n",
    "\n",
    "from matplotlib.lines import Line2D\n",
    "a = Line2D([0], [0], color='tab:red', lw=4)\n",
    "b = Line2D([0], [0], color='tab:blue', lw=4)\n",
    "c = Line2D([0], [0], color='tab:olive', lw=4)\n",
    "\n",
    "plt.legend([b, a,  c], [ 'both', '2000x only', '150x only'], bbox_to_anchor=(1,1), loc=\"upper left\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = pd.read_csv('figures/figure2b/gt_986_exome_150x_atleast5callersinundilutedsample_snv.csv', index_col=0)\n",
    "B = pd.read_csv('figures/figure2b/gt_986_exome_2000x_atleast5callersinundilutedsample_snv.csv', index_col=0)\n",
    "print(A.shape[0], B.shape[0])\n",
    "ab = list(set(set(list(A.index)) & set(list(B.index))))\n",
    "len(ab)\n",
    "Awithoutab = A.loc[list(set(A.index) - set(ab))]\n",
    "Bwithoutab = B.loc[list(set(B.index) - set(ab))]\n",
    "\n",
    "plt.figure(figsize=(10,10))\n",
    "ac = pd.DataFrame(calltablesseries.loc[[a for a in list(Bwithoutab.index) if a in calltablesseries.index]][['{:.2f}_{}_vaf'.format(aux['tf'].max(), m) for m in config.methods if ('{:.2f}_{}'.format(aux['tf'].max(), m) in calltablesseries.columns) and (m != 'smurf')]].median(skipna=True, axis=1))\n",
    "ac = ac[~ac.index.duplicated()]\n",
    "Bwithoutab = Bwithoutab[~Bwithoutab.index.duplicated()]\n",
    "comp = pd.concat([ac, Bwithoutab], axis=1)\n",
    "comp.fillna(0, inplace=True)\n",
    "comp.columns = ['vaf 150x', 'vaf 2000x']\n",
    "print(Bwithoutab.shape[0], ac.shape[0])\n",
    "sns.histplot(x='vaf 150x', y='vaf 2000x', data=comp, binwidth=0.01, binrange=[0,1], alpha=1, color='g', label='2000x only')\n",
    "\n",
    "comp = pd.concat([A.loc[ab], B.loc[ab]], axis=1)\n",
    "comp.columns = ['vaf 150x', 'vaf 2000x']\n",
    "sns.histplot(x='vaf 150x', y='vaf 2000x', data=comp, binwidth=0.01, binrange=[0,1], alpha=1, label='both')\n",
    "\n",
    "table_ultradeep = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_snv_calls_all.csv'), index_col=0, memory_map=True)\n",
    "aux_ultradeep = pd.read_csv(os.path.join('data', 'mixtures_ultradeep', 'mixtures_chrall', 'mixtures_chrall_CRC-986_100215-CW-T_CRC-986_300316-CW-T' , 'calls', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T_tf_cov.csv'), index_col=0)\n",
    "table_ultradeep = generate_groundtruth(config, table_ultradeep, aux_ultradeep['tf'], ground_truth_method=5, muttype='snv', matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "\n",
    "ac = calltablesseries.loc[[a for a in list(Awithoutab.index) if a in table_ultradeep.index]][['{:.2f}_{}_vaf'.format(aux_ultradeep['tf'].max(), m) for m in config.methods if ('{:.2f}_{}'.format(aux_ultradeep['tf'].max(), m) in table_ultradeep.columns) and (m != 'smurf')]].median(skipna=True, axis=1)\n",
    "comp = pd.concat([Awithoutab, ac], axis=1)\n",
    "comp.fillna(0, inplace=True)\n",
    "comp.columns = ['vaf 150x', 'vaf 2000x']\n",
    "print(Awithoutab.shape[0], ac.shape[0])\n",
    "sns.histplot(x='vaf 150x', y='vaf 2000x', data=comp, binwidth=0.01, binrange=[0,1], alpha=1, color='r', label='150x only')\n",
    "\n",
    "from matplotlib.lines import Line2D\n",
    "a = Line2D([0], [0], color='g', lw=4)\n",
    "b = Line2D([0], [0], color='b', lw=4)\n",
    "c = Line2D([0], [0], color='r', lw=4)\n",
    "\n",
    "plt.legend([b, a,  c], [ 'both', '2000x only', '150x only'], bbox_to_anchor=(1,1), loc=\"upper left\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy.stats import fisher_exact\n",
    "from statsmodels.sandbox.stats.multicomp import multipletests\n",
    "\n",
    "# contingency table and Fisher exact test\n",
    "\n",
    "print(len(list(A.index)), len(list(B.index)))\n",
    "A = A[~A.index.duplicated()]\n",
    "B = B[~B.index.duplicated()]\n",
    "print(len(list(A.index)), len(list(B.index)))\n",
    "res = pd.concat([A, B], axis=1)\n",
    "res[res.isna()] = False\n",
    "res[res != False] = True\n",
    "res.columns = ['150x', '2000x']\n",
    "\n",
    "testres = {}\n",
    "a, b = '150x', '2000x'\n",
    "print(\"############\")\n",
    "print(a, 'VS', b)\n",
    "print(\"############\")\n",
    "ctable = [[res[(res[a]==True) & (res[b]==True)].shape[0], res[(res[a]==False) & (res[b]==True)].shape[0]],\n",
    "          [res[(res[a]==True) & (res[b]==False)].shape[0], int(20000 - res[(res[a]==True) | (res[b]==True)].shape[0])]]\n",
    "\n",
    "#p_adjusted = multipletests(Column6, method='bonferroni')\n",
    "odd_ratio, p_value = fisher_exact(ctable, alternative='two-sided')\n",
    "print('Contingency table:')\n",
    "print(pd.DataFrame(ctable, columns=[a, '!'+a], index=[b, '!'+b]))\n",
    "print('Fisher exact test (two-sided): odd_ratio = {}, p_value = {}'.format(odd_ratio, p_value))\n",
    "testres[a+' VS '+b] = [p_value]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    " int(res[(res[a]==True) | (res[b]==True)].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "list(calltablesseries[calltablesseries['truth'] == True][['{:.2f}_{}_score'.format(aux['tf'].max(), m) for m in config.methods]].index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.transforms as transforms\n",
    "\n",
    "patient = mixtureid.split('-')[1].split('_')[0]\n",
    "print(patient)\n",
    "gtanalysis = calltablesseries[calltablesseries['truth'] == True][['{:.2f}_{}'.format(aux['tf'].max(), m) for m in config.methods]]\n",
    "initialanalysis = calltablesseries[calltablesseries['truth'] == False][['{:.2f}_{}'.format(aux['tf'].max(), m) for m in config.methods]]\n",
    "ngt = gtanalysis.shape[0]\n",
    "print(ngt)\n",
    "gtanalysis = pd.DataFrame(gtanalysis.sum()).T\n",
    "gtanalysis.columns = [l.split('_')[1] for l in list(gtanalysis.columns)]\n",
    "gtanalysis.index = ['calls in GT']\n",
    "initialanalysis = pd.DataFrame(initialanalysis.sum()).T\n",
    "initialanalysis.columns = [l.split('_')[1] for l in list(initialanalysis.columns)]\n",
    "initialanalysis.index = ['calls not in GT']\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10,5))\n",
    "plt.bar(gtanalysis.columns, gtanalysis.values.flatten(), bottom=1, color=[config.colors[config.methods.index(m)] for m in gtanalysis.columns], width=1)\n",
    "plt.axhline(y=ngt, c='blue', lw='3')\n",
    "trans = transforms.blended_transform_factory(\n",
    "    ax.get_yticklabels()[0].get_transform(), ax.transData)\n",
    "ax.text(0, ngt, \"{:.0f}\".format(ngt), color=\"blue\", transform=trans, ha=\"right\", va=\"center\")\n",
    "for col in initialanalysis.columns:\n",
    "    print(col, initialanalysis[col].values[0], gtanalysis[col].values[0])\n",
    "    plt.bar(col, initialanalysis[col].values[0], bottom=gtanalysis[col].values[0], label=col, color='grey', alpha=0.5, width=1)\n",
    "        \n",
    "for pi, p in enumerate(ax.patches):\n",
    "    print(pi, p)\n",
    "    if pi < (len(ax.patches)/2):\n",
    "        width, height = p.get_width(), p.get_height()\n",
    "        x, y = p.get_xy() \n",
    "        ax.text(x+width/2, \n",
    "                20, \n",
    "                '{:.0f}'.format(height), \n",
    "                horizontalalignment='center')\n",
    "    else:\n",
    "        width, height = p.get_width(), p.get_height()\n",
    "        x, y = p.get_xy() \n",
    "        ax.text(x+width/2, \n",
    "                y+height/10, \n",
    "                '{:.0f}'.format(height), \n",
    "                horizontalalignment='center')\n",
    "ax.set_yscale('log')\n",
    "plt.ylabel('# SNV')\n",
    "ax.grid(axis='y')\n",
    "plt.ylim([1, 5e5])\n",
    "ax = plt.gca()\n",
    "ax.set_xticklabels(labels=gtanalysis.columns,rotation=90)\n",
    "plt.savefig(os.path.join(*config.outputpath, 'figure2a', 'gtanalysis_barplot_'+patient+'_SNV.svg'), bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.transforms as transforms\n",
    "\n",
    "gtanalysis = calltablesseries[calltablesseries['truth'] == True][['{:.2f}_{}'.format(aux['tf'].max(), m) for m in config.methods]]\n",
    "initialanalysis = calltablesseries[calltablesseries['truth'] == False][['{:.2f}_{}'.format(aux['tf'].max(), m) for m in config.methods]]\n",
    "ngt = gtanalysis.shape[0]\n",
    "print(ngt)\n",
    "gtanalysis = pd.DataFrame(gtanalysis.sum()).T\n",
    "gtanalysis.columns = [l.split('_')[1] for l in list(gtanalysis.columns)]\n",
    "gtanalysis.index = ['calls in GT']\n",
    "initialanalysis = pd.DataFrame(initialanalysis.sum()).T\n",
    "initialanalysis.columns = [l.split('_')[1] for l in list(initialanalysis.columns)]\n",
    "initialanalysis.index = ['calls not in GT']\n",
    "gtanalysis = pd.concat([gtanalysis, initialanalysis])\n",
    "#ax = gtanalysis.T.plot(kind='bar', stacked=True, figsize=(12, 5), rot=0, color=[config.colors[config.methods.index(m)] for m in gtanalysis.columns])\n",
    "ax = sns.barplot(x=gtanalysis.columns, y='calls in GT', data=gtanalysis.T) #, figsize=(12, 5), rot=0) #, color=[config.colors[config.methods.index(m)] for m in gtanalysis.columns])\n",
    "ax.set_yscale('log')\n",
    "for ci, c in enumerate(ax.containers):\n",
    "    if ci == 0:\n",
    "        ax.bar_label(c, label_type='center')\n",
    "    else:\n",
    "        ax.bar_label(c, label_type='edge')\n",
    "ax.grid(axis='y')\n",
    "plt.ylim([1, 5e5])\n",
    "ax = plt.gca()\n",
    "ax.set_xticklabels(labels=gtanalysis.columns,rotation=90)\n",
    "hand, labl = ax.get_legend_handles_labels()\n",
    "ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "\"\"\"\n",
    "fig, ax = plt.subplots(figsize=(15,5))\n",
    "plt.bar(gtanalysis.columns, gtanalysis.values.flatten(), color=[config.colors[config.methods.index(m)] for m in gtanalysis.columns], width=.9)\n",
    "plt.axhline(y=ngt, c='blue', lw='3')\n",
    "trans = transforms.blended_transform_factory(\n",
    "    ax.get_yticklabels()[0].get_transform(), ax.transData)\n",
    "ax.text(0, ngt, \"{:.0f}\".format(ngt), color=\"blue\", transform=trans, ha=\"right\", va=\"center\")\n",
    "for col in initialanalysis.columns:\n",
    "    print(col, initialanalysis[col].values[0], gtanalysis[col].values[0], bottom=0)\n",
    "    plt.bar(col, initialanalysis[col].values[0], bottom=gtanalysis[col].values[0], label=col, color='grey', alpha=0.5, width=.9)\n",
    "ax.set_yscale('log')\n",
    "plt.ylabel('# SNV')\n",
    "ax.grid(axis='y')\n",
    "plt.ylim([1, 5e5])\n",
    "#plt.savefig(os.path.join(*config.outputpath, 'figure2a', 'gtanalysis_barplot_986_SNV.svg'), bbox_inches='tight')\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall', ground_truth_method=gtm,\n",
    "                                       refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "results_maxf1_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method=gtm,\n",
    "                    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part II: (4) Load back metric results and plot combined metric plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for fixedvar in fixedvars:\n",
    "#fixedvar = 'coverage'\n",
    "    if fixedvar == 'coverage':\n",
    "        xaxis = 'tumor burden'\n",
    "    elif fixedvar == 'ctdna':\n",
    "        xaxis = 'coverage'\n",
    "    #for mt in muttypes:\n",
    "    mt = 'indel'\n",
    "    if mt == 'snv':\n",
    "        gtm = 4\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    else:  # elif mt == 'indel':\n",
    "        gtm = 2\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    print(refname)\n",
    "    # for metric in metrics:\n",
    "    metric = 'auprc'\n",
    "    # load results tables\n",
    "    restables = {'snv': [], 'indel': []}\n",
    "    mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "    if mixtureid == 'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        gtm = 3\n",
    "        refname = 'intissuesamplebyatleast'+str(gtm)+'callers'\n",
    "    else:\n",
    "        gtm = 4\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "    print(mixtureid, plasmasample)\n",
    "    xa = xaxis if xaxis != 'tumor burden' else 'tb'\n",
    "    print(xa)\n",
    "    restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_allchr', 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar+'_'+ xa +'.csv'), index_col=0)\n",
    "    #restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_chr22_wgs', 'mixtures_chr22_'+mixtureid, 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar +'.csv'), index_col=0)\n",
    "    restable['plasma sample'] = plasmasample\n",
    "    restables[mt].append(restable)\n",
    "    restables[mt] = pd.concat(restables[mt])\n",
    "    res1 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=False, save=False)\n",
    "    #res2 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=True, save=True)\n",
    "    #res3 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True)\n",
    "    #res4 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=True, save=True)\n",
    "    #plt.grid(linewidth=1)\n",
    "    plt.ylim([0, .8])\n",
    "    #plt.xlim([20, 0])\n",
    "    #plt.xlim([40, 250])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plt.grid(linewidth=1)\n",
    "\n",
    "#for fixedvar in fixedvars:\n",
    "fixedvar = 'coverage'\n",
    "if fixedvar == 'coverage':\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    xaxis = 'coverage'\n",
    "#for mt in muttypes:\n",
    "mt = 'snv'\n",
    "if mt == 'snv':\n",
    "    gtm = 5\n",
    "    refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "else:  # elif mt == 'indel':\n",
    "    gtm = 4\n",
    "    refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "print(refname)\n",
    "# for metric in metrics:\n",
    "metric = 'auprc'\n",
    "#metric = 'maxrecallatleast0_03precision'\n",
    "# load results tables\n",
    "restables = {'snv': [], 'indel': []}\n",
    "#for mixtureid in mixtureids:\n",
    "print(mixtureid)\n",
    "refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(mixtureid, plasmasample)\n",
    "xa = xaxis if xaxis != 'tumor burden' else 'tb'\n",
    "print(xa)\n",
    "restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_allchr', 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar+'_'+ xa +'.csv'), index_col=0)\n",
    "#restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_chr22_wgs', 'mixtures_chr22_'+mixtureid, 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar +'.csv'), index_col=0)\n",
    "restable['plasma sample'] = plasmasample\n",
    "restables[mt].append(restable)\n",
    "\n",
    "restables[mt] = pd.concat(restables[mt])\n",
    "res1 = plot_metricsseries(config, restables, [mixtureid], 'all', metric=metric, muttype=mt,\n",
    "                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True)\n",
    "#res1 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=False, save=False)\n",
    "#res2 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=True, save=True)\n",
    "#res3 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True)\n",
    "#res4 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=True, save=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "resx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tfdict = {'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T': [40.3, 3.2], \n",
    "         'CRC-123_310715-CW-T_CRC-123_121115-CW-T': [62.1, 2.2],\n",
    "         'CRC-986_100215-CW-T_CRC-986_300316-CW-T': [32.6, 0]}\n",
    "\n",
    "for mi, mixtureid in enumerate(mixtureids):\n",
    "    print(mixtureid)\n",
    "    seriesorder = [(70, 0), (70, 80), (50, 100), (30, 120), (20, 130), (10, 140), (5, 145)]\n",
    "    print(seriesorder)\n",
    "    esttflist = []\n",
    "    for s in seriesorder[1:]:\n",
    "        sh, sl = s\n",
    "        esttf = round((tfdict[mixtureid][0]*sh + tfdict[mixtureid][1]*sl)/(sl+sh), 2)\n",
    "        print(esttf)\n",
    "        esttflist.append(esttf)\n",
    "    esttflist =[esttflist] * len(config.methods)\n",
    "    print(esttflist)\n",
    "    resx[mi*len(config.methods):(mi+1)*len(config.methods), :] = esttflist\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "resx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#resx = np.array([rx.values for rx in res1['x']])\n",
    "resx.mean(axis=0)\n",
    "resy = np.array([ry.values for ry in res1['y']])\n",
    "print(resy)\n",
    "res = {m: [] for m in config.methods}\n",
    "for mi, m in enumerate(config.methods):\n",
    "    resmean = np.mean([resy[mi], resy[mi+len(config.methods)], resy[mi+2*len(config.methods)]], axis=0)\n",
    "    resstd = np.std([resy[mi], resy[mi+len(config.methods)], resy[mi+2*len(config.methods)]], axis=0)\n",
    "    #reslabel = m\n",
    "    res[m] = [resmean, resstd, resx.mean(axis=0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res[m]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "plt.figure(figsize=(15, 10))\n",
    "#plt.grid(linewidth=1)\n",
    "plt.grid()\n",
    "for m in config.methods:\n",
    "    #plt.plot(res['freebayes'][2], res['freebayes'][0], c=color_dict['freebayes'], markersize=15, lw=2)\n",
    "    plt.errorbar(res[m][2], res[m][0], xerr = resx.std(axis=0)/np.sqrt(int(resx.shape[0]/(len(config.methods)))), yerr=res[m][1]/int(resx.shape[0]/(len(config.methods))), marker='s',  c=color_dict[m], label=m,  markersize=15, lw=2, fmt='-o')\n",
    "ax = plt.gca()\n",
    "if fixedvar == 'coverage':\n",
    "    plt.gca().invert_xaxis()\n",
    "    xlab='tumor burden'\n",
    "else:\n",
    "    xlab = 'coverage or added noise'\n",
    "hand, labl = ax.get_legend_handles_labels()\n",
    "ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.xlabel(xlab)\n",
    "plt.ylabel(metric.upper()+' score')\n",
    "plt.grid()\n",
    "plt.title(metric.upper() + \" score for {} calling in chr{} with ref {}\".format(mt.upper(), chrom, refname))\n",
    "plt.ylim([0,0.5])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for m in config.methods:\n",
    "    res[m][2] = res[m][2][::-1]\n",
    "    res[m][0] = res[m][0][::-1]\n",
    "    res[m][1] = res[m][1][::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15, 10))\n",
    "#plt.grid(linewidth=1)\n",
    "#res[m][2] = res[m][2][::-1]\n",
    "#res[m][0] = res[m][0][::-1]\n",
    "#res[m][1] = res[m][1][::-1]\n",
    "rankres = []\n",
    "for m in config.methods:\n",
    "    xstart = 0\n",
    "    ystart = 0\n",
    "    sum_all = []\n",
    "    for i in range(len(res[m][2])):\n",
    "        #if i >0:\n",
    "        #print(res[m][2][i], xstart)\n",
    "        #print((res[m][2][i]-xstart) * res[m][0][i] )\n",
    "        #print(((res[m][2][i]-xstart) * (res[m][0][i]-ystart)/2))\n",
    "        sum_all.append(((res[m][2][i]-xstart) * res[m][0][i]) - ((res[m][2][i]-xstart) * (res[m][0][i]-ystart)/2))\n",
    "        xstart = res[m][2][i]\n",
    "        ystart = res[m][0][i]\n",
    "    print(m, np.sum(sum_all) )\n",
    "    rankres.append(np.sum(sum_all))\n",
    "    #plt.plot(res['freebayes'][2], res['freebayes'][0], c=color_dict['freebayes'], markersize=15, lw=2)\n",
    "    #plt.errorbar(res[m][2], res[m][0], xerr = resx.std(axis=0), yerr=res[m][1], marker=config.markers[0], label=m,  markersize=15, lw=2, fmt='-o')\n",
    "    #plt.gca().invert_xaxis()\n",
    "#ax = plt.gca()\n",
    "#xlab = 'coverage or added noise'\n",
    "#xlab='tumor burden'\n",
    "#hand, labl = ax.get_legend_handles_labels()\n",
    "#ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "#plt.xlabel(xlab)\n",
    "#plt.ylabel(metric.upper()+' score')\n",
    "#plt.title(metric.upper() + \" score for {} calling in chr{} with ref {}\".format(mt.upper(), chrom, refname))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patches as mpatches\n",
    "rankres_sort = np.argsort(rankres)[::-1]\n",
    "rankres_sort\n",
    "plt.figure(figsize=(14, 5))\n",
    "method_order =[]\n",
    "rank_order =[]\n",
    "for i in rankres_sort:\n",
    "    method_order.append(config.methods[i])\n",
    "    rank_order.append(rankres[i])\n",
    "plt.bar(method_order, rank_order, color=['b', 'b', 'b', 'b', 'b', 'r', 'r'])\n",
    "blue_patch = mpatches.Patch(color='b', label='DNA methods')\n",
    "red_patch = mpatches.Patch(color='r', label='cfDNA methods')\n",
    "plt.legend(handles=[blue_patch, red_patch])\n",
    "#plt.legend(['DNA methods', 'cfDNA methods'])\n",
    "plt.ylabel('area under AUPRC curve')\n",
    "plt.title('Caller ranking mixtures with decreasing tumor burden')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
