{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Series analysis\n",
    "\n",
    "\n",
    "Analysis updates:\n",
    "- retrieve calls rejected only because VAF<0.01 (bcbio lower acceptable threshold) for Mutect2, Strelka2 and Vardict\n",
    "- correct for germline mutations using GATK haplotype calls\n",
    "- get PR curves for sinvict by encoding the 6 files as thresholds. Assumption: linear filters\n",
    "- plotting: PR curves stops at 10e-2 on the left\n",
    "- ground truths:\n",
    "\n",
    "    1) Consensus: build using\n",
    "        majority of 5/8 callers for SNV and 3/5 callers for INDELS\n",
    "\n",
    "\n",
    "    2) Ranked mutations: metascore built using\n",
    "        weigthed sum of normalised scores between 0 and 1 for each caller\n",
    "        with weigths = inversially proportional to number of calls made by caller (if a caller calls few mutations higher weight, it many calls low weight)\n",
    "        threshold = 1/ncallers\n",
    "        interpretation: if 1 caller is sure (score = 1) of calling this position, add it to GT\n",
    "        interpretation: if 2 callers are quite sure of calling this position (score > 0,5 each), add it to GT\n",
    "        \n",
    "- integrate VAF approx\n",
    "\n",
    "    1) mixture with VAF instead of tumor burden\n",
    "    \n",
    "    2) correct for mutations non present in diluted samples (vaf = 0)\n",
    "    \n",
    "    3) pool patients together using VAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "import warnings\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, average_precision_score\n",
    "warnings.filterwarnings('ignore')\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "# set working directory\n",
    "if not os.getcwd().endswith('cfdna_snv_benchmark'):\n",
    "    os.chdir('../')\n",
    "print('Current working directory: {}'.format(os.getcwd()))\n",
    "\n",
    "from utils.config import Config\n",
    "from utils.viz import *\n",
    "from utils.table import *\n",
    "from utils.metrics import *\n",
    "from utils.calltable import *\n",
    "from utils.calltableseries import *\n",
    "from utils.groundtruth import *\n",
    "from utils.metricsseries import *\n",
    "from utils.venn import venn6, get_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Config and Display paramaters\n",
    "\n",
    "config = Config(\"config/\", \"config_viz.yaml\")\n",
    "set_display_params(config)\n",
    "print(config.methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chomosome\n",
    "\n",
    "#mixtureids =  ['CRC-1014_180816-CW-T_CRC-1014_090516-CW-T', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T', 'CRC-123_310715-CW-T_CRC-123_121115-CW-T']\n",
    "mixtureid = 'BRA-412_240820-CW-T_BRA-412_060220-CW-T'\n",
    "#mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "#mixtureid = 'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T'\n",
    "#mixtureid = 'CRC-123_310715-CW-T_CRC-123_121115-CW-T'\n",
    "reload = False\n",
    "save = True\n",
    "fixedvars=['coverage', 'ctdna']\n",
    "filterparam = 'all'\n",
    "\n",
    "markers = ['o', '^', 'X']\n",
    "linestyles = ['-', '-', '-']\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "\n",
    "muttypes = ['snv', 'indel']\n",
    "metrics = ['auprc', 'precision', 'recall']\n",
    "\n",
    "chrom = '22'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part I: (1) Load/Generate call tables, (2) Generate Ground truths and (3) Compute/Save metrics per patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixedvar = 'ctdna'\n",
    "#for fixedvar in fixedvars:\n",
    "if fixedvar == 'coverage':\n",
    "    seriesorder = [(70, 0), (70, 80), (50, 100), (30, 120), (20, 130), (10, 140), (5, 145)]\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    seriesorder = [(70, 0), (70, 80), (70, 180)]\n",
    "    xaxis = 'coverage'\n",
    "#for mixtureid in mixtureids:\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "aux_all = []\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom, muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='mixture_wgs')\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom, muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='mixture_wgs')\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom, muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='mixture_wgs')\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(plasmasample)\n",
    "healthysample = '_'.join(mixtureid.split('_')[2:])\n",
    "print(healthysample)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "dilutionseries = aux.T[['mixture_chr'+str(chrom) + '_'+ '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "#for muttype in muttypes:\n",
    "muttype = 'snv'\n",
    "refsample = 'undiluted'\n",
    "if muttype == 'snv':\n",
    "    gtm = 5\n",
    "else:  # elif muttype == 'indel':\n",
    "    gtm = 3\n",
    "print(max(aux['tf']))\n",
    "#if mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "#    gtm = 3\n",
    "#    refsample = 'tissue'\n",
    "#    calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "#                                        matchedtissuepath=os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-T1W', 'calls', 'NCC_CRC-986_100215-T1W_snv_calls_PASS.csv'))\n",
    "#else:\n",
    "#    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype, \n",
    "#                                                matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict', 'cfsnv'])\n",
    "if mixtureid !=  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                            matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict'])\n",
    "else:\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict']) # missing varnet\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wgs')\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall', ground_truth_method=gtm,\n",
    "                                       refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wgs')\n",
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wgs')\n",
    "results_maxrecallatlast0_05precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_01precision', ground_truth_method=gtm,\n",
    "                                      refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wgs')\n",
    "results_maxrecallatlast0_05precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_05precision', ground_truth_method=gtm,\n",
    "                                  refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wgs')\n",
    "results_maxrecallatlast0_10precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_10precision', ground_truth_method=gtm,\n",
    "                                  refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis, diltype='mixture_wgs')\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method=gtm,\n",
    "                    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save, diltype='mixture_wgs')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part II: (4) Load back metric results and plot combined metric plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for fixedvar in fixedvars:\n",
    "    print(fixedvar)\n",
    "    if fixedvar == 'coverage':\n",
    "        xaxis = 'tumor burden'\n",
    "    elif fixedvar == 'ctdna':\n",
    "        xaxis = 'coverage'\n",
    "    #for mt in muttypes:\n",
    "    mt = 'snv'\n",
    "    if mt == 'snv':\n",
    "        gtm = 4\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    else:  # elif mt == 'indel':\n",
    "        gtm = 2\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    print(refname)\n",
    "    # for metric in metrics:\n",
    "    metric = 'auprc'\n",
    "    # load results tables\n",
    "    restables = {'snv': [], 'indel': []}\n",
    "    mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "    if mixtureid == 'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        gtm = 3\n",
    "        refname = 'intissuesamplebyatleast'+str(gtm)+'callers'\n",
    "    else:\n",
    "        gtm = 4\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "    print(mixtureid, plasmasample)\n",
    "    xa = xaxis if xaxis != 'tumor burden' else 'tb'\n",
    "    print(xa)\n",
    "    restable = pd.read_csv(os.path.join(*config.mixturefolderwholegenome, 'mixtures_allchr', 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar+'_'+ xa +'.csv'), index_col=0)\n",
    "    #restable = pd.read_csv(os.path.join(*config.mixturefolderwholegenome, 'mixtures_chr22_wgs', 'mixtures_chr22_'+mixtureid, 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar +'.csv'), index_col=0)\n",
    "    restable['plasma sample'] = plasmasample\n",
    "    restables[mt].append(restable)\n",
    "    restables[mt] = pd.concat(restables[mt])\n",
    "    res1 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=False, save=False, diltype='mixtures_wgs')\n",
    "    #plt.grid(linewidth=1)\n",
    "    plt.ylim([0, 0.4])\n",
    "    #plt.xlim([40, 250])\n",
    "    #plt.xlim([20, 0])\n",
    "    #res2 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=True, save=True, diltype='mixtures_wgs')\n",
    "    #res3 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True, diltype='mixtures_wgs')\n",
    "    #res4 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=True, save=True, diltype='mixtures_wgs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
