{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Series analysis with respect to VAF\n",
    "\n",
    "\n",
    "Analysis updates:\n",
    "- retrieve calls rejected only because VAF<0.01 (bcbio lower acceptable threshold) for Mutect2, Strelka2 and Vardict\n",
    "- correct for germline mutations using GATK haplotype calls\n",
    "- get PR curves for sinvict by encoding the 6 files as thresholds. Assumption: linear filters\n",
    "- plotting: PR curves stops at 10e-2 on the left\n",
    "- ground truths:\n",
    "\n",
    "    1) Consensus: build using\n",
    "        majority of 5/8 callers for SNV and 3/5 callers for INDELS\n",
    "\n",
    "\n",
    "    2) Ranked mutations: metascore built using\n",
    "        weigthed sum of normalised scores between 0 and 1 for each caller\n",
    "        with weigths = inversially proportional to number of calls made by caller (if a caller calls few mutations higher weight, it many calls low weight)\n",
    "        threshold = 1/ncallers\n",
    "        interpretation: if 1 caller is sure (score = 1) of calling this position, add it to GT\n",
    "        interpretation: if 2 callers are quite sure of calling this position (score > 0,5 each), add it to GT\n",
    "        \n",
    "- integrate VAF approx\n",
    "\n",
    "    1) mixture with VAF instead of tumor burden\n",
    "    \n",
    "    2) correct for mutations non present in diluted samples (vaf = 0)\n",
    "    \n",
    "    3) pool patients together using VAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "import warnings\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, average_precision_score\n",
    "warnings.filterwarnings('ignore')\n",
    "from sklearn.metrics import confusion_matrix\n",
    "from tqdm.notebook import trange, tqdm\n",
    "\n",
    "# set working directory\n",
    "if not os.getcwd().endswith('cfdna_snv_benchmark'):\n",
    "    os.chdir('../')\n",
    "print('Current working directory: {}'.format(os.getcwd()))\n",
    "\n",
    "from utils.config import Config\n",
    "from utils.viz import *\n",
    "from utils.table import *\n",
    "from utils.metrics import *\n",
    "from utils.calltable import *\n",
    "from utils.calltableseries import *\n",
    "from utils.groundtruth import *\n",
    "from utils.metricsseries import *\n",
    "from utils.venn import venn6, get_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Config and Display paramaters\n",
    "\n",
    "config = Config(\"config/\", \"config_viz.yaml\")\n",
    "set_display_params(config)\n",
    "print(config.methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chomosome\n",
    "\n",
    "mixtureids =  ['CRC-123_310715-CW-T_CRC-123_121115-CW-T', 'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T', 'CRC-986_100215-CW-T_CRC-986_300316-CW-T', ]\n",
    "mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "#mixtureid = 'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T'\n",
    "#mixtureid = 'CRC-123_310715-CW-T_CRC-123_121115-CW-T'\n",
    "reload = False\n",
    "save = False\n",
    "fixedvars=['coverage', 'ctdna']\n",
    "filterparam = 'all'\n",
    "\n",
    "markers = ['o', '^', 'X']\n",
    "linestyles = ['-', '-', '-']\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "\n",
    "muttypes = ['snv', 'indel']\n",
    "metrics = ['auprc', 'precision', 'recall']\n",
    "\n",
    "chrom = 'all'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part I: Load/Generate call tables, Generate Ground truths and Compute metrics per patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixedvar = 'coverage'\n",
    "#for fixedvar in fixedvars:\n",
    "if fixedvar == 'coverage':\n",
    "    seriesorder = [(70, 0), (70, 80), (50, 100), (30, 120), (20, 130), (10, 140), (5, 145)]\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    seriesorder = [(70, 0), (70, 80), (70, 180)]\n",
    "    xaxis = 'coverage'\n",
    "calltables_snv = pd.DataFrame()\n",
    "#for mixtureid in mixtureids:\n",
    "mixtureid = mixtureid\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "if mixtureid ==  'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T':\n",
    "    chroms = [str(c) for c in range(1,23) if c != 2 and c!=6 and c !=17 and c!=19 and c!=20 and c!=21]\n",
    "    chroms = [str(c) for c in range(1,9) if c != 2 and c!=6]\n",
    "elif mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    chroms = [str(c) for c in range(1,23) if c !=1 and c!= 2 and c !=8 and c!=20 and c!=21 and c!=22] \n",
    "else:\n",
    "    chroms = [str(c) for c in range(1,23) if c !=6 and c!=19 and c!=20]  # c !=1 and c!= 2 and\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "aux_all = []\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chroms, muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='mixture', concat='tf')\n",
    "#calltable_indel, aux = get_calltableseries(config, mixtureid, chroms, muttype='indel', filterparam=filterparam, reload=reload, save=save,  diltype='mixture', concat='vaf')\n",
    "#calltable_snp, aux = get_calltableseries(config, mixtureid, chroms, muttype='snp', filterparam=filterparam, reload=reload, save=save,  diltype='mixture',  concat='vaf')\n",
    "#print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(plasmasample)\n",
    "healthysample = '_'.join(mixtureid.split('_')[2:])\n",
    "print(healthysample)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "#calltables['indel'].append(calltable_indel)\n",
    "#calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "#calltables['indel'] = pd.concat(calltables['indel'])\n",
    "#calltables['snp'] = pd.concat(calltables['snp'])\n",
    "dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "#for muttype in muttypes:\n",
    "muttype = 'snv'\n",
    "refsample = 'undiluted'\n",
    "if muttype == 'snv':\n",
    "    gtm = 5\n",
    "else:  # elif muttype == 'indel':\n",
    "    gtm = 3\n",
    "print(max(aux['tf']))\n",
    "if mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                            matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "else:\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                            matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict'])\n",
    "#results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "#                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis='tf')\n",
    "calltables_snv = pd.concat([calltables_snv, calltables['snv']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux['tf']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = calltablesseries[['{:.2f}_{}'.format(aux['tf'][0], m) for m in config.methods] + ['{:.2f}_{}_vaf'.format(aux['tf'][0], m) for m in config.methods] + ['truth']]\n",
    "A['median_vaf'] = A[['{:.2f}_{}_vaf'.format(aux['tf'][0], m) for m in config.methods]].median(axis=1, skipna=True)\n",
    "A = A[['{:.2f}_{}'.format(aux['tf'][0], m) for m in config.methods] + ['median_vaf', 'truth']]\n",
    "A.dropna(how='all', subset=A.columns[:-1], inplace=True)\n",
    "print(A['truth'].value_counts())\n",
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "B = A[(A['truth'] == True) | (A[['{:.2f}_{}'.format(aux['tf'][0], m) for m in config.methods]].sum(axis=1) > 0)]\n",
    "B.reset_index(inplace=True)\n",
    "B.set_index(['chrom_pos_ref_alt', 'median_vaf', 'truth'], inplace=True)\n",
    "B.columns = [m for m in config.methods]\n",
    "B = B.stack().reset_index()\n",
    "B.columns = ['chrom_pos_ref_alt', 'median_vaf', 'truth', 'caller', 'value']\n",
    "B.loc[(B['value'] == True) & (B['truth'] == True), 'value'] = 'TP'\n",
    "B.loc[(B['value'] == True) & (B['truth'] == False), 'value'] = 'FP'\n",
    "B.loc[(B['value'] == False) & (B['truth'] == True), 'value'] = 'FN'\n",
    "B.loc[(B['value'] == False) & (B['truth'] == False), 'value'] = 'TN'\n",
    "B = B[B['value'] != 'TN']\n",
    "B"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for m in config.methods:\n",
    "    plt.figure(figsize=(15, 5))\n",
    "    sns.histplot(x='median_vaf', hue='value', multiple=\"stack\", hue_order=['FP', 'FN', 'TP'], data=B[B['caller'] == m], binwidth=0.01) #, log_scale=True)\n",
    "    plt.title(m)\n",
    "    plt.xlim([0, 0.4])\n",
    "    plt.ylim([0, 100])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux['tf'][2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "i = 3\n",
    "C = calltablesseries[['{:.2f}_{}_score'.format(aux['tf'][i], m) for m in config.methods] + ['{:.2f}_{}_vaf'.format(aux['tf'][i], m) for m in config.methods] + ['truth']]\n",
    "C['median_vaf'] = C[['{:.2f}_{}_vaf'.format(aux['tf'][i], m) for m in config.methods]].median(axis=1, skipna=True)\n",
    "C = C[['{:.2f}_{}_score'.format(aux['tf'][i], m) for m in config.methods] + ['median_vaf', 'truth']]\n",
    "print(aux['tf'][i])\n",
    "print(C.shape[0])\n",
    "C.dropna(how='all', subset=C.columns[:-1], inplace=True)\n",
    "print(C.shape[0])\n",
    "print(C['truth'].value_counts())\n",
    "C.columns = [m for m in config.methods] + ['median_vaf', 'truth']\n",
    "C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[['freebayes', 'truth']].dropna()['truth'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(22, 22))\n",
    "for mi, m in enumerate(config.methods):\n",
    "    plt.subplot(3, 3, mi+1)\n",
    "    sns.histplot(x='median_vaf', y=m, data=C[C['truth'] == False], binwidth=(0.01, 0.01), legend=False, color='grey', pmax=1)\n",
    "    sns.histplot(x='median_vaf', y=m, data=C[C['truth'] == True], binwidth=(0.01, 0.01), legend=False, color=color_dict[m], pmax=0.8)\n",
    "    plt.title('{} _ nTP={}'.format(m, C[[m, 'truth']].dropna()['truth'].sum()))\n",
    "    #plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    plt.xlim([0, 1.01])\n",
    "    plt.ylim([0, 1.01])\n",
    "    plt.axvline(x=aux['tf'][i]/100/2, c='k', ls='--')\n",
    "plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'vafhist_986_{:.0f}tf_150x.svg'.format(aux['tf'][2])), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mi, m in enumerate(config.methods):\n",
    "    plt.figure(figsize=(8, 8))\n",
    "    # plt.subplot(3, 3, mi+1)\n",
    "    sns.histplot(x='median_vaf', y=m, data=C[C['truth'] == False], binwidth=(0.01, 0.01), legend=False, color='grey', pmax=1)\n",
    "    sns.histplot(x='median_vaf', y=m, data=C[C['truth'] == True], binwidth=(0.01, 0.01), legend=False, color=color_dict[m], pmax=0.8)\n",
    "    plt.title('{} _ nTP={}'.format(m, C[[m, 'truth']].dropna()['truth'].sum()))\n",
    "    #plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    plt.xlim([0, 1.01])\n",
    "    plt.ylim([0, 1.01])\n",
    "    plt.grid()\n",
    "    plt.axvline(x=aux['tf'][i]/100/2, c='k', ls='--')\n",
    "    plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'vafhist_986_{:.0f}tf_150x_{}.svg'.format(aux['tf'][2], m)), bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D['median_vaf']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "D = C[['median_vaf', 'truth']]\n",
    "D['ideal caller'] = 0\n",
    "D.loc[D['truth'] == False, 'ideal caller'] = [random.uniform(0, 0.4) for _ in range(D[D['truth'] == False].shape[0])]\n",
    "D.loc[D['truth'] == True, 'ideal caller'] = [random.uniform(0.6, 1) for _ in range(D[D['truth'] == True].shape[0])]\n",
    "D['random caller'] = [random.uniform(0, 1) for _ in range(C.shape[0])]\n",
    "D['good caller for unguided analysis'] = 0\n",
    "D.loc[D['truth'] == False, 'good caller for unguided analysis'] = [random.uniform(0, 0.5) for _ in range(D[D['truth'] == False].shape[0])]\n",
    "D.loc[D['truth'] == True, 'good caller for unguided analysis'] = [random.uniform(0.3, 1) for _ in range(D[D['truth'] == True].shape[0])]\n",
    "for idx in D.loc[(D['truth'] == False) & (D['median_vaf'] > 0.025)].index[:400]:\n",
    "    D.loc[idx, 'good caller for unguided analysis'] = random.uniform(0.5, 1)\n",
    "D['good caller for guided analysis'] = 0\n",
    "D.loc[D['truth'] == False, 'good caller for guided analysis'] = [random.uniform(0, 0.5) for _ in range(D[D['truth'] == False].shape[0])]\n",
    "D.loc[D['truth'] == True, 'good caller for guided analysis'] = [random.uniform(0.6, 1) for _ in range(D[D['truth'] == True].shape[0])]\n",
    "for idx in D.loc[(D['truth'] == False) & (D['median_vaf'] > 0.02) & (D['median_vaf'] < 0.5)].index[:4000]:\n",
    "    D.loc[idx, 'good caller for guided analysis'] = random.uniform(0.5, 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# simulation \n",
    "for m in ['ideal caller', 'random caller', 'good caller for unguided analysis', 'good caller for guided analysis']:\n",
    "    #print(m.replace(' ', ''))\n",
    "    plt.figure(figsize=(12, 12))\n",
    "    # plt.subplot(3, 3, mi+1)\n",
    "    sns.histplot(x='median_vaf', y=m, data=D[D['truth'] == False], binwidth=(0.01, 0.01), legend=False, color='grey', pmax=1)\n",
    "    sns.histplot(x='median_vaf', y=m, data=D[D['truth'] == True], binwidth=(0.01, 0.01), legend=False, color='darkred', pmax=0.8)\n",
    "    plt.axhline(y=0.5, c='r', lw=3)\n",
    "    plt.title('{} _ nTP={}'.format(m, D[[m, 'truth']].dropna()['truth'].sum()))\n",
    "    #plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    plt.xlim([0, 1.01])\n",
    "    plt.ylim([0, 1.01])\n",
    "    plt.grid()\n",
    "    plt.axvline(x=aux['tf'][i]/100/2, c='k', ls='--')\n",
    "    plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'vafhist_986_{:.0f}tf_150x_{}.svg'.format(aux['tf'][2], m.replace(' ', ''))), bbox_inches='tight')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for tf in aux['tf'].values:\n",
    "    calltablesseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fixedvar = 'coverage'\n",
    "#for fixedvar in fixedvars:\n",
    "if fixedvar == 'coverage':\n",
    "    seriesorder = [(70, 0), (70, 80), (50, 100), (30, 120), (20, 130), (10, 140), (5, 145)]\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    seriesorder = [(70, 0), (70, 80), (70, 180)]\n",
    "    xaxis = 'coverage'\n",
    "calltables_snv = pd.DataFrame()\n",
    "#for mixtureid in mixtureids:\n",
    "mixtureid = mixtureid\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "if mixtureid ==  'CRC-1014_180816-CW-T_CRC-1014_090516-CW-T':\n",
    "    chroms = [str(c) for c in range(1,23) if c != 2 and c!=6 and c !=17 and c!=19 and c!=20 and c!=21]\n",
    "    chroms = [str(c) for c in range(1,9) if c != 2 and c!=6]\n",
    "elif mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    chroms = [str(c) for c in range(1,23) if c !=1 and c!= 2 and c !=8 and c!=20 and c!=21 and c!=22] \n",
    "else:\n",
    "    chroms = [str(c) for c in range(1,23) if c !=6 and c!=19 and c!=20]  # c !=1 and c!= 2 and\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "aux_all = []\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chroms, muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='mixture', concat='vaf')\n",
    "#calltable_indel, aux = get_calltableseries(config, mixtureid, chroms, muttype='indel', filterparam=filterparam, reload=reload, save=save,  diltype='mixture', concat='vaf')\n",
    "#calltable_snp, aux = get_calltableseries(config, mixtureid, chroms, muttype='snp', filterparam=filterparam, reload=reload, save=save,  diltype='mixture',  concat='vaf')\n",
    "#print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "print(plasmasample)\n",
    "healthysample = '_'.join(mixtureid.split('_')[2:])\n",
    "print(healthysample)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "#calltables['indel'].append(calltable_indel)\n",
    "#calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "#calltables['indel'] = pd.concat(calltables['indel'])\n",
    "#calltables['snp'] = pd.concat(calltables['snp'])\n",
    "dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "#for muttype in muttypes:\n",
    "muttype = 'snv'\n",
    "refsample = 'undiluted'\n",
    "if muttype == 'snv':\n",
    "    gtm = 5\n",
    "else:  # elif muttype == 'indel':\n",
    "    gtm = 3\n",
    "print(max(aux['tf']))\n",
    "#if mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    #gtm = 3\n",
    "    #refsample = 'tissue'\n",
    "    #calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "    #                                    matchedtissuepath=os.path.join('data', 'matchedtissue', 'NCC_CRC-986_100215-T1W', 'calls', 'NCC_CRC-986_100215-T1W_snv_calls_PASS_exome.csv'))\n",
    "    #calltablesseries  = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method='tissue', muttype=muttype,\n",
    "    #                                    matchedtissuepath=os.path.join('data', 'matchedtissue_ultradeep', '986_100215_T1-E', 'calls', '986_100215_T1-E_snv_calls_all.csv'))\n",
    "#else:\n",
    "    #calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "    #                                        matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict'])\n",
    "if mixtureid ==  'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                            matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'varnet', 'abemus', 'sinvict'])\n",
    "else:\n",
    "    calltablesseries = generate_groundtruth(config, calltables[muttype], aux['tf'], ground_truth_method=gtm, muttype=muttype,\n",
    "                                            matchedtissuepath=None, methods=['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'abemus', 'sinvict'])\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method=gtm,\n",
    "                                 refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis='vaf')\n",
    "calltables_snv = pd.concat([calltables_snv, calltables['snv']])\n",
    "#results_maxf1_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1', ground_truth_method=gtm,\n",
    "#                                     refsample=refsample, muttype=muttype, methods=config.methods, fixedvar=fixedvar, xaxis=xaxis)\n",
    "#figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method=gtm,\n",
    "#                    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltables_snv[[c for c in list(calltables_snv.columns) if 'vaf' in c ]].dropna()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure(figsize=(15, 20))\n",
    "for i in calltables['snv'].index.unique()[:200]:\n",
    "    #print(i)\n",
    "    aux = calltables['snv'].loc[i]\n",
    "    if isinstance(aux, pd.DataFrame):\n",
    "       # try:\n",
    "        aux = aux.sort_values(by='sampletf')\n",
    "        #except:\n",
    "        #    print(aux)\n",
    "        #    stop\n",
    "        plt.plot(aux['sampletf'], aux['median_vaf'], label=i)\n",
    "plt.gca().invert_xaxis()\n",
    "#plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tflist = np.sort(calltables['snv']['sampletf'].unique())\n",
    "print(tflist)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vafevolution = pd.DataFrame(columns=tflist)\n",
    "#plt.figure(figsize=(15, 20))\n",
    "for i in calltables['snv'].index.unique()[:1000]:\n",
    "    aux = calltables['snv'].loc[i]\n",
    "    if isinstance(aux, pd.DataFrame):\n",
    "        aux = aux.sort_values(by='sampletf')\n",
    "        aux = aux[['sampletf', 'median_vaf']].reset_index().set_index('sampletf').reindex(tflist)\n",
    "        vafevolution.loc[i] = aux['median_vaf'].values\n",
    "#plt.gca().invert_xaxis()\n",
    "#plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "#plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vafevolution.dropna().T.plot()\n",
    "plt.gca().invert_xaxis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from statannot import add_stat_annotation\n",
    "df = calltables['snv'][['sampletf', 'median_vaf']].sort_values(by='sampletf')\n",
    "df = df[df['median_vaf'] > 0]\n",
    "sns.set(rc={\"figure.figsize\":(10, 5)}) \n",
    "ax = sns.boxplot(x=\"sampletf\", y=\"median_vaf\", data=df)\n",
    "plt.yscale(\"log\")\n",
    "#plt.ylim([0.001, 0.045])\n",
    "order = [1.3959, 2.79245, 5.5863, 8.37985, 11.7316, 13.96675, 19.55335, 41.8999]\n",
    "add_stat_annotation(ax, data=df, x='sampletf', y='median_vaf', order=order,\n",
    "                    box_pairs=[(1.3959, 2.79245), (2.79245, 5.5863), (5.5863, 8.37985), (8.37985, 11.7316), (11.7316, 13.96675), (13.96675, 19.55335), (19.55335, 41.8999)],\n",
    "                    test='t-test_welch', text_format='star', loc='outside', verbose=2)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df.groupby('sampletf').median().plot()\n",
    "plt.gca().invert_xaxis()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.xticks()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.009, 0.008, 0.007, 0.006, 0.005, 0.004, 0.003, 0.002, 0.001]\n",
    "vafranges = [1., 0.4, 0.3, 0.2, 0.15, 0.1, 0.075, 0.05, 0.04, 0.03, 0.02, 0.01, 0.001]\n",
    "vafranges = [1., 0.4, 0.3, 0.2, 0.1, 0.09, 0.08, 0.07, 0.06, 0.05, 0.04, 0.03, 0.02, 0.001]\n",
    "\n",
    "vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "cond = True\n",
    "for i in range(len(mixtureids)):\n",
    "    cond = cond & (calltables_snv['sampletf'] != np.sort(calltables_snv['sampletf'].unique())[-i])\n",
    "calltable_snv = calltables_snv[cond]\n",
    "# calltable_snv = calltables_snv[(calltables_snv['sampletf'] != tfmax for np.sort(calltables_snv['sampletf'].unique())[-3:]]\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltable_snv[(calltable_snv['median_vaf'] >= vafranges[vi-1]) & (calltable_snv['median_vaf'] < vafranges[vi])]\n",
    "            if method == 'varnet':\n",
    "                aux = aux[aux['sampleid'] == 'CRC-986_100215-CW-T_CRC-986_300316-CW-T']\n",
    "            print(aux['truth'].sum(), aux.shape[0])\n",
    "            if not aux.empty and aux[method + '_score'].isna().sum() != aux.shape[0]:\n",
    "                #print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "                precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "                f1list = 2*(precision * recall)/(precision + recall)\n",
    "                #print(len(f1list), len(precision))\n",
    "                #max(f1list)\n",
    "                #print(precision, recall, thresholds)\n",
    "                #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #y.append(max(f1list))\n",
    "                y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                #y.append(roc_auc_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                \n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "plt.figure(figsize=(15,8))\n",
    "for k,v in res.items():\n",
    "    plt.plot(v['x'], v['y'], marker='s', label=k, color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "#plt.ylim([0.4, 1])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calltables_snv[['sampleid', 'sampletf']]\n",
    "np.sort(calltables_snv['sampletf'].unique())[-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable_snv.median_vaf.describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.009, 0.008, 0.007, 0.006, 0.005, 0.004, 0.003, 0.002, 0.001]\n",
    "vafranges = [1., 0.4, 0.3, 0.2, 0.15, 0.1, 0.075, 0.05, 0.04, 0.03, 0.02, 0.01, 0.001]\n",
    "vafranges = [0.3, 0.2, 0.1, 0.09, 0.08, 0.07, 0.06, 0.05, 0.04, 0.03, 0.02, 0.005]\n",
    "\n",
    "vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "cond = True\n",
    "i = 4\n",
    "cond = cond & (calltables_snv['sampletf'] == np.sort(calltables_snv['sampletf'].unique())[-i])\n",
    "print(np.sort(calltables_snv['sampletf'].unique())[-i])\n",
    "#for i in range(len(mixtureids)):\n",
    "#    cond = cond & (calltables_snv['sampletf'] != np.sort(calltables_snv['sampletf'].unique())[-i])\n",
    "calltable_snv = calltables_snv[cond]\n",
    "out, bins = pd.qcut(calltables_snv[calltables_snv['truth']==True]['median_vaf'], q=10, retbins=True)\n",
    "#out, bins = pd.qcut(calltables_snv[calltables_snv['median_vaf']>0]['median_vaf'], q=10, retbins=True)\n",
    "vafranges = list(bins)\n",
    "#vafranges = [0.0, 0.0286567622950819, 0.0543806646525679, 0.07850000000000001, 0.104, 0.2303213957759411, 0.975]\n",
    "print(vafranges)\n",
    "# calltable_snv = calltables_snv[(calltables_snv['sampletf'] != tfmax for np.sort(calltables_snv['sampletf'].unique())[-3:]]\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    xbaseline, ybaseline = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltable_snv[(calltable_snv['median_vaf'] >= vafranges[vi-1]) & (calltable_snv['median_vaf'] < vafranges[vi])]\n",
    "            #if method == 'varnet':\n",
    "                #aux = aux[aux['sampleid'].str.contains('CRC-986_100215-CW-T')]\n",
    "            print(aux['truth'].sum(), aux.shape[0])\n",
    "            if not aux.empty and aux[method + '_score'].isna().sum() != aux.shape[0]:\n",
    "                #print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "                precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "                f1list = 2*(precision * recall)/(precision + recall)\n",
    "                #print(len(f1list), len(precision))\n",
    "                #max(f1list)\n",
    "                #print(precision, recall, thresholds)\n",
    "                #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                xbaseline.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #y.append(max(f1list))\n",
    "                #idxa = min(range(len(precision)), key=lambda i: abs(precision[i]-0.7))\n",
    "                #if not  precision[idxa] < 0.65 and precision[idxa] != 1 :\n",
    "                #    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #    y.append(recall[idxa])\n",
    "                #if np.nanmax(f1list) >= 0.1 and aux[method + '_score'].isna().sum() != aux.shape[0]:\n",
    "                #    print(method, np.nanmax(f1list), vafranges[vi-1], vafranges[vi])\n",
    "                #    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #    y.append(recall[list(f1list).index(np.nanmax(f1list))])\n",
    "                y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                ybaseline.append(aux['truth'].sum()/aux.shape[0])\n",
    "                #p = np.where(precision >= 0.01, 1, 0)\n",
    "                #rlist = recall * p\n",
    "                #y.append(rlist.max())\n",
    "                #y.append(roc_auc_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "fig, ax = plt.subplots(1, 1, figsize=(20,8))\n",
    "for k,v in res.items():\n",
    "    plt.plot(v['x'], v['y'], marker='o', label=k, color=color_dict[k])\n",
    "plt.plot(xbaseline, ybaseline, marker='o', ls='--', color='k', label='baseline')\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "ax.set_xticks(np.array(vafranges).round(2))\n",
    "import matplotlib.ticker\n",
    "ax.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "ax.get_xaxis().set_minor_formatter(matplotlib.ticker.NullFormatter())\n",
    "xlim = ax.set_xlim(1., 0.01)\n",
    "plt.axvline(x=np.sort(calltables_snv['sampletf'].unique())[-i]/2/100, c='k', ls='--', label='expected cancer VAF = ctDNA/2')\n",
    "#plt.ylim([0.4, 1])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ybaseline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.009, 0.008, 0.007, 0.006, 0.005, 0.004, 0.003, 0.002, 0.001]\n",
    "vafranges = [1., 0.4, 0.3, 0.2, 0.15, 0.1, 0.075, 0.05, 0.04, 0.03, 0.02, 0.01, 0.001]\n",
    "vafranges = [0.3, 0.2, 0.1, 0.09, 0.08, 0.07, 0.06, 0.05, 0.04, 0.03, 0.02, 0.005]\n",
    "\n",
    "vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "cond = True\n",
    "i = 4\n",
    "cond = cond & (calltables_snv['sampletf'] == np.sort(calltables_snv['sampletf'].unique())[-i])\n",
    "print(np.sort(calltables_snv['sampletf'].unique())[-i])\n",
    "#for i in range(len(mixtureids)):\n",
    "#    cond = cond & (calltables_snv['sampletf'] != np.sort(calltables_snv['sampletf'].unique())[-i])\n",
    "calltable_snv = calltables_snv[cond]\n",
    "out, bins = pd.qcut(calltables_snv[calltables_snv['truth']==True]['median_vaf'], q=10, retbins=True)\n",
    "vafranges = list(bins)\n",
    "print(vafranges)\n",
    "# calltable_snv = calltables_snv[(calltables_snv['sampletf'] != tfmax for np.sort(calltables_snv['sampletf'].unique())[-3:]]\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltable_snv[(calltable_snv['median_vaf'] >= vafranges[vi-1]) & (calltable_snv['median_vaf'] < vafranges[vi])]\n",
    "            #if method == 'varnet':\n",
    "                #aux = aux[aux['sampleid'].str.contains('CRC-986_100215-CW-T')]\n",
    "            print(aux['truth'].sum(), aux.shape[0])\n",
    "            if not aux.empty and aux[method + '_score'].isna().sum() != aux.shape[0]:\n",
    "                #print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "                precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "                f1list = 2*(precision * recall)/(precision + recall)\n",
    "                #print(len(f1list), len(precision))\n",
    "                #max(f1list)\n",
    "                #print(precision, recall, thresholds)\n",
    "                #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #y.append(max(f1list))\n",
    "                #idxa = min(range(len(precision)), key=lambda i: abs(precision[i]-0.7))\n",
    "                #if not  precision[idxa] < 0.65 and precision[idxa] != 1 :\n",
    "                #    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #    y.append(recall[idxa])\n",
    "                #if np.nanmax(f1list) >= 0.05 and aux[method + '_score'].isna().sum() != aux.shape[0]:\n",
    "                #    print(method, np.nanmax(f1list), vafranges[vi-1], vafranges[vi])\n",
    "                #    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #    y.append(recall[list(f1list).index(np.nanmax(f1list))])\n",
    "                #y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                p = np.where(precision >= 0.05, 1, 0)\n",
    "                rlist = recall * p\n",
    "                if np.max(rlist) > 0:\n",
    "                    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                    y.append(np.max(rlist))\n",
    "                #y.append(recall_score(aux['truth'], aux[method].fillna(0)))\n",
    "                \n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "fig, ax = plt.subplots(1, 1, figsize=(20,8))\n",
    "for k,v in res.items():\n",
    "    plt.plot(v['x'], v['y'], marker='o', label=k, color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "ax.set_xticks(np.array(vafranges).round(2))\n",
    "import matplotlib.ticker\n",
    "ax.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "ax.get_xaxis().set_minor_formatter(matplotlib.ticker.NullFormatter())\n",
    "xlim = ax.set_xlim(1., 0.01)\n",
    "plt.axvline(x=np.sort(calltables_snv['sampletf'].unique())[-i]/2/100, c='k', ls='--', label='expected cancer VAF = ctDNA/2')\n",
    "#plt.ylim([0.4, 1])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sort(calltables_snv['sampletf'].unique())[-2]/2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.array(vafranges).round(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bins"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "[500*np.round(bins/500, 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltables_snv_vaf = calltables_snv[calltables_snv['sampletf'] != calltables_snv['sampletf'].unique().max()]\n",
    "out, bins = pd.qcut(calltables_snv_vaf[calltables_snv_vaf['truth']==True]['median_totcov'], q=5, retbins=True)\n",
    "vafranges = list(bins)# + [1]\n",
    "print(vafranges)\n",
    "#np.sort(pd.qcut(calltable_snv[calltable_snv['truth']==True]['median_vaf'], q=10).unique())\n",
    "#vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.0075, 0.005, 0.0025, 0.001]\n",
    "#vafranges = [1., 0.016, 0.012, 0.0093, 0.0073, 0.0063, 0.0056, 0.0051, 0.0045, 0.0038, 0]\n",
    "#vafranges = [.195, 0.0894, 0.0758, 0.0597, 0.0495, 0.0429, 0.0342, 0.0281, 0.024, 0.0172, 0.0052]\n",
    "#vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltables_snv_vaf[(calltables_snv_vaf['median_totcov'] >= vafranges[vi-1]) & (calltables_snv_vaf['median_totcov'] < vafranges[vi])]\n",
    "            if method == 'varnet':\n",
    "                aux = aux[aux['sampleid'].str.contains('CRC-986_100215-CW-T')]\n",
    "            #print(aux['truth'].sum())\n",
    "            #if not aux.empty:\n",
    "            print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "            #precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "            #f1list = 2*(precision * recall)/(precision + recall)\n",
    "            #print(len(f1list), len(precision))\n",
    "            #max(f1list)\n",
    "            #print(precision, recall, thresholds)\n",
    "            #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "            #if np.nanmax(f1list) > 0:\n",
    "            #    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #    y.append(recall[list(f1list).index(np.nanmax(f1list))])\n",
    "            x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "            #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #y.append(np.nanmax(f1list))\n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15,8))\n",
    "fig, ax = plt.subplots(1, 1, figsize=(12,8))\n",
    "for k,v in res.items():\n",
    "    plt.plot(v['x'], v['y'], marker='o', label=k, color=color_dict[k])\n",
    "#plt.gca().invert_xaxis()\n",
    "#plt.xscale(\"log\")\n",
    "plt.axhline(y=1/aux['truth'].sum(), ls='--', c='k')\n",
    "#plt.xlim([0.5, 0.01])\n",
    "#ax.set_xticks([500*np.round(bins/500, 1)])\n",
    "#xlim = ax.set_xlim(100, 1000)\n",
    "import matplotlib.ticker\n",
    "#ax.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "#ax.get_xaxis().set_minor_formatter(matplotlib.ticker.NullFormatter())\n",
    "plt.xlabel('Local depth of coverage')\n",
    "plt.ylabel('AUPRC score')\n",
    "plt.title(\"Callers' AUPRC performance w.r.t local depth of coverage\", pad=50)\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "if not os.path.exists(os.path.join(*config.outputpath, 'figure2c')):\n",
    "        os.mkdir(os.path.join(*config.outputpath, 'figure2c'))\n",
    "#plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'depthanalysis_auprc_986.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltables_snv['sampletf'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cond = True\n",
    "for i in range(len(mixtureids)):\n",
    "    cond = cond & (calltables_snv['sampletf'] != np.sort(calltables_snv['sampletf'].unique())[-i]) & (calltables_snv['sampletf'] == 11.7316)\n",
    "calltables_snv_vaf = calltables_snv[cond]\n",
    "#calltables_snv_vaf = calltables_snv[calltables_snv['sampletf'] != calltables_snv['sampletf'].unique().max()]\n",
    "print(calltables_snv_vaf[calltables_snv_vaf['truth']==True].shape[0])\n",
    "out, bins = pd.qcut(calltables_snv_vaf[calltables_snv_vaf['truth']==True]['median_vaf'], q=10, retbins=True)\n",
    "#out, bins = pd.qcut(calltables_snv_vaf[calltables_snv_vaf['median_vaf']>0]['median_vaf'], q=10, retbins=True)\n",
    "vafranges = list(bins)# + [1]\n",
    "print(vafranges)\n",
    "if vafranges[0] == 0:\n",
    "    vafranges[0] = 0.001\n",
    "#np.sort(pd.qcut(calltable_snv[calltable_snv['truth']==True]['median_vaf'], q=10).unique())\n",
    "#vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.0075, 0.005, 0.0025, 0.001]\n",
    "#vafranges = [1., 0.016, 0.012, 0.0093, 0.0073, 0.0063, 0.0056, 0.0051, 0.0045, 0.0038, 0]\n",
    "#vafranges = [.195, 0.0894, 0.0758, 0.0597, 0.0495, 0.0429, 0.0342, 0.0281, 0.024, 0.0172, 0.0052]\n",
    "#vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltables_snv_vaf[(calltables_snv_vaf['median_vaf'] >= vafranges[vi-1]) & (calltables_snv_vaf['median_vaf'] < vafranges[vi])]\n",
    "            #if method == 'varnet':\n",
    "            #    aux = aux[aux['sampleid'].str.contains('CRC-986_100215-CW-T')]\n",
    "            #print(aux['truth'].sum())\n",
    "            #if not aux.empty:\n",
    "            if not aux.empty: #and method != 'varnet':\n",
    "                print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "                #precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "                #f1list = 2*(precision * recall)/(precision + recall)\n",
    "                #print(len(f1list), len(precision))\n",
    "                #max(f1list)\n",
    "                #print(precision, recall, thresholds)\n",
    "                #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                #if np.nanmax(f1list) > 0:\n",
    "                #    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #    y.append(recall[list(f1list).index(np.nanmax(f1list))])\n",
    "                x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #y.append(np.nanmax(f1list))\n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15,8))\n",
    "fig, ax = plt.subplots(1, 1, figsize=(22,8))\n",
    "for k,v in res.items():\n",
    "    plt.plot(v['x'], v['y'], marker='s', markersize=10, label=k, color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.axhline(y=1/aux['truth'].sum(), ls='--', c='k', label='baseline AUPRC')\n",
    "#plt.xlim([0.5, 0.001])\n",
    "plt.xscale(\"log\")\n",
    "ax.set_xticks([b.round(2) if b > 0.025 else b.round(3) for b in bins])\n",
    "xlim = ax.set_xlim(0.75, 0.006)\n",
    "import matplotlib.ticker\n",
    "ax.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "ax.get_xaxis().set_minor_formatter(matplotlib.ticker.NullFormatter())\n",
    "plt.xlabel('Variant Allele Frequency (VAF)')\n",
    "plt.ylabel('AUPRC score')\n",
    "plt.title(\"Callers' AUPRC performance w.r.t VAF\", pad=50)\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "if not os.path.exists(os.path.join(*config.outputpath, 'figure2c')):\n",
    "        os.mkdir(os.path.join(*config.outputpath, 'figure2c'))\n",
    "#plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'vafanalysis_auprc_3patients.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cond = True\n",
    "for i in range(len(mixtureids)):\n",
    "    cond = cond & (calltables_snv['sampletf'] != np.sort(calltables_snv['sampletf'].unique())[-i])  & (calltables_snv['sampletf'] == 11.7316)\n",
    "calltables_snv_vaf = calltables_snv[cond]\n",
    "#calltables_snv_vaf = calltables_snv[calltables_snv['sampletf'] != calltables_snv['sampletf'].unique().max()]\n",
    "out, bins = pd.qcut(calltables_snv_vaf[calltables_snv_vaf['truth']==True]['median_vaf'], q=10, retbins=True)\n",
    "#out, bins = pd.qcut(calltables_snv_vaf[calltables_snv_vaf['median_vaf']>0]['median_vaf'], q=10, retbins=True)\n",
    "vafranges = list(bins)# + [1]\n",
    "print(vafranges)\n",
    "#np.sort(pd.qcut(calltable_snv[calltable_snv['truth']==True]['median_vaf'], q=10).unique())\n",
    "#vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.0075, 0.005, 0.0025, 0.001]\n",
    "#vafranges = [1., 0.016, 0.012, 0.0093, 0.0073, 0.0063, 0.0056, 0.0051, 0.0045, 0.0038, 0]\n",
    "#vafranges = [.195, 0.0894, 0.0758, 0.0597, 0.0495, 0.0429, 0.0342, 0.0281, 0.024, 0.0172, 0.0052]\n",
    "#vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltables_snv_vaf[(calltables_snv_vaf['median_vaf'] >= vafranges[vi-1]) & (calltables_snv_vaf['median_vaf'] < vafranges[vi])]\n",
    "            #print(aux['truth'].sum())\n",
    "            #if not aux.empty:\n",
    "            #print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "            precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "            f1list = 2*(precision * recall)/(precision + recall)\n",
    "            #print(len(f1list), len(precision))\n",
    "            #max(f1list)\n",
    "            #print(precision, recall, thresholds)\n",
    "            #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "            if np.nanmax(f1list) >= 0.05: # and method != 'varnet':\n",
    "                print(method, np.nanmax(f1list), vafranges[vi-1], vafranges[vi])\n",
    "                x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                y.append(recall[list(f1list).index(np.nanmax(f1list))])\n",
    "            #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "            #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #y.append(np.nanmax(f1list))\n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15,8))\n",
    "fig, ax = plt.subplots(1, 1, figsize=(22,8))\n",
    "for k,v in res.items():\n",
    "    if v['x'] != []:\n",
    "        print(k)\n",
    "        plt.plot(v['x'], v['y'], marker='s', markersize=10, label=k, color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "plt.axhline(y=1/aux['truth'].sum(), ls='--', c='k')\n",
    "#plt.xlim([0.5, 0.01])\n",
    "ax.set_xticks(bins.round(3))\n",
    "xlim = ax.set_xlim(0.75, 0.006)\n",
    "#xlim = ax.set_xlim(100, 1000)\n",
    "import matplotlib.ticker\n",
    "ax.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "ax.get_xaxis().set_minor_formatter(matplotlib.ticker.NullFormatter())\n",
    "plt.xlabel('Variant Allele Frequency (VAF)')\n",
    "plt.ylabel('Recall value corresponding to max F1-score (when  0.05)')\n",
    "plt.title(\"Callers' recall of max F1-score w.r.t VAF\", pad=50)\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "if not os.path.exists(os.path.join(*config.outputpath, 'figure2c')):\n",
    "        os.mkdir(os.path.join(*config.outputpath, 'figure2c'))\n",
    "#plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'vafanalysis_recallmaxf1_986.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cond = True\n",
    "for i in range(len(mixtureids)):\n",
    "    cond = cond & (calltables_snv['sampletf'] != np.sort(calltables_snv['sampletf'].unique())[-i])  & (calltables_snv['sampletf'] == 11.7316)\n",
    "calltables_snv_vaf = calltables_snv[cond]\n",
    "#calltables_snv_vaf = calltables_snv[calltables_snv['sampletf'] != calltables_snv['sampletf'].unique().max()]\n",
    "out, bins = pd.qcut(calltables_snv_vaf[calltables_snv_vaf['truth']==True]['median_vaf'], q=10, retbins=True)\n",
    "vafranges = list(bins)# + [1]\n",
    "print(vafranges)\n",
    "#np.sort(pd.qcut(calltable_snv[calltable_snv['truth']==True]['median_vaf'], q=10).unique())\n",
    "#vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.0075, 0.005, 0.0025, 0.001]\n",
    "#vafranges = [1., 0.016, 0.012, 0.0093, 0.0073, 0.0063, 0.0056, 0.0051, 0.0045, 0.0038, 0]\n",
    "#vafranges = [.195, 0.0894, 0.0758, 0.0597, 0.0495, 0.0429, 0.0342, 0.0281, 0.024, 0.0172, 0.0052]\n",
    "#vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltables_snv_vaf[(calltables_snv_vaf['median_vaf'] >= vafranges[vi-1]) & (calltables_snv_vaf['median_vaf'] < vafranges[vi])]\n",
    "            #print(aux['truth'].sum())\n",
    "            #if not aux.empty:\n",
    "            #print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "            precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "            f1list = 2*(precision * recall)/(precision + recall)\n",
    "            if np.nanmax(f1list) >= 0.05 and method != 'varnet':\n",
    "                ar = np.argwhere(f1list == np.nanmax(f1list)).flatten()\n",
    "                br = [recall[a] for a in ar]\n",
    "                idxa = ar[br.index(max(br))]\n",
    "                #print(len(f1list), len(precision))\n",
    "                #max(f1list)\n",
    "                #print(precision, recall, thresholds)\n",
    "                #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                print(method, np.nanmax(f1list), vafranges[vi-1], vafranges[vi])\n",
    "                x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                y.append(np.nanmax(f1list))\n",
    "            #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "            #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #y.append(np.nanmax(f1list))\n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15,8))\n",
    "fig, ax = plt.subplots(1, 1, figsize=(22,8))\n",
    "for k,v in res.items():\n",
    "    if v['x'] != []:\n",
    "        print(k)\n",
    "        plt.plot(v['x'], v['y'], marker='s', markersize=10, label=k, color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "#plt.axhline(y=1/aux['truth'].sum(), ls='--', c='k')\n",
    "#plt.xlim([0.5, 0.01])\n",
    "ax.set_xticks(bins.round(3))\n",
    "xlim = ax.set_xlim(0.75, 0.006)\n",
    "#xlim = ax.set_xlim(100, 1000)\n",
    "import matplotlib.ticker\n",
    "ax.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "ax.get_xaxis().set_minor_formatter(matplotlib.ticker.NullFormatter())\n",
    "plt.xlabel('Variant Allele Frequency (VAF)')\n",
    "plt.ylabel('Max F1-score (when  0.05)')\n",
    "plt.title(\"Callers' max F1-score w.r.t VAF\", pad=50)\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "if not os.path.exists(os.path.join(*config.outputpath, 'figure2c')):\n",
    "        os.mkdir(os.path.join(*config.outputpath, 'figure2c'))\n",
    "#plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'vafanalysis_recallmaxf1_986.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "1/aux['truth'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "precisionthreshold = 0.1\n",
    "\n",
    "cond = True\n",
    "for i in range(len(mixtureids)):\n",
    "    cond = cond & (calltables_snv['sampletf'] != np.sort(calltables_snv['sampletf'].unique())[-i])  & (calltables_snv['sampletf'] == 11.7316)\n",
    "calltables_snv_vaf = calltables_snv[cond]\n",
    "#calltables_snv_vaf = calltables_snv[calltables_snv['sampletf'] != calltables_snv['sampletf'].unique().max()]\n",
    "out, bins = pd.qcut(calltables_snv_vaf[calltables_snv_vaf['truth']==True]['median_vaf'], q=10, retbins=True)\n",
    "vafranges = list(bins)# + [1]\n",
    "print(vafranges)\n",
    "#np.sort(pd.qcut(calltable_snv[calltable_snv['truth']==True]['median_vaf'], q=10).unique())\n",
    "#vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.0075, 0.005, 0.0025, 0.001]\n",
    "#vafranges = [1., 0.016, 0.012, 0.0093, 0.0073, 0.0063, 0.0056, 0.0051, 0.0045, 0.0038, 0]\n",
    "#vafranges = [.195, 0.0894, 0.0758, 0.0597, 0.0495, 0.0429, 0.0342, 0.0281, 0.024, 0.0172, 0.0052]\n",
    "#vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltables_snv_vaf[(calltables_snv_vaf['median_vaf'] >= vafranges[vi-1]) & (calltables_snv_vaf['median_vaf'] < vafranges[vi])]\n",
    "            #print(aux['truth'].sum())\n",
    "            #if not aux.empty:\n",
    "            #print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "            precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "            #idxa = min(range(len(precision)), key=lambda i: abs(precision[i]-precisionthreshold))\n",
    "            a = [(lambda pi: abs(pi-precisionthreshold))(pi) for pi in precision]\n",
    "            #print(a)\n",
    "            ar = np.argwhere(a == np.amin(a)).flatten()\n",
    "            br = [recall[a] for a in ar]\n",
    "            idxa = ar[br.index(max(br))]\n",
    "            plt.plot(recall, precision, label=np.round((vafranges[vi-1]+ vafranges[vi])/2, 3))\n",
    "            \n",
    "            #print(ar, br, idxa)\n",
    "            #idxa = min(range(len(precision[::-1])), key=lambda i: abs(precision[::-1][i]-precisionthreshold))\n",
    "            print(precision[idxa].round(3), recall[idxa].round(3), vafranges[vi-1].round(3), vafranges[vi].round(3))\n",
    "            #f1list = 2*(precision * recall)/(precision + recall)\n",
    "            #print(len(f1list), len(precision))\n",
    "            #max(f1list)\n",
    "            #print(precision, recall, thresholds)\n",
    "            #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "            #if np.nanmax(f1list) >= 0.01:\n",
    "            #    print(method, np.nanmax(f1list), vafranges[vi-1], vafranges[vi])\n",
    "            #    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #    y.append(recall[list(f1list).index(np.nanmax(f1list))])\n",
    "            if precision[idxa] >= precisionthreshold-0.05 and recall[idxa] != 0: #and method != 'varnet':\n",
    "                x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                y.append(recall[idxa])\n",
    "            #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "            #x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "            #y.append(np.nanmax(f1list))\n",
    "    plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    plt.show()\n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15,8))\n",
    "fig, ax = plt.subplots(1, 1, figsize=(20,8))\n",
    "for k,v in res.items():\n",
    "    plt.plot(v['x'], v['y'], marker='s', markersize=10, label=k, color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "plt.axhline(y=1/aux['truth'].sum(), ls='--', c='k')\n",
    "#plt.xlim([0.5, 0.01])\n",
    "ax.set_xticks([b.round(2) if b > 0.025 else b.round(3) for b in bins])\n",
    "#xlim = ax.set_xlim(100, 1000)\n",
    "xlim = ax.set_xlim(0.75, 0.006)\n",
    "import matplotlib.ticker\n",
    "ax.get_xaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())\n",
    "ax.get_xaxis().set_minor_formatter(matplotlib.ticker.NullFormatter())\n",
    "plt.xlabel('Variant Allele Frequency (VAF)')\n",
    "plt.ylabel('Recall value for precision fixed at '+str(precisionthreshold))\n",
    "plt.title(\"Callers' recall when precision = {} w.r.t VAF\".format(precisionthreshold), pad=50)\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "if not os.path.exists(os.path.join(*config.outputpath, 'figure2c')):\n",
    "        os.mkdir(os.path.join(*config.outputpath, 'figure2c'))\n",
    "#plt.savefig(os.path.join(*config.outputpath, 'figure2c', 'vafanalysis_recallatfixedprecision_'+str(precisionthreshold)+'_3partients.svg'), bbox_inches='tight')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = [(lambda pi: abs(pi-precisionthreshold))(pi) for pi in precision]\n",
    "ar = np.argwhere(a == np.amin(a)).flatten()\n",
    "print(ar)\n",
    "br = [recall[a] for a in ar]\n",
    "idx = ar[br.index(max(br))]\n",
    "idx\n",
    "#print(list(ar).index(max([recall[a] for a in ar])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.009, 0.008, 0.007, 0.006, 0.005, 0.004, 0.003, 0.002, 0.001]\n",
    "vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "calltable_snv = calltables['snv'][calltables['snv']['sampletf'] != calltables['snv']['sampletf'].unique().max()]\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    x, y  = [], []\n",
    "    for vi, vafrange in enumerate(vafranges):\n",
    "        #print(vafranges[vi-1], vafranges[vi])\n",
    "        if vi > 0:\n",
    "            aux = calltable_snv[(calltable_snv['median_vaf'] >= vafranges[vi-1]) & (calltable_snv['median_vaf'] < vafranges[vi])]\n",
    "            if not aux.empty:\n",
    "                #print(vafranges[vi-1], vafranges[vi], aux.shape[0])\n",
    "                precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method + '_score'].fillna(0))\n",
    "                f1list = 2*(precision * recall)/(precision + recall)\n",
    "                #print(len(f1list), len(precision))\n",
    "                #max(f1list)\n",
    "                #print(precision, recall, thresholds)\n",
    "                #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                #y.append(max(f1list))\n",
    "                y.append(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "    res[method] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "plt.figure(figsize=(15,8))\n",
    "for k,v in res.items():\n",
    "    plt.plot(v['x'], v['y'], marker='o', label=k, color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "plt.legend()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vafranges = [1., 0.1, 0.075, 0.05, 0.025, 0.01, 0.009, 0.008, 0.007, 0.006, 0.005, 0.004, 0.003, 0.002, 0.001]\n",
    "vafranges = vafranges[::-1]\n",
    "res = {}\n",
    "calltable_snv = calltables['snv'][calltables['snv']['sampletf'] != calltables['snv']['sampletf'].unique().max()]\n",
    "for method in config.methods:\n",
    "    print(method)\n",
    "    res[method] = {}\n",
    "    x, y  = [], []\n",
    "    for tf in aux['sampletf'].unique():\n",
    "        for vi, vafrange in enumerate(vafranges):\n",
    "            if vi > 0:\n",
    "                aux = calltable_snv[(calltable_snv['median_vaf'] >= vafranges[vi-1]) & (calltable_snv['median_vaf'] < vafranges[vi])]\n",
    "                if not aux[aux['sampletf'] == tf].empty:\n",
    "                    #precision, recall, thresholds = precision_recall_curve(aux[aux['sampletf']==tf]['truth'], aux[aux['sampletf']==tf][method + '_score'].fillna(0))\n",
    "                    #print(precision, recall, thresholds)\n",
    "                    #print(average_precision_score(aux['truth'], aux[method + '_score'].fillna(0)))\n",
    "                    x.append((vafranges[vi-1]+ vafranges[vi])/2)\n",
    "                    y.append(average_precision_score(aux[aux['sampletf']==tf]['truth'], aux[aux['sampletf']==tf][method + '_score'].fillna(0)))\n",
    "        res[method][tf] = {'x': x, 'y': y}\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "plt.figure(figsize=(15,8))\n",
    "for k,v in res.items():\n",
    "    for vk, vv in v.items():\n",
    "        plt.plot(vv['x'], vv['y'], color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lwlist = list(np.sort(calltables['snv']['sampletf'].unique()))\n",
    "plt.figure(figsize=(15,8))\n",
    "for k,v in res.items():\n",
    "    for vk, vv in v.items():\n",
    "        #print(np.ceil(lwlist.index(vk)/2))\n",
    "        #print(int(np.ceil(lwlist.index(vk)/2)), 1-vk/50)\n",
    "        plt.plot(vv['x'], vv['y'], lw=int(np.ceil(lwlist.index(vk)/2)), color=color_dict[k])\n",
    "plt.gca().invert_xaxis()\n",
    "plt.xscale(\"log\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "caux"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part II: Group per VAF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for dil in dilutionseries['tf'].values:\n",
    "    dil = '{:.2f}'.format(np.round(100*float(dil), 4)/100)\n",
    "    print(dil)\n",
    "    print([str(dil)+'_'+method+'_vaf' for method in config.methods])\n",
    "    calltablesseries[str(dil)+'_vaf'] = calltablesseries[[str(dil)+'_'+method+'_vaf' for method in config.methods]].median(skipna=True, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries['vaf'] = [l.tolist() for l in calltablesseries[colvaf].values]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colvaf = ['{:.2f}'.format(np.round(100*float(dil), 4)/100)+'_vaf' for dil in dilutionseries['tf'].values][1:]\n",
    "aux = calltablesseries.iloc[:2000][colvaf]\n",
    "#aux.columns = [float(c.split('_')[0]) for c in colvaf]\n",
    "aux['direction'] = np.nan\n",
    "aux.at[(aux[colvaf[1]] <= aux[colvaf[0]]) & (aux[colvaf[2]] <= aux[colvaf[1]]) &\n",
    "    (aux[colvaf[3]] <= aux[colvaf[2]]) & (aux[colvaf[4]] <= aux[colvaf[3]]) & (aux[colvaf[5]] <= aux[colvaf[4]]), 'direction'] = 'strict decrease'\n",
    "aux.at[(aux[colvaf[1]] >= aux[colvaf[0]]) & (aux[colvaf[2]] >= aux[colvaf[1]]) & \n",
    "    (aux[colvaf[3]] >= aux[colvaf[2]]) & (aux[colvaf[4]] >= aux[colvaf[3]]) & (aux[colvaf[5]] >= aux[colvaf[4]]), 'direction'] = 'strict increase'\n",
    "aux.at[(aux[colvaf[1]] <= aux[colvaf[0]]) | (aux[colvaf[2]] <= aux[colvaf[1]]) | \n",
    "    (aux[colvaf[3]] <= aux[colvaf[2]]) | (aux[colvaf[4]] <= aux[colvaf[3]]) | (aux[colvaf[5]] <= aux[colvaf[4]]), 'direction'] = 'decrease'\n",
    "aux.at[((aux[colvaf[1]] >= aux[colvaf[0]]) | (aux[colvaf[2]] >= aux[colvaf[1]]) | (aux[colvaf[3]] >= aux[colvaf[2]])  |\n",
    "       (aux[colvaf[4]] >= aux[colvaf[3]]) | (aux[colvaf[5]] >= aux[colvaf[4]])) & (aux['direction'] != 'decrease'), 'direction'] = 'increase'\n",
    "aux.at[((aux[colvaf[1]] >= aux[colvaf[0]]) | (aux[colvaf[2]] >= aux[colvaf[1]]) | (aux[colvaf[3]] >= aux[colvaf[2]])  |\n",
    "       (aux[colvaf[4]] >= aux[colvaf[3]]) | (aux[colvaf[5]] >= aux[colvaf[4]])) & (aux['direction'] == 'decrease'), 'direction'] = 'up and down'\n",
    "aux['direction'].fillna('other', inplace=True)\n",
    "print(aux['direction'].value_counts())\n",
    "fig, ax = plt.subplots(figsize=(10,10))\n",
    "aux[aux['direction'] == 'other'][colvaf].T.plot(kind='line', color='k', ax=ax, legend=False)\n",
    "aux[aux['direction'] == 'increase'][colvaf].T.plot(kind='line', color='r',  ax=ax,  legend=False)\n",
    "aux[aux['direction'] == 'up and down'][colvaf].T.plot(kind='line', color='b',  ax=ax,  legend=False)\n",
    "aux[aux['direction'] == 'decrease'][colvaf].T.plot(kind='line', color='g',  ax=ax,  legend=False)\n",
    "\n",
    "#plt.plt(c='k', hue='direction', data=aux[colvaf])\n",
    "#ax.invert_xaxis()\n",
    "plt.xlabel('sample TF')\n",
    "plt.ylabel('median VAF found')\n",
    "#plt.ylim([0, 0.02])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries['sampleTF'] = \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "aux.explode('vaf')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for fixedvar in fixedvars:\n",
    "#fixedvar = 'coverage'\n",
    "    if fixedvar == 'coverage':\n",
    "        xaxis = 'tumor burden'\n",
    "    elif fixedvar == 'ctdna':\n",
    "        xaxis = 'coverage'\n",
    "    #for mt in muttypes:\n",
    "    mt = 'snv'\n",
    "    if mt == 'snv':\n",
    "        gtm = 4\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    else:  # elif mt == 'indel':\n",
    "        gtm = 2\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    print(refname)\n",
    "    # for metric in metrics:\n",
    "    metric = 'auprc'\n",
    "    # load results tables\n",
    "    restables = {'snv': [], 'indel': []}\n",
    "    mixtureid = 'CRC-986_100215-CW-T_CRC-986_300316-CW-T'\n",
    "    if mixtureid == 'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        gtm = 3\n",
    "        refname = 'intissuesamplebyatleast'+str(gtm)+'callers'\n",
    "    else:\n",
    "        gtm = 4\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "    print(mixtureid, plasmasample)\n",
    "    xa = xaxis if xaxis != 'tumor burden' else 'tb'\n",
    "    print(xa)\n",
    "    restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_allchr', 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar+'_'+ xa +'.csv'), index_col=0)\n",
    "    #restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_chr22_wgs', 'mixtures_chr22_'+mixtureid, 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar +'.csv'), index_col=0)\n",
    "    restable['plasma sample'] = plasmasample\n",
    "    restables[mt].append(restable)\n",
    "    restables[mt] = pd.concat(restables[mt])\n",
    "    res1 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=False, save=False)\n",
    "    #res2 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=True, save=True)\n",
    "    #res3 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True)\n",
    "    #res4 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "    #                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=True, save=True)\n",
    "    #plt.grid(linewidth=1)\n",
    "    plt.ylim([0, .8])\n",
    "    #plt.xlim([20, 0])\n",
    "    #plt.xlim([40, 250])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#plt.grid(linewidth=1)\n",
    "\n",
    "#for fixedvar in fixedvars:\n",
    "fixedvar = 'coverage'\n",
    "if fixedvar == 'coverage':\n",
    "    xaxis = 'tumor burden'\n",
    "elif fixedvar == 'ctdna':\n",
    "    xaxis = 'coverage'\n",
    "#for mt in muttypes:\n",
    "mt = 'snv'\n",
    "if mt == 'snv':\n",
    "    gtm = 4\n",
    "    refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "else:  # elif mt == 'indel':\n",
    "    gtm = 2\n",
    "    refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "print(refname)\n",
    "# for metric in metrics:\n",
    "metric = 'auprc'\n",
    "# load results tables\n",
    "restables = {'snv': [], 'indel': []}\n",
    "for mixtureid in mixtureids:\n",
    "    if mixtureid == 'CRC-986_100215-CW-T_CRC-986_300316-CW-T':\n",
    "        gtm = 3\n",
    "        refname = 'intissuesamplebyatleast'+str(gtm)+'callers'\n",
    "    else:\n",
    "        gtm = 4\n",
    "        refname = 'inundilutedsamplebyatleast'+str(gtm)+'callers'\n",
    "    plasmasample = '_'.join(mixtureid.split('_')[:2])\n",
    "    print(mixtureid, plasmasample)\n",
    "    xa = xaxis if xaxis != 'tumor burden' else 'tb'\n",
    "    print(xa)\n",
    "    restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_allchr', 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar+'_'+ xa +'.csv'), index_col=0)\n",
    "    #restable = pd.read_csv(os.path.join(*config.mixturefolder, 'mixtures_chr22_wgs', 'mixtures_chr22_'+mixtureid, 'results', mixtureid+'_'+mt+'_'+metric+'_'+refname+'_fixed'+fixedvar +'.csv'), index_col=0)\n",
    "    restable['plasma sample'] = plasmasample\n",
    "    restables[mt].append(restable)\n",
    "restables[mt] = pd.concat(restables[mt])\n",
    "res1 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=False, save=False)\n",
    "#res2 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=True, logscale=True, save=True)\n",
    "#res3 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                   ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=False, save=True)\n",
    "#res4 = plot_metricsseries(config, restables, mixtureids, 'all', metric=metric, muttype=mt,\n",
    "#                       ground_truth_method='mixture', fixedvar=fixedvar, refname=refname, allpatients=False, logscale=True, save=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "resx = np.array([rx.values for rx in res1['x']])\n",
    "resx.mean(axis=0)\n",
    "resy = np.array([ry.values for ry in res1['y']])\n",
    "print(resy)\n",
    "res = {m: [] for m in config.methods}\n",
    "for mi, m in enumerate(config.methods):\n",
    "    resmean = np.mean([resy[mi], resy[mi+len(config.methods)], resy[mi+2*len(config.methods)]], axis=0)\n",
    "    resstd = np.std([resy[mi], resy[mi+len(config.methods)], resy[mi+2*len(config.methods)]], axis=0)\n",
    "    #reslabel = m\n",
    "    res[m] = [resmean, resstd, resx.mean(axis=0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "plt.figure(figsize=(15, 10))\n",
    "#plt.grid(linewidth=1)\n",
    "plt.grid()\n",
    "for m in config.methods:\n",
    "    #plt.plot(res['freebayes'][2], res['freebayes'][0], c=color_dict['freebayes'], markersize=15, lw=2)\n",
    "    plt.errorbar(res[m][2], res[m][0], xerr = resx.std(axis=0), yerr=res[m][1], marker=config.markers[0],  c=color_dict[m], label=m,  markersize=15, lw=2, fmt='-o')\n",
    "ax = plt.gca()\n",
    "if fixedvar == 'coverage':\n",
    "    plt.gca().invert_xaxis()\n",
    "    xlab='tumor burden'\n",
    "else:\n",
    "    xlab = 'coverage or added noise'\n",
    "hand, labl = ax.get_legend_handles_labels()\n",
    "ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.xlabel(xlab)\n",
    "plt.ylabel(metric.upper()+' score')\n",
    "plt.grid()\n",
    "plt.title(metric.upper() + \" score for {} calling in chr{} with ref {}\".format(mt.upper(), chrom, refname))\n",
    "plt.ylim([0,1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for m in config.methods:\n",
    "    res[m][2] = res[m][2][::-1]\n",
    "    res[m][0] = res[m][0][::-1]\n",
    "    res[m][1] = res[m][1][::-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "#plt.figure(figsize=(15, 10))\n",
    "#plt.grid(linewidth=1)\n",
    "#res[m][2] = res[m][2][::-1]\n",
    "#res[m][0] = res[m][0][::-1]\n",
    "#res[m][1] = res[m][1][::-1]\n",
    "rankres = []\n",
    "for m in config.methods:\n",
    "    xstart = 0\n",
    "    ystart = 0\n",
    "    sum_all = []\n",
    "    for i in range(len(res[m][2])):\n",
    "        #if i >0:\n",
    "        #print(res[m][2][i], xstart)\n",
    "        #print((res[m][2][i]-xstart) * res[m][0][i] )\n",
    "        #print(((res[m][2][i]-xstart) * (res[m][0][i]-ystart)/2))\n",
    "        sum_all.append(((res[m][2][i]-xstart) * res[m][0][i]) - ((res[m][2][i]-xstart) * (res[m][0][i]-ystart)/2))\n",
    "        xstart = res[m][2][i]\n",
    "        ystart = res[m][0][i]\n",
    "    print(m, np.sum(sum_all) )\n",
    "    rankres.append(np.sum(sum_all))\n",
    "    #plt.plot(res['freebayes'][2], res['freebayes'][0], c=color_dict['freebayes'], markersize=15, lw=2)\n",
    "    #plt.errorbar(res[m][2], res[m][0], xerr = resx.std(axis=0), yerr=res[m][1], marker=config.markers[0], label=m,  markersize=15, lw=2, fmt='-o')\n",
    "    #plt.gca().invert_xaxis()\n",
    "#ax = plt.gca()\n",
    "#xlab = 'coverage or added noise'\n",
    "#xlab='tumor burden'\n",
    "#hand, labl = ax.get_legend_handles_labels()\n",
    "#ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "#plt.xlabel(xlab)\n",
    "#plt.ylabel(metric.upper()+' score')\n",
    "#plt.title(metric.upper() + \" score for {} calling in chr{} with ref {}\".format(mt.upper(), chrom, refname))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.patches as mpatches\n",
    "rankres_sort = np.argsort(rankres)[::-1]\n",
    "rankres_sort\n",
    "plt.figure(figsize=(14, 5))\n",
    "method_order =[]\n",
    "rank_order =[]\n",
    "for i in rankres_sort:\n",
    "    method_order.append(config.methods[i])\n",
    "    rank_order.append(rankres[i])\n",
    "plt.bar(method_order, rank_order, color=['b', 'b', 'b', 'b', 'b', 'r', 'r'])\n",
    "blue_patch = mpatches.Patch(color='b', label='DNA methods')\n",
    "red_patch = mpatches.Patch(color='r', label='cfDNA methods')\n",
    "plt.legend(handles=[blue_patch, red_patch])\n",
    "#plt.legend(['DNA methods', 'cfDNA methods'])\n",
    "plt.ylabel('area under AUPRC curve')\n",
    "plt.title('Caller ranking mixtures with decreasing tumor burden')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
