{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SEQC2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Imports\n",
    "\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "import io\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "import warnings\n",
    "from tqdm.notebook import tqdm\n",
    "from sklearn.metrics import precision_recall_curve, f1_score, average_precision_score\n",
    "warnings.filterwarnings('ignore')\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n",
    "# set working directory\n",
    "if not os.getcwd().endswith('cfdna_snv_benchmark'):\n",
    "    os.chdir('../')\n",
    "print('Current working directory: {}'.format(os.getcwd()))\n",
    "\n",
    "from utils.config import Config\n",
    "from utils.viz import *\n",
    "from utils.table import *\n",
    "from utils.metrics import *\n",
    "from utils.calltable import *\n",
    "from utils.calltableseries import *\n",
    "from utils.groundtruth import *\n",
    "from utils.metricsseries import *\n",
    "from utils.util_lift_over import *\n",
    "from utils.venn import venn6, get_labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Config and Display paramaters\n",
    "\n",
    "config = Config(\"config/\", \"config_viz.yaml\")\n",
    "set_display_params(config)\n",
    "print(config.methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Chomosome\n",
    "\n",
    "mixtureids =  ['BRP2_ST26_25ng_LIB1-P', 'IDT2_ST05_25ng_LIB1-P', 'ILM2_ST29_25ng_LIB1-P', 'ROC2_ST21_25ng_LIB1-PH']\n",
    "mixtureid = 'BRP2_ST26_25ng_LIB1-P'\n",
    "reload = False\n",
    "save = False\n",
    "fixedvars=['coverage', 'ctdna']\n",
    "filterparam = 'all'\n",
    "\n",
    "markers = ['o', '^', 'X']\n",
    "linestyles = ['-', '-', '-']\n",
    "color_dict = {config.methods[i]: config.colors[i] for i in range(len(config.methods))}\n",
    "\n",
    "muttypes = ['snv', 'indel']\n",
    "metrics = ['auprc', 'precision', 'recall']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bed files intersection"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# load bedfile for method BRP\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'intersect_BRP_IDT_ILM_ROC_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bedfile.shape[0]\n",
    "germlinetruthslistall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load bedfile for method BRP\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'intersect_BRP_IDT_ILM_ROC_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "# load ground truths (cerified mutations either somatic or diluted germline in the sample A)\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "print(groundtruths.shape[0])\n",
    "\n",
    "bedfilelistall = []\n",
    "for il in range(bedfile.shape[0]):\n",
    "    for i in list(np.arange(bedfile.startpos.iloc[il], bedfile.endpos.iloc[il])):\n",
    "        bedfilelistall.append(bedfile.chrom.iloc[il]+'_'+str(i))\n",
    "print(len(bedfilelistall))\n",
    "\n",
    "evaluation_bed = pd.DataFrame(index=bedfilelistall)\n",
    "evaluation_bed['truth'] = False\n",
    "gtchrompos = set(set(groundtruths['chrom_pos'].values) & set(evaluation_bed.index))\n",
    "evaluation_bed.loc[gtchrompos, 'truth'] = True\n",
    "evaluation_bed.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load ground truths (cerified mutations either somatic or diluted germline in the sample A)\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "# load known negatives (no mutation in sample A and sample B)\n",
    "germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_hg19.bed'), sep='\\t', header=None)\n",
    "germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "\n",
    "# load bedfile for method BRP\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'intersect_BRP_IDT_ILM_ROC_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "# keep only ground truths in bed\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "\n",
    "bedfilelistall = []\n",
    "for il in range(bedfile.shape[0]):\n",
    "    for i in list(np.arange(bedfile.startpos.iloc[il], bedfile.endpos.iloc[il])):\n",
    "        bedfilelistall.append(bedfile.chrom.iloc[il]+'_'+str(bedfile.startpos.iloc[il]))\n",
    "print(len(bedfilelistall))\n",
    "germlinetruthslistall = []\n",
    "for il in tqdm(range(germlinetruths.shape[0])):\n",
    "    for i in list(np.arange(germlinetruths.startpos.iloc[il], germlinetruths.endpos.iloc[il])):\n",
    "        germlinetruthslistall.append(germlinetruths.chrom.iloc[il]+'_'+str(germlinetruths.startpos.iloc[il]))\n",
    "germlinetruths_in_bed = list(set(bedfilelistall) & set(germlinetruthslistall))\n",
    "print(len(germlinetruths_in_bed))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluation_bed = list(set(set(germlinetruths_in_bed) | set(list(groundtruths_in_bed.index.str.split('_').str[:2].str.join('_')))))\n",
    "print(len(evaluation_bed))\n",
    "evaluation_bed_df = pd.DataFrame(index=evaluation_bed, columns=['truth'])\n",
    "evaluation_bed_df.loc[germlinetruths_in_bed] = False\n",
    "evaluation_bed_df.loc[list(groundtruths_in_bed.index.str.split('_').str[:2].str.join('_'))] = True\n",
    "print(evaluation_bed_df.isna().sum())\n",
    "print(evaluation_bed_df.value_counts())\n",
    "#np.save(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg19.npy'), evaluation_bed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluation_bed_df.to_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg19.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#evalaution_bed = np.load(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg19.npy'))\n",
    "evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg19.csv'), index_col=0)\n",
    "evaluation_bed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluation_bed_hg38 = []\n",
    "for eb in evaluation_bed:\n",
    "    #print(eb)\n",
    "    eb = 'chr'+eb\n",
    "    evaluation_bed_hg38.append(liftover(eb, os.path.join('data', 'extdata', 'hg19ToHg38.over.chain.gz')))\n",
    "evaluation_bed_hg38"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groundtruths_annotated = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19_nochr_annotated.vcf.gz'))\n",
    "a = groundtruths_annotated['ID'].str.contains('rs').sum()\n",
    "b = groundtruths_annotated.shape[0]\n",
    "print(a, b, round(100*a/b, 2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groundtruths_annotated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groundtruths_annotated['chrom_pos'] = groundtruths_annotated['CHROM'].astype(str).str.cat(groundtruths_annotated['POS'].astype(str), sep='_')\n",
    "groundtruths_annotated_intersect = groundtruths_annotated[groundtruths_annotated['chrom_pos'].isin(evaluation_bed_df.index)]\n",
    "print(groundtruths_annotated_intersect.shape[0])\n",
    "print(groundtruths_annotated_intersect['ID'].str.contains('rs').sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for eb in evaluation_bed_df.index:\n",
    "    print(eb)\n",
    "    print(groundtruths_annotated[groundtruths_annotated['chrom_pos'] == eb])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(evaluation_bed_hg38))\n",
    "evaluation_bed_hg38.count(np.nan)\n",
    "#np.save(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg38.npy'), evaluation_bed_hg38)\n",
    "evaluation_bed_hg38_df = evaluation_bed_df.copy()\n",
    "evaluation_bed_hg38_df.index = evaluation_bed_hg38\n",
    "evaluation_bed_hg38_df\n",
    "evaluation_bed_hg38_df.to_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg38.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluation_bed_hg38_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg38.csv'), index_col=0)\n",
    "evaluation_bed_hg38_df[evaluation_bed_hg38_df['truth'] == True]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixtureid = 'BRP2_ST26_25ng_LIB1-P'\n",
    "seriesorder = ['SampleDf', 'SampleEf']\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01)\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01)\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01)\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 0: Reproduce paper figure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_hg19.bed'), sep='\\t', header=None)\n",
    "germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "germlinetruths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load ground truths (cerified mutations either somatic or diluted germline in the sample A)\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos', inplace=True)\n",
    "groundtruths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load ground truths (cerified mutations either somatic or diluted germline in the sample A)\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "# load known negatives (no mutation in sample A and sample B)\n",
    "germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_hg19.bed'), sep='\\t', header=None)\n",
    "germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "\n",
    "# load bedfile for method BRP\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_BRP_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "# keep only ground truths in bed\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "\n",
    "bedfilelistall = []\n",
    "for il in range(bedfile.shape[0]):\n",
    "    for i in list(np.arange(bedfile.startpos.iloc[il], bedfile.endpos.iloc[il])):\n",
    "        bedfilelistall.append(bedfile.chrom.iloc[il]+'_'+str(bedfile.startpos.iloc[il]))\n",
    "print(len(bedfilelistall))\n",
    "germlinetruthslistall = []\n",
    "for il in tqdm(range(germlinetruths.shape[0])):\n",
    "    for i in list(np.arange(germlinetruths.startpos.iloc[il], germlinetruths.endpos.iloc[il])):\n",
    "        germlinetruthslistall.append(germlinetruths.chrom.iloc[il]+'_'+str(germlinetruths.startpos.iloc[il]))\n",
    "print(len(germlinetruthslistall))\n",
    "germlinetruths_in_bed = list(set(bedfilelistall) & set(germlinetruthslistall))\n",
    "print(len(germlinetruths_in_bed))\n",
    "      \n",
    "# load table\n",
    "calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_BRP2_ST26_25ng_LIB1-P', 'SampleEf_BRP2_ST26_25ng_LIB1-P'), ['varnet', 'BRP'], save=False, filter='PASS')\n",
    "calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "\n",
    "# build lists method VS ground truths\n",
    "calltable['truth'] = False\n",
    "print(calltable.shape[0])\n",
    "calltable = calltable.reindex(list(set(list(groundtruths_in_bed.index)) | set([list(germlinetruths_in_bed)))) ### get known negatives\n",
    "#calltable = calltable.reindex(list(set(list(calltable.index))))\n",
    "#calltable.loc[calltable.index.isin(list(calltable.index)), 'truth'] = True\n",
    "#calltable = calltable.reindex(list(set(list(calltable.index))))\n",
    "#calltable.loc[calltable.index.isin(list(calltable.index)), 'truth'] = True\n",
    "print(calltable.shape[0])\n",
    "calltable = calltable[~calltable['chrom_pos'].isin(germlinetruths_in_bed)]\n",
    "print(calltable.shape[0])\n",
    "calltable.loc[calltable.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "calltable['truth'].fillna(True, inplace=True)\n",
    "calltable['BRP'].fillna(False, inplace=True)\n",
    "print(calltable['truth'].value_counts())\n",
    "print(calltable.shape[0])\n",
    "print(calltable['BRP'].value_counts())\n",
    "\n",
    "\n",
    "print(\"germline\")\n",
    "print(germlinetruths.shape[0])\n",
    "print(calltable[(calltable['truth'] == False) & (calltable['BRP_vaf'] > 0)].shape[0])\n",
    "print(list(set(list(calltable['chrom_pos'].values)) & set(list(germlinetruths.index))))\n",
    "\n",
    "\n",
    "aux = calltable[['truth', 'BRP_vaf', 'BRP_score']]\n",
    "aux['BRP_score'] =  aux['BRP_vaf']\n",
    "aux['BRP_score'][(aux['BRP_score'] < 0.001)] = 0\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.001) & (aux['BRP_score'] < 0.0025)] = 0.001\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.0025) & (aux['BRP_score'] < 0.003)] = 0.0025\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.003) & (aux['BRP_score'] < 0.004)] = 0.003\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.004) & (aux['BRP_score'] < 0.005)] = 0.004\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.005) & (aux['BRP_score'] < 0.01)] = 0.005\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.01) & (aux['BRP_score'] < 0.015)] = 0.01\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.015) & (aux['BRP_score'] < 0.02)] = 0.015\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.02) & (aux['BRP_score'] < 0.025)] = 0.02\n",
    "aux['BRP_score'][(aux['BRP_score'] >= 0.025)] = 0.025\n",
    "aux['BRP_score'].fillna(0, inplace=True)\n",
    "aux['BRP_vaf'].fillna(0, inplace=True)\n",
    "#aux.dropna(inplace=True)\n",
    "precision, recall, thresholds = precision_recall_curve(aux['truth'], aux['BRP_score'])\n",
    "idxa = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.001))\n",
    "idxb = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.025))\n",
    "\n",
    "# VAF cutoffs 0.1–0.5%, 0.5–2.5% and >2.5%\n",
    "plot_pr_curve(precision[idxa:idxb], recall[idxa:idxb], estimator_name='BRP', f1_score=None, figax=None, kwargs={'color': 'darkgreen'})\n",
    "plt.ylim([0, 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(thresholds))\n",
    "print(idxa, idxb)\n",
    "print(thresholds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(calltable['BRP_vaf'][~calltable['BRP_vaf'].isna()].shape[0])\n",
    "print(calltable['truth'].sum())\n",
    "print(calltable['truth'][calltable['truth'] == False].shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "precision, recall"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load ground truths (somatic mutations)\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "# load known negatives (germline mutations)\n",
    "germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_hg19.bed'), sep='\\t', header=None)\n",
    "germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "\n",
    "# load bedfile for method BRP\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_BRP_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "# keep only ground truths in bed\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "\n",
    "\"\"\"\n",
    "germlinetruth_outside_bed = []\n",
    "for idx, row in tqdm(germlinetruths.iterrows(), total=germlinetruths.shape[0]):\n",
    "    chrom = row['chrom']\n",
    "    pos = row['startpos']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        germlinetruth_outside_bed.append(chrom + '_'+ str(pos))\n",
    "print(len(germlinetruth_outside_bed))\n",
    "germlinetruths_in_bed = groundtruths.drop(germlinetruth_outside_bed)\n",
    "print(germlinetruths_in_bed.shape[0])\n",
    "\"\"\"\n",
    "\"\"\"\n",
    "bedfilelistall = []\n",
    "for il in range(bedfile.shape[0]):\n",
    "    for i in list(np.arange(bedfile.startpos.iloc[il], bedfile.endpos.iloc[il])):\n",
    "        bedfilelistall.append(i)\n",
    "print(len(bedfilelistall))\n",
    "germlinetruthslistall = []\n",
    "for il in tqdm(range(germlinetruths.shape[0])):\n",
    "    for i in list(np.arange(germlinetruths.startpos.iloc[il], germlinetruths.endpos.iloc[il])):\n",
    "        germlinetruthslistall.append(i)\n",
    "print(len(germlinetruthslistall))\n",
    "germlinetruths_in_bed = list(set(bedfilelistall) & set(germlinetruthslistall))\n",
    "print(len(germlinetruths_in_bed))\n",
    "\"\"\"\n",
    "      \n",
    "# load table\n",
    "calltable = pd.DataFrame()\n",
    "for st in ['25', '26']:\n",
    "    for lib in ['1', '2', '3', '4']:\n",
    "#for sample in ['SampleEf_BRP2_ST25_25ng_LIB1-P',  'SampleEf_BRP2_ST25_25ng_LIB2-P', 'SampleEf_BRP2_ST25_25ng_LIB3-P',  'SampleEf_BRP2_ST25_25ng_LIB4-P', 'SampleEf_BRP2_ST26_25ng_LIB1-P',  'SampleEf_BRP2_ST26_25ng_LIB2-P',  'SampleEf_BRP2_ST26_25ng_LIB3-P',  'SampleEf_BRP2_ST26_25ng_LIB4-P']:\n",
    "        calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_BRP2_ST'+st+'_25ng_LIB'+lib+'-P', 'SampleEf_BRP2_ST'+st+'_25ng_LIB'+lib+'-P'), ['varnet', 'BRP'], save=False, filter='PASS')\n",
    "        calltable_tmp = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "        calltable = pd.concat([calltable, calltable_tmp])\n",
    "        #calltable = calltable[(calltable['BRP_vaf'] <= 0.025) & (calltable['BRP_vaf'] > 0.001)]\n",
    "\n",
    "# build lists method VS ground truths\n",
    "calltable['truth'] = False\n",
    "print(calltable.shape[0])\n",
    "#calltable = calltable.reindex(list(set(list(calltable.index)) | set(list(groundtruths_in_bed.index))))\n",
    "print(calltable.shape[0])\n",
    "calltable.loc[calltable.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "calltable['truth'].fillna(True, inplace=True)\n",
    "calltable['BRP'].fillna(False, inplace=True)\n",
    "print(calltable['truth'].value_counts())\n",
    "print(calltable.shape[0])\n",
    "print(calltable['BRP'].value_counts())\n",
    "\n",
    "\n",
    "print(\"germline\")\n",
    "print(germlinetruths.shape[0])\n",
    "print(calltable[(calltable['truth'] == False) & (calltable['BRP_vaf'] > 0)].shape[0])\n",
    "calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "print(list(set(list(calltable['chrom_pos'].values)) & set(list(germlinetruths.index))))\n",
    "\n",
    "\n",
    "aux = calltable[['truth', 'BRP_vaf', 'BRP_score']]\n",
    "aux['BRP_score'] = aux['BRP_vaf']\n",
    "#aux['BRP_score'][aux['BRP_score'] > 0.025] = 0 #0.03\n",
    "#aux['BRP_score'][aux['BRP_score'] <= 0.005] = 0 # 0.005\n",
    "#aux['BRP_score'][(aux['BRP_score'] > 0.005) & (aux['BRP_score'] <= 0.025)] = 0.01\n",
    "aux['BRP_score'].fillna(0, inplace=True)\n",
    "aux['BRP_vaf'].fillna(0, inplace=True)\n",
    "#aux.dropna(inplace=True)\n",
    "precision, recall, thresholds = precision_recall_curve(aux['truth'], aux['BRP_score'])\n",
    "idxa = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.001))\n",
    "idxb = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.025))\n",
    "\n",
    "# VAF cutoffs 0.1–0.5%, 0.5–2.5% and >2.5%\n",
    "plot_pr_curve(precision[idxa:idxb], recall[idxa:idxb], estimator_name='BRP', f1_score=None, figax=None, kwargs={'color': 'darkgreen'})\n",
    "plt.ylim([0, 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load table\n",
    "for st in ['25', '26']:\n",
    "    for lib in ['1', '2', '3', '4']:\n",
    "        calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_BRP2_ST'+st+'_25ng_LIB'+lib+'-P', 'SampleEf_BRP2_ST'+st+'_25ng_LIB'+lib+'-P'), ['varnet', 'BRP'], save=False, filter='PASS')\n",
    "        calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "        ###\n",
    "        # build lists method VS ground truths\n",
    "        calltable['truth'] = False\n",
    "        print(calltable.shape[0])\n",
    "        calltable.loc[calltable.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "        calltable['truth'].fillna(True, inplace=True)\n",
    "        calltable['truth_vaf'] = 0\n",
    "        calltable['truth_vaf'] = groundtruths['INFO'].str.split(';').str[0].str.split('VAF=').str[1].astype(float)\n",
    "        calltable['truth_vaf'].fillna(0, inplace=True)\n",
    "        calltable['BRP'].fillna(False, inplace=True)\n",
    "        print(calltable['truth'].value_counts())\n",
    "        print(calltable.shape[0])\n",
    "        print(calltable['BRP'].value_counts())\n",
    "        calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "\n",
    "aux = calltable[['truth', 'truth_vaf', 'BRP_vaf', 'BRP_score']]\n",
    "aux['BRP_score'] = aux['BRP_vaf']\n",
    "#aux['BRP_score'][aux['BRP_score'] > 0.025] = 0 #0.03\n",
    "#aux['BRP_score'][aux['BRP_score'] <= 0.005] = 0 # 0.005\n",
    "#aux['BRP_score'][(aux['BRP_score'] > 0.005) & (aux['BRP_score'] <= 0.025)] = 0.01\n",
    "aux['BRP_score'].fillna(0, inplace=True)\n",
    "aux['BRP_vaf'].fillna(0, inplace=True)\n",
    "#aux.dropna(inplace=True)\n",
    "precision, recall, thresholds = precision_recall_curve(aux['truth'], aux['BRP_score'])\n",
    "idxa = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.001))\n",
    "idxb = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.025))\n",
    "\n",
    "# VAF cutoffs 0.1–0.5%, 0.5–2.5% and >2.5%\n",
    "plot_pr_curve(precision[idxa:idxb], recall[idxa:idxb], estimator_name='BRP', f1_score=None, figax=None, kwargs={'color': 'darkgreen'})\n",
    "plt.ylim([0, 1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "\n",
    "method_sampleid_dict = {\n",
    "    'BRP': 'SampleDf_BRP2_ST26_25ng_LIB1-P',\n",
    "    'IDT': 'SampleDfIS_IDT2_ST05_25ng_LIB1-P',\n",
    "    'ILM': 'SampleDf_ILM2_ST29_25ng_LIB1-P',\n",
    "    'ROC': 'SampleDf_ROC2_ST21_25ng_LIB1-P',\n",
    "    'TFS': 'SampleDf_TFS2_ST24_25ng_LIB1-P',\n",
    "}\n",
    "\n",
    "\n",
    "method_sampleid_dict = {\n",
    "    'BRP': 'SampleEf_BRP2_ST26_25ng_LIB1-P',\n",
    "    'IDT': 'SampleEfIS_IDT2_ST05_25ng_LIB1-P',\n",
    "    'ILM': 'SampleEf_ILM2_ST29_25ng_LIB1-P',\n",
    "    'ROC': 'SampleEf_ROC2_ST21_25ng_LIB1-P',\n",
    "    'TFS': 'SampleEf_TFS2_ST24_25ng_LIB1-P',\n",
    "}\n",
    "\n",
    "\n",
    "for method in ['ROC', 'BRP']: #, 'IDT', 'ILM']: #'ROC', 'TFS']:\n",
    "    \n",
    "    print('######## '+ method + ' ########')\n",
    "    \n",
    "    if method == 'ROC':\n",
    "        refgenome = 'hg38'\n",
    "    else:\n",
    "        refgenome = 'hg19'\n",
    "    \n",
    "    groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_'+refgenome+'.vcf'))\n",
    "    if method == 'BRP':\n",
    "        groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "    groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "    groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "    germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_'+refgenome+'.bed'), sep='\\t', header=None)\n",
    "    germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "    if method == 'BRP':\n",
    "        germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "    germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "    germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "\n",
    "    bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_'+method+'_'+refgenome+'.bed'), sep='\\t', header=None)\n",
    "    bedfile = bedfile.iloc[:,:3]\n",
    "    bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "    if method == 'BRP':\n",
    "        bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "    bedfile['diff'] = bedfile['endpos'] - bedfile['startpos'] + 1\n",
    "    N = bedfile['diff'].sum()\n",
    "    print('reportable region in bp')\n",
    "    print(N)\n",
    "    \n",
    "    germlinetruths['chrom_pos'] = germlinetruths['chrom']+'_'+germlinetruths['startpos'].astype(str)\n",
    "    germlinetruths_all = list(germlinetruths['chrom_pos'].values)\n",
    "    bed_all = []\n",
    "    for ind in tqdm(list(bedfile.index)):\n",
    "        for i in range(int(bedfile['startpos'][ind]), int(bedfile['endpos'][ind])+1):\n",
    "            bed_all.append(bedfile['chrom'][ind]+'_'+str(i))\n",
    "    print(len(germlinetruths_all))\n",
    "    print(len(bed_all))\n",
    "    germlinetruths_in_bed = list(set(germlinetruths_all) & set(bed_all))\n",
    "    print(len(germlinetruths_in_bed))\n",
    "\n",
    "    groundtruth_outside_bed = []\n",
    "    for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "        chrom = row['CHROM']\n",
    "        pos = row['POS']\n",
    "        ref = row['REF']\n",
    "        alt = row['ALT']\n",
    "        aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "        count = False\n",
    "        for ind in list(aux.index):\n",
    "            #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "            if aux['endpos'][ind] >= int(pos):\n",
    "                if aux['startpos'][ind] <= int(pos):\n",
    "                    count = True\n",
    "        if not count:\n",
    "            groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "    print('groundtruth')\n",
    "    print(len(groundtruth_outside_bed))\n",
    "    groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "    print(groundtruths_in_bed.shape[0])\n",
    "\n",
    "    \"\"\"\n",
    "    germlinetruths['chrom_pos'] = germlinetruths['chrom']+'_'+germlinetruths['startpos']\n",
    "    germlinetruths_all = list(germlinetruths['chrom_pos'].values)\n",
    "    bed_all = []\n",
    "    for ind in list(befile.index):\n",
    "        bed_all.append([befile['chrom'][ind]+'_'+str(i) for i in range(int(befile['startpos'][ind]), int(befile['endpos'][ind])+1)])\n",
    "    germlinetruth_outside_bed = []\n",
    "    for idx, row in tqdm(germlinetruths.iterrows(), total=germlinetruths.shape[0]):\n",
    "        chrom = row['chrom']\n",
    "        pos = row['startpos']\n",
    "        aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "        count = False\n",
    "        for ind in list(aux.index):\n",
    "            #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "            if aux['endpos'][ind] >= int(pos):\n",
    "                if aux['startpos'][ind] <= int(pos):\n",
    "                    count = True\n",
    "        if not count:\n",
    "            germlinetruth_outside_bed.append(chrom + '_'+ str(pos))\n",
    "    print(len(germlinetruth_outside_bed))\n",
    "    germlinetruths_in_bed = groundtruths.drop(germlinetruth_outside_bed)\n",
    "    print(germlinetruths_in_bed.shape[0])\n",
    "    \"\"\"\n",
    "    bedfilelistall = []\n",
    "    for il in range(bedfile.shape[0]):\n",
    "        for i in list(np.arange(bedfile.startpos.iloc[il], bedfile.endpos.iloc[il])):\n",
    "            bedfilelistall.append(bedfile.chrom.iloc[il]+'_'+str(bedfile.startpos.iloc[il]))\n",
    "    print(len(bedfilelistall))\n",
    "    germlinetruthslistall = []\n",
    "    for il in tqdm(range(germlinetruths.shape[0])):\n",
    "        for i in list(np.arange(germlinetruths.startpos.iloc[il], germlinetruths.endpos.iloc[il])):\n",
    "            germlinetruthslistall.append(germlinetruths.chrom.iloc[il]+'_'+str(germlinetruths.startpos.iloc[il]))\n",
    "    print(len(germlinetruthslistall))\n",
    "    germlinetruths_in_bed = list(set(bedfilelistall) & set(germlinetruthslistall))\n",
    "    print(len(germlinetruths_in_bed))\n",
    "\n",
    "    calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+'_'.join(method_sampleid_dict[method].split('_')[1:]), method_sampleid_dict[method]), ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method], save=False, filter='all')\n",
    "    #calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+'_'.join(method_sampleid_dict[method].split('_')[1:]), method_sampleid_dict[method]), ['varnet', method], save=False, filter='PASS')\n",
    "    print(calltable_snv.shape[0], calltable_indel.shape[0], calltable_snp.shape[0])\n",
    "    calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "    calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "\n",
    "    calltable['truth'] = False\n",
    "    print(calltable.shape[0])\n",
    "    calltable = calltable.reindex(list(set(list(calltable.index)) | set(list(groundtruths_in_bed.index))))\n",
    "    print('TP: {}'.format(len(list(set(list(calltable.index)) & set(list(groundtruths_in_bed.index))))))\n",
    "    print('FP: {}'.format(len(list(set(list(calltable.index)) ^ set(list(groundtruths_in_bed.index))))))\n",
    "    print(calltable.shape[0])\n",
    "    calltable = calltable[~calltable['chrom_pos'].isin(germlinetruths_in_bed)]\n",
    "    print(calltable.shape[0])\n",
    "    calltable.loc[calltable.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "    calltable['truth'].fillna(True, inplace=True)\n",
    "    #for m in ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method]:\n",
    "    #    print(m)\n",
    "    #    calltable[m].fillna(False, inplace=True)\n",
    "    #    print(calltable[m].value_counts())\n",
    "    m = method\n",
    "    calltable[m].fillna(False, inplace=True)\n",
    "    print(calltable[m].value_counts())\n",
    "    print(calltable['truth'].value_counts())\n",
    "    print(calltable.shape[0])\n",
    "\n",
    "    print(\"germline\")\n",
    "    print(germlinetruths.shape[0])\n",
    "    print(calltable[(calltable['truth'] == False) & (calltable[method+'_vaf'] > 0)].shape[0])\n",
    "    print('germline in calls')\n",
    "    print(len(list(set(list(calltable['chrom_pos'].values)) & set(list(germlinetruths.index)))))\n",
    "    \n",
    "    if method == 'BRP':\n",
    "        for m in ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan']:\n",
    "            aux = calltable[['truth', m+'_vaf', m+'_score']]\n",
    "            print(aux[m+'_score'].describe())\n",
    "            aux[m+'_score'].fillna(0, inplace=True)\n",
    "            aux[m+'_vaf'].fillna(0, inplace=True)\n",
    "            precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[m+'_score'])\n",
    "            precisionbis = []\n",
    "            Nneg = len(germlinetruths_in_bed)\n",
    "            print('Nneg')\n",
    "            print(Nneg)\n",
    "            plot_pr_curve(precision, recall, estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "\n",
    "    m = method\n",
    "    aux = calltable[['truth', m+'_vaf', m+'_score']]\n",
    "    #if m == 'BRP':\n",
    "    aux[m+'_score'] = aux[m+'_vaf']\n",
    "    print(aux[m+'_vaf'].describe())\n",
    "    aux[m+'_score'].fillna(0, inplace=True)\n",
    "    aux[m+'_vaf'].fillna(0, inplace=True)\n",
    "    #aux.dropna(inplace=True)\n",
    "    precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[m+'_vaf'])\n",
    "    precisionbis = []\n",
    "    Nneg = len(germlinetruths_in_bed)\n",
    "    print('Nneg')\n",
    "    print(Nneg)\n",
    "    for thres in thresholds:\n",
    "        y_pred = aux[m+'_vaf'].copy()\n",
    "        y_pred[y_pred >= thres] = 1\n",
    "        y_pred[y_pred != 1] = 0\n",
    "        _, fp, _, tp = confusion_matrix(aux['truth'], y_pred).ravel()\n",
    "        precisionbis.append(tp/(tp+(fp/Nneg)))\n",
    "\n",
    "    if m == 'BRP' or m == 'IDT' or m =='ILM' or m=='ROC':\n",
    "        idxa = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.001))\n",
    "        idxb = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.025))\n",
    "    else:\n",
    "        idxa = 0\n",
    "        idxb = len(thresholds)\n",
    "\n",
    "    # VAF cutoffs 0.1–0.5%, 0.5.–2.5% and >2.5%\n",
    "    # precision divided by the size of on-target negative positions for each panel\n",
    "    #print(precision[idxa:idxb])\n",
    "    #print(precisionbis[idxa:idxb])\n",
    "    #print(recall[idxa:idxb])\n",
    "    plot_pr_curve(precision[idxa:idxb], recall[idxa:idxb], estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "#plt.ylim([0.97, 1.0])\n",
    "plt.ylim([0.0, 1.0])\n",
    "plt.xlim([0.0, 1.0])\n",
    "#plt.ylim([0., 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable['truth'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "\n",
    "# taking one sample of each company\n",
    "method_sampleid_dict = {\n",
    "    'BRP': 'SampleEf_BRP2_ST26_25ng_LIB1-P',\n",
    "    'IDT': 'SampleEfIS_IDT2_ST05_25ng_LIB1-P',\n",
    "    'ILM': 'SampleEf_ILM2_ST29_25ng_LIB1-P',\n",
    "    'ROC': 'SampleEf_ROC2_ST21_25ng_LIB1-P',\n",
    "    'TFS': 'SampleEf_TFS2_ST24_25ng_LIB1-P',\n",
    "}\n",
    "\n",
    "\n",
    "for method in ['ROC', 'BRP', 'IDT', 'ILM']: #, 'TFS']: \n",
    "    \n",
    "    print('######## '+ method + ' ########')\n",
    "    \n",
    "    if method == 'ROC':\n",
    "        refgenome = 'hg38'\n",
    "    else:\n",
    "        refgenome = 'hg19'\n",
    "    \n",
    "    # read ground truths vcf file 'known positives'\n",
    "    groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_'+refgenome+'.vcf'))\n",
    "    if method == 'BRP':\n",
    "        groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "    groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "    groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "    # read germline truths bed file 'known negatives'\n",
    "    germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_'+refgenome+'.bed'), sep='\\t', header=None)\n",
    "    germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "    if method == 'BRP':\n",
    "        germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "    germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "    germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "\n",
    "    # get bed file corresponding to each company\n",
    "    if method == 'TFS':\n",
    "        bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_TFS_pre-defined_Hotspots_hg19.bed'), sep='\\t', header=None)\n",
    "    else:\n",
    "        bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_'+method+'_'+refgenome+'.bed'), sep='\\t', header=None)\n",
    "    bedfile = bedfile.iloc[:,:3]\n",
    "    bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "    if method == 'BRP':\n",
    "        bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "    bedfile['diff'] = bedfile['endpos'] - bedfile['startpos'] + 1\n",
    "    N = bedfile['diff'].sum()\n",
    "    print('reportable region in bp')\n",
    "    print(N)\n",
    "    \n",
    "    # keep germline inside current bed\n",
    "    germlinetruths['chrom_pos'] = germlinetruths['chrom']+'_'+germlinetruths['startpos'].astype(str)\n",
    "    germlinetruths_all = list(germlinetruths['chrom_pos'].values)\n",
    "    bed_all = []\n",
    "    for ind in tqdm(list(bedfile.index)):\n",
    "        for i in range(int(bedfile['startpos'][ind]), int(bedfile['endpos'][ind])+1):\n",
    "            bed_all.append(bedfile['chrom'][ind]+'_'+str(i))\n",
    "    print('germline all')\n",
    "    print(len(germlinetruths_all))\n",
    "    #print(len(bed_all))\n",
    "    germlinetruths_in_bed = list(set(germlinetruths_all) & set(bed_all))\n",
    "    print('germline in bed')\n",
    "    print(len(germlinetruths_in_bed))\n",
    "\n",
    "    # keep ground truths inside current bed\n",
    "    groundtruth_outside_bed = []\n",
    "    for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "        chrom = row['CHROM']\n",
    "        pos = row['POS']\n",
    "        ref = row['REF']\n",
    "        alt = row['ALT']\n",
    "        aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "        count = False\n",
    "        for ind in list(aux.index):\n",
    "            #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "            if aux['endpos'][ind] >= int(pos):\n",
    "                if aux['startpos'][ind] <= int(pos):\n",
    "                    count = True\n",
    "        if not count:\n",
    "            groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "    print('groundtruth in bed')\n",
    "    #print(len(groundtruth_outside_bed))\n",
    "    groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "    print(groundtruths_in_bed.shape[0])\n",
    "\n",
    "    # read company vcf file (snvs + indels)\n",
    "    #calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+method+'2_ST26_25ng_LIB1-P', 'SampleDf_BRP2_ST26_25ng_LIB1-P'), ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method], save=False, filter='all')\n",
    "    calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+'_'.join(method_sampleid_dict[method].split('_')[1:]), method_sampleid_dict[method]), ['varnet', method], save=False, filter='PASS')\n",
    "    print(calltable_snv.shape[0], calltable_indel.shape[0], calltable_snp.shape[0])\n",
    "    calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "    #calltable =  calltable_snv\n",
    "\n",
    "    calltable['truth'] = False\n",
    "    print('calltable shape')\n",
    "    print(calltable.shape[0])\n",
    "    calltable = calltable.reindex(list(set(list(calltable.index)) | set(list(groundtruths_in_bed.index))))\n",
    "    print('TP: {}'.format(len(list(set(list(calltable.index)) & set(list(groundtruths_in_bed.index))))))\n",
    "    print('FP: {}'.format(len(list(set(list(calltable.index)) ^ set(list(groundtruths_in_bed.index))))))\n",
    "    print(calltable.shape[0])\n",
    "    calltable.loc[calltable.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "    calltable['truth'].fillna(True, inplace=True)\n",
    "    #for m in ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method]:\n",
    "    #    print(m)\n",
    "    #    calltable[m].fillna(False, inplace=True)\n",
    "    #    print(calltable[m].value_counts())\n",
    "    m = method\n",
    "    calltable[m].fillna(False, inplace=True)\n",
    "    print(calltable[m].value_counts())\n",
    "    print(calltable['truth'].value_counts())\n",
    "    print(calltable.shape[0])\n",
    "\n",
    "    print(\"germline\")\n",
    "    print(germlinetruths.shape[0])\n",
    "    print(calltable[(calltable['truth'] == False) & (calltable[method+'_vaf'] > 0)].shape[0])\n",
    "    calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "    print('germline in calls')\n",
    "    print(len(list(set(list(calltable['chrom_pos'].values)) & set(list(germlinetruths.index)))))\n",
    "    calltable.set_index('chrom_pos').drop(list(set(list(calltable['chrom_pos'].values)) & set(list(germlinetruths.index))), inplace=True)\n",
    "    calltable['chrom_pos_ref_alt'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_').str.cat(calltable['ref'].astype('str'), sep='_').str.cat(calltable['alt'].astype('str'), sep='_')\n",
    "    calltable.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "    #for m in ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method]:\n",
    "    m = method\n",
    "    aux = calltable[['truth', m+'_vaf', m+'_score']]\n",
    "    #if m == 'BRP':\n",
    "    aux[m+'_score'] = aux[m+'_vaf']\n",
    "    print(aux[m+'_vaf'].describe())\n",
    "    #aux['BRP_score'] = aux['BRP_vaf']\n",
    "    #aux['BRP_score'][aux['BRP_score'] > 0.025] = 0.03\n",
    "    #aux['BRP_score'][aux['BRP_score'] <= 0.005] = 0.005\n",
    "    #aux['BRP_score'][(aux['BRP_score'] > 0.005) & (aux['BRP_score'] <= 0.025)] = 0.01\n",
    "    aux[m+'_score'].fillna(0, inplace=True)\n",
    "    aux[m+'_vaf'].fillna(0, inplace=True)\n",
    "    #aux.dropna(inplace=True)\n",
    "    precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[m+'_vaf'])\n",
    "    precisionbis = []\n",
    "    Nneg = len(germlinetruths_in_bed)\n",
    "    print('Nneg')\n",
    "    print(Nneg)\n",
    "    for thres in thresholds:\n",
    "        y_pred = aux[m+'_vaf'].copy()\n",
    "        y_pred[y_pred >= thres] = 1\n",
    "        y_pred[y_pred != 1] = 0\n",
    "        _, fp, _, tp = confusion_matrix(aux['truth'], y_pred).ravel()\n",
    "        precisionbis.append(tp/(tp+(fp/Nneg)))\n",
    "                     \n",
    "    \n",
    "    idxa = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.001))\n",
    "    idxb = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.025))\n",
    "    #else:\n",
    "    #    idxa = 0\n",
    "    #    idxb = len(thresholds)\n",
    "\n",
    "    # VAF cutoffs 0.1–0.5%, 0.5.–2.5% and >2.5%\n",
    "    # precision divided by the size of on-target negative positions for each panel\n",
    "    #print(precision[idxa:idxb])\n",
    "    #print(precisionbis[idxa:idxb])\n",
    "    #print(recall[idxa:idxb])\n",
    "    plot_pr_curve(precision[idxa:idxb], recall[idxa:idxb], estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "#plt.ylim([0.97, 1.0])\n",
    "plt.ylim([0.0, 1.0])\n",
    "plt.xlim([0.0, 1.0])\n",
    "#plt.ylim([0., 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# compare bed files\n",
    "\n",
    "groundtruths_bedfiles = {}\n",
    "\n",
    "for method in ['ROC', 'BRP', 'IDT', 'ILM']: #, 'TFS']: \n",
    "    \n",
    "    print('######## '+ method + ' ########')\n",
    "    \n",
    "    if method == 'ROC':\n",
    "        refgenome = 'hg38'\n",
    "    else:\n",
    "        refgenome = 'hg19'\n",
    "    \n",
    "    # read ground truths vcf file 'known positives'\n",
    "    groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_'+refgenome+'.vcf'))\n",
    "    if method == 'BRP':\n",
    "        groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "    groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "    groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "    # get bed file corresponding to each company\n",
    "    if method == 'TFS':\n",
    "        bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_TFS_pre-defined_Hotspots_hg19.bed'), sep='\\t', header=None)\n",
    "    else:\n",
    "        bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_'+method+'_'+refgenome+'.bed'), sep='\\t', header=None)\n",
    "    bedfile = bedfile.iloc[:,:3]\n",
    "    bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "    if method == 'BRP':\n",
    "        bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "    bedfile['diff'] = bedfile['endpos'] - bedfile['startpos'] + 1\n",
    "    N = bedfile['diff'].sum()\n",
    "    print('reportable region in bp')\n",
    "    print(N)\n",
    "\n",
    "    # keep ground truths inside current bed\n",
    "    groundtruth_outside_bed = []\n",
    "    for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "        chrom = row['CHROM']\n",
    "        pos = row['POS']\n",
    "        ref = row['REF']\n",
    "        alt = row['ALT']\n",
    "        aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "        count = False\n",
    "        for ind in list(aux.index):\n",
    "            #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "            if aux['endpos'][ind] >= int(pos):\n",
    "                if aux['startpos'][ind] <= int(pos):\n",
    "                    count = True\n",
    "        if not count:\n",
    "            groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "    print('groundtruth in bed')\n",
    "    #print(len(groundtruth_outside_bed))\n",
    "    groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "    print(groundtruths_in_bed.shape[0])\n",
    "    \n",
    "    groundtruths_bedfiles[method] = list(groundtruths_in_bed.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#list(groundtruths_bedfiles.values())[0]\n",
    "print(groundtruths_bedfiles.keys())\n",
    "print(len(set(set(groundtruths_bedfiles['BRP']) & set([c[3:] for c in groundtruths_bedfiles['IDT']]) & set([c[3:] for c in groundtruths_bedfiles['ILM']]))))\n",
    "print([len(i) for i in list(groundtruths_bedfiles.values())])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set.intersection(*map(set,list(groundtruths_bedfiles.values())[1:-1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "\n",
    "# taking one sample of each company\n",
    "method_sampleid_dict = {\n",
    "    'BRP': 'SampleEf_BRP2_ST26_25ng_LIB1-P',\n",
    "    'IDT': 'SampleEfIS_IDT2_ST05_25ng_LIB1-P',\n",
    "    'ILM': 'SampleEf_ILM2_ST29_25ng_LIB1-P',\n",
    "    'ROC': 'SampleEf_ROC2_ST21_25ng_LIB1-P',\n",
    "    'TFS': 'SampleEf_TFS2_ST24_25ng_LIB1-P',\n",
    "}\n",
    "\n",
    "\n",
    "for method in ['ROC', 'BRP', 'IDT', 'ILM']: \n",
    "    \n",
    "    print('######## '+ method + ' ########')\n",
    "    if method == 'ROC':\n",
    "        refgenome = 'hg38'\n",
    "    else:\n",
    "        refgenome = 'hg19'\n",
    "    \n",
    "    # read ground truths vcf file 'known positives'\n",
    "    groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_'+refgenome+'.vcf'))\n",
    "    if method == 'BRP':\n",
    "        groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "    groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "    groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "    # read germline truths bed file 'known negatives'\n",
    "    germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_'+refgenome+'.bed'), sep='\\t', header=None)\n",
    "    germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "    if method == 'BRP':\n",
    "        germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "    germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "    germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "\n",
    "    # get bed file corresponding to each company\n",
    "    bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_'+method+'_'+refgenome+'.bed'), sep='\\t', header=None)\n",
    "    bedfile = bedfile.iloc[:,:3]\n",
    "    bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "    if method == 'BRP':\n",
    "        bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "    bedfile['diff'] = bedfile['endpos'] - bedfile['startpos'] + 1\n",
    "    N = bedfile['diff'].sum()\n",
    "    print('reportable region in bp')\n",
    "    print(N)\n",
    "    \n",
    "    # keep germline inside current bed\n",
    "    germlinetruths['chrom_pos'] = germlinetruths['chrom']+'_'+germlinetruths['startpos'].astype(str)\n",
    "    germlinetruths_all = list(germlinetruths['chrom_pos'].values)\n",
    "    bed_all = []\n",
    "    for ind in tqdm(list(bedfile.index)):\n",
    "        for i in range(int(bedfile['startpos'][ind]), int(bedfile['endpos'][ind])+1):\n",
    "            bed_all.append(bedfile['chrom'][ind]+'_'+str(i))\n",
    "    print('germline all')\n",
    "    print(len(germlinetruths_all))\n",
    "    germlinetruths_in_bed = list(set(germlinetruths_all) & set(bed_all))\n",
    "    print('germline in bed')\n",
    "    print(len(germlinetruths_in_bed))\n",
    "\n",
    "    # keep ground truths inside current bed\n",
    "    groundtruth_outside_bed = []\n",
    "    for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "        chrom = row['CHROM']\n",
    "        pos = row['POS']\n",
    "        ref = row['REF']\n",
    "        alt = row['ALT']\n",
    "        aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "        count = False\n",
    "        for ind in list(aux.index):\n",
    "            #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "            if aux['endpos'][ind] >= int(pos):\n",
    "                if aux['startpos'][ind] <= int(pos):\n",
    "                    count = True\n",
    "        if not count:\n",
    "            groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "    print('groundtruth in bed')\n",
    "    #print(len(groundtruth_outside_bed))\n",
    "    groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "    print(groundtruths_in_bed.shape[0])\n",
    "\n",
    "    # read company vcf file (snvs + indels)\n",
    "    calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+'_'.join(method_sampleid_dict[method].split('_')[1:]), method_sampleid_dict[method]), ['varnet', method], save=False, filter='PASS')\n",
    "    print(calltable_snv.shape[0], calltable_indel.shape[0], calltable_snp.shape[0])\n",
    "    calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "\n",
    "    calltable['truth'] = False\n",
    "    print('calltable shape')\n",
    "    print(calltable.shape[0])\n",
    "    calltable = calltable.reindex(list(set(list(calltable.index)) | set(list(groundtruths_in_bed.index))))\n",
    "    #print('TP: {}'.format(len(list(set(list(calltable.index)) & set(list(groundtruths_in_bed.index))))))\n",
    "    #print('FP: {}'.format(len(list(set(list(calltable.index)) ^ set(list(groundtruths_in_bed.index))))))\n",
    "    print(calltable.shape[0])\n",
    "    calltable.loc[calltable.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "    calltable['truth'].fillna(True, inplace=True)\n",
    "    calltable[method].fillna(False, inplace=True)\n",
    "    print(calltable[method].value_counts())\n",
    "    print(calltable['truth'].value_counts())\n",
    "\n",
    "    print(\"germline\")\n",
    "    print(germlinetruths.shape[0])\n",
    "    print(calltable[(calltable['truth'] == False) & (calltable[method+'_vaf'] > 0)].shape[0])\n",
    "    calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "    print('germline in calls')\n",
    "    print(len(list(set(list(calltable['chrom_pos'].values)) & set(list(germlinetruths.index)))))\n",
    "    calltable.set_index('chrom_pos').drop(list(set(list(calltable['chrom_pos'].values)) & set(list(germlinetruths.index))), inplace=True)\n",
    "    calltable['chrom_pos_ref_alt'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_').str.cat(calltable['ref'].astype('str'), sep='_').str.cat(calltable['alt'].astype('str'), sep='_')\n",
    "    calltable.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "    aux = calltable[['truth', method+'_vaf', method+'_score']]\n",
    "    aux[method+'_vaf'].fillna(0, inplace=True)\n",
    "    precision, recall, thresholds = precision_recall_curve(aux['truth'], aux[method+'_vaf'])    \n",
    "    idxa = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.001))\n",
    "    idxb = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.025))\n",
    "    plot_pr_curve(precision[idxa:idxb], recall[idxa:idxb], estimator_name=method, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(method)]})\n",
    "\n",
    "plt.ylim([0.0, 1.0])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groundtruths[groundtruths['INFO'].str.split(';').str[1] == 'TYPE=INDEL'].value_counts() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable[calltable['truth']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable[calltable['truth']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable.loc[list(set(list(calltable.index)) & set(list(groundtruths_in_bed.index)))]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bed file study"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bedfile_ROC = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_ROC_hg38.bed'), sep='\\t', header=None)\n",
    "bedfile_ROC.columns = ['chrom', 'startpos', 'endpos']\n",
    "\n",
    "bedfile_BRP = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_BRP_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile_BRP.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "\n",
    "bedfile_ILM = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_ILM_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile_ILM.columns = ['chrom', 'startpos', 'endpos']\n",
    "\n",
    "bedfile_IDT = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_IDT_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile_IDT.columns = ['chrom', 'startpos', 'endpos', 'A', 'B', 'C']\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for mixtureid in mixtureids:\n",
    "mixtureid = 'ROC2_ST21_25ng_LIB1-PH'\n",
    "seriesorder = ['SampleDf', 'SampleEf']\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg38.vcf'))\n",
    "#groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_ROC_hg38.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "#bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['truth'] = False\n",
    "calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDf_ROC_score'] = calltablesseries['SampleDf_ROC_vaf']\n",
    "calltablesseries['SampleEf_ROC_score'] = calltablesseries['SampleEf_ROC_vaf']\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'ROC']\n",
    "#results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methodss, splitby='dilution')\n",
    "#results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "##                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    " #                                refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods, splitby='dilution')\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "method_sampleid_dict = {\n",
    "    'BRP': 'BRP2_ST26_25ng_LIB1-P',\n",
    "    'IDT': 'IDT2_ST05_25ng_LIB1-P',\n",
    "    'ILM': 'ILM2_ST29_25ng_LIB1-P',\n",
    "    'ROC': 'ROC2_ST21_25ng_LIB1-PH',\n",
    "    'TFS': 'TFS2_ST24_25ng_LIB1-P',\n",
    "}\n",
    "seriesorder = ['SampleDf', 'SampleEf']\n",
    "\n",
    "calltablesseries_df = {}\n",
    "\n",
    "for method in ['ROC', 'BRP', 'IDT', 'ILM']: \n",
    "    print('######## '+ method + ' ########')\n",
    "    if method == 'ROC':\n",
    "        refgenome = 'hg38'\n",
    "    else:\n",
    "        refgenome = 'hg19'\n",
    "    mixtureid = method_sampleid_dict[method]\n",
    "    print('############# {} ############'.format(mixtureid))\n",
    "    calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "    calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01)\n",
    "    calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01)\n",
    "    calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01)\n",
    "    print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "    print(aux)\n",
    "    calltables['snv'].append(calltable_snv)\n",
    "    calltables['indel'].append(calltable_indel)\n",
    "    calltables['snp'].append(calltable_snp)\n",
    "    calltables['sampleid'] = mixtureid \n",
    "    #calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "    calltables['snv'] = pd.concat(calltables['snv'])\n",
    "    calltables['indel'] = pd.concat(calltables['indel'])\n",
    "    calltables['snp'] = pd.concat(calltables['snp'])\n",
    "    #dilutionseries\n",
    "    dilutionseries = aux\n",
    "    #dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "    muttype = 'snv'\n",
    "    refsample = 'SEQC2'\n",
    "    # ground truth\n",
    "    #groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg38.vcf'))\n",
    "    #groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "    #groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "    #groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "    #bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_ROC_hg38.bed'), sep='\\t', header=None)\n",
    "    #bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "    evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_'+refgenome+'.csv'), index_col=0)\n",
    "    if refgenome == 'hg38':\n",
    "        evaluation_bed_df.index = ['chr'+i for i in list(evaluation_bed_df.index)]\n",
    "    #bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "    #groundtruth_outside_bed = []\n",
    "    #for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    #    chrom = row['CHROM']\n",
    "    #    pos = row['POS']\n",
    "    #    ref = row['REF']\n",
    "    #    alt = row['ALT']\n",
    "    #    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    #    count = False\n",
    "    #    for ind in list(aux.index):\n",
    "    #        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "    #        if aux['endpos'][ind] >= int(pos):\n",
    "    #            if aux['startpos'][ind] <= int(pos):\n",
    "    #                count = True\n",
    "    #    if not count:\n",
    "    #        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "    #print(len(groundtruth_outside_bed))\n",
    "    #groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "    #print(groundtruths_in_bed.shape[0])\n",
    "    calltablesseries = calltables[muttype]\n",
    "    calltablesseries['chrom_pos'] = calltablesseries['chrom'].astype(str).str.cat(calltablesseries['pos'].astype(str), sep='_')\n",
    "    calltablesseries.set_index('chrom_pos', inplace=True)\n",
    "    #print(calltablesseries['chrom_pos'])\n",
    "    #calltablesseries['truth'] = False\n",
    "    calltablesseries = calltablesseries[~calltablesseries.index.duplicated()] ### TODO\n",
    "    calltablesseries = calltablesseries.reindex(evaluation_bed_df.index)\n",
    "    calltablesseries['truth'] = evaluation_bed_df['truth']\n",
    "    #calltablesseries = calltablesseries[calltablesseries['chrom_pos'].isin(list(evaluation_bed_df.index))]\n",
    "    #calltablesseries = pd.concat([calltablesseries, evaluation_bed_df], axis=1)\n",
    "    #calltablesseries = calltablesseries[~calltablesseries['truth'].isna()]\n",
    "    print(calltablesseries.shape[0])\n",
    "    print(calltablesseries.head())\n",
    "    #calltablesseries['truth'] = False\n",
    "    #calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "    print(calltablesseries['truth'].value_counts())\n",
    "    #calltablesseries['SampleDf_'+method+'_score'] = calltablesseries['SampleDf_'+method+'_vaf']\n",
    "    #calltablesseries['SampleEf_'+method+'_score'] = calltablesseries['SampleEf_'+method+'_vaf']\n",
    "    # metrics\n",
    "    methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method]\n",
    "    #results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #results_maxf1_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1',  ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "                    #    ground_truth_method=gtm,\n",
    "                    #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)\n",
    "    calltablesseries_df[method] = calltablesseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.transforms as transforms\n",
    "\n",
    "for met in ['TP', 'FP']:\n",
    "    metbool = False if met == 'FP' else True \n",
    "    for m, calltablesseries in calltablesseries_df.items():\n",
    "        print(m)\n",
    "        if m == 'IDT':\n",
    "            suffix = 'lS'\n",
    "        else:\n",
    "            suffix = ''\n",
    "        print(suffix)\n",
    "        print(calltablesseries[calltablesseries['truth'] == metbool].shape[0])\n",
    "        A = calltablesseries[calltablesseries['truth'] == metbool][['SampleDf'+suffix+'_'+m  for m in config.methods]]\n",
    "        print(A.sum())\n",
    "        B = calltablesseries[calltablesseries['truth'] == metbool][['SampleEf'+suffix+'_'+m  for m in config.methods]]\n",
    "        print(B.sum())\n",
    "        ngt = calltablesseries[calltablesseries['truth'] == metbool].shape[0]\n",
    "        fig, ax = plt.subplots(figsize=(10,5))\n",
    "        plt.bar(x=np.arange(len(config.methods))-0.2, height=A.sum().values, width=0.4, color=[config.colors[config.methods.index(m)] for m in config.methods], label=A.sum().index[0].split('_')[0], alpha=0.6)\n",
    "        plt.bar(x=np.arange(len(config.methods))+0.2, height=B.sum().values, width=0.4, color=[config.colors[config.methods.index(m)] for m in config.methods], label=B.sum().index[0].split('_')[0], alpha=1)\n",
    "        plt.xticks(np.arange(len(config.methods)), [a.split('_')[1] for a in A.sum().index])\n",
    "        plt.axhline(y=ngt, c='red', lw=3)\n",
    "        trans = transforms.blended_transform_factory(ax.get_yticklabels()[0].get_transform(), ax.transData)\n",
    "        ax.text(0, ngt, \"{:.0f}\".format(ngt), color=\"red\", transform=trans, ha=\"right\", va=\"center\")\n",
    "        plt.xticks(rotation=90)\n",
    "        plt.legend(bbox_to_anchor=(1,1))\n",
    "        plt.title(met)\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.transforms as transforms\n",
    "\n",
    "methodcalls_dict = {}\n",
    "for me in config.methods:\n",
    "        methodcalls_dict[me] = {}\n",
    "        methodcalls_dict[me]['SampleDf'] = []\n",
    "        methodcalls_dict[me]['SampleEf'] = []\n",
    "\n",
    "met = 'TP'\n",
    "#for met in ['TP', 'FP']:\n",
    "metbool = False if met == 'FP' else True \n",
    "for m, calltablesseries in calltablesseries_df.items():\n",
    "    print(m)\n",
    "    if m == 'IDT':\n",
    "        suffix = 'lS'\n",
    "    else:\n",
    "        suffix = ''\n",
    "    ngt = calltablesseries[calltablesseries['truth'] == metbool].shape[0]\n",
    "    print(ngt)\n",
    "    A = calltablesseries[calltablesseries['truth'] == metbool][['SampleDf'+suffix+'_'+m  for m in config.methods]]\n",
    "    B = calltablesseries[calltablesseries['truth'] == metbool][['SampleEf'+suffix+'_'+m  for m in config.methods]]\n",
    "    for me in config.methods:\n",
    "        if m != 'ROC':\n",
    "            methodcalls_dict[me]['SampleDf'].append(list(A[A['SampleDf'+suffix+'_'+me] == True].index))\n",
    "            methodcalls_dict[me]['SampleEf'].append(list(B[B['SampleEf'+suffix+'_'+me] == True].index))\n",
    "        else:\n",
    "            a = list(A[A['SampleDf'+suffix+'_'+me] == True].index)\n",
    "            for ebi, eb in enumerate(a):\n",
    "                print(eb)\n",
    "                a[ebi] = liftover(eb, os.path.join('data', 'extdata', 'hg38ToHg19.over.chain.gz'))\n",
    "            methodcalls_dict[me]['SampleDf'].append(a)\n",
    "            b = list(B[B['SampleEf'+suffix+'_'+me] == True].index)\n",
    "            for ebi, eb in enumerate(b):\n",
    "                print(eb)\n",
    "                b[ebi] = liftover(eb, os.path.join('data', 'extdata', 'hg38ToHg19.over.chain.gz'))\n",
    "            methodcalls_dict[me]['SampleEf'].append(b)\n",
    "    \n",
    "methodcalls_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = {}\n",
    "res2 = {}\n",
    "for m in config.methods:\n",
    "    aux = methodcalls_dict[m]['SampleDf'].copy()\n",
    "    aux = pd.Series([item for sublist in aux for item in sublist])\n",
    "    aux = aux.value_counts()\n",
    "    aux = aux.value_counts()\n",
    "    aux2 = methodcalls_dict[m]['SampleEf'].copy()\n",
    "    aux2 = pd.Series([item for sublist in aux2 for item in sublist])\n",
    "    aux2 = aux2.value_counts()\n",
    "    aux2 = aux2.value_counts()\n",
    "    res[m] = aux\n",
    "    res2[m] = aux2\n",
    "print(res)\n",
    "res = pd.DataFrame(res)\n",
    "res.fillna(0, inplace=True)\n",
    "res = res.astype(int)\n",
    "res = res.reindex(index=res.index[::-1])\n",
    "res2 = pd.DataFrame(res2)\n",
    "res2.fillna(0, inplace=True)\n",
    "res2 = res2.astype(int)\n",
    "res2 = res2.reindex(index=res2.index[::-1])\n",
    "res['sample'] = 0\n",
    "res2['sample'] = 1\n",
    "res = pd.concat([res, res2])\n",
    "res['count'] = res.index\n",
    "res.reset_index(inplace=True, drop=True)\n",
    "print(res.columns)\n",
    "\n",
    "res3 = res.set_index(['sample', 'count']).stack().reset_index(level=2, drop=False).reset_index()\n",
    "res3.columns = ['sample', 'count', 'method', 'value']\n",
    "res3.sort_values(['sample', 'count', 'method'],  ascending=[True, False, True], inplace=True)\n",
    "res3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_pivot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#sns.barplot(x='method', y='value', hue='sample', data=res3)\n",
    "\n",
    "df_pivot = pd.pivot_table(res3[res3['sample'] == 0], index='method', columns='count', values='value')\n",
    "df_pivot = df_pivot.reindex(config.methods)\n",
    "df_pivot = df_pivot[df_pivot.columns[::-1]] \n",
    "df_pivot.plot.bar(stacked=True)\n",
    "\n",
    "df_pivot = pd.pivot_table(res3[res3['sample'] == 1], index='method', columns='count', values='value')\n",
    "df_pivot = df_pivot.reindex(config.methods)\n",
    "df_pivot = df_pivot[df_pivot.columns[::-1]] \n",
    "df_pivot.plot.bar(stacked=True)\n",
    "\n",
    "#res3.set_index('method').plot(kind='bar', hue='sample', stacked=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FP default threshold\n",
    "if method == 'IDT':\n",
    "    suffix = 'lS'\n",
    "else:\n",
    "    suffix = ''\n",
    "print(suffix)\n",
    "print(calltablesseries[calltablesseries['truth'] == False].shape[0])\n",
    "A = calltablesseries[calltablesseries['truth'] == False][['SampleDf'+suffix+'_'+m  for m in config.methods]]\n",
    "print(A.sum())\n",
    "B = calltablesseries[calltablesseries['truth'] == False][['SampleEf'+suffix+'_'+m  for m in config.methods]]\n",
    "print(B.sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FP \n",
    "\"\"\"\n",
    "print(calltablesseries[calltablesseries['truth'] == False].shape[0])\n",
    "A = calltablesseries[calltablesseries['truth'] == False][['SampleDf'+suffix+'_'+m+'_score'  for m in config.methods]]\n",
    "A[A < 0.8] = 0\n",
    "A[A >= 0.8] = 1\n",
    "print(A.sum())\n",
    "B = calltablesseries[calltablesseries['truth'] == False][['SampleEf'+suffix+'_'+m+'_score'  for m in config.methods]]\n",
    "B[B < 0.8] = 0\n",
    "B[B >= 0.8] = 1\n",
    "print(B.sum())\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.transforms as transforms\n",
    "\n",
    "for m, calltablesseries in calltablesseries_df.items():\n",
    "    print(m)\n",
    "    if m == 'IDT':\n",
    "        suffix = 'lS'\n",
    "    else:\n",
    "        suffix = ''\n",
    "    print(suffix)\n",
    "    print(calltablesseries[calltablesseries['truth'] == False].shape[0])\n",
    "    A = calltablesseries[calltablesseries['truth'] == False][['SampleDf'+suffix+'_'+m  for m in config.methods]]\n",
    "    print(A.sum())\n",
    "    B = calltablesseries[calltablesseries['truth'] == False][['SampleEf'+suffix+'_'+m  for m in config.methods]]\n",
    "    print(B.sum())\n",
    "    ngt = calltablesseries[calltablesseries['truth'] == False].shape[0]\n",
    "    fig, ax = plt.subplots(figsize=(10,5))\n",
    "    plt.bar(x=np.arange(len(config.methods))-0.2, height=A.sum().values, width=0.4, color=[config.colors[config.methods.index(m)] for m in config.methods], label=A.sum().index[0].split('_')[0], alpha=0.6)\n",
    "    plt.bar(x=np.arange(len(config.methods))+0.2, height=B.sum().values, width=0.4, color=[config.colors[config.methods.index(m)] for m in config.methods], label=B.sum().index[0].split('_')[0], alpha=1)\n",
    "    plt.xticks(np.arange(len(config.methods)), [a.split('_')[1] for a in A.sum().index])\n",
    "    plt.axhline(y=ngt, c='red', lw=3)\n",
    "    trans = transforms.blended_transform_factory(ax.get_yticklabels()[0].get_transform(), ax.transData)\n",
    "    ax.text(0, ngt, \"{:.0f}\".format(ngt), color=\"red\", transform=trans, ha=\"right\", va=\"center\")\n",
    "    plt.xticks(rotation=90)\n",
    "    plt.legend(bbox_to_anchor=(1,1))\n",
    "    plt.title('FP')\n",
    "    plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TP default threshold\n",
    "print(calltablesseries[calltablesseries['truth'] == True].shape[0])\n",
    "A = calltablesseries[calltablesseries['truth'] == True][['SampleDf'+suffix+'_'+m  for m in config.methods]]\n",
    "print(A.sum())\n",
    "B = calltablesseries[calltablesseries['truth'] == True][['SampleEf'+suffix+'_'+m for m in config.methods]]\n",
    "print(B.sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# TP\n",
    "\"\"\"\n",
    "threshold = 0.8\n",
    "print(calltablesseries[calltablesseries['truth'] == True].shape[0])\n",
    "A = calltablesseries[calltablesseries['truth'] == True][['SampleDf'+suffix+'_'+m+'_score'  for m in config.methods]]\n",
    "A[A < threshold] = 0\n",
    "A[A >= threshold] = 1\n",
    "print(A.sum())\n",
    "B = calltablesseries[calltablesseries['truth'] == True][['SampleEf'+suffix+'_'+m+'_score'  for m in config.methods]]\n",
    "B[B < threshold] = 0\n",
    "B[B >= threshold] = 1\n",
    "print(B.sum())\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10,5))\n",
    "ngt = calltablesseries[calltablesseries['truth'] == True].shape[0]\n",
    "plt.bar(x=np.arange(len(config.methods))-0.2, height=A.sum().values, width=0.4, color=[config.colors[config.methods.index(m)] for m in config.methods], label=A.sum().index[0].split('_')[0], alpha=0.6)\n",
    "plt.bar(x=np.arange(len(config.methods))+0.2, height=B.sum().values, width=0.4, color=[config.colors[config.methods.index(m)] for m in config.methods], label=B.sum().index[0].split('_')[0], alpha=1)\n",
    "plt.xticks(np.arange(len(config.methods)), [a.split('_')[1] for a in A.sum().index])\n",
    "plt.axhline(y=calltablesseries[calltablesseries['truth'] == True].shape[0], c='red', lw=3)\n",
    "plt.xticks(rotation=90)\n",
    "plt.legend(bbox_to_anchor=(1,1))\n",
    "plt.axhline(y=ngt, c='red', lw=3)\n",
    "trans = transforms.blended_transform_factory(ax.get_yticklabels()[0].get_transform(), ax.transData)\n",
    "ax.text(0, ngt, \"{:.0f}\".format(ngt), color=\"red\", transform=trans, ha=\"right\", va=\"center\")\n",
    "plt.title('TP')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries.head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FP default threshold\n",
    "calltablesseries[calltablesseries['truth'] == False][['SampleDf_'+m+'_score' for m in config.methods]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for m in config.methods:\n",
    "    if method == 'IDT':\n",
    "        suffix = 'lS'\n",
    "    else:\n",
    "        suffix = ''\n",
    "    print(suffix)\n",
    "    plt.figure()\n",
    "    if calltablesseries['SampleDf'+suffix+'_'+m+'_vaf'].isna().sum() < calltablesseries.shape[0]:\n",
    "        sns.scatterplot(x='SampleDf'+suffix+'_'+m+'_vaf', y='SampleDf'+suffix+'_'+m+'_score', hue='truth', marker='o', data=calltablesseries)\n",
    "        #plt.xlim([-.01, .21])\n",
    "    if calltablesseries['SampleEf'+suffix+'_'+m+'_vaf'].isna().sum() < calltablesseries.shape[0]:\n",
    "        sns.scatterplot(x='SampleEf'+suffix+'_'+m+'_vaf', y='SampleEf'+suffix+'_'+m+'_score', hue='truth', marker='s', data=calltablesseries)\n",
    "        #plt.xlim([-.01, .21])\n",
    "        plt.ylim([-.01, 1.02])\n",
    "        plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluation_bed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calltablesseries.set_index('chrom_pos', inplace=True)\n",
    "calltablesseries.loc[calltablesseries[calltablesseries.index.duplicated()].index][['chrom', 'pos', 'ref', 'alt'] + ['SampleDf_'+m for m in config.methods]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries[calltablesseries['chrom_pos'].isin(list(evaluation_bed_df.index))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries['pos'].isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries.tail(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(dilutionseries.iloc[1]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "figure_curve_allchr(config, calltablesseries, pd.DataFrame(dilutionseries.iloc[1]).T, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries['chrom_pos'] = calltablesseries['chrom'].str[3:].str.cat(calltablesseries['pos'].astype(str), sep='_')\n",
    "print(calltablesseries.shape[0])\n",
    "A = calltablesseries[calltablesseries['chrom_pos'].isin(evaluation_bed_hg38)]\n",
    "missed = set(evaluation_bed_hg38) - set(A['chrom_pos'].isin(evaluation_bed_hg38))\n",
    "print(len(missed))\n",
    "missed_index = ['chr'+i for i in missed]\n",
    "#calltablesseries = calltablesseries.reindex(set(set(calltablesseries.index) | set(missed)))\n",
    "print(A.shape[0])\n",
    "#evaluation_bed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A = A.reindex(set(set(A.index) | set(missed_index)))\n",
    "A.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "A['chrom'] = A.index.str.split('_').str[0]\n",
    "A['pos'] =  A.index.str.split('_').str[1]\n",
    "A['chrom_pos'] = A['chrom'].str[3:]+'_'+A['pos']\n",
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for a in A[A['truth'].isna()]['chrom_pos'].values:\n",
    "    #print(evaluation_bed_hg38_df.loc[a].values[0])\n",
    "    A.loc[(A['chrom_pos'] == a), 'truth'] = evaluation_bed_hg38_df.loc[a].values[0]\n",
    "A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_auprc_df = metric_curve_allchr(config, A, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_precision_df = metric_curve_allchr(config, A, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_recall_df = metric_curve_allchr(config, A, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1precision_df = metric_curve_allchr(config, A, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1recall_df = metric_curve_allchr(config, A, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "figure_curve_allchr(config, A, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part I: (1) Load/Generate call tables, (2) Generate Ground truths and (3) Compute/Save metrics per patient"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for mixtureid in mixtureids:\n",
    "mixtureid = 'ROC2_ST21_25ng_LIB1-PH'\n",
    "seriesorder = ['SampleDf', 'SampleEf']\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg38.vcf'))\n",
    "#groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_ROC_hg38.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "#bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['truth'] = False\n",
    "calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDf_ROC_score'] = calltablesseries['SampleDf_ROC_vaf']\n",
    "calltablesseries['SampleEf_ROC_score'] = calltablesseries['SampleEf_ROC_vaf']\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'ROC']\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#for mixtureid in mixtureids:\n",
    "mixtureid = 'BRP2_ST26_25ng_LIB1-P'\n",
    "seriesorder = ['SampleDf', 'SampleEf']\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_BRP_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['truth'] = False\n",
    "calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDf_BRP_score'] = calltablesseries['SampleDf_BRP_vaf']\n",
    "calltablesseries['SampleEf_BRP_score'] = calltablesseries['SampleEf_BRP_vaf']\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'BRP']\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixtureids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "evaluation_bed_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fig, ax = plt.subplots(figsize=(10, 10))\n",
    "res = {}\n",
    "mixtureid = 'ROC2_ST21_25ng_LIB1-PH'\n",
    "#for mixtureid in mixtureids:\n",
    "#mixtureid = 'BRP2_ST26_25ng_LIB1-P'\n",
    "method = mixtureid.split('_')[0][:-1]\n",
    "suffix = 'lS' if method == 'IDT' else ''\n",
    "seriesorder = ['SampleDf'+suffix, 'SampleEf'+suffix]\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "if mixtureid == 'ROC2_ST21_25ng_LIB1-PH':\n",
    "    evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg38.csv'), index_col=0)\n",
    "    evaluation_bed_df.index = ['chr'+i for i in list(evaluation_bed_df.index)]\n",
    "    print(evaluation_bed_df.shape[0])\n",
    "else:\n",
    "    evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg19.csv'), index_col=0)\n",
    "    print(evaluation_bed_df.shape[0])\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['chrom_pos'] = calltablesseries['chrom'].astype('str').str.cat(calltablesseries['pos'].astype('str'), sep=\"_\")\n",
    "print(calltablesseries.shape[0])\n",
    "\n",
    "calltablesseries.reset_index(inplace=True)\n",
    "calltablesseries.set_index('chrom_pos', inplace=True)\n",
    "calltablesseries = calltablesseries.loc[~calltablesseries.index.duplicated()] ### TODO improve, merge pred for instance for '9_21971120'\n",
    "calltablesseries = calltablesseries.reindex(evaluation_bed_df.index)\n",
    "print(calltablesseries.shape[0])\n",
    "calltablesseries['truth'] = evaluation_bed_df['truth']\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDf'+suffix+'_'+method+'_score'] = calltablesseries['SampleDf'+suffix+'_'+method].fillna(0).astype(int)\n",
    "calltablesseries['SampleEf'+suffix+'_'+method+'_score'] = calltablesseries['SampleEf'+suffix+'_'+method].fillna(0).astype(int)\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method]\n",
    "for m in methods:\n",
    "    calltablesseries['SampleDf'+suffix+'_'+m+'_score'].fillna(0, inplace=True)\n",
    "    calltablesseries['SampleEf'+suffix+'_'+m+'_score'].fillna(0, inplace=True)\n",
    "#results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_maxf1_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "pr = figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods, splitby='dilution', plot='partial', figax=None)\n",
    "res[mixtureid] = pr\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixtureids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fig, ax = plt.subplots(figsize=(10, 10))\n",
    "res = {}\n",
    "res_auprc = {}\n",
    "res_maxrecallatleast010precision = {}\n",
    "res_maxrecallatleast050precision = {}\n",
    "res_maxrecallatleast090precision = {}\n",
    "\n",
    "\n",
    "for mixtureid in mixtureids:\n",
    "    reload = False\n",
    "    save = False\n",
    "    #mixtureid = 'BRP2_ST26_25ng_LIB1-P'\n",
    "    method = mixtureid.split('_')[0][:-1]\n",
    "    suffix = 'lS' if method == 'IDT' else ''\n",
    "    seriesorder = ['SampleDf'+suffix, 'SampleEf'+suffix]\n",
    "    print('############# {} ############'.format(mixtureid))\n",
    "    calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "    calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam='PASSREJECT', reload=reload, save=reload, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "    calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam='PASSREJECT', reload=reload, save=reload, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "    calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam='PASSREJECT', reload=reload, save=reload, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "    print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "    print(aux)\n",
    "    calltables['snv'].append(calltable_snv)\n",
    "    calltables['indel'].append(calltable_indel)\n",
    "    calltables['snp'].append(calltable_snp)\n",
    "    calltables['sampleid'] = mixtureid \n",
    "    #calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "    calltables['snv'] = pd.concat(calltables['snv'])\n",
    "    calltables['indel'] = pd.concat(calltables['indel'])\n",
    "    calltables['snp'] = pd.concat(calltables['snp'])\n",
    "    #dilutionseries\n",
    "    dilutionseries = aux\n",
    "    #dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "    muttype = 'snv'\n",
    "    refsample = 'SEQC2'\n",
    "    # ground truth\n",
    "    #if mixtureid == 'ROC2_ST21_25ng_LIB1-PH':\n",
    "        #evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg38.csv'), index_col=0)\n",
    "        #evaluation_bed_df.index = ['chr'+i for i in list(evaluation_bed_df.index)]\n",
    "        #print(evaluation_bed_df.shape[0])\n",
    "    #else:\n",
    "        #evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg19.csv'), index_col=0)\n",
    "        #print(evaluation_bed_df.shape[0])\n",
    "        # load bedfile for method BRP\n",
    "    genomeref = 'hg19' if mixtureid != 'ROC2_ST21_25ng_LIB1-PH' else 'hg38'\n",
    "    print(genomeref)\n",
    "    bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'intersect_BRP_IDT_ILM_ROC_'+genomeref+'.bed'), sep='\\t', header=None)\n",
    "    bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "    if genomeref == 'hg19':\n",
    "        bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "    # load ground truths (cerified mutations either somatic or diluted germline in the sample A)\n",
    "    groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_'+genomeref+'.vcf'))\n",
    "    if genomeref == 'hg19':\n",
    "        groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "    groundtruths['chrom_pos'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_')\n",
    "    groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "    groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "    print(groundtruths.shape[0])\n",
    "    bedfilelistall = []\n",
    "    for il in range(bedfile.shape[0]):\n",
    "        for i in list(np.arange(bedfile.startpos.iloc[il], bedfile.endpos.iloc[il])):\n",
    "            bedfilelistall.append(bedfile.chrom.iloc[il]+'_'+str(i))\n",
    "    print(len(bedfilelistall))\n",
    "    evaluation_bed_df = pd.DataFrame(index=bedfilelistall)\n",
    "    evaluation_bed_df['truth'] = False\n",
    "    gtchrompos = set(set(groundtruths['chrom_pos'].values) & set(evaluation_bed_df.index))\n",
    "    evaluation_bed_df.loc[gtchrompos, 'truth'] = True\n",
    "    print(evaluation_bed_df.value_counts())\n",
    "    calltablesseries = calltables[muttype]\n",
    "    calltablesseries['chrom_pos'] = calltablesseries['chrom'].astype('str').str.cat(calltablesseries['pos'].astype('str'), sep=\"_\")\n",
    "    print(calltablesseries.shape[0])\n",
    "\n",
    "    calltablesseries.reset_index(inplace=True)\n",
    "    calltablesseries.set_index('chrom_pos', inplace=True)\n",
    "    calltablesseries = calltablesseries.loc[~calltablesseries.index.duplicated()] ### TODO improve, merge pred for instance for '9_21971120'\n",
    "    calltablesseries = calltablesseries.reindex(evaluation_bed_df.index)\n",
    "    print(calltablesseries.shape[0])\n",
    "    calltablesseries['truth'] = evaluation_bed_df['truth']\n",
    "    print(calltablesseries['truth'].value_counts())\n",
    "    calltablesseries['SampleDf'+suffix+'_'+method+'_score'] = calltablesseries['SampleDf'+suffix+'_'+method].fillna(0).astype(int)\n",
    "    calltablesseries['SampleEf'+suffix+'_'+method+'_score'] = calltablesseries['SampleEf'+suffix+'_'+method].fillna(0).astype(int)\n",
    "    # metrics\n",
    "    methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method]\n",
    "    for m in methods:\n",
    "        calltablesseries['SampleDf'+suffix+'_'+m+'_score'].fillna(0, inplace=True)\n",
    "        calltablesseries['SampleEf'+suffix+'_'+m+'_score'].fillna(0, inplace=True)\n",
    "    results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                     refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    results_maxrecallatlast0_10precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_10precision', ground_truth_method='SEQC2',\n",
    "                                      refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    results_maxrecallatlast0_50precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_50precision', ground_truth_method='SEQC2',\n",
    "                                      refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    results_maxrecallatlast0_90precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxrecallatleast0_90precision', ground_truth_method='SEQC2',\n",
    "                                      refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #results_maxf1_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1',  ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    #results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "    #                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "    pr = figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                     refsample='SEQC2', muttype=muttype, methods=methods, splitby='dilution', plot='all', figax=None)\n",
    "    res[mixtureid] = pr\n",
    "    res_auprc[mixtureid] = results_auprc_df\n",
    "    res_maxrecallatleast010precision[mixtureid] = results_maxrecallatlast0_10precision_df\n",
    "    res_maxrecallatleast050precision[mixtureid] = results_maxrecallatlast0_50precision_df\n",
    "    res_maxrecallatleast090precision[mixtureid] = results_maxrecallatlast0_90precision_df\n",
    "\n",
    "                    #    ground_truth_method=gtm,\n",
    "                    #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#fig, ax = plt.subplots(figsize=(10, 10))\n",
    "res = {}\n",
    "mixtureid = 'ROC2_ST21_25ng_LIB1-PH'\n",
    "reload = True\n",
    "save = True\n",
    "filterparam = 'PASSREJECT' # 'PASSREJECT'\n",
    "#for mixtureid in mixtureids:\n",
    "#mixtureid = 'BRP2_ST26_25ng_LIB1-P'\n",
    "method = mixtureid.split('_')[0][:-1]\n",
    "suffix = 'lS' if method == 'IDT' else ''\n",
    "seriesorder = ['SampleDf'+suffix, 'SampleEf'+suffix]\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=reload, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=reload, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=reload, diltype='SEQC2', concat='tf', bcbiovaf=0.01, gatkcorr=False)\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "#if mixtureid == 'ROC2_ST21_25ng_LIB1-PH':\n",
    "    #evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg38.csv'), index_col=0)\n",
    "    #evaluation_bed_df.index = ['chr'+i for i in list(evaluation_bed_df.index)]\n",
    "    #print(evaluation_bed_df.shape[0])\n",
    "#else:\n",
    "    #evaluation_bed_df = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'evaluationbed_hg19.csv'), index_col=0)\n",
    "    #print(evaluation_bed_df.shape[0])\n",
    "    # load bedfile for method BRP\n",
    "genomeref = 'hg19' if mixtureid != 'ROC2_ST21_25ng_LIB1-PH' else 'hg38'\n",
    "print(genomeref)\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'intersect_BRP_IDT_ILM_ROC_'+genomeref+'.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "if genomeref == 'hg19':\n",
    "    bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "# load ground truths (cerified mutations either somatic or diluted germline in the sample A)\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_'+genomeref+'.vcf'))\n",
    "if genomeref == 'hg19':\n",
    "    groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "print(groundtruths.shape[0])\n",
    "bedfilelistall = []\n",
    "for il in range(bedfile.shape[0]):\n",
    "    for i in list(np.arange(bedfile.startpos.iloc[il], bedfile.endpos.iloc[il])):\n",
    "        bedfilelistall.append(bedfile.chrom.iloc[il]+'_'+str(i))\n",
    "print(len(bedfilelistall))\n",
    "evaluation_bed_df = pd.DataFrame(index=bedfilelistall)\n",
    "evaluation_bed_df['truth'] = False\n",
    "gtchrompos = set(set(groundtruths['chrom_pos'].values) & set(evaluation_bed_df.index))\n",
    "evaluation_bed_df.loc[gtchrompos, 'truth'] = True\n",
    "print(evaluation_bed_df.value_counts())\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['chrom_pos'] = calltablesseries['chrom'].astype('str').str.cat(calltablesseries['pos'].astype('str'), sep=\"_\")\n",
    "print(calltablesseries.shape[0])\n",
    "\n",
    "calltablesseries.reset_index(inplace=True)\n",
    "calltablesseries.set_index('chrom_pos', inplace=True)\n",
    "calltablesseries = calltablesseries.loc[~calltablesseries.index.duplicated()] ### TODO improve, merge pred for instance for '9_21971120'\n",
    "calltablesseries = calltablesseries.reindex(evaluation_bed_df.index)\n",
    "print(calltablesseries.shape[0])\n",
    "calltablesseries['truth'] = evaluation_bed_df['truth']\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDf'+suffix+'_'+method+'_score'] = calltablesseries['SampleDf'+suffix+'_'+method].fillna(0).astype(int)\n",
    "calltablesseries['SampleEf'+suffix+'_'+method+'_score'] = calltablesseries['SampleEf'+suffix+'_'+method].fillna(0).astype(int)\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', method]\n",
    "for m in methods:\n",
    "    calltablesseries['SampleDf'+suffix+'_'+m+'_score'].fillna(0, inplace=True)\n",
    "    calltablesseries['SampleEf'+suffix+'_'+m+'_score'].fillna(0, inplace=True)\n",
    "#results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_maxf1_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "#results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "#                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "pr = figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods, splitby='dilution', plot='all', figax=None)\n",
    "res[mixtureid] = pr\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def function_to_split_ls(hand,labl,dividor):\n",
    "\n",
    "    Hand_L=[]\n",
    "    Hand_M=[]\n",
    "    Labl_L=[]\n",
    "    Labl_M=[]\n",
    "\n",
    "    for h,l in zip(hand,labl):\n",
    "        co=h.get_color()\n",
    "        ls=h.get_linestyle()\n",
    "        lw=h.get_linewidth()\n",
    "        mk=h.get_marker()\n",
    "        mew=h.get_markeredgewidth()\n",
    "        ms=h.get_markersize()\n",
    "\n",
    "        LABS=l.split(dividor)\n",
    "\n",
    "        if len(LABS) != 2:\n",
    "            print('Split Legends Error: Only exactly 1 Dividor is accepted.')\n",
    "            print('                     Currently ' + str(len(LABS)-1) + ' dividors were given')\n",
    "            return hand,labl\n",
    "\n",
    "        #Color\n",
    "        LICO = plt.Line2D((0,1),(0,0), color=co, marker='', linestyle='solid',linewidth=lw)\n",
    "        #Linestyle\n",
    "        STYLE = plt.Line2D((0,1),(0,0), color='k', linestyle=ls)\n",
    "\n",
    "        if LABS[0] not in Labl_L:\n",
    "            Hand_L.append(LICO)\n",
    "            Labl_L.append(LABS[0])\n",
    "\n",
    "        if LABS[1] not in Labl_M:\n",
    "            Hand_M.append(STYLE)\n",
    "            Labl_M.append(LABS[1])\n",
    "\n",
    "    return Hand_L+Hand_M,Labl_L+Labl_M"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "ls = ['solid', 'dashdot', (0, (5, 1)), (0, (1,1))]\n",
    "for pref in ['Df', 'Ef']: # sample Df and sample Ef\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))\n",
    "    plt.grid()\n",
    "    for si, s in enumerate(list(res.keys())):\n",
    "        print(s, [k for k in list(res[s].keys()) if pref in k][0])\n",
    "        aux = res[s][[k for k in list(res[s].keys()) if pref in k][0]]\n",
    "        for k, v in aux.items():\n",
    "            recall, precision = v[0], v[1]\n",
    "            kwargs = {}\n",
    "            kwargs[\"label\"] = k + ' * ' + s\n",
    "            print(kwargs[\"label\"])\n",
    "            kwargs[\"color\"] = config.colors[config.methods.index(k)]\n",
    "            #kwargs[\"marker\"] = markers[si]\n",
    "            kwargs[\"linestyle\"] = ls[si]\n",
    "            kwargs[\"lw\"] = 3\n",
    "            # print(precision, recall)\n",
    "            if (len(recall) == 2) and (recall[0] == 1) and (recall[1] != 0):\n",
    "                #ax.scatter(recall[1], precision[1],  **kwargs)\n",
    "                kwargs[\"markersize\"]=15\n",
    "                ax.plot(recall[1], precision[1],  **kwargs)\n",
    "            else:\n",
    "                kwargs[\"drawstyle\"] = \"steps-post\"\n",
    "                kwargs[\"markersize\"]=10\n",
    "                ax.plot(recall, precision, **kwargs)\n",
    "    #plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    plt.semilogx()\n",
    "    plt.xlim([0.01, 1.01])\n",
    "    ax = plt.gca()\n",
    "    hand, labl = ax.get_legend_handles_labels()\n",
    "    hand, labl = function_to_split_ls(hand, labl, '*')\n",
    "    ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    print(k, v)\n",
    "    plt.title('Sample'+pref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for k, v in res_auprc.items():\n",
    "    res_auprc[k]['sampleid'] = k\n",
    "res_auprc_tot = pd.concat(res_auprc.values())\n",
    "res_auprc_tot['samplename'] = res_auprc_tot['samplename'].str.replace('lS', '')\n",
    "res_auprc_tot\n",
    "\n",
    "sns.catplot(x='samplename', y=\"AUPRC score\", hue=\"caller\",\n",
    "                capsize=.1, height=6, aspect=1.5, kind=\"point\",\n",
    "                order=sorted(res_auprc_tot['samplename'].unique()),\n",
    "                palette=[config.colors[config.methods.index(m)] for m in list(res_auprc_tot['caller'].unique())], data=res_auprc_tot)\n",
    "\n",
    "for k, v in res_maxrecallatleast010precision.items():\n",
    "    res_maxrecallatleast010precision[k]['sampleid'] = k\n",
    "res_maxrecallatleast010precision_tot = pd.concat(res_maxrecallatleast010precision.values())\n",
    "res_maxrecallatleast010precision_tot['samplename'] = res_maxrecallatleast010precision_tot['samplename'].str.replace('lS', '')\n",
    "res_maxrecallatleast010precision_tot\n",
    "\n",
    "sns.catplot(x='samplename', y=\"MAXRECALLATLEAST0_10PRECISION score\", hue=\"caller\",\n",
    "                capsize=.1, height=6, aspect=1.5, kind=\"point\",\n",
    "                order=sorted(res_maxrecallatleast010precision_tot['samplename'].unique()),\n",
    "                palette=[config.colors[config.methods.index(m)] for m in list(res_maxrecallatleast010precision_tot['caller'].unique())], data=res_maxrecallatleast010precision_tot)\n",
    "\n",
    "\n",
    "for k, v in res_maxrecallatleast050precision.items():\n",
    "    res_maxrecallatleast050precision[k]['sampleid'] = k\n",
    "res_maxrecallatleast050precision_tot = pd.concat(res_maxrecallatleast050precision.values())\n",
    "res_maxrecallatleast050precision_tot['samplename'] = res_maxrecallatleast050precision_tot['samplename'].str.replace('lS', '')\n",
    "res_maxrecallatleast050precision_tot\n",
    "\n",
    "sns.catplot(x='samplename', y=\"MAXRECALLATLEAST0_50PRECISION score\", hue=\"caller\",\n",
    "                capsize=.1, height=6, aspect=1.5, kind=\"point\",\n",
    "                order=sorted(res_maxrecallatleast050precision_tot['samplename'].unique()),\n",
    "                palette=[config.colors[config.methods.index(m)] for m in list(res_maxrecallatleast050precision_tot['caller'].unique())], data=res_maxrecallatleast050precision_tot)\n",
    "\n",
    "\n",
    "for k, v in res_maxrecallatleast090precision.items():\n",
    "    res_maxrecallatleast090precision[k]['sampleid'] = k\n",
    "res_maxrecallatleast090precision_tot = pd.concat(res_maxrecallatleast090precision.values())\n",
    "res_maxrecallatleast090precision_tot['samplename'] = res_maxrecallatleast090precision_tot['samplename'].str.replace('lS', '')\n",
    "res_maxrecallatleast090precision_tot\n",
    "\n",
    "sns.catplot(x='samplename', y=\"MAXRECALLATLEAST0_90PRECISION score\", hue=\"caller\",\n",
    "                capsize=.1, height=6, aspect=1.5, kind=\"point\",\n",
    "                order=sorted(res_maxrecallatleast090precision_tot['samplename'].unique()),\n",
    "                palette=[config.colors[config.methods.index(m)] for m in list(res_maxrecallatleast090precision_tot['caller'].unique())], data=res_maxrecallatleast090precision_tot)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries['truth'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "markers = ['D', 's', 'o', '^']\n",
    "for pref in ['Df', 'Ef']: # sample Df and sample Ef\n",
    "    fig, ax = plt.subplots(figsize=(10, 10))\n",
    "    plt.grid()\n",
    "    for si, s in enumerate(list(res.keys())):\n",
    "        print(s, [k for k in list(res[s].keys()) if pref in k][0])\n",
    "        aux = res[s][[k for k in list(res[s].keys()) if pref in k][0]]\n",
    "        for k, v in aux.items():\n",
    "            recall, precision = v[0], v[1]\n",
    "            kwargs = {}\n",
    "            kwargs[\"label\"] = k + ' * ' + s\n",
    "            print(kwargs[\"label\"])\n",
    "            kwargs[\"color\"] = config.colors[config.methods.index(k)]\n",
    "            kwargs[\"marker\"] = markers[si]\n",
    "            kwargs[\"lw\"] = 3\n",
    "            print(precision, recall)\n",
    "            if (len(recall) == 2) and (recall[0] == 1) and (recall[1] != 0):\n",
    "                #ax.scatter(recall[1], precision[1],  **kwargs)\n",
    "                kwargs[\"markersize\"]=15\n",
    "                ax.plot(recall[1], precision[1],  **kwargs)\n",
    "            else:\n",
    "                kwargs[\"drawstyle\"] = \"steps-post\"\n",
    "                kwargs[\"markersize\"]=10\n",
    "                ax.plot(recall, precision, **kwargs)\n",
    "    #plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    #plt.semilogx()\n",
    "    #plt.xlim([0.01, 1.01])\n",
    "    ax = plt.gca()\n",
    "    hand, labl = ax.get_legend_handles_labels()\n",
    "    hand, labl = function_to_split(hand, labl, '*')\n",
    "    ax.legend(hand, labl, bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    print(k, v)\n",
    "    plt.title('Sample'+pref)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries[['SampleDf_'+m+'_score' for m in config.methods] + ['truth']].tail(25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#calltablesseries.reset_index(inplace=True)\n",
    "#calltablesseries.set_index('chrom_pos', inplace=True)\n",
    "calltablesseries.loc[calltablesseries.index[calltablesseries.index.duplicated()]]\n",
    "#calltablesseries.reindex(evaluation_bed_df.index)\n",
    "#calltablesseries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixtureid = 'ILM2_ST29_25ng_LIB1-P'\n",
    "seriesorder = ['SampleDf', 'SampleEf']\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_ILM_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['truth'] = False\n",
    "calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDf_ILM_score'] = calltablesseries['SampleDf_ILM_vaf']\n",
    "calltablesseries['SampleEf_ILM_score'] = calltablesseries['SampleEf_ILM_vaf']\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'ILM']\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixtureid = 'IDT2_ST05_25ng_LIB1-P'\n",
    "seriesorder = ['SampleDflS', 'SampleEflS']\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_IDT_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene', 'a','b']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['truth'] = False\n",
    "calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDflS_IDT_score'] = calltablesseries['SampleDflS_IDT_vaf']\n",
    "calltablesseries['SampleEflS_IDT_score'] = calltablesseries['SampleEflS_IDT_vaf']\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'IDT']\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mixtureid = 'TFS2_ST24_25ng_LIB1-P'\n",
    "seriesorder = ['SampleDf', 'SampleEf']\n",
    "print('############# {} ############'.format(mixtureid))\n",
    "calltables = {'sampleid':[], 'tf':[], 'cov':[], 'snv':[], 'indel':[], 'snp':[]}\n",
    "calltable_snv, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snv', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_indel, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='indel', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "calltable_snp, aux = get_calltableseries(config, mixtureid, chrom='all', muttype='snp', filterparam=filterparam, reload=reload, save=save, diltype='SEQC2', concat='tf')\n",
    "print(calltable_snv.shape, calltable_indel.shape, calltable_snp.shape)\n",
    "print(aux)\n",
    "calltables['snv'].append(calltable_snv)\n",
    "calltables['indel'].append(calltable_indel)\n",
    "calltables['snp'].append(calltable_snp)\n",
    "calltables['sampleid'] = mixtureid \n",
    "#calltables['tf'] = np.unique([cn.split('_')[0] for cn in list(calltable_snv.columns)])[:-5].astype(float)\n",
    "calltables['snv'] = pd.concat(calltables['snv'])\n",
    "calltables['indel'] = pd.concat(calltables['indel'])\n",
    "calltables['snp'] = pd.concat(calltables['snp'])\n",
    "#dilutionseries\n",
    "dilutionseries = aux\n",
    "#dilutionseries = aux.T[['mixture_' + '_'.join(mixtureid.split('_')[:2]) + '_' + str(s[0]) + 'x_' + '_'.join(mixtureid.split('_')[2:4]) + '_' + str(s[1]) + 'x' for s in seriesorder]].T\n",
    "muttype = 'snv'\n",
    "refsample = 'SEQC2'\n",
    "# ground truth\n",
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_TFS_pre-defined_Hotspots_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene', 'a','b', 'c', 'd']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')\n",
    "\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])\n",
    "calltablesseries = calltables[muttype]\n",
    "calltablesseries['truth'] = False\n",
    "calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth'] = True\n",
    "print(calltablesseries['truth'].value_counts())\n",
    "calltablesseries['SampleDf_TFS_score'] = calltablesseries['SampleDf_TFS_vaf']\n",
    "calltablesseries['SampleEf_TFS_score'] = calltablesseries['SampleEf_TFS_vaf']\n",
    "# metrics\n",
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'TFS']\n",
    "results_auprc_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='auprc', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "results_maxf1recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='maxf1recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "figure_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, xy='pr', ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=methods)\n",
    "                #    ground_truth_method=gtm,\n",
    "                #    refsample=refsample, muttype=muttype.upper(), methods=None, fixedvar=fixedvar, save=save)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries[(calltablesseries['truth']) | (calltablesseries[['SampleDf_'+m+'_score' for m in ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan']]].isna().sum() < 5) | (calltablesseries['SampleDf_TFS_score'] > 0)][['SampleDf_'+m+'_score' for m in ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan', 'TFS']]+['truth']]#, 'truth']]#.sum(axis=1).value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "results_precision_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='precision',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=config.methods)\n",
    "results_recall_df = metric_curve_allchr(config, calltablesseries, dilutionseries, mixtureid, metric='recall',  ground_truth_method='SEQC2',\n",
    "                                 refsample='SEQC2', muttype=muttype, methods=config.methods)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltablesseries.loc[calltablesseries.index.isin(list(groundtruths_in_bed.index)), 'truth']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# From Excel sheet "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = {}\n",
    "m = 'BRP'\n",
    "\n",
    "for lab in ['25', '26']:\n",
    "    for rep in ['1', '2', '3', '4']:\n",
    "        \n",
    "        A = pd.read_excel(os.path.join(*config.mixturefolderSEQC2, 'excelreport', '41587_2021_857_MOESM6_ESM.xlsx'),  'BRP_Ef25', index_col=0)\n",
    "        A.index = [str(a[3:]) for a in list(A.index)]\n",
    "        A = A[['On-target', 'Known position', 'TP', 'FP', 'SampleB', 'BRP2_Ef25_ST'+lab+'_R'+rep]]\n",
    "        print(A.shape)\n",
    "\n",
    "        # load table\n",
    "        calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_BRP2_ST'+lab+'_25ng_LIB'+rep+'-P', 'SampleEf_BRP2_ST'+lab+'_25ng_LIB'+rep+'-P'), ['varnet', 'BRP'], save=False, filter='PASS')\n",
    "        calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "        calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "\n",
    "        B = calltable\n",
    "        B = B[['BRP', 'BRP_score', 'BRP_vaf', 'BRP_altcov', 'BRP_totcov']]\n",
    "        B.index = B.index.astype(str)\n",
    "        print(B.shape)\n",
    "        \n",
    "        C = pd.concat([A, B], axis=1)\n",
    "        Nneg = C[C['FP'] == 1].shape[0]\n",
    "        print('Nneg')\n",
    "        print(Nneg)\n",
    "        #C = C[(C['BRP_vaf'] <= 0.025) & (C['BRP_vaf'] > 0.001)]\n",
    "        C['truth'] = C[\"TP\"]\n",
    "        C.dropna(inplace=True)\n",
    "        print(C.shape)\n",
    "        C['BRP_vaf_range'] = pd.cut(C['BRP_vaf'], [0.001, 0.002, 0.003, 0.005, 0.025])\n",
    "        C['BRP_vaf_range_num'] = 0\n",
    "        C['BRP_vaf_range_num'][C['BRP_vaf'] < 0.001] = 0\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.0025) & (C['BRP_vaf'] > 0.001)] = 0.0025\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.003) & (C['BRP_vaf'] > 0.0025)] = 0.003\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.004) & (C['BRP_vaf'] > 0.003)] = 0.004\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.005) & (C['BRP_vaf'] > 0.004)] = 0.005\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.01) & (C['BRP_vaf'] > 0.005)] = 0.01\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.015) & (C['BRP_vaf'] > 0.01)] = 0.015\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.02) & (C['BRP_vaf'] > 0.015)] = 0.02\n",
    "        C['BRP_vaf_range_num'][(C['BRP_vaf'] <= 0.025) & (C['BRP_vaf'] > 0.02)] = 0.025\n",
    "        C['BRP_vaf_range_num'][C['BRP_vaf'] > 0.025] = 1\n",
    "        print(C['BRP_vaf_range_num'].value_counts())\n",
    "    \n",
    "        fig, ax = plt.subplots(figsize=(10, 10))\n",
    "        #precision, recall, thresholds = precision_recall_curve(C['truth'], C['BRP_vaf_range_num'])\n",
    "        precision, recall, thresholds = precision_recall_curve(C['truth'], C['BRP_vaf_range_num'])\n",
    "        print(thresholds)\n",
    "        precisionbis = []\n",
    "        for thres in [0] + list(thresholds):\n",
    "            y_pred = C[m+'_vaf'].copy()\n",
    "            y_pred[y_pred >= thres] = 1\n",
    "            y_pred[y_pred != 1] = 0\n",
    "            _, fp, _, tp = confusion_matrix(C['truth'], y_pred).ravel()\n",
    "            precisionbis.append(tp/(tp+(fp/Nneg)))\n",
    "        \n",
    "        m = 'BRP'\n",
    "        plot_pr_curve(precision[1:-1], recall[1:-1], estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "        plt.ylim([0.0, 1.0])\n",
    "        plt.xlim([0.0, 1.0])\n",
    "        plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "\n",
    "        #res['BRP2_Ef25_ST'+lab+'_R'+rep] = (precisionbis, recall, thresholds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = {}\n",
    "m = 'BRP'\n",
    "recallall, precisionall = [], []\n",
    "Call = pd.DataFrame()\n",
    "for lab in ['25', '26']:\n",
    "    for rep in ['1', '2', '3', '4']:\n",
    "        A = pd.read_excel(os.path.join(*config.mixturefolderSEQC2, 'excelreport', '41587_2021_857_MOESM6_ESM.xlsx'),  'BRP_Ef25', index_col=0)\n",
    "        A.index = [str(a[3:]) for a in list(A.index)]\n",
    "        A = A[['On-target', 'Known position', 'TP', 'FP', 'SampleB', 'BRP2_Ef25_ST'+lab+'_R'+rep]]\n",
    "        print(A.shape)\n",
    "        # load table\n",
    "        calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_BRP2_ST'+lab+'_25ng_LIB'+rep+'-P', 'SampleEf_BRP2_ST'+lab+'_25ng_LIB'+rep+'-P'), ['varnet', 'BRP'], save=False, filter='all')\n",
    "        calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "        calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "        B = calltable\n",
    "        B = B[['BRP', 'BRP_score', 'BRP_vaf', 'BRP_altcov', 'BRP_totcov']]\n",
    "        B.index = B.index.astype(str)\n",
    "        print(B.shape)\n",
    "        C = pd.concat([A, B], axis=1)\n",
    "        print(C.shape)\n",
    "        Call = pd.concat([Call, C])\n",
    "\n",
    "        Call = C\n",
    "        Call['BRP_vaf_filter_2.5%'] = (Call['BRP_vaf'] > 0.025)\n",
    "        Call['BRP_vaf_filter_1.5%'] = (Call['BRP_vaf'] > 0.015)\n",
    "        Call['BRP_vaf_filter_1.0%'] = (Call['BRP_vaf'] > 0.01)\n",
    "        Call['BRP_vaf_filter_0.5%'] = (Call['BRP_vaf'] > 0.005)\n",
    "        Call['BRP_vaf_filter_0.4%'] = (Call['BRP_vaf'] > 0.004)\n",
    "        Call['BRP_vaf_filter_0.3%'] = (Call['BRP_vaf'] > 0.003)\n",
    "        Call['BRP_vaf_filter_0.25%'] = (Call['BRP_vaf'] > 0.0025)\n",
    "        Call['BRP_vaf_filter_0.2%'] = (Call['BRP_vaf'] > 0.002)\n",
    "        Call['BRP_vaf_filter_0.15%'] = (Call['BRP_vaf'] > 0.0015)\n",
    "        Call['BRP_vaf_filter_0.1%'] = (Call['BRP_vaf'] > 0.001)\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(10, 10))\n",
    "        #precision, recall, thresholds = precision_recall_curve(C['truth'], C['BRP_vaf_range_num'])\n",
    "        precision, recall = [], []\n",
    "        thresholds = ['2.5%', '1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.2%', '0.15%', '0.1%']\n",
    "        #thresholds = ['2.5%', '1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.1%']\n",
    "        Call = Call[Call['SampleB'] == 0]\n",
    "        for th in thresholds:\n",
    "            #TP = float(Call[(Call['TP'] == 1) & (Call['BRP_vaf_filter_'+th] == 1)].shape[0])\n",
    "            TP = float(Call[(Call['Known position'] == 1) & (Call['BRP_vaf_filter_'+th] == 1)].shape[0])\n",
    "            #FP = float(Call[(Call['FP'] == 1)].shape[0])\n",
    "            FP = float(Call[(Call['FP'] == 1) & (Call['BRP_vaf_filter_'+th] == 1)].shape[0])\n",
    "            #FP = float(Call[(Call['Known position'] == 0) & (Call['BRP_vaf_filter_'+th] == 1)].shape[0])\n",
    "            FN = float(Call[(Call['TP'] == 1) & (Call['BRP_vaf_filter_'+th] == 0)].shape[0])\n",
    "            #FN = float(Call[(Call['Known position'] == 1) & (Call['BRP_vaf_filter_'+th] == 0)].shape[0])\n",
    "            print(TP, FP, FN)\n",
    "            p = TP/(TP + FP)\n",
    "            r = TP/(TP + FN)\n",
    "            precision.append(p)\n",
    "            recall.append(r)\n",
    "            print(th, p, r)\n",
    "        print(thresholds)\n",
    "        plt.plot(recall, precision, '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "        #plot_pr_curve(precision, recall, estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "        plt.ylim([0.97, 1.0001])\n",
    "        plt.xlim([0.0, 1.0])\n",
    "        plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "        recallall.append(recall)\n",
    "        precisionall.append(precision)\n",
    "        #res['BRP2_Ef25_ST'+lab+'_R'+rep] = (precisionbis, recall, thresholds)\n",
    "\n",
    "precisionall = np.array(precisionall)\n",
    "recallall = np.array(recallall)\n",
    "print(recallall.shape, precisionall.shape)\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "print(np.mean(recallall, axis=0), np.mean(precisionall, axis=0))\n",
    "plt.plot(np.mean(recallall, axis=0), np.mean(precisionall, axis=0), '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "#plot_pr_curve(precision, recall, estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "plt.ylim([0.97, 1.0001])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = 'BRP'\n",
    "recallall, precisionall = [], []\n",
    "\n",
    "A = pd.read_excel(os.path.join(*config.mixturefolderSEQC2, 'excelreport', '41587_2021_857_MOESM6_ESM.xlsx'),  m+'_Ef25', index_col=0)\n",
    "A.index = [str(a[3:]) for a in list(A.index)]\n",
    "print(A.shape)\n",
    "print(A['FP'].sum(), A['TP'].sum(), A['Known position'].sum())\n",
    "#lab = '25'\n",
    "#rep = '1'\n",
    "\n",
    "for lab in ['25', '26']:\n",
    "    for rep in ['1', '2', '3', '4']:\n",
    "        \n",
    "        \n",
    "        calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+m+'2_ST'+lab+'_25ng_LIB'+rep+'-P', 'SampleEf_'+m+'2_ST'+lab+'_25ng_LIB'+rep+'-P'), ['varnet', m], save=False, filter='all')\n",
    "        calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "        calltable.drop(['varnet', 'varnet_score', 'varnet_vaf', 'varnet_altcov', 'varnet_totcov'], axis=1, inplace=True)\n",
    "        calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "        B = calltable\n",
    "        B.rename(columns={m: m+'2_Ef25_ST'+lab+'_R'+rep+'_pass', m+'_score': m+'2_Ef25_ST'+lab+'_R'+rep+'_score', m+'_vaf': m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf', m+'_altcov': m+'2_Ef25_ST'+lab+'_R'+rep+'_altcov',  m+'_totcov' : m+'2_Ef25_ST'+lab+'_R'+rep+'_totcov'}, inplace=True)\n",
    "        B.index = B.index.astype(str)\n",
    "\n",
    "        ### RQ : 4_1803407_T_C missing (SNV number 230 !)\n",
    "\n",
    "        C = pd.concat([A, B], axis=1)\n",
    "\n",
    "        C[(C[m+'2_Ef25_ST'+lab+'_R'+rep].isna()) & ~C[m+'2_Ef25_ST'+lab+'_R'+rep].isna()]\n",
    "        print(C[C['SampleB'] == 1][m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'].min())\n",
    "        C = C[C['SampleB']!=1]\n",
    "        #C = C[C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] <= 0.025]\n",
    "        #C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_2.5%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.025)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_1.5%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.015)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_1.0%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.01)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.5%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.005)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.4%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.004)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.3%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.003)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.25%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.0025)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.2%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.002)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.15%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.0015)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.1%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.001)\n",
    "\n",
    "\n",
    "        precision, recall = [], []\n",
    "        #thresholds = ['2.5%', '1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.2%', '0.15%', '0.1%']\n",
    "        thresholds = ['1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.2%', '0.15%', '0.1%']\n",
    "        for th in thresholds:\n",
    "            TP = float(C[(C['TP'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            #TP = float(C[(C['Known position'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            FP = float(C[(C['FP'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            #FP = float(C[(C['Known position'] == 0) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            FN = float(C[(C['TP'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 0)].shape[0])\n",
    "            #FN = float(C[(C['Known position'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 0)].shape[0])\n",
    "            print(TP, FP, FN)\n",
    "            p = TP/(TP + FP)\n",
    "            r = TP/(TP + FN)\n",
    "            precision.append(p)\n",
    "            recall.append(r)\n",
    "            print(th, round(p, 3), round(r, 3))\n",
    "        plt.plot(recall, precision, '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "        plt.ylim([0.97, 1.0001])\n",
    "        plt.xlim([0.0, 1.0])\n",
    "        plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    \n",
    "        recallall.append(recall)\n",
    "        precisionall.append(precision)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C.columns[12]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "D = C[list(C.columns[:4]) + [C.columns[12]] + list(C.columns[20:])]\n",
    "print(D[(D['BRP2_Ef25_ST26_R4'] == 0) & (D['BRP2_Ef25_ST26_R4_vaf'].isna())].shape[0], D[(D['BRP2_Ef25_ST26_R4'] == 0)].shape[0], D[(D['BRP2_Ef25_ST26_R4_vaf'].isna())].shape[0])\n",
    "D"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "print(np.mean(recallall, axis=0), np.mean(precisionall, axis=0))\n",
    "plt.plot(np.mean(recallall, axis=0)[1:], np.mean(precisionall, axis=0)[1:], '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "#plot_pr_curve(precision, recall, estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "plt.ylim([0.97, 1.0001])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.title(m+' PR curve average')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "groundtruths = read_vcf(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownPositives_hg19.vcf'))\n",
    "groundtruths['CHROM'] = groundtruths['CHROM'].str.replace('chr', '')\n",
    "groundtruths['chrom_pos_ref_alt'] = groundtruths['CHROM'].astype(str).str.cat(groundtruths['POS'].astype('str'), sep='_').str.cat(groundtruths['REF'].astype('str'), sep='_').str.cat(groundtruths['ALT'].astype('str'), sep='_')\n",
    "groundtruths.set_index('chrom_pos_ref_alt', inplace=True)\n",
    "groundtruths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load bedfile for method BRP\n",
    "bedfile = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'bedfiles', 'LBx_BRP_hg19.bed'), sep='\\t', header=None)\n",
    "bedfile.columns = ['chrom', 'startpos', 'endpos', 'gene']\n",
    "bedfile['chrom'] = bedfile['chrom'].str.replace('chr', '')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# keep only ground truths in bed\n",
    "groundtruth_outside_bed = []\n",
    "for idx, row in tqdm(groundtruths.iterrows(), total=groundtruths.shape[0]):\n",
    "    chrom = row['CHROM']\n",
    "    pos = row['POS']\n",
    "    ref = row['REF']\n",
    "    alt = row['ALT']\n",
    "    aux = bedfile[bedfile['chrom'] == str(chrom)]\n",
    "    count = False\n",
    "    for ind in list(aux.index):\n",
    "        #print(aux['endpos'][ind], aux['startpos'][ind])\n",
    "        if aux['endpos'][ind] >= int(pos):\n",
    "            if aux['startpos'][ind] <= int(pos):\n",
    "                count = True\n",
    "    if not count:\n",
    "        groundtruth_outside_bed.append(chrom + '_'+ str(pos) + '_'+ ref + '_'+ alt)\n",
    "print(len(groundtruth_outside_bed))\n",
    "groundtruths_in_bed = groundtruths.drop(groundtruth_outside_bed)\n",
    "print(groundtruths_in_bed.shape[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(set(set(list(C[C['TP'] == 1].index)) & set(groundtruths_in_bed.index))))\n",
    "lother = list(set(set(groundtruths_in_bed.index)) - set(list(C[C['TP'] == 1].index)))\n",
    "print(len(lother))\n",
    "print(lother[1])\n",
    "C[C['TP'] == 0].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bedfile[bedfile['chrom'] == '10']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load known negatives (germline mutations)\n",
    "germlinetruths = pd.read_csv(os.path.join(*config.mixturefolderSEQC2, 'ground_truths', 'KnownNegatives_hg19.bed'), sep='\\t', header=None)\n",
    "germlinetruths.columns = ['chrom', 'startpos', 'endpos']\n",
    "germlinetruths['chrom'] = germlinetruths['chrom'].str.replace('chr', '')\n",
    "germlinetruths['chrom_pos'] = germlinetruths['chrom'].astype(str).str.cat(germlinetruths['startpos'].astype('str'), sep='_')\n",
    "germlinetruths.set_index('chrom_pos', inplace=True)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "germlinetruths.index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "set(C['chrom_pos'].values) & set(germlinetruths.index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(len(set(list(C[C['TP'] == 1].index))))\n",
    "print(len(set(set(list(C[C['TP'] == 1].index)) & set(groundtruths.index))))\n",
    "C.loc[list(set(list(C[C['TP'] != 1].index)) & set(groundtruths.index))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "m = 'IDT'\n",
    "recallall, precisionall = [], []\n",
    "\n",
    "A = pd.read_excel(os.path.join(*config.mixturefolderSEQC2, 'excelreport', '41587_2021_857_MOESM6_ESM.xlsx'),  m+'_Ef25', index_col=0)\n",
    "A.index = [str(a[3:]) for a in list(A.index)]\n",
    "print(A.shape)\n",
    "print(A['FP'].sum(), A['TP'].sum(), A['Known position'].sum())\n",
    "#lab = '25'\n",
    "#rep = '1'\n",
    "for lab in ['05', '06']:\n",
    "    for rep in ['1', '2', '3', '4']:\n",
    "        if lab == '04':\n",
    "            lab = '05'\n",
    "        calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+m+'2_ST'+lab+'_25ng_LIB'+rep+'-P', 'SampleEflS_'+m+'2_ST'+lab+'_25ng_LIB'+rep+'-P'), ['varnet', m], save=False, filter='all')\n",
    "        calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "        calltable.drop(['varnet', 'varnet_score', 'varnet_vaf', 'varnet_altcov', 'varnet_totcov'], axis=1, inplace=True)\n",
    "        calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "        B = calltable\n",
    "        if lab == '05':\n",
    "            lab = '04'\n",
    "        B.rename(columns={m: m+'2_Ef25_ST'+lab+'_R'+rep+'_pass', m+'_score': m+'2_Ef25_ST'+lab+'_R'+rep+'_score', m+'_vaf': m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf', m+'_altcov': m+'2_Ef25_ST'+lab+'_R'+rep+'_altcov',  m+'_totcov' : m+'2_Ef25_ST'+lab+'_R'+rep+'_totcov'}, inplace=True)\n",
    "        B.index = B.index.astype(str)\n",
    "\n",
    "        C = pd.concat([A, B], axis=1)\n",
    "\n",
    "        C[(C[m+'2_Ef25_ST'+lab+'_R'+rep].isna()) & ~C[m+'2_Ef25_ST'+lab+'_R'+rep].isna()]\n",
    "        print(C[C['SampleB'] == 1][m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'].min())\n",
    "        print(C[C['SampleB']==1].shape[0])\n",
    "        C = C[C['SampleB']!=1]\n",
    "        C = C[C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] <= 0.025]\n",
    "        print(C.shape)\n",
    "        #C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_2.5%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.025)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_1.5%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.015)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_1.0%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.01)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.5%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.005)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.4%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.004)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.3%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.003)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.25%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.0025)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.2%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.002)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.15%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.0015)\n",
    "        C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_0.1%'] = (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'] > 0.001)\n",
    "\n",
    "\n",
    "        precision, recall = [], []\n",
    "        #thresholds = ['2.5%', '1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.2%', '0.15%', '0.1%']\n",
    "        thresholds = ['1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.2%', '0.15%', '0.1%']\n",
    "        for th in thresholds:\n",
    "            TP = float(C[(C['TP'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            #TP = float(C[(C['Known position'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            FP = float(C[(C['FP'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            #FP = float(C[(C['Known position'] == 0) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            FN = float(C[(C['TP'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 0)].shape[0])\n",
    "            #FN = float(C[(C['Known position'] == 1) & (C[m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf_filter_'+th] == 0)].shape[0])\n",
    "            print(TP, FP, FN)\n",
    "            p = TP/(TP + FP)\n",
    "            r = TP/(TP + FN)\n",
    "            precision.append(p)\n",
    "            recall.append(r)\n",
    "            print(th, round(p, 3), round(r, 3))\n",
    "        plt.plot(recall, precision, '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "        #plt.ylim([0.97, 1.0001])\n",
    "        plt.xlim([0.0, 1.0])\n",
    "        plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "    \n",
    "        recallall.append(recall)\n",
    "        precisionall.append(precision)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[(C[m+'2_Ef25_ST'+lab+'_R'+rep] == 1) & C[m+'2_Ef25_ST'+lab+'_R'+rep+\"_vaf\"].isna()][[m+'2_Ef25_ST'+lab+'_R'+rep, m+'2_Ef25_ST'+lab+'_R'+rep+\"_vaf\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[C['chrom'] == '12'].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[(C[m+'2_Ef25_ST'+lab+'_R'+rep].isna()) & ~C[m+'2_Ef25_ST'+lab+'_R'+rep+\"_vaf\"].isna()][[m+'2_Ef25_ST'+lab+'_R'+rep, m+'2_Ef25_ST'+lab+'_R'+rep+\"_vaf\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "print(np.mean(recallall, axis=0), np.mean(precisionall, axis=0))\n",
    "plt.plot(np.mean(recallall, axis=0)[1:], np.mean(precisionall, axis=0)[1:], '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "#plot_pr_curve(precision, recall, estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "#plt.ylim([0.97, 1.0001])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.title(m+' PR curve average')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[C['SampleB'] == 1][m+'2_Ef25_ST'+lab+'_R'+rep+'_vaf'].min()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[(C['BRP2_Ef25_ST25_R1'] == 0) & ~C['BRP2_Ef25_ST25_R1_vaf'].isna()]#.dropna()\n",
    "C[(C['BRP2_Ef25_ST25_R1'] == 1) & C['BRP2_Ef25_ST25_R1_vaf'].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = {}\n",
    "m = 'IDT'\n",
    "precisionall, recallall = [], []\n",
    "Call = pd.DataFrame()\n",
    "for lab in ['05', '06']:\n",
    "    for rep in ['1', '2', '3', '4']:\n",
    "        A = pd.read_excel(os.path.join(*config.mixturefolderSEQC2, 'excelreport', '41587_2021_857_MOESM6_ESM.xlsx'),  m+'_Ef25', index_col=0)\n",
    "        A.index = [str(a[3:]) for a in list(A.index)]\n",
    "        laba = lab if lab == '06' else '04'\n",
    "        A = A[['On-target', 'Known position', 'TP', 'FP', 'SampleB', m+'2_Ef25_ST'+laba+'_R'+rep]]\n",
    "        print(A.shape)\n",
    "        # load table\n",
    "        calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_IDT2_ST'+lab+'_25ng_LIB'+rep+'-P', 'SampleEflS_IDT2_ST'+lab+'_25ng_LIB'+rep+'-P'), ['varnet', m], save=False, filter='all')\n",
    "        calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "        calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "        B = calltable\n",
    "        B = B[[m, m+'_score', m+'_vaf', m+'_altcov', m+'_totcov']]\n",
    "        B.index = B.index.astype(str)\n",
    "        print(B.shape)\n",
    "        C = pd.concat([A, B], axis=1)\n",
    "        print(C.shape)\n",
    "        Call = pd.concat([Call, C])\n",
    "        Call = C\n",
    "\n",
    "        Call[m+'_vaf_filter_2.5%'] = (Call[m+'_vaf'] > 0.025)\n",
    "        Call[m+'_vaf_filter_1.5%'] = (Call[m+'_vaf'] > 0.015)\n",
    "        Call[m+'_vaf_filter_1.0%'] = (Call[m+'_vaf'] > 0.01)\n",
    "        Call[m+'_vaf_filter_0.5%'] = (Call[m+'_vaf'] > 0.005)\n",
    "        Call[m+'_vaf_filter_0.4%'] = (Call[m+'_vaf'] > 0.004)\n",
    "        Call[m+'_vaf_filter_0.3%'] = (Call[m+'_vaf'] > 0.003)\n",
    "        Call[m+'_vaf_filter_0.25%'] = (Call[m+'_vaf'] > 0.0025)\n",
    "        Call[m+'_vaf_filter_0.2%'] = (Call[m+'_vaf'] > 0.002)\n",
    "        Call[m+'_vaf_filter_0.15%'] = (Call[m+'_vaf'] > 0.0015)\n",
    "        Call[m+'_vaf_filter_0.1%'] = (Call[m+'_vaf'] > 0.001)\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(10, 10))\n",
    "        #precision, recall, thresholds = precision_recall_curve(C['truth'], C['BRP_vaf_range_num'])\n",
    "        precision, recall = [], []\n",
    "        #thresholds = ['2.5%', '1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.1%']\n",
    "        thresholds = ['2.5%', '1.5%', '1.0%', '0.5%', '0.4%', '0.3%', '0.25%', '0.2%', '0.15%', '0.1%']\n",
    "        Call = Call[Call['SampleB'] == 0]\n",
    "        for th in thresholds:\n",
    "            TP = float(Call[(Call['TP'] == 1) & (Call[m+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            #FP = float(Call[(Call['FP'] == 1)].shape[0])\n",
    "            FP = float(Call[(Call['FP'] == 1) & (Call[m+'_vaf_filter_'+th] == 1)].shape[0])\n",
    "            FN = float(Call[(Call['TP'] == 1) & (Call[m+'_vaf_filter_'+th] == 0)].shape[0])\n",
    "            print(TP, FP, FN)\n",
    "            p = TP/(TP + FP)\n",
    "            r = TP/(TP + FN)\n",
    "            precision.append(p)\n",
    "            recall.append(r)\n",
    "            print(th, p, r)\n",
    "        print(thresholds)\n",
    "        plt.plot(recall, precision, '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "        #plot_pr_curve(precision, recall, estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "        plt.ylim([0.97, 1.0001])\n",
    "        plt.xlim([0.0, 1.0])\n",
    "        plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "        \n",
    "        recallall.append(recall)\n",
    "        precisionall.append(precision)\n",
    "\n",
    "precisionall = np.array(precisionall)\n",
    "recallall = np.array(recallall)\n",
    "print(recallall.shape, precisionall.shape)\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "print(np.mean(recallall, axis=0), np.mean(precisionall, axis=0))\n",
    "plt.plot(np.mean(recallall, axis=0), np.mean(precisionall, axis=0), '.-', label=m, c=config.colors[config.methods.index(m)])\n",
    "#plot_pr_curve(precision, recall, estimator_name=m, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(m)]})\n",
    "plt.ylim([0.97, 1.0001])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")\n",
    "plt.title(m+' PR curve average')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[~C['Known position'].isna()].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Call[(~Call['IDT_vaf'].isna()) & (Call['SampleB'] == 0)].head(50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Call[Call['IDT_vaf'] <= 0.025]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable['truth'][calltable['truth'] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "res = {}\n",
    "\n",
    "method_sampleid_dict = {\n",
    "    'BRP': ['SampleEf_BRP2_ST26_25ng_LIB1-P', 'BRP2_Ef25_ST26_R1'],\n",
    "    'IDT': ['SampleEfIS_IDT2_ST05_25ng_LIB1-P', 'IDT2_Ef25_ST05_R1'],\n",
    "    'ILM': ['SampleEf_ILM2_ST29_25ng_LIB1-P', 'ILM2_Ef25_ST29_R1'],\n",
    "    'ROC': ['SampleEf_ROC2_ST21_25ng_LIB1-P','ROC2_Ef25_ST21_R1'],\n",
    "    'TFS': ['SampleEf_TFS2_ST24_25ng_LIB1-P','TFS2_Ef25_ST24_R1'],\n",
    "}\n",
    "\n",
    "fig, ax = plt.subplots(figsize=(10, 10))\n",
    "\n",
    "\n",
    "for method in ['BRP', 'ROC', 'ILM', 'IDT']: #,  'IDT']: #'ROC', 'TFS']:\n",
    "    \n",
    "    print('######## '+ method + ' ########')\n",
    "\n",
    "    #Call = pd.DataFrame()\n",
    "    #for lab in ['25', '26']:\n",
    "    #    for rep in ['1', '2', '3', '4']:\n",
    "            #lab = '25'\n",
    "            #rep = '1'\n",
    "\n",
    "    A = pd.read_excel(os.path.join(*config.mixturefolderSEQC2, 'excelreport', '41587_2021_857_MOESM6_ESM.xlsx'), method + '_' + method_sampleid_dict[method][1].split('_')[1], index_col=0)\n",
    "    A.index = [str(a[3:]) for a in list(A.index)]\n",
    "    #A = A[['On-target', 'Known position', 'TP', 'FP', 'SampleB', 'BRP2_Ef25_ST'+lab+'_R'+rep]]\n",
    "    A = A[['On-target', 'Known position', 'TP', 'FP', 'SampleB', method_sampleid_dict[method][1]]]\n",
    "    print(A.shape)\n",
    "\n",
    "    # load table\n",
    "    #calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_BRP2_ST'+lab+'_25ng_LIB'+rep+'-P', 'SampleEf_BRP2_ST'+lab+'_25ng_LIB'+rep+'-P'), ['varnet', 'BRP'], save=False, filter='PASS')\n",
    "    calltable_snv, calltable_indel, calltable_snp = get_calltable(os.path.join(*config.mixturefolderSEQC2, 'SEQC2s_chrall', 'SEQC2s_chrall_'+'_'.join(method_sampleid_dict[method][0].split('_')[1:]), method_sampleid_dict[method][0]), ['varnet', method], save=False, filter='PASS')\n",
    "    calltable = pd.concat([calltable_snv, calltable_indel, calltable_snp])\n",
    "    if method == 'ROC' or method == 'ILM':\n",
    "        calltable.index = calltable.index.astype(str).str[3:]\n",
    "    calltable['chrom_pos'] = calltable['chrom'].astype(str).str.cat(calltable['pos'].astype('str'), sep='_')\n",
    "   \n",
    "    B = calltable\n",
    "    #B = B[['BRP', 'BRP_score', 'BRP_vaf', 'BRP_altcov', 'BRP_totcov']]\n",
    "    B = B[[method, method+'_score', method+'_vaf', method+'_altcov', method+'_totcov']]\n",
    "    B.index = B.index.astype(str)\n",
    "    print(B.shape)\n",
    "\n",
    "    C = pd.concat([A, B], axis=1)\n",
    "    #print('Nneg')\n",
    "    #print(Nneg)\n",
    "    #C = C[(C['BRP_vaf'] <= 0.025) & (C['BRP_vaf'] > 0.001)]\n",
    "    #C.dropna(inplace=True)\n",
    "    C = C[C['SampleB'] != 1]\n",
    "    C['TP'].fillna(0, inplace=True)\n",
    "    #C['BRP2_Ef25_ST25_R1'].fillna(1, inplace=True)\n",
    "    C[method].fillna(True, inplace=True)\n",
    "    C[method+'_vaf'].fillna(0, inplace=True)\n",
    "    C['truth'] = C[\"TP\"]\n",
    "    print(C.shape)\n",
    "    C[method+'_vaf_range'] = pd.cut(C[method+'_vaf'], [0.001, 0.002, 0.003, 0.005, 0.025])\n",
    "    C[method+'_vaf_range_num'] = 5\n",
    "    C[method+'_vaf_range_num'][C[method+'_vaf'] <= 0.001] = 0\n",
    "    C[method+'_vaf_range_num'][(C[method+'_vaf'] <= 0.002) & (C[method+'_vaf'] > 0.001)] = 0.002\n",
    "    C[method+'_vaf_range_num'][(C[method+'_vaf'] <= 0.003) & (C[method+'_vaf'] > 0.002)] = 0.003\n",
    "    C[method+'_vaf_range_num'][(C[method+'_vaf'] <= 0.005) & (C[method+'_vaf'] > 0.003)] = 0.005\n",
    "    C[method+'_vaf_range_num'][(C[method+'_vaf'] <= 0.025) & (C[method+'_vaf'] > 0.005)] = 0.025\n",
    "    C[method+'_vaf_range_num'][C[method+'_vaf'] > 0.025] = 0\n",
    "    print(C[method+'_vaf_range_num'].value_counts())\n",
    "\n",
    "    #Call = pd.concat([Call, C])\n",
    "        \n",
    "    #print(Call.shape)\n",
    "\n",
    "    precision, recall, thresholds = precision_recall_curve(C['truth'], C[method+'_vaf'])\n",
    "    #precision, recall, thresholds = precision_recall_curve(Call['truth'], Call[method+'_vaf_range_num'])\n",
    "    #print(thresholds)\n",
    "    precisionbis = []\n",
    "    for thres in [0] + list(thresholds):\n",
    "        #Neg = Call[(Call['On-target']==1) & (Call['SampleB']==1)].shape[0]\n",
    "        #Neg = C[(C['On-target']==1) & (C['SampleB']==1)].shape[0]\n",
    "        #print(Neg)\n",
    "        #y_pred = Call[m+'_vaf'].copy()\n",
    "        y_pred = C[method+'_vaf'].copy()\n",
    "        y_pred[y_pred >= thres] = 1\n",
    "        y_pred[y_pred != 1] = 0\n",
    "        #_, fp, _, tp = confusion_matrix(Call['truth'], y_pred).ravel()\n",
    "        _, fp, _, tp = confusion_matrix(C['truth'], y_pred).ravel()\n",
    "        #precisionbis.append(tp/(tp+(fp/Nneg)))\n",
    "        \n",
    "    idxa = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.001))\n",
    "    idxb = min(range(len(thresholds)), key=lambda i: abs(thresholds[i]-0.025))\n",
    "    if method == 'IDT':\n",
    "        idxa, idxb = 0, len(thresholds)\n",
    "    \n",
    "    print(thresholds[idxa:idxb])\n",
    "\n",
    "    plot_pr_curve(precision[idxa:idxb], recall[idxa:idxb], estimator_name=method, f1_score=None, figax=(fig, ax), kwargs={'color': config.colors[config.methods.index(method)]})\n",
    "plt.ylim([0.0, 1.0])\n",
    "plt.xlim([0.0, 1.0])\n",
    "plt.legend(bbox_to_anchor=(1, 1), loc=\"upper left\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "calltable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "recall[idxa:idxb]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[(C['IDT_vaf'] > 0) & (C['IDT_vaf'] <= 0.025)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sns.catplot(x='TP', y='ILM_vaf', data=C[['TP', 'ILM_vaf']])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C['TP'].fillna(0, inplace=True)\n",
    "C['BRP2_Ef25_ST25_R1'].fillna(1, inplace=True)\n",
    "C['BRP'].fillna(True, inplace=True)\n",
    "C['BRP_vaf'].fillna(0, inplace=True)\n",
    "C[['TP', 'BRP', 'BRP_vaf']]\n",
    "for vaf in np.sort(C['BRP_vaf'].unique()):\n",
    "    if vaf > 0.001 and vaf <= 0.025:\n",
    "        C['BRP_tmp'] = C['BRP'].copy()\n",
    "        C['BRP_tmp'][C['BRP_vaf'] < vaf] = False\n",
    "        P = precision_score(C['TP'], C['BRP_tmp'])\n",
    "        R = recall_score(C['TP'], C['BRP_tmp'])\n",
    "        Neg = C[(C['On-target']==1) & (C['SampleB']==1)].shape[0]\n",
    "        #TP\n",
    "        TP = C[(C['TP'] == 1) & (C['BRP_tmp'] == True)].shape[0] \n",
    "        #FP\n",
    "        FP = C[((C['TP'] == 0) & (C['BRP_tmp'] == True)) | ((C['FP'] == 1) & (C['BRP_tmp'] == True))].shape[0]\n",
    "        Pbis = TP/(TP+(FP/Neg))\n",
    "        print(vaf, P, Pbis,  R)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C['TP'], C['BRP_tmp']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[(C['BRP_vaf'].isna()) & (C['BRP2_Ef25_ST25_R1'] == 1)] # weird"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C['TP'].fillna(0, inplace=True)\n",
    "C['FP'].fillna(1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[C['BRP_vaf'] < 0.001]['BRP'] = False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TP\n",
    "TP = C[(C['TP'] == 1) & (C['BRP'] == True)].shape[0] \n",
    "#FN\n",
    "FN = C[(C['TP'] == 1) & (C['BRP'] == False)].shape[0]\n",
    "#FP\n",
    "FP = C[((C['TP'] == 0) & (C['BRP'] == True)) | ((C['FP'] == 1) & (C['BRP'] == True))].shape[0]\n",
    "#TN\n",
    "TN = C[((C['FP'] == 1) & (C['BRP'] == False)) | ((C['TP'] == 0) & (C['BRP'] == False))].shape[0]\n",
    "\n",
    "print(TP, FN, FP, TN)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Neg = C[(C['On-target']==1) & (C['SampleB']==1)].shape[0]\n",
    "precisionbis = TP/(TP+(FP/Neg))\n",
    "recall = TP/(TP+FN)\n",
    "precision = TP/(TP+FP)\n",
    "print(precision, recall, precisionbis)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Neg = C[(C['On-target']==1) & (C['SampleB']==1)].shape[0]\n",
    "print(Neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C.shape, A.shape, B.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C['BRP2_Ef25_ST25_R1'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "C[C[\"BRP\"] == False]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
