{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "from collections import Counter\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "import io\n",
    "from pysam import VariantFile\n",
    "\n",
    "from sklearn.metrics import precision_score, recall_score, average_precision_score, precision_recall_curve"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tab20 = cm.get_cmap('tab10', 8)\n",
    "newcmap_list = []\n",
    "for i in range(4):\n",
    "    newcmap_list.append(tab20.colors[i])\n",
    "for i in range(4):\n",
    "    newcmap_list.append(np.array(list(tab20.colors[i][:-1]) +[0.3]))\n",
    "\n",
    "newcmap = ListedColormap(newcmap_list, name='newcmap')\n",
    "color_list = ['tab:blue', 'tab:orange', 'tab:green', 'tab:red', 'tab:purple']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    res = pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    return res\n",
    "\n",
    "foo = lambda x: pd.Series(x.split('CALLERS=')[1].split(';')[0])\n",
    "foo2 = lambda x: pd.Series(x.split('TYPE=')[1].split(';')[0] if len(x.split('TYPE=')) > 1 else np.nan)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "methods = ['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Select plasma sample"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "sample = '809'\n",
    "#sample = '986'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if sample == '809':\n",
    "    plasmasample1 = '809_110914'\n",
    "    plasmasample2 = '809_030915'\n",
    "    tumorsample1 = '809_290714-T1W'\n",
    "elif sample == '986':\n",
    "    plasmasample1 = '986_100215'\n",
    "    plasmasample2 = '986_261016'\n",
    "    tumorsample1 = '986_100215-T1W'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Load SNV calls for plasma sample and matching mixed samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_pd_0 = read_vcf(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-0/CRC-\"+plasmasample1+\"-1-0-ensemble-annotated.vcf\")\n",
    "vcf_pd_0.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_pd_0 = read_vcf(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-0/CRC-\"+plasmasample1+\"-1-0-ensemble-annotated.vcf\")\n",
    "vcf_pd_1 = read_vcf(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-05775/CRC-\"+plasmasample1+\"-1-05775-ensemble-annotated.vcf\")\n",
    "vcf_pd_2 = read_vcf(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-075-06738/CRC-\"+plasmasample1+\"-075-06738-ensemble-annotated.vcf\")\n",
    "vcf_pd_3 = read_vcf(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-05-07701/CRC-\"+plasmasample1+\"-05-07701-ensemble-annotated.vcf\")\n",
    "vcf_pd_4 = read_vcf(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-025-08663/CRC-\"+plasmasample1+\"-025-08663-ensemble-annotated.vcf\")\n",
    "\n",
    "vcf_pd_0['callers'] = vcf_pd_0['INFO'].apply(foo)\n",
    "vcf_pd_0['type'] = np.nan\n",
    "vcf_pd_0['type'][vcf_pd_0['ALT'].str.len() - vcf_pd_0['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_0['type'][vcf_pd_0['ALT'].str.len() - vcf_pd_0['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_0['type'][vcf_pd_0['ALT'].str.len() - vcf_pd_0['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_0['type'][vcf_pd_0['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_0['freebayes'] = vcf_pd_0['INFO'].str.contains('freebayes')\n",
    "vcf_pd_0['vardict'] = vcf_pd_0['INFO'].str.contains('vardict')\n",
    "vcf_pd_0['varscan'] = vcf_pd_0['INFO'].str.contains('varscan')\n",
    "vcf_pd_0['mutect2'] = vcf_pd_0['INFO'].str.contains('mutect2')\n",
    "vcf_pd_0['strelka2'] = vcf_pd_0['INFO'].str.contains('strelka2')\n",
    "\n",
    "vcf_pd_1['callers'] = vcf_pd_1['INFO'].apply(foo)\n",
    "vcf_pd_1['type'] = np.nan\n",
    "vcf_pd_1['type'][vcf_pd_1['ALT'].str.len() - vcf_pd_1['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_1['type'][vcf_pd_1['ALT'].str.len() - vcf_pd_1['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_1['type'][vcf_pd_1['ALT'].str.len() - vcf_pd_1['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_1['type'][vcf_pd_1['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_1['freebayes'] = vcf_pd_1['INFO'].str.contains('freebayes')\n",
    "vcf_pd_1['vardict'] = vcf_pd_1['INFO'].str.contains('vardict')\n",
    "vcf_pd_1['varscan'] = vcf_pd_1['INFO'].str.contains('varscan')\n",
    "vcf_pd_1['mutect2'] = vcf_pd_1['INFO'].str.contains('mutect2')\n",
    "vcf_pd_1['strelka2'] = vcf_pd_1['INFO'].str.contains('strelka2')\n",
    "\n",
    "vcf_pd_2['callers'] = vcf_pd_2['INFO'].apply(foo)\n",
    "vcf_pd_2['type'] = np.nan\n",
    "vcf_pd_2['type'][vcf_pd_2['ALT'].str.len() - vcf_pd_2['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_2['type'][vcf_pd_2['ALT'].str.len() - vcf_pd_2['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_2['type'][vcf_pd_2['ALT'].str.len() - vcf_pd_2['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_2['type'][vcf_pd_2['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_2['freebayes'] = vcf_pd_2['INFO'].str.contains('freebayes')\n",
    "vcf_pd_2['vardict'] = vcf_pd_2['INFO'].str.contains('vardict')\n",
    "vcf_pd_2['varscan'] = vcf_pd_2['INFO'].str.contains('varscan')\n",
    "vcf_pd_2['mutect2'] = vcf_pd_2['INFO'].str.contains('mutect2')\n",
    "vcf_pd_2['strelka2'] = vcf_pd_2['INFO'].str.contains('strelka2')\n",
    "\n",
    "vcf_pd_3['callers'] = vcf_pd_3['INFO'].apply(foo)\n",
    "vcf_pd_3['type'] = np.nan\n",
    "vcf_pd_3['type'][vcf_pd_3['ALT'].str.len() - vcf_pd_3['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_3['type'][vcf_pd_3['ALT'].str.len() - vcf_pd_3['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_3['type'][vcf_pd_3['ALT'].str.len() - vcf_pd_3['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_3['type'][vcf_pd_3['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_3['freebayes'] = vcf_pd_3['INFO'].str.contains('freebayes')\n",
    "vcf_pd_3['vardict'] = vcf_pd_3['INFO'].str.contains('vardict')\n",
    "vcf_pd_3['varscan'] = vcf_pd_3['INFO'].str.contains('varscan')\n",
    "vcf_pd_3['mutect2'] = vcf_pd_3['INFO'].str.contains('mutect2')\n",
    "vcf_pd_3['strelka2'] = vcf_pd_3['INFO'].str.contains('strelka2')\n",
    "\n",
    "vcf_pd_4['callers'] = vcf_pd_4['INFO'].apply(foo)\n",
    "vcf_pd_4['type'] = np.nan\n",
    "vcf_pd_4['type'][vcf_pd_4['ALT'].str.len() - vcf_pd_4['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_4['type'][vcf_pd_4['ALT'].str.len() - vcf_pd_4['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_4['type'][vcf_pd_4['ALT'].str.len() - vcf_pd_4['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_4['type'][vcf_pd_4['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_4['freebayes'] = vcf_pd_4['INFO'].str.contains('freebayes')\n",
    "vcf_pd_4['vardict'] = vcf_pd_4['INFO'].str.contains('vardict')\n",
    "vcf_pd_4['varscan'] = vcf_pd_4['INFO'].str.contains('varscan')\n",
    "vcf_pd_4['mutect2'] = vcf_pd_4['INFO'].str.contains('mutect2')\n",
    "vcf_pd_4['strelka2'] = vcf_pd_4['INFO'].str.contains('strelka2')\n",
    "\n",
    "sample_0 = vcf_pd_0[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'FILTER', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_0['CHROM_POS'] = sample_0['CHROM'].astype('str').str.cat(sample_0['POS'].astype('str'),sep=\"_\")\n",
    "sample_0.set_index('CHROM_POS', inplace = True)\n",
    "\n",
    "sample_1 = vcf_pd_1[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_1['CHROM_POS'] = sample_1['CHROM'].astype('str').str.cat(sample_1['POS'].astype('str'),sep=\"_\")\n",
    "sample_1.set_index('CHROM_POS', inplace = True)\n",
    "\n",
    "sample_2 = vcf_pd_2[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_2['CHROM_POS'] = sample_2['CHROM'].astype('str').str.cat(sample_2['POS'].astype('str'),sep=\"_\")\n",
    "sample_2.set_index('CHROM_POS', inplace = True)\n",
    "\n",
    "sample_3 = vcf_pd_3[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_3['CHROM_POS'] = sample_3['CHROM'].astype('str').str.cat(sample_3['POS'].astype('str'),sep=\"_\")\n",
    "sample_3.set_index('CHROM_POS', inplace = True)\n",
    "\n",
    "sample_4 = vcf_pd_4[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_4['CHROM_POS'] = sample_4['CHROM'].astype('str').str.cat(sample_4['POS'].astype('str'),sep=\"_\")\n",
    "sample_4.set_index('CHROM_POS', inplace = True)\n",
    "\n",
    "sample_4.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_pd_0_bis = read_vcf(\"../data/2015-07-31_CRC-\"+plasmasample2+\"-1-0/CRC-\"+plasmasample2+\"-1-0-ensemble-annotated.vcf\")\n",
    "\n",
    "vcf_pd_0_bis['type'] = np.nan\n",
    "vcf_pd_0_bis['type'][vcf_pd_0_bis['ALT'].str.len() - vcf_pd_0_bis['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_0_bis['type'][vcf_pd_0_bis['ALT'].str.len() - vcf_pd_0_bis['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_0_bis['type'][vcf_pd_0_bis['ALT'].str.len() - vcf_pd_0_bis['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_0_bis['type'][vcf_pd_0_bis['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_0_bis['freebayes'] = vcf_pd_0_bis['INFO'].str.contains('freebayes')\n",
    "vcf_pd_0_bis['vardict'] = vcf_pd_0_bis['INFO'].str.contains('vardict')\n",
    "vcf_pd_0_bis['varscan'] = vcf_pd_0_bis['INFO'].str.contains('varscan')\n",
    "vcf_pd_0_bis['mutect2'] = vcf_pd_0_bis['INFO'].str.contains('mutect2')\n",
    "vcf_pd_0_bis['strelka2'] = vcf_pd_0_bis['INFO'].str.contains('strelka2')\n",
    "\n",
    "sample_0_bis = vcf_pd_0_bis[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_0_bis['CHROM_POS'] = sample_0_bis['CHROM'].astype('str').str.cat(sample_0_bis['POS'].astype('str'),sep=\"_\")\n",
    "sample_0_bis.set_index('CHROM_POS', inplace = True)\n",
    "sample_0_bis.head()\n",
    "\n",
    "sample0 = sample_0.copy()\n",
    "sample0['date'] = '110914'\n",
    "sample0bis = sample_0_bis.copy()\n",
    "sample0bis['date'] = '030915'\n",
    "sample_ref = pd.concat([sample0, sample0bis], join='inner')\n",
    "l1 = list(sample_ref[sample_ref['date'] == '110914'].index)\n",
    "l2 = list(sample_ref[sample_ref['date'] == '030915'].index)\n",
    "print(len(l1), len(l2))\n",
    "print(len(list(set(set(l1) ^ set(l2)))))\n",
    "print(len(list(set(set(l1) & set(l2)))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_pd_h = read_vcf(\"../data/2015-07-31_Merged_Healthy/Merged_Healthy-ensemble-annotated.vcf\")\n",
    "\n",
    "vcf_pd_h['type'] = np.nan\n",
    "vcf_pd_h['type'][vcf_pd_h['ALT'].str.len() - vcf_pd_h['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_h['type'][vcf_pd_h['ALT'].str.len() - vcf_pd_h['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_h['type'][vcf_pd_h['ALT'].str.len() - vcf_pd_h['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_h['type'][vcf_pd_h['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_h['freebayes'] = vcf_pd_h['INFO'].str.contains('freebayes')\n",
    "vcf_pd_h['vardict'] = vcf_pd_h['INFO'].str.contains('vardict')\n",
    "vcf_pd_h['varscan'] = vcf_pd_h['INFO'].str.contains('varscan')\n",
    "vcf_pd_h['mutect2'] = vcf_pd_h['INFO'].str.contains('mutect2')\n",
    "vcf_pd_h['strelka2'] = vcf_pd_h['INFO'].str.contains('strelka2')\n",
    "\n",
    "sample_h = vcf_pd_h[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_h['CHROM_POS'] = sample_h['CHROM'].astype('str').str.cat(sample_h['POS'].astype('str'),sep=\"_\")\n",
    "sample_h.set_index('CHROM_POS', inplace = True)\n",
    "sample_h = sample_h[sample_h['CHROM'] == '22']\n",
    "sample_h.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_pd_t = read_vcf(\"../data/2015-07-31_NCC_CRC-\"+tumorsample1+\"/NCC_CRC-\"+tumorsample1+\"-ensemble-annotated.vcf\")\n",
    "\n",
    "vcf_pd_t['type'] = np.nan\n",
    "vcf_pd_t['type'][vcf_pd_t['ALT'].str.len() - vcf_pd_t['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd_t['type'][vcf_pd_t['ALT'].str.len() - vcf_pd_t['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd_t['type'][vcf_pd_t['ALT'].str.len() - vcf_pd_t['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd_t['type'][vcf_pd_t['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd_t['freebayes'] = vcf_pd_t['INFO'].str.contains('freebayes')\n",
    "vcf_pd_t['vardict'] = vcf_pd_t['INFO'].str.contains('vardict')\n",
    "vcf_pd_t['varscan'] = vcf_pd_t['INFO'].str.contains('varscan')\n",
    "vcf_pd_t['mutect2'] = vcf_pd_t['INFO'].str.contains('mutect2')\n",
    "vcf_pd_t['strelka2'] = vcf_pd_t['INFO'].str.contains('strelka2')\n",
    "\n",
    "sample_t = vcf_pd_t[['CHROM', 'POS', 'REF', 'ALT', 'QUAL', 'type', 'freebayes', 'vardict', 'varscan', 'mutect2', 'strelka2']]\n",
    "sample_t['CHROM_POS'] = sample_t['CHROM'].astype('str').str.cat(sample_t['POS'].astype('str'),sep=\"_\")\n",
    "sample_t.set_index('CHROM_POS', inplace = True)\n",
    "sample_t = sample_t[sample_t['CHROM'] == '22']\n",
    "print(sample_t.shape)\n",
    "sample_t = sample_t.loc[list(sample_t.index[~sample_t.index.isin(sample_h.index)])]\n",
    "print(sample_t.shape)\n",
    "\n",
    "l1 = list(sample_t.index)\n",
    "l2 = list(sample_0.index)\n",
    "print(len(l1), len(l2))\n",
    "print(len(list(set(set(l1) & set(l2)))))\n",
    "\n",
    "sample_t.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Estimated tumor burden of mixed samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_tf = {\n",
    "    'sample_0': float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-0/estimated_tf.txt\").columns)[0]),\n",
    "    'sample_1': float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-05775/estimated_tf.txt\").columns)[0]),\n",
    "    'sample_2': float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-075-06738/estimated_tf.txt\").columns)[0]),\n",
    "    'sample_3': float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-05-07701/estimated_tf.txt\").columns)[0]),\n",
    "    'sample_4': float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-025-08663/estimated_tf.txt\").columns)[0]),\n",
    "}\n",
    "\n",
    "print(samples_tf)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Number of detections detected"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbersnvs_pd = pd.DataFrame()\n",
    "\n",
    "for si, s in enumerate([sample_0, sample_1, sample_2, sample_3, sample_4]):\n",
    "    nb_snv = []\n",
    "    for method in methods:\n",
    "        #if si == 4:\n",
    "        #    si = 't'\n",
    "       # print('sample '+ str(si) + ': ', method, s[s[method] == True].shape[0])\n",
    "        nb_snv.append(s[s[method] == True].shape[0])\n",
    "    if si == 0:\n",
    "        numbersnvs_pd = pd.DataFrame.from_dict({'sample_'+ str(si): nb_snv}).T\n",
    "        numbersnvs_pd.columns = methods\n",
    "    else:\n",
    "        numbersnvs_pd.loc['sample_'+ str(si)] = nb_snv\n",
    "    numbersnvs_pd = numbersnvs_pd.rename(index=samples_tf)\n",
    "\n",
    "numbersnvs_pd.plot(style='.-', logx=True, xlim=(10e-1, 10e-3))\n",
    "#sns.catplot(x='tumor burden', y='coverage', kind='point', data=coverage_pd,\n",
    "#            order=sorted(coverage_pd['tumor burden'].unique(), reverse=True))\n",
    "        \n",
    "numbersnvs_pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbersnvs_pd = pd.DataFrame()\n",
    "\n",
    "for si, s in enumerate([sample_0, sample_1, sample_2, sample_3, sample_4]):\n",
    "    nb_snv = []\n",
    "    for method in methods:\n",
    "        #if si == 4:\n",
    "        #    si = 't'\n",
    "       # print('sample '+ str(si) + ': ', method, s[s[method] == True].shape[0])\n",
    "        nb_snv.append(s[(s[method] == True) & ((s['type'] == 'INS') | (s['type'] == 'DEL'))].shape[0])\n",
    "    if si == 0:\n",
    "        numbersnvs_pd = pd.DataFrame.from_dict({'sample_'+ str(si): nb_snv}).T\n",
    "        numbersnvs_pd.columns = methods\n",
    "    else:\n",
    "        numbersnvs_pd.loc['sample_'+ str(si)] = nb_snv\n",
    "    numbersnvs_pd = numbersnvs_pd.rename(index=samples_tf)\n",
    "\n",
    "numbersnvs_pd.plot(style='.-', logx=True, xlim=(10e-1, 10e-3))\n",
    "#sns.catplot(x='tumor burden', y='coverage', kind='point', data=coverage_pd,\n",
    "#            order=sorted(coverage_pd['tumor burden'].unique(), reverse=True))\n",
    "        \n",
    "numbersnvs_pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbersnvs_pd = pd.DataFrame()\n",
    "\n",
    "for si, s in enumerate([sample_0, sample_1, sample_2, sample_3, sample_4]):\n",
    "    nb_snv = []\n",
    "    for method in methods:\n",
    "        #if si == 4:\n",
    "        #    si = 't'\n",
    "       # print('sample '+ str(si) + ': ', method, s[s[method] == True].shape[0])\n",
    "        nb_snv.append(s[(s[method] == True) & (s['type'] == 'SNV')].shape[0])\n",
    "    if si == 0:\n",
    "        numbersnvs_pd = pd.DataFrame.from_dict({'sample_'+ str(si): nb_snv}).T\n",
    "        numbersnvs_pd.columns = methods\n",
    "    else:\n",
    "        numbersnvs_pd.loc['sample_'+ str(si)] = nb_snv\n",
    "    numbersnvs_pd = numbersnvs_pd.rename(index=samples_tf)\n",
    "\n",
    "numbersnvs_pd.plot(style='.-', logx=True, xlim=(10e-1, 10e-3))\n",
    "#sns.catplot(x='tumor burden', y='coverage', kind='point', data=coverage_pd,\n",
    "#            order=sorted(coverage_pd['tumor burden'].unique(), reverse=True))\n",
    "        \n",
    "numbersnvs_pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "numbersnvs_pd = pd.DataFrame()\n",
    "\n",
    "for si, s in enumerate([sample_0, sample_1, sample_2, sample_3, sample_4]):\n",
    "    nb_snv = []\n",
    "    for method in methods:\n",
    "        #if si == 4:\n",
    "        #    si = 't'\n",
    "       # print('sample '+ str(si) + ': ', method, s[s[method] == True].shape[0])\n",
    "        nb_snv.append(s[(s[method] == True) & (s['type'] == 'SNP')].shape[0])\n",
    "    if si == 0:\n",
    "        numbersnvs_pd = pd.DataFrame.from_dict({'sample_'+ str(si): nb_snv}).T\n",
    "        numbersnvs_pd.columns = methods\n",
    "    else:\n",
    "        numbersnvs_pd.loc['sample_'+ str(si)] = nb_snv\n",
    "    numbersnvs_pd = numbersnvs_pd.rename(index=samples_tf)\n",
    "\n",
    "numbersnvs_pd.plot(style='.-', logx=True, xlim=(10e-1, 10e-3))\n",
    "#sns.catplot(x='tumor burden', y='coverage', kind='point', data=coverage_pd,\n",
    "#            order=sorted(coverage_pd['tumor burden'].unique(), reverse=True))\n",
    "        \n",
    "numbersnvs_pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Coverage variability"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samples_cov_tf = {\n",
    "    'sample_0': [float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-0/coverage.txt\").columns)[0]),\n",
    "                np.round(float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-0/estimated_tf.txt\").columns)[0]),2)],\n",
    "    'sample_1': [float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-05775/coverage.txt\").columns)[0]),\n",
    "                np.round(float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-1-05775/estimated_tf.txt\").columns)[0]),2)],\n",
    "    'sample_2': [float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-075-06738/coverage.txt\").columns)[0]),\n",
    "                np.round(float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-075-06738/estimated_tf.txt\").columns)[0]),2)],\n",
    "    'sample_3': [float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-05-07701/coverage.txt\").columns)[0]),\n",
    "                np.round(float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-05-07701/estimated_tf.txt\").columns)[0]),2)],\n",
    "    'sample_4': [float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-025-08663/coverage.txt\").columns)[0]),\n",
    "                np.round(float(list(pd.read_csv(\"../data/2015-07-31_CRC-\"+plasmasample1+\"-025-08663/estimated_tf.txt\").columns)[0]),2)],\n",
    "    \n",
    "}\n",
    "\n",
    "print(samples_cov_tf)\n",
    "\n",
    "coverage_pd = pd.DataFrame.from_dict(samples_cov_tf).T\n",
    "coverage_pd.columns = ['coverage', 'tumor burden']\n",
    "sns.catplot(x='tumor burden', y='coverage', data=coverage_pd,\n",
    "            order=sorted(coverage_pd['tumor burden'].unique(), reverse=True), color='magenta')\n",
    "\n",
    "        \n",
    "coverage_pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AUPRC with reference = SNV found in undiluted plasma sample by the same method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd_methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mutation in ['snv', 'indel']:\n",
    "    pd_methods = pd.DataFrame\n",
    "    count = 0\n",
    "\n",
    "    for mi, method in enumerate(methods):\n",
    "        #print(method)\n",
    "\n",
    "        s0 = sample_0[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s0.rename(columns = {method:'sample_0'},  inplace = True)\n",
    "        s1 = sample_1[['QUAL', 'type', method]]\n",
    "        s1.rename(columns = {method:'sample_1'},  inplace = True)\n",
    "        s2 = sample_2[['QUAL','type', method]]\n",
    "        s2.rename(columns = {method:'sample_2'},  inplace = True)\n",
    "        s3 = sample_3[['QUAL','type', method]]\n",
    "        s3.rename(columns = {method:'sample_3'},  inplace = True)\n",
    "        s4 = sample_4[['QUAL','type', method]]\n",
    "        s4.rename(columns = {method:'sample_4'},  inplace = True)\n",
    "\n",
    "        pd_method = pd.concat([s0, s1, s2, s3, s4], axis=1)\n",
    "\n",
    "        pd_method.columns = ['REF', 'ALT', 'QUAL_0', 'type_0', 'sample_0', 'QUAL_1', 'type_1', 'sample_1', 'QUAL_2', 'type_2', 'sample_2',  'QUAL_3', 'type_3', 'sample_3', 'QUAL_4', 'type_4', 'sample_4']\n",
    "        pd_method[['sample_0', 'sample_1', 'sample_2', 'sample_3', 'sample_4']] = pd_method[['sample_0','sample_1', 'sample_2', 'sample_3', 'sample_4']].fillna(value=False)\n",
    "        pd_method.drop(pd_method[pd_method['sample_0'] + pd_method['sample_1'] + pd_method['sample_2']  + pd_method['sample_3'] + pd_method['sample_4'] == False].index, axis=0, inplace=True)\n",
    "        pd_method['caller'] = method\n",
    "        pd_method['type'] = pd_method[['type_0','type_1', 'type_2', 'type_3', 'type_4']].fillna('').sum(axis=1).apply(lambda x: pd.Series(x[:3]))\n",
    "\n",
    "        if mutation == 'snv':\n",
    "            pd_method = pd_method[pd_method['type'] == 'SNV']\n",
    "        elif mutation == 'indel':\n",
    "            pd_method = pd_method[(pd_method['type'] == 'INS') | (pd_method['type'] == 'DEL')]\n",
    "                                \n",
    "        if count == 0:\n",
    "            pd_methods = pd_method\n",
    "            count = 1\n",
    "        else:\n",
    "            pd_methods = pd.concat([pd_methods, pd_method], join='inner')\n",
    "    #print(pd_methods.shape)\n",
    "\n",
    "    pd_results = pd.DataFrame\n",
    "    pd_results_PR = pd.DataFrame\n",
    "    count = 0\n",
    "\n",
    "    fig, axs = plt.subplots(1,5,figsize=(30, 4))\n",
    "    fig.suptitle('Precision-Recall curves')\n",
    "    \n",
    "    y_true = pd_methods[['sample_0', 'caller']]\n",
    "    y_true.index.name = 'CHROM_POS'\n",
    "    y_true = y_true.groupby(['CHROM_POS'])['sample_0'].sum()\n",
    "    y_true[y_true <= 2] = 0\n",
    "    y_true = y_true.astype(bool)\n",
    "    print(y_true.shape[0])\n",
    "    print(pd_methods.shape)\n",
    "\n",
    "\n",
    "    baselineAUPRC = {}\n",
    "\n",
    "\n",
    "    for mi, method in enumerate(methods):\n",
    "\n",
    "        res_df = (100*pd.Series(samples_tf)).round(decimals=2).to_frame(name='tumor burden')\n",
    "        res_PR_df = (100*pd.Series(samples_tf)).round(decimals=2).to_frame(name='tumor burden')\n",
    "\n",
    "       # y_true = pd_methods[pd_methods['caller'] == method]['sample_0'].astype(bool)\n",
    "        baselineAUPRC[method] = len(y_true[y_true == True])/len(y_true)\n",
    "        y_0 = pd_methods[pd_methods['caller'] == method]['sample_0'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_0 = y_0.fillna(False)\n",
    "        y_1 = pd_methods[pd_methods['caller'] == method]['sample_1'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_1 = y_1.fillna(False)\n",
    "        y_2 = pd_methods[pd_methods['caller'] == method]['sample_2'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_2 = y_2.fillna(False)\n",
    "        y_3 = pd_methods[pd_methods['caller'] == method]['sample_3'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_3 = y_3.fillna(False)\n",
    "        y_4 = pd_methods[pd_methods['caller'] == method]['sample_4'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_4 = y_4.fillna(False)\n",
    "\n",
    "\n",
    "        res_df['AUPRC'] = [average_precision_score(y_true, y_0),\n",
    "                          average_precision_score(y_true, y_1),\n",
    "                          average_precision_score(y_true, y_2),\n",
    "                          average_precision_score(y_true, y_3),\n",
    "                          average_precision_score(y_true, y_4),\n",
    "                         ]\n",
    "        res_df['AUPRC - baseline AUPRC'] = [average_precision_score(y_true, y_0) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_1) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_2) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_3) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_4) - baselineAUPRC[method],\n",
    "                     ]\n",
    "        res_PR_df['precision'] = [precision_score(y_true, y_0),\n",
    "                          precision_score(y_true, y_1),\n",
    "                          precision_score(y_true, y_2),\n",
    "                          precision_score(y_true, y_3),\n",
    "                          precision_score(y_true, y_4),\n",
    "                         ]\n",
    "\n",
    "        res_PR_df['recall'] = [recall_score(y_true, y_0),\n",
    "                          recall_score(y_true, y_1),\n",
    "                          recall_score(y_true, y_2),\n",
    "                          recall_score(y_true, y_3),\n",
    "                          recall_score(y_true, y_4),\n",
    "                         ]\n",
    "        res_PR_df = pd.melt(res_PR_df, id_vars =['tumor burden'], value_vars =['precision', 'recall'],\n",
    "                        var_name='metric', value_name='value')\n",
    "        res_df['caller'] = method\n",
    "        res_PR_df['caller'] = method\n",
    "\n",
    "        res_df.drop(['sample_0'], inplace=True)\n",
    "        res_PR_df.drop([0, 5], inplace=True)\n",
    "\n",
    "        # Plot Precision-Recall curve\n",
    "        alpha_list = [0.4, 1, .75, .5, .3, .1]\n",
    "        for i in range(1,5):\n",
    "            y_i = pd_methods[pd_methods['caller'] == method]['sample_'+str(i)].astype(bool).reindex(y_true.index).squeeze()\n",
    "            y_i = y_i.fillna(False)\n",
    "            precision, recall, _ = precision_recall_curve(y_true, y_i)\n",
    "            axs[mi].plot(recall, precision, 'o-',\n",
    "                                  label='tf='+str(res_df['tumor burden'].loc['sample_'+str(i)])+ ', AP='+str(round(average_precision_score(y_true, y_i), 2)),\n",
    "                                  c=color_list[mi], alpha=alpha_list[i])\n",
    "        axs[mi].set_xlabel('Recall')\n",
    "        axs[mi].set_ylabel('Precision')\n",
    "        axs[mi].set_ylim([0.0, 1.05])\n",
    "        axs[mi].set_xlim([0.0, 1.05])\n",
    "        axs[mi].set_title(method)\n",
    "        axs[mi].legend()\n",
    "\n",
    "        if count == 0:\n",
    "            pd_results = res_df\n",
    "            pd_results_PR = res_PR_df\n",
    "            count = 1\n",
    "        else:\n",
    "            pd_results = pd.concat([pd_results, res_df], join='inner')\n",
    "            pd_results_PR = pd.concat([pd_results_PR, res_PR_df], join='inner')\n",
    "    plt.show()\n",
    "\n",
    "    plt.figure()\n",
    "    sns.catplot(x=\"tumor burden\", y=\"AUPRC\", hue=\"caller\",  \n",
    "                    capsize=.2, height=4, aspect=1.5,\n",
    "                    kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results)\n",
    "    for mi, method in enumerate(methods):\n",
    "        plt.axhline(y = baselineAUPRC[method], color = 'k', linestyle = '--') \n",
    "    #plt.ylim([0.1, 0.5])\n",
    "    \n",
    "    sns.catplot(x=\"tumor burden\", y=\"AUPRC - baseline AUPRC\", hue=\"caller\",  \n",
    "                capsize=.2, height=6, aspect=.75,\n",
    "                kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results)\n",
    "    plt.axhline(y = 0, color = 'k', linestyle = '--', label='baseline AUPRC') \n",
    "\n",
    "    plt.figure()\n",
    "    sns.catplot(x=\"tumor burden\", y=\"value\", col=\"metric\", hue=\"caller\",  \n",
    "                    capsize=.2, height=6, aspect=.75,\n",
    "                    kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results_PR)\n",
    "    #plt.ylim([0, 0.5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# AUPRC with reference = SNV found in both undiluted plasma samples by the same method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mutation in ['snv', 'indel']:\n",
    "    pd_methods = pd.DataFrame\n",
    "    count = 0\n",
    "\n",
    "    for mi, method in enumerate(['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan']):\n",
    "\n",
    "        s0 = sample_0[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s0.rename(columns = {method:'sample_0'},  inplace = True)\n",
    "        s0_bis = sample_0_bis[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s0_bis.rename(columns = {method:'sample_0_bis'},  inplace = True)\n",
    "        s1 = sample_1[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s1.rename(columns = {method:'sample_1'},  inplace = True)\n",
    "        s2 = sample_2[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s2.rename(columns = {method:'sample_2'},  inplace = True)\n",
    "        s3 = sample_3[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s3.rename(columns = {method:'sample_3'},  inplace = True)\n",
    "        s4 = sample_4[['REF', 'ALT', 'QUAL','type', method]]\n",
    "        s4.rename(columns = {method:'sample_4'},  inplace = True)\n",
    "\n",
    "        pd_method = pd.concat([s0, s0_bis, s1, s2, s3, s4], axis=1)\n",
    "\n",
    "        pd_method['REF'] = [list(np.unique([i for i in list(ai) if str(i) != 'nan']))[-1] for ai in list(pd_method['REF'].values)]\n",
    "        pd_method['ALT'] = [list(np.unique([i for i in list(ai) if str(i) != 'nan']))[-1] for ai in list(pd_method['ALT'].values)]\n",
    "        pd_method = pd_method.T.drop_duplicates().T\n",
    "        pd_method.columns = ['REF', 'ALT', 'QUAL_0', 'type_0', 'sample_0','QUAL_0_bis', 'type_0_bis', 'sample_0_bis', 'QUAL_1', 'type_1', 'sample_1', 'QUAL_2', 'type_2', 'sample_2',  'QUAL_3', 'type_3', 'sample_3', 'QUAL_4', 'type_4', 'sample_4']\n",
    "        pd_method[['sample_0', 'sample_0_bis', 'sample_1', 'sample_2', 'sample_3', 'sample_4']] = pd_method[['sample_0', 'sample_0_bis', 'sample_1', 'sample_2', 'sample_3', 'sample_4']].fillna(value=False)\n",
    "        #pd_method.drop(pd_method[pd_method['sample_0'] + pd_method['sample_0_bis']+ pd_method['sample_1'] + pd_method['sample_2'] + pd_method['sample_3'] + pd_method['sample_4'] == False].index, axis=0, inplace=True)\n",
    "        pd_method[['REF', 'sample_0', 'sample_0_bis', 'sample_1', 'sample_2', 'sample_3', 'sample_4']].head()\n",
    "        pd_method['caller'] = method\n",
    "        pd_method['type'] = pd_method[['type_0','type_0_bis','type_1', 'type_2', 'type_3', 'type_4']].fillna('').sum(axis=1).apply(lambda x: pd.Series(x[:3]))\n",
    "\n",
    "        if mutation == 'snv':\n",
    "            pd_method = pd_method[pd_method['type'] == 'SNV']\n",
    "        elif mutation == 'indel':\n",
    "            pd_method = pd_method[(pd_method['type'] == 'INS') | (pd_method['type'] == 'DEL')]\n",
    "\n",
    "        if count == 0:\n",
    "            pd_methods = pd_method\n",
    "            count = 1\n",
    "        else:\n",
    "            pd_methods = pd.concat([pd_methods, pd_method], join='inner')\n",
    "\n",
    "    pd_results = pd.DataFrame.empty\n",
    "    pd_results_PR = pd.DataFrame.empty\n",
    "    count = 0\n",
    "\n",
    "    fig, axs = plt.subplots(1,5,figsize=(30, 4))\n",
    "    fig.suptitle('Precision-Recall curves')\n",
    "\n",
    "\n",
    "    y_true = pd_methods[['sample_0', 'sample_0_bis', 'caller']]\n",
    "    y_true['ref'] = pd_methods['sample_0'] & pd_methods['sample_0_bis']\n",
    "    y_true.index.name = 'CHROM_POS'\n",
    "    y_true = y_true.groupby(['CHROM_POS'])['ref'].sum()\n",
    "    y_true = y_true.astype(bool)\n",
    "\n",
    "\n",
    "    baselineAUPRC = {}\n",
    "\n",
    "    for mi, method in enumerate(methods):\n",
    "\n",
    "        res_df = (100*pd.Series(samples_tf)).round(decimals=2).to_frame(name='tumor burden')\n",
    "        res_PR_df = (100*pd.Series(samples_tf)).round(decimals=2).to_frame(name='tumor burden')\n",
    "\n",
    "        #y_true = pd_methods[pd_methods['caller'] == method]['sample_0'].astype(bool) + pd_methods[pd_methods['caller'] == method]['sample_0_bis'].astype(bool)\n",
    "        baselineAUPRC[method] = len(y_true[y_true == True])/len(y_true)\n",
    "\n",
    "        y_0 = pd_methods[pd_methods['caller'] == method]['sample_0'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_0 = y_0.fillna(False)\n",
    "        y_1 = pd_methods[pd_methods['caller'] == method]['sample_1'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_1 = y_1.fillna(False)\n",
    "        y_2 = pd_methods[pd_methods['caller'] == method]['sample_2'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_2 = y_2.fillna(False)\n",
    "        y_3 = pd_methods[pd_methods['caller'] == method]['sample_3'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_3 = y_3.fillna(False)\n",
    "        y_4 = pd_methods[pd_methods['caller'] == method]['sample_4'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_4 = y_4.fillna(False)\n",
    "\n",
    "        res_df['AUPRC'] = [average_precision_score(y_true, y_0),\n",
    "                          average_precision_score(y_true, y_1),\n",
    "                          average_precision_score(y_true, y_2),\n",
    "                          average_precision_score(y_true, y_3),\n",
    "                          average_precision_score(y_true, y_4),\n",
    "                         ]\n",
    "        res_df['AUPRC - baseline AUPRC'] = [average_precision_score(y_true, y_0) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_1) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_2) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_3) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_4) - baselineAUPRC[method],\n",
    "                     ]\n",
    "        res_PR_df['precision'] = [precision_score(y_true, y_0),\n",
    "                          precision_score(y_true, y_1),\n",
    "                          precision_score(y_true, y_2),\n",
    "                          precision_score(y_true, y_3),\n",
    "                          precision_score(y_true, y_4),\n",
    "                         ]\n",
    "\n",
    "        res_PR_df['recall'] = [recall_score(y_true, y_0),\n",
    "                          recall_score(y_true, y_1),\n",
    "                          recall_score(y_true, y_2),\n",
    "                          recall_score(y_true, y_3),\n",
    "                          recall_score(y_true, y_4),\n",
    "                         ]\n",
    "        res_PR_df = pd.melt(res_PR_df, id_vars =['tumor burden'], value_vars =['precision', 'recall'],\n",
    "                        var_name='metric', value_name='value')\n",
    "        res_df['caller'] = method\n",
    "        res_PR_df['caller'] = method\n",
    "        res_df.drop(['sample_0'], inplace=True)\n",
    "        res_PR_df.drop([0, 5], inplace=True)\n",
    "\n",
    "        # Plot Precision-Recall curve\n",
    "        alpha_list = [0, 1, .75, .5, .2, .1]\n",
    "        for i in range(1,5):\n",
    "            y_i = pd_methods[pd_methods['caller'] == method]['sample_'+str(i)].astype(bool).reindex(y_true.index).squeeze()\n",
    "            y_i = y_i.fillna(False)\n",
    "            precision, recall, _ = precision_recall_curve(y_true, y_i)\n",
    "            axs[mi].plot(recall, precision, 'o-',\n",
    "                                  label='tf='+str(res_df['tumor burden'].loc['sample_'+str(i)])+ ', AP='+str(round(average_precision_score(y_true, y_i), 2)),\n",
    "                                  c=color_list[mi], alpha=alpha_list[i])\n",
    "        axs[mi].set_xlabel('Recall')\n",
    "        axs[mi].set_ylabel('Precision')\n",
    "        axs[mi].set_ylim([0.0, 1.05])\n",
    "        axs[mi].set_xlim([0.0, 1.05])\n",
    "        axs[mi].set_title(method)\n",
    "        axs[mi].legend()\n",
    "\n",
    "        if count == 0:\n",
    "            pd_results = res_df\n",
    "            pd_results_PR = res_PR_df\n",
    "            count = 1\n",
    "        else:\n",
    "            pd_results = pd.concat([pd_results, res_df], join='inner')\n",
    "            pd_results_PR = pd.concat([pd_results_PR, res_PR_df], join='inner')\n",
    "    plt.show()\n",
    "\n",
    "    sns.catplot(x=\"tumor burden\", y=\"AUPRC\", hue=\"caller\",  \n",
    "                    capsize=.2, height=4, aspect=1.5,\n",
    "                    kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results, legend=False)\n",
    "    for mi, method in enumerate(methods):\n",
    "        plt.axhline(y = baselineAUPRC[method], color = 'k', linestyle = '--') \n",
    "    #plt.ylim([0, 0.5])\n",
    "    plt.legend(bbox_to_anchor=(1,1), loc=\"upper left\")\n",
    "    \n",
    "    sns.catplot(x=\"tumor burden\", y=\"AUPRC - baseline AUPRC\", hue=\"caller\",  \n",
    "                capsize=.2, height=6, aspect=.75,\n",
    "                kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results)\n",
    "    plt.axhline(y = 0, color = 'k', linestyle = '--', label='baseline AUPRC') \n",
    "\n",
    "    sns.catplot(x=\"tumor burden\", y=\"value\", col=\"metric\", hue=\"caller\",  \n",
    "                    capsize=.2, height=6, aspect=.75,\n",
    "                    kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results_PR)\n",
    "    #plt.ylim([0, 0.5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## AUPRC with reference = SNV found in the tumor by the same method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for mutation in ['snv', 'indel']:\n",
    "    pd_methods = pd.DataFrame\n",
    "    count = 0\n",
    "\n",
    "    for mi, method in enumerate(['freebayes', 'mutect2', 'strelka2', 'vardict', 'varscan']):\n",
    "\n",
    "        s0 = sample_0[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s0.rename(columns = {method:'sample_0'},  inplace = True)\n",
    "        st = sample_t[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        st.rename(columns = {method:'sample_t'},  inplace = True)\n",
    "        s1 = sample_1[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s1.rename(columns = {method:'sample_1'},  inplace = True)\n",
    "        s2 = sample_2[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s2.rename(columns = {method:'sample_2'},  inplace = True)\n",
    "        s3 = sample_3[['REF', 'ALT', 'QUAL', 'type', method]]\n",
    "        s3.rename(columns = {method:'sample_3'},  inplace = True)\n",
    "        s4 = sample_4[['REF', 'ALT', 'QUAL','type', method]]\n",
    "        s4.rename(columns = {method:'sample_4'},  inplace = True)\n",
    "\n",
    "        pd_method = pd.concat([s0, st, s1, s2, s3, s4], axis=1)\n",
    "\n",
    "        pd_method['REF'] = [list(np.unique([i for i in list(ai) if str(i) != 'nan']))[-1] for ai in list(pd_method['REF'].values)]\n",
    "        pd_method['ALT'] = [list(np.unique([i for i in list(ai) if str(i) != 'nan']))[-1] for ai in list(pd_method['ALT'].values)]\n",
    "        pd_method = pd_method.T.drop_duplicates().T\n",
    "        pd_method.columns = ['REF', 'ALT', 'QUAL_0', 'type_0', 'sample_0','QUAL_t', 'type_t', 'sample_t', 'QUAL_1', 'type_1', 'sample_1', 'QUAL_2', 'type_2', 'sample_2',  'QUAL_3', 'type_3', 'sample_3', 'QUAL_4', 'type_4', 'sample_4']\n",
    "        pd_method[['sample_0', 'sample_t', 'sample_1', 'sample_2', 'sample_3', 'sample_4']] = pd_method[['sample_0', 'sample_t', 'sample_1', 'sample_2', 'sample_3', 'sample_4']].fillna(value=False)\n",
    "        #pd_method.drop(pd_method[pd_method['sample_0'] + pd_method['sample_0_bis']+ pd_method['sample_1'] + pd_method['sample_2'] + pd_method['sample_3'] + pd_method['sample_4'] == False].index, axis=0, inplace=True)\n",
    "        pd_method[['REF', 'sample_0', 'sample_t', 'sample_1', 'sample_2', 'sample_3', 'sample_4']].head()\n",
    "        pd_method['caller'] = method\n",
    "        pd_method['type'] = pd_method[['type_0','type_t','type_1', 'type_2', 'type_3', 'type_4']].fillna('').sum(axis=1).apply(lambda x: pd.Series(x[:3]))\n",
    "\n",
    "        if mutation == 'snv':\n",
    "            pd_method = pd_method[pd_method['type'] == 'SNV']\n",
    "        elif mutation == 'indel':\n",
    "            pd_method = pd_method[(pd_method['type'] == 'INS') | (pd_method['type'] == 'DEL')]\n",
    "\n",
    "        if count == 0:\n",
    "            pd_methods = pd_method\n",
    "            count = 1\n",
    "        else:\n",
    "            pd_methods = pd.concat([pd_methods, pd_method], join='inner')\n",
    "\n",
    "    pd_results = pd.DataFrame.empty\n",
    "    pd_results_PR = pd.DataFrame.empty\n",
    "    count = 0\n",
    "\n",
    "    fig, axs = plt.subplots(1,5,figsize=(30, 4))\n",
    "    fig.suptitle('Precision-Recall curves')\n",
    "\n",
    "\n",
    "    y_true = pd_methods[['sample_t', 'caller']]\n",
    "    y_true.index.name = 'CHROM_POS'\n",
    "    y_true = y_true.groupby(['CHROM_POS'])['sample_t'].sum()\n",
    "    y_true[y_true <= 2] = 0\n",
    "    y_true = y_true.astype(bool)\n",
    "    print(y_true.shape[0])\n",
    "\n",
    "\n",
    "\n",
    "    baselineAUPRC = {}\n",
    "\n",
    "    for mi, method in enumerate(methods):\n",
    "\n",
    "        res_df = (100*pd.Series(samples_tf)).round(decimals=2).to_frame(name='tumor burden')\n",
    "        res_PR_df = (100*pd.Series(samples_tf)).round(decimals=2).to_frame(name='tumor burden')\n",
    "\n",
    "        #y_true = pd_methods[pd_methods['caller'] == method]['sample_0'].astype(bool) + pd_methods[pd_methods['caller'] == method]['sample_0_bis'].astype(bool)\n",
    "        baselineAUPRC[method] = len(y_true[y_true == True])/len(y_true)\n",
    "\n",
    "        y_0 = pd_methods[pd_methods['caller'] == method]['sample_0'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_0 = y_0.fillna(False)\n",
    "        y_1 = pd_methods[pd_methods['caller'] == method]['sample_1'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_1 = y_1.fillna(False)\n",
    "        y_2 = pd_methods[pd_methods['caller'] == method]['sample_2'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_2 = y_2.fillna(False)\n",
    "        y_3 = pd_methods[pd_methods['caller'] == method]['sample_3'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_3 = y_3.fillna(False)\n",
    "        y_4 = pd_methods[pd_methods['caller'] == method]['sample_4'].astype(bool).reindex(y_true.index).squeeze()\n",
    "        y_4 = y_4.fillna(False)\n",
    "\n",
    "        res_df['AUPRC'] = [average_precision_score(y_true, y_0),\n",
    "                          average_precision_score(y_true, y_1),\n",
    "                          average_precision_score(y_true, y_2),\n",
    "                          average_precision_score(y_true, y_3),\n",
    "                          average_precision_score(y_true, y_4),\n",
    "                         ]\n",
    "        res_df['AUPRC - baseline AUPRC'] = [average_precision_score(y_true, y_0) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_1) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_2) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_3) - baselineAUPRC[method],\n",
    "                      average_precision_score(y_true, y_4) - baselineAUPRC[method],\n",
    "                     ]\n",
    "        res_PR_df['precision'] = [precision_score(y_true, y_0),\n",
    "                          precision_score(y_true, y_1),\n",
    "                          precision_score(y_true, y_2),\n",
    "                          precision_score(y_true, y_3),\n",
    "                          precision_score(y_true, y_4),\n",
    "                         ]\n",
    "\n",
    "        res_PR_df['recall'] = [recall_score(y_true, y_0),\n",
    "                          recall_score(y_true, y_1),\n",
    "                          recall_score(y_true, y_2),\n",
    "                          recall_score(y_true, y_3),\n",
    "                          recall_score(y_true, y_4),\n",
    "                         ]\n",
    "        res_PR_df = pd.melt(res_PR_df, id_vars =['tumor burden'], value_vars =['precision', 'recall'],\n",
    "                        var_name='metric', value_name='value')\n",
    "        res_df['caller'] = method\n",
    "        res_PR_df['caller'] = method\n",
    "        res_df.drop(['sample_0'], inplace=True)\n",
    "        res_PR_df.drop([0, 5], inplace=True)\n",
    "\n",
    "        # Plot Precision-Recall curve\n",
    "        alpha_list = [0, 1, .75, .5, .2, .1]\n",
    "        for i in range(1,5):\n",
    "            y_i = pd_methods[pd_methods['caller'] == method]['sample_'+str(i)].astype(bool).reindex(y_true.index).squeeze()\n",
    "            y_i = y_i.fillna(False)\n",
    "            precision, recall, _ = precision_recall_curve(y_true, y_i)\n",
    "            axs[mi].plot(recall, precision, 'o-',\n",
    "                                  label='tf='+str(res_df['tumor burden'].loc['sample_'+str(i)])+ ', AP='+str(round(average_precision_score(y_true, y_i), 2)),\n",
    "                                  c=color_list[mi], alpha=alpha_list[i])\n",
    "        axs[mi].set_xlabel('Recall')\n",
    "        axs[mi].set_ylabel('Precision')\n",
    "        axs[mi].set_ylim([0.0, 1.05])\n",
    "        axs[mi].set_xlim([0.0, 1.05])\n",
    "        axs[mi].set_title(method)\n",
    "        axs[mi].legend()\n",
    "\n",
    "        if count == 0:\n",
    "            pd_results = res_df\n",
    "            pd_results_PR = res_PR_df\n",
    "            count = 1\n",
    "        else:\n",
    "            pd_results = pd.concat([pd_results, res_df], join='inner')\n",
    "            pd_results_PR = pd.concat([pd_results_PR, res_PR_df], join='inner')\n",
    "    plt.show()\n",
    "\n",
    "    sns.catplot(x=\"tumor burden\", y=\"AUPRC\", hue=\"caller\",  \n",
    "                    capsize=.2, height=4, aspect=1.5,\n",
    "                    kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results, legend=False)\n",
    "    for mi, method in enumerate(methods):\n",
    "        plt.axhline(y = baselineAUPRC[method], color = 'k', linestyle = '--') \n",
    "    #plt.ylim([0, 0.5])\n",
    "    plt.legend(bbox_to_anchor=(1,1), loc=\"upper left\")\n",
    "    \n",
    "    sns.catplot(x=\"tumor burden\", y=\"AUPRC - baseline AUPRC\", hue=\"caller\",  \n",
    "                capsize=.2, height=6, aspect=.75,\n",
    "                kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results)\n",
    "    plt.axhline(y = 0, color = 'k', linestyle = '--', label='baseline AUPRC') \n",
    "\n",
    "    sns.catplot(x=\"tumor burden\", y=\"value\", col=\"metric\", hue=\"caller\",  \n",
    "                    capsize=.2, height=6, aspect=.75,\n",
    "                    kind=\"point\", order=sorted(pd_results['tumor burden'].unique(), reverse=True), data=pd_results_PR)\n",
    "    #plt.ylim([0, 0.5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Histograms of true allele frequencies in each tumor sample. Note how increasing admixture increases the prevalence of low- frequency variants.\n",
    "- Benchmarking results for germline SNVs\n",
    "- Benchmarking results for somatic SNVs on exome data.\n",
    "- averaged over the four replicates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- add fake mutations on healthy mixtures"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
