{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import re\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import pysam\n",
    "from collections import Counter\n",
    "from matplotlib import cm\n",
    "from matplotlib.colors import ListedColormap\n",
    "\n",
    "import io\n",
    "from pysam import VariantFile\n",
    "from matplotlib_venn import venn3\n",
    "import venn"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "tab20 = cm.get_cmap('tab10', 8)\n",
    "newcmap_list = []\n",
    "for i in range(4):\n",
    "    newcmap_list.append(tab20.colors[i])\n",
    "for i in range(4):\n",
    "    newcmap_list.append(np.array(list(tab20.colors[i][:-1]) +[0.3]))\n",
    "\n",
    "newcmap = ListedColormap(newcmap_list, name='newcmap')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_vcf(path):\n",
    "    with open(path, 'r') as f:\n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "    res = pd.read_csv(\n",
    "        io.StringIO(''.join(lines[1:])),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': str, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})\n",
    "    return res"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_pd = read_vcf(\"../data/2015-07-31_PATIENTID/PATIENTID-ensemble-annotated.vcf\")\n",
    "\n",
    "\n",
    "vcf_pd['callers'] = vcf_pd['INFO'].apply(foo)\n",
    "vcf_pd['type'] = np.nan\n",
    "vcf_pd['type'][vcf_pd['ALT'].str.len() - vcf_pd['REF'].str.len() == 0] = 'SNV'\n",
    "vcf_pd['type'][vcf_pd['ALT'].str.len() - vcf_pd['REF'].str.len() > 0] = 'INS'\n",
    "vcf_pd['type'][vcf_pd['ALT'].str.len() - vcf_pd['REF'].str.len() < 0] = 'DEL'\n",
    "vcf_pd['type'][vcf_pd['ID'].str.contains('rs')] = 'SNP'\n",
    "vcf_pd['freebayes'] = vcf_pd['INFO'].str.contains('freebayes')\n",
    "vcf_pd['vardict'] = vcf_pd['INFO'].str.contains('vardict')\n",
    "vcf_pd['varscan'] = vcf_pd['INFO'].str.contains('varscan')\n",
    "\n",
    "vcf_pd[vcf_pd['type'] == 'INS']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vcf_pd.groupby('CHROM').size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bam File"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter Bam w.r.t. VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile(\"../data/healthy_chr22_merged-ready.bam\", \"rb\")\n",
    "\n",
    "# ititiate list of reads to remove\n",
    "reads2remove = []\n",
    "\n",
    "# iterate over positions\n",
    "for count, mutation in vcf_pd[vcf_pd['type'] == 'INS'].iterrows():\n",
    "    print('#####')\n",
    "    print(count)\n",
    "#mutation = vcf_pd.iloc[94198]\n",
    "    #print(mutation)\n",
    "\n",
    "    c = 0\n",
    "    t = 0\n",
    "\n",
    "    # get reads in this position\n",
    "    if mutation['type'] == 'INS':\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "            # get supporting reads\n",
    "            # print(read) \n",
    "            seq = read.query_alignment_sequence\n",
    "            pos = (mutation['POS'] -1) - read.reference_start + 1 \n",
    "            #print(mutation['POS'], read.reference_start)\n",
    "            #print(pos)\n",
    "            #print(seq[: pos+len(mutation['ALT'])], seq[pos-1:  pos-1+len(mutation['ALT'])],mutation['ALT'])\n",
    "            #print(read.query_alignment_length, len(seq), read.cigarstring)\n",
    "            cigar = read.cigarstring\n",
    "            #print(cigar)\n",
    "            if cigar is not None:\n",
    "                cigar_pos = re.split('M|I|D|N|S|H|P|=|X',cigar)[:-1]\n",
    "                cigar_states = re.split('[0-9]+',cigar)[1:]\n",
    "                cond2 = False\n",
    "                if 'I' in cigar_states:\n",
    "                    cumul = 0\n",
    "                    indel_pos = None\n",
    "                    for i, cp in enumerate(cigar_pos):\n",
    "                        if (cp != 'S') and (cumul <= pos):\n",
    "                            cumul += -int(cp) if cp == 'D' else int(cp)\n",
    "                            indel_pos = i\n",
    "                    #print(indel_pos)\n",
    "                    #print(cigar_pos, indel_pos)\n",
    "                    if cigar_states[indel_pos] == 'I':\n",
    "                        #print(cigar_states[indel_pos])\n",
    "                        cond2 = True\n",
    "                #print(cigar_states, cigar_pos)\n",
    "                #print(cond2)\n",
    "                if (seq[pos-1:  pos-1+len(mutation['ALT'])] == mutation['ALT']) and cond2:\n",
    "                    print(read.query_name) \n",
    "                reads2remove.append(read.query_name)\n",
    "    elif (mutation['type'] == 'SNV') or (mutation['type'] == 'SNP'):\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "            t += 1\n",
    "            # print('yes')\n",
    "            # get supporting reads\n",
    "            # print(read)\n",
    "            seq = read.query_alignment_sequence\n",
    "            # print(seq)\n",
    "            pos = (mutation['POS'] -1) - read.reference_start + 1\n",
    "            #print(seq[: pos], seq[pos-1], mutation['ALT'])\n",
    "            if len(seq) <= pos-1:\n",
    "                print('length issue')\n",
    "            else:\n",
    "                #print(mutation['ALT'], mutation['REF'])\n",
    "                #print(read.cigarstring)\n",
    "                if (seq[pos-1] == mutation['ALT']):\n",
    "                    c += 1\n",
    "                    #print(read.query_name) \n",
    "                    reads2remove.append(read.query_name)\n",
    "        print(100*c/t)\n",
    "    elif mutation['type'] == 'DEL':\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "        # get supporting reads\n",
    "            #print(read) \n",
    "            #print(read.cigarstring)\n",
    "            seq = read.query_alignment_sequence\n",
    "            #print(seq)\n",
    "            pos = (mutation['POS'] -1) - read.reference_start + 1 \n",
    "            #print(pos)\n",
    "            #print(seq[: pos-1+len(mutation['REF'])], seq[pos-1:  pos-1+len(mutation['REF'])], mutation['REF'], mutation['ALT'])\n",
    "            #print(read.query_alignment_length, len(seq), read.cigarstring)\n",
    "            cigar = read.cigarstring\n",
    "            if cigar is None:\n",
    "                print('cigar is None')\n",
    "            else:\n",
    "                #print(cigar)\n",
    "                cigar_pos = re.split('M|I|D|N|S|H|P|=|X',cigar)[:-1]\n",
    "                cigar_states = re.split('[0-9]+',cigar)[1:]\n",
    "                cond2 = False\n",
    "                if 'D' in cigar_states:\n",
    "                    cumul = 0\n",
    "                    indel_pos = None\n",
    "                    for i, cp in enumerate(cigar_pos):\n",
    "                        if (cp != 'S') and (cumul <= pos):\n",
    "                            cumul += -int(cp) if cp == 'D' else int(cp)\n",
    "                            indel_pos = i\n",
    "                    #print(indel_pos)\n",
    "                    if cigar_states[indel_pos] == 'D':\n",
    "                        #print(cigar_states[indel_pos])\n",
    "                        cond2 = True\n",
    "                #print(cigar_states, cigar_pos)\n",
    "                #print(cond2)\n",
    "                if (seq[pos-1:  pos-1+len(mutation['REF'])] != mutation['REF']) and cond2:\n",
    "                    print(read.query_name) \n",
    "                    reads2remove.append(read.query_name)\n",
    "\n",
    "\n",
    "samfile.close()\n",
    "\n",
    "# filter bam with list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile(\"../data/healthy_chr22_merged-ready.bam\", \"rb\")\n",
    "\n",
    "# ititiate list of reads to remove\n",
    "reads2remove = []\n",
    "\n",
    "# iterate over positions\n",
    "for _, mutation in vcf_pd.iterrows():\n",
    "\n",
    "    c = 0\n",
    "    t = 0\n",
    "\n",
    "    if mutation['type'] == 'INS':\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']): \n",
    "            t += 1\n",
    "            seq = read.query_alignment_sequence\n",
    "            pos = (mutation['POS'] -1) - read.reference_start + 1 \n",
    "            cond1 = (seq[pos-1:  pos-1+len(mutation['ALT'])] == mutation['ALT'])\n",
    "            cigar = read.cigarstring\n",
    "            if cigar is None:\n",
    "                print('INS cigar is None')\n",
    "            else:\n",
    "                cigar_pos = re.split('M|I|D|N|S|H|P|=|X',cigar)[:-1]\n",
    "                cigar_states = re.split('[0-9]+',cigar)[1:]\n",
    "                cond2 = False\n",
    "                if 'I' in cigar_states:\n",
    "                    cumul = 0\n",
    "                    indel_pos = None\n",
    "                    for i, cp in enumerate(cigar_pos):\n",
    "                        if (cp != 'S') and (cumul <= pos):\n",
    "                            cumul += -int(cp) if cp == 'D' else int(cp)\n",
    "                            indel_pos = i\n",
    "                    if cigar_states[indel_pos] == 'I':\n",
    "                        cond2 = True\n",
    "                if cond1 and cond2:\n",
    "                    c += 1\n",
    "                    reads2remove.append(read.query_name)\n",
    "        print(mutation['type'], 100*c/t, t)\n",
    "\n",
    "    elif (mutation['type'] == 'SNV') or (mutation['type'] == 'SNP'):\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "            t += 1\n",
    "            seq = read.query_alignment_sequence\n",
    "            pos = (mutation['POS'] -1) - read.reference_start + 1\n",
    "            if len(seq) <= pos-1:\n",
    "                pass\n",
    "                #print('length issue')\n",
    "            else:\n",
    "                if (seq[pos-1] == mutation['ALT']):\n",
    "                    c += 1\n",
    "                    reads2remove.append(read.query_name)\n",
    "        #print(mutation['type'], 100*c/t, t)\n",
    "\n",
    "    elif mutation['type'] == 'DEL':\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "            t += 1\n",
    "            seq = read.query_alignment_sequence\n",
    "            pos = (mutation['POS'] -1) - read.reference_start + 1 \n",
    "            cond1 = (seq[pos-1:  pos-1+len(mutation['REF'])] != mutation['REF'])\n",
    "            cigar = read.cigarstring\n",
    "            if cigar is None:\n",
    "                print('cigar is None')\n",
    "            else:\n",
    "                cigar_pos = re.split('M|I|D|N|S|H|P|=|X',cigar)[:-1]\n",
    "                cigar_states = re.split('[0-9]+',cigar)[1:]\n",
    "                cond2 = False\n",
    "                if 'D' in cigar_states:\n",
    "                    cumul = 0\n",
    "                    indel_pos = None\n",
    "                    for i, cp in enumerate(cigar_pos):\n",
    "                        if (cp != 'S') and (cumul <= pos):\n",
    "                            cumul += -int(cp) if cp == 'D' else int(cp)\n",
    "                            indel_pos = i\n",
    "                    if cigar_states[indel_pos] == 'D':\n",
    "                        cond2 = True\n",
    "                if cond1 and cond2:\n",
    "                    c += 1\n",
    "                    reads2remove.append(read.query_name)\n",
    "        print(mutation['type'], 100*c/t, t)\n",
    "\n",
    "samfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "print(vcf_pd.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "samfile = pysam.AlignmentFile(\"../data/healthy_chr22_merged-ready.bam\", \"rb\")\n",
    "\n",
    "# ititiate list of reads to remove\n",
    "reads2remove = []\n",
    "\n",
    "# iterate over positions\n",
    "for n, mutation in vcf_pd.iterrows():\n",
    "    print('###############')\n",
    "    print(n)\n",
    "    print(mutation['type'])\n",
    "\n",
    "    c = 0\n",
    "    t = 0\n",
    "\n",
    "    if mutation['type'] == 'INS': # INSERTION\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "            t += 1\n",
    "            seq = read.query_alignment_sequence\n",
    "            pos = mutation['POS'] - read.reference_start\n",
    "            cond1 = seq[pos-1:  pos-1+len(mutation['ALT'])] == mutation['ALT']\n",
    "            cond2 = False\n",
    "            cigar = read.cigarstring\n",
    "            cigar_pos = re.split('M|I|D|N|S|H|P|=|X',cigar)[:-1]\n",
    "            cigar_states = re.split('[0-9]+',cigar)[1:]\n",
    "            if 'I' in cigar_states:\n",
    "                cumul = 0\n",
    "                indel_pos = None\n",
    "                for i, cp in enumerate(cigar_pos):\n",
    "                    if (cigar_states[i] != 'S') and (cumul <= pos):\n",
    "                        cumul += int(cigar_pos[i])\n",
    "                        indel_pos = i\n",
    "                if indel_pos is not None:\n",
    "                    if cigar_states[indel_pos] == 'I':\n",
    "                        cond2 = True\n",
    "            if cond1 and cond2:\n",
    "                c += 1 \n",
    "                reads2remove.append(read.query_name)\n",
    "        print(100*c/t, t)\n",
    "    elif (mutation['type'] == 'SNV') or (mutation['type'] == 'SNP'): # SNP/SNV\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "            t += 1\n",
    "            seq = read.query_sequence\n",
    "            pos = mutation['POS'] - read.reference_start + 1\n",
    "            if len(seq) >= pos:\n",
    "                if (seq[pos-1] == mutation['ALT']):\n",
    "                    c += 1\n",
    "                    reads2remove.append(read.query_name)\n",
    "            else:\n",
    "                print(len(seq), pos)\n",
    "        print(100*c/t, t)\n",
    "    elif mutation['type'] == 'DEL': # DELETION\n",
    "        for read in samfile.fetch(mutation['CHROM'], mutation['POS']-1, mutation['POS']):\n",
    "            t += 1\n",
    "            seq = read.query_alignment_sequence\n",
    "            pos = mutation['POS'] - read.reference_start\n",
    "            cond1 = (seq[pos-1:  pos-1+len(mutation['REF'])] != mutation['REF'])\n",
    "            cigar = read.cigarstring\n",
    "            if cigar is not None:\n",
    "                cigar_pos = re.split('M|I|D|N|S|H|P|=|X',cigar)[:-1]\n",
    "                cigar_states = re.split('[0-9]+',cigar)[1:]\n",
    "                cond2 = False\n",
    "                if 'D' in cigar_states:\n",
    "                    cumul = 0\n",
    "                    indel_pos = None\n",
    "                    for i, cp in enumerate(cigar_pos):\n",
    "                        if (cp != 'S') and (cumul <= pos):\n",
    "                            cumul += -int(cp) if cp == 'D' else int(cp)\n",
    "                            indel_pos = i\n",
    "                    if indel_pos is not None:\n",
    "                        if cigar_states[indel_pos] == 'D':\n",
    "                            cond2 = True\n",
    "                if cond1 and cond2:\n",
    "                    c += 1\n",
    "                    reads2remove.append(read.query_name)\n",
    "        print(100*c/t, t)\n",
    "\n",
    "    \n",
    "samfile.close()\n",
    "\n",
    "# filter bam with list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "    for read in samfile.fetch(mutation['CHROM'], mutation['POS'], mutation['POS']+len(mutation['REF'])):\n",
    "        # get supporting reads\n",
    "        print(read)\n",
    "        seq = read.query_alignment_sequence\n",
    "        print(seq)\n",
    "        if (seq[mutation['POS']:  mutation['POS']+len(mutation['REF'])] != mutation['REF']) and (read.query_alignment_length - len(get_forward_sequence) > 0):\n",
    "            print(read.query_name) \n",
    "            reads2remove.append(read.query_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "a = 'GCAGCCACTCAGGATGTTGGAACCTGGCCATCCCTGCTTCTTTCAGTGGGTGAGGTTGGTGGCTGCTCCACCTGTTCCAGGCACACCCTTAACAGAGGTGGCTGCTTGCTCTTTAAGCCAGCTTGGCCTTGCCTGGCATGCACAGGCCCCG'\n",
    "b = 'GCAGCCACTCAGGATGTTGGAACCTGGCCATCCCTGCTTCTTTCAGTGGGTGAGGTTGGTGGCTGCTCCACCTGTTCCAGGCACACCCTTAACAGAGGTGGCTGCTTGCTCTTTAAGCCAGCTTGGCCTTGCCTGGCATGCACAGGCCCCG'\n",
    "\n",
    "a == b"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
